<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Johannes Zauner">
<meta name="author" content="Aaron Nicholls">
<meta name="author" content="Lisa A. Ostrin">
<meta name="author" content="Manuel Spitschan">
<meta name="dcterms.date" content="2025-12-13">

<title>Supplement 1 – Analysis of human visual experience data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-0b37c64f34216b628666a8dac638b53b.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>
.quarto-notebook .cell-container .cell-decorator {
  display: none;
}
</style>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Analysis of human visual experience data</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Manuscript</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./supplement.html" aria-current="page"> 
<span class="menu-text">Supplement</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/tscnlab/ZaunerEtAl_JVis_2026"> <i class="bi bi-github" role="img" aria-label="Quarto GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#abstract" id="toc-abstract" class="nav-link active" data-scroll-target="#abstract">Abstract</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#clouclip-data-raw-structure-and-import" id="toc-clouclip-data-raw-structure-and-import" class="nav-link" data-scroll-target="#clouclip-data-raw-structure-and-import"><span class="header-section-number">2</span> <code>Clouclip</code> Data: Raw Structure and Import</a></li>
  <li><a href="#veet-data-ambient-light-illuminance-processing" id="toc-veet-data-ambient-light-illuminance-processing" class="nav-link" data-scroll-target="#veet-data-ambient-light-illuminance-processing"><span class="header-section-number">3</span> <code>VEET</code> Data: Ambient Light (Illuminance) Processing</a>
  <ul class="collapse">
  <li><a href="#identifying-non-wear-times" id="toc-identifying-non-wear-times" class="nav-link" data-scroll-target="#identifying-non-wear-times"><span class="header-section-number">3.1</span> Identifying non-wear times</a></li>
  </ul></li>
  <li><a href="#veet-data-spectral-data-processing" id="toc-veet-data-spectral-data-processing" class="nav-link" data-scroll-target="#veet-data-spectral-data-processing"><span class="header-section-number">4</span> <code>VEET</code> Data: Spectral Data Processing</a>
  <ul class="collapse">
  <li><a href="#import" id="toc-import" class="nav-link" data-scroll-target="#import"><span class="header-section-number">4.1</span> Import</a></li>
  <li><a href="#spectral-calibration" id="toc-spectral-calibration" class="nav-link" data-scroll-target="#spectral-calibration"><span class="header-section-number">4.2</span> Spectral calibration</a></li>
  <li><a href="#spectral-reconstruction" id="toc-spectral-reconstruction" class="nav-link" data-scroll-target="#spectral-reconstruction"><span class="header-section-number">4.3</span> Spectral reconstruction</a></li>
  </ul></li>
  <li><a href="#veet-data-time-of-flight-distance" id="toc-veet-data-time-of-flight-distance" class="nav-link" data-scroll-target="#veet-data-time-of-flight-distance"><span class="header-section-number">5</span> <code>VEET</code> Data: Time of flight (distance)</a>
  <ul class="collapse">
  <li><a href="#removing-non-wear-times" id="toc-removing-non-wear-times" class="nav-link" data-scroll-target="#removing-non-wear-times"><span class="header-section-number">5.1</span> Removing non-wear times</a></li>
  </ul></li>
  <li><a href="#saving-the-cleaned-data" id="toc-saving-the-cleaned-data" class="nav-link" data-scroll-target="#saving-the-cleaned-data"><span class="header-section-number">6</span> Saving the Cleaned Data</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">7</span> References</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/tscnlab/ZaunerEtAl_JVis_2026/edit/main/supplement.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/tscnlab/ZaunerEtAl_JVis_2026/issues" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Supplement 1</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
<p class="subtitle lead">Analysis of human visual experience data</p>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Johannes Zauner <a href="mailto:johannes.zauner@tum.de" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0003-2171-4566" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Technical University of Munich, Germany
          </p>
        <p class="affiliation">
            Max Planck Institute for Biological Cybernetics, Tübingen, Germany
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Aaron Nicholls <a href="https://orcid.org/0009-0001-6683-6826" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Reality Labs Research, USA
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Lisa A. Ostrin <a href="https://orcid.org/0000-0001-8675-1478" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            University of Houston College of Optometry, USA
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Manuel Spitschan <a href="https://orcid.org/0000-0002-8572-9268" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Technical University of Munich, Germany
          </p>
        <p class="affiliation">
            Max Planck Institute for Biological Cybernetics, Tübingen, Germany
          </p>
        <p class="affiliation">
            Technical University of Munich Institute of advanced study (TUM-IAS), Munich, Germany
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Last modified:</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 13, 2025</p>
    </div>
  </div>
  
    
    <div>
    <div class="quarto-title-meta-heading">Doi</div>
    <div class="quarto-title-meta-contents">
      <p class="doi">
        <a href="https://doi.org/10.5281/zenodo.16566014">10.5281/zenodo.16566014</a>
      </p>
    </div>
  </div>
  </div>
  


</header>


<section id="abstract" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="abstract">Abstract</h2>
<p>This supplementary document provides a detailed, step-by-step tutorial on importing and preprocessing raw data from two wearable devices: <code>Clouclip</code> and <code>VEET</code>. We describe the structure of the raw datasets recorded by each device and explain how to parse these data, specify time zones, handle special sentinel values, clean missing observations, regularize timestamps to fixed intervals, and aggregate data as needed. All original R code from the <a href="./index.html">main tutorial</a> is shown here for transparency, with additional guidance intended for a broad research audience. We demonstrate how to detect gaps and irregular sampling, convert implicit missing periods into explicit missing values, and address device-specific quirks such as the <code>Clouclip</code>’s use of sentinel codes for “sleep mode” and “out of range” readings. Special procedures for processing the <code>VEET</code>’s rich spectral data (e.g.&nbsp;normalizing sensor counts and reconstructing full spectra from multiple sensor channels) are also outlined. Finally, we show how to save the cleaned datasets from both devices into a single R data file for downstream analysis. This comprehensive walkthrough is designed to ensure reproducibility and to assist researchers in understanding and adapting the data pipeline for their own visual experience datasets.</p>
</section>
<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>Wearable sensors like the <code>Clouclip</code> and the Visual Environment Evaluation Tool (<code>VEET</code>) produce high-dimensional time-series data on viewing distance and light exposure. Proper handling of these raw data is essential before any analysis of visual experience metrics. In the <a href="./index.html">main tutorial</a>, we introduce an analysis pipeline using the open-source R package LightLogR <span class="citation" data-cites="Zauner2025JOpenSourceSoftw">(<a href="#ref-Zauner2025JOpenSourceSoftw" role="doc-biblioref">Zauner, Hartmeyer, and Spitschan 2025</a>)</span> to calculate various distance and light exposure metrics. Here, we present a full account of the data import and preparation steps as a supplement to the methods, with an emphasis on clarity for researchers who may be less familiar with data processing in R.</p>
<p>We use example datasets from a <code>Clouclip</code> device <span class="citation" data-cites="Wen2021ActaOphtalmol Wen2020bjophthalmol">(<a href="#ref-Wen2021ActaOphtalmol" role="doc-biblioref">Wen et al. 2021</a>, <a href="#ref-Wen2020bjophthalmol" role="doc-biblioref">2020</a>)</span> and from a <code>VEET</code> device <span class="citation" data-cites="Sah2025OphtalmicPhysiolOpt">(<a href="#ref-Sah2025OphtalmicPhysiolOpt" role="doc-biblioref">Sah, Narra, and Ostrin 2025</a>)</span> (both provided in the accompanying <a href="https://github.com/tscnlab/ZaunerEtAl_JVis_2026">repository</a>). The <code>Clouclip</code> is a glasses-mounted sensor that records working distance (distance from eyes to object, in cm) and ambient illuminance (in lux) at 5-second intervals. The <code>VEET</code> is a head-mounted multi-modal sensor that logs ambient light and spectral information (along with other data like motion and distance via a depth sensor) - in this exemplary case at 2-second intervals. A single week of continuous data illustrates the contrast in complexity: approximately 1.6&nbsp;MB for the <code>Clouclip</code>’s simple two-column output versus up to 270&nbsp;MB for the <code>VEET</code>’s multi-channel output (due to its higher sampling rate and richer sensor modalities).</p>
<p>In the following sections, we detail how these raw data are structured and how to import and preprocess them using LightLogR. We cover device-specific considerations such as file format quirks and sensor range limitations, as well as general best practices like handling missing timestamps and normalizing sensor readings. All code blocks can be executed in R (with the required packages loaded) to reproduce the cleaning steps. The end result will be clean, regularized datasets (<code>dataCC</code> for Clouclip, <code>dataVEET</code> for VEET light data, <code>dataVEET2</code> for VEET spectral data, and <code>dataVEET3</code> for VEET distance) ready for calculating visual experience metrics. We conclude by saving these cleaned datasets into a single file for convenient reuse.</p>
</section>
<section id="clouclip-data-raw-structure-and-import" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="clouclip-data-raw-structure-and-import"><span class="header-section-number">2</span> <code>Clouclip</code> Data: Raw Structure and Import</h2>
<p>The <code>Clouclip</code> device exports its data as a text file (not a true Excel file despite sometimes using an .xls extension), which is actually tab-separated values. See <a href="#fig-Clouclip_file" class="quarto-xref">Figure&nbsp;1</a> for the raw file.</p>
<div id="fig-Clouclip_file" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-Clouclip_file-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/Clouclip_fileformat.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="Figure&nbsp;1: Screenshot of the first rows of the Clouclip export file format as seen in a text editor"><img src="assets/Clouclip_fileformat.png" class="img-fluid figure-img" style="width:35.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Clouclip_file-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Screenshot of the first rows of the <code>Clouclip</code> export file format as seen in a text editor
</figcaption>
</figure>
</div>
<p>Each record corresponds to one timestamped observation (nominally every 5 seconds) and includes two measured variables: <strong>distance</strong> and <strong>illuminance</strong>. In the sample dataset (<code>Sample_Clouclip.csv</code> provided), the columns are:</p>
<ul>
<li><p><code>Date</code> – the date and time of the observation (in the device’s local time, here one week in 2021).</p></li>
<li><p><code>Dis</code> – the viewing distance in centimeters.</p></li>
<li><p><code>Lux</code> – the ambient illuminance in lux.</p></li>
</ul>
<p>For example, a raw data line might look like:</p>
<pre><code>2021-07-01 08:00:00 45  320</code></pre>
<p>indicating that at <em>2021-07-01 08:00:00</em> local time, the device recorded a working distance of 45&nbsp;cm and illuminance of 320&nbsp;lx. The Clouclip uses special <strong>sentinel values</strong><a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> in these measurement columns to denote certain device states. Specifically, a distance (<code>Dis</code>) value of <strong>204</strong> is a code meaning the object is <em>out of the sensor’s range</em>, and a value of <strong>-1</strong> in either <code>Dis</code> or <code>Lux</code> indicates the device was in <em>sleep mode</em> (not actively recording). During normal operation, distance measurements are limited by the device’s range, and illuminance readings are positive lux values. Any sentinel codes in the raw file need to be handled appropriately, as described below.</p>
<p>We will use <code>LightLogR</code>’s built-in import function for <code>Clouclip</code>, which automatically reads the file, parses the timestamps, and converts sentinel codes into a separate status annotation. To begin, we load the necessary libraries and import the raw <code>Clouclip</code> dataset:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Load required packages</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-2"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-2-1" class="code-annotation-target"><a href="#annotated-cell-2-1"></a><span class="fu">library</span>(tidyverse)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-2-2" class="code-annotation-target"><a href="#annotated-cell-2-2"></a><span class="fu">library</span>(LightLogR)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-2-3" class="code-annotation-target"><a href="#annotated-cell-2-3"></a><span class="fu">library</span>(gt)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-2-4" class="code-annotation-target"><a href="#annotated-cell-2-4"></a><span class="fu">library</span>(ggridges)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-2-5" class="code-annotation-target"><a href="#annotated-cell-2-5"></a><span class="fu">library</span>(downlit)</span>
<span id="annotated-cell-2-6"><a href="#annotated-cell-2-6"></a><span class="fu">library</span>(magick)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-2" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="1" data-code-annotation="1">For tidy data science</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="2" data-code-annotation="2">Wearable analysis package</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="3" data-code-annotation="3">For great tables</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="4" data-code-annotation="4">For ridgeline plots</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="5,6" data-code-annotation="5">These packages are not used, but are needed for dependencies</span>
</dd>
</dl>
</div>
</div>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Import of Clouclip data</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-3"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-3-1" class="code-annotation-target"><a href="#annotated-cell-3-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/Sample_Clouclip.csv"</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-3-2" class="code-annotation-target"><a href="#annotated-cell-3-2"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-3-3" class="code-annotation-target"><a href="#annotated-cell-3-3"></a>dataCC <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">Clouclip</span>(path, <span class="at">tz =</span> tz, <span class="at">manual.id =</span> <span class="st">"Clouclip"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-3" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="1" data-code-annotation="1">Define file path</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="2" data-code-annotation="2">Time zone in which device was recording (e.g., US Central Time)</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="3" data-code-annotation="3">Import Clouclip data</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Successfully read in 58'081 observations across 1 Ids from 1 Clouclip-file(s).
Timezone set is US/Central.
The system timezone is UTC. Please correct if necessary!

First Observation: 2021-02-06 17:12:47
Last Observation: 2021-02-14 17:12:36
Timespan: 8 days

Observation intervals: 
  Id       interval.time            n pct     
1 Clouclip 5s                   54572 93.9601%
2 Clouclip 17s                     12 0.0207% 
3 Clouclip 18s                     14 0.0241% 
4 Clouclip 120s (~2 minutes)     3479 5.9900% 
5 Clouclip 128s (~2.13 minutes)     1 0.0017% 
6 Clouclip 132s (~2.2 minutes)      1 0.0017% 
7 Clouclip 133s (~2.22 minutes)     1 0.0017% </code></pre>
</div>
<div class="cell-output-display">
<div id="fig-importCC" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-importCC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-importCC-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Figure&nbsp;2: Overview plot of imported Clouclip data"><img src="supplement_files/figure-html/fig-importCC-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-importCC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Overview plot of imported Clouclip data
</figcaption>
</figure>
</div>
</div>
</div>
<p>In this code, <code>import$Clouclip()</code> reads the tab-delimited file and returns a <strong>tibble</strong><a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> (saved in the variable <code>dataCC</code>) containing the data. We specify&nbsp;<code>tz = "US/Central"</code>&nbsp;because the device’s clock was set to U.S. Central time; this ensures that the&nbsp;<code>Datetime</code>values are properly interpreted with the correct time zone. The argument&nbsp;<code>manual.id = "Clouclip"</code>&nbsp;simply tags the dataset with an identifier (useful if combining data from multiple devices).</p>
<p>During import, LightLogR automatically handles the <code>Clouclip</code>’s sentinel codes. The&nbsp;<code>Date</code>&nbsp;column from the raw file is parsed into a <code>POSIXct</code> date-time (<code>Datetime</code>) with the specified time zone. The&nbsp;<code>Lux</code>&nbsp;and&nbsp;<code>Dis</code>&nbsp;columns are read as numeric, but any occurrences of&nbsp;<strong>-1</strong>&nbsp;or&nbsp;<strong>204</strong>&nbsp;are treated specially: these are replaced with&nbsp;<code>NA</code>&nbsp;(missing values) in the numeric columns, and corresponding status columns&nbsp;<code>Lux_status</code>&nbsp;and&nbsp;<code>Dis_status</code>&nbsp;are created to indicate the reason for those <code>NA</code> values. For example, if a&nbsp;<code>Dis</code>&nbsp;value of <em>204</em> was encountered, that row’s&nbsp;<code>Dis</code>&nbsp;will be set to NA and&nbsp;<code>Dis_status</code>&nbsp;will contain&nbsp;<code>"out_of_range"</code>; if&nbsp;<code>Lux</code>&nbsp;or&nbsp;<code>Dis</code>&nbsp;was -1, the status is&nbsp;<code>"sleep_mode"</code>. We will set all other readings to &nbsp;<code>"operational"</code> (meaning the device was functioning normally at that time) for visualisation purposes.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Print the first 6 rows of the Clouclip dataset</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
# Groups:   Id [1]
  Id       file.name       Datetime              Dis   Lux Lux_status Dis_status
  &lt;fct&gt;    &lt;chr&gt;           &lt;dttm&gt;              &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;     
1 Clouclip Sample_Clouclip 2021-02-06 17:12:47    92    92 &lt;NA&gt;       &lt;NA&gt;      
2 Clouclip Sample_Clouclip 2021-02-06 17:12:52    92    92 &lt;NA&gt;       &lt;NA&gt;      
3 Clouclip Sample_Clouclip 2021-02-06 17:12:57    92    92 &lt;NA&gt;       &lt;NA&gt;      
4 Clouclip Sample_Clouclip 2021-02-06 17:13:02    92    92 &lt;NA&gt;       &lt;NA&gt;      
5 Clouclip Sample_Clouclip 2021-02-06 17:13:07    92    92 &lt;NA&gt;       &lt;NA&gt;      
6 Clouclip Sample_Clouclip 2021-02-06 17:13:12    92    92 &lt;NA&gt;       &lt;NA&gt;      </code></pre>
</div>
</div>
<p>After import, it is good practice to get an overview of the data. The import function by default prints a brief summary (and generates an overview plot of the timeline) showing the number of observations, the time span, and any irregularities or large gaps. In our case, the <code>Clouclip</code> summary indicates the data spans one week and reveals that there are&nbsp;<strong>irregular intervals</strong>&nbsp;in the timestamps. This means some observations do not occur exactly every 5 seconds as expected. We can programmatically check for irregular timing:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Check if data are on a perfectly regular 5-second schedule</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">has_irregulars</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>If the result is&nbsp;<code>TRUE</code>, it confirms that the time sequence has deviations from the regular interval. Indeed, our example dataset has many small timing irregularities and gaps (periods with no data). Understanding the pattern of these missing or irregular readings is important. We can visualize them using a gap plot:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Plot gaps and irregular timestamps for Clouclip data</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-6"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-6-1"><a href="#annotated-cell-6-1"></a>y.label <span class="ot">&lt;-</span> <span class="st">"Distance (cm)"</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-6-2" class="code-annotation-target"><a href="#annotated-cell-6-2"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">gg_gaps</span>(Dis,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-6-3" class="code-annotation-target"><a href="#annotated-cell-6-3"></a>                  <span class="at">include.implicit.gaps =</span> <span class="cn">FALSE</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-6" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-6-4" class="code-annotation-target"><a href="#annotated-cell-6-4"></a>                  <span class="at">show.irregulars =</span> <span class="cn">TRUE</span>,</span>
<span id="annotated-cell-6-5"><a href="#annotated-cell-6-5"></a>                  <span class="at">y.axis.label =</span> y.label,</span>
<span id="annotated-cell-6-6"><a href="#annotated-cell-6-6"></a>                  <span class="at">group.by.days =</span> <span class="cn">TRUE</span>, <span class="at">col.irregular =</span> <span class="fu">alpha</span>(<span class="st">"red"</span>, <span class="fl">0.03</span>)</span>
<span id="annotated-cell-6-7"><a href="#annotated-cell-6-7"></a>                  ) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="cn">NULL</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-6" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="2" data-code-annotation="1">Basing the gap-figure on the <code>Dis</code> distance variable</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="3" data-code-annotation="2">Only show missing observations (<code>NA</code> values)</span>
</dd>
<dt data-target-cell="annotated-cell-6" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-6" data-code-lines="4" data-code-annotation="3">Highlight irregular timing</span>
</dd>
</dl>
</div>
<div class="cell-output-display">
<div id="fig-irregular" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-irregular-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-irregular-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Figure&nbsp;3: Visualization of gaps and irregular data. Black traces show available data. Red shaded areas show times of missing data. Red dots show instances where observations occur off the regular interval from start to finish, i.e., irregular data."><img src="supplement_files/figure-html/fig-irregular-1.png" class="img-fluid figure-img" width="672"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-irregular-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Visualization of gaps and irregular data. Black traces show available data. Red shaded areas show times of missing data. Red dots show instances where observations occur off the regular interval from start to finish, i.e., irregular data.
</figcaption>
</figure>
</div>
</div>
</div>
<p>In <a href="#fig-irregular" class="quarto-xref">Figure&nbsp;3</a>, time periods where data are missing appear as red-shaded areas, and any off-schedule observation times are marked with red dots. The <code>Clouclip</code> example shows extensive gaps (red blocks) on certain days and irregular timing on all days except the first and last. These irregular timestamps likely arise from the device’s logging process (e.g.&nbsp;slight clock drift or buffering when the device was turned on/off). Such issues must be addressed before further analysis.</p>
<p>When faced with irregular or gapped data, we recommend a few strategies:</p>
<ul>
<li><p><em>Remove leading/trailing segments that cause irregularity.</em>&nbsp;For example, if only the first day is regular and subsequent days drift, one might exclude the problematic portion using date filters (see&nbsp;<a href="https://tscnlab.github.io/LightLogR/reference/filter_Datetime.html">filter_Date() / filter_Datetime()</a>&nbsp;in LightLogR).</p></li>
<li><p><em>Round timestamps to the nearest regular interval.</em>&nbsp;This relabels slightly off-schedule times back to the 5-second grid (using&nbsp;<a href="https://tscnlab.github.io/LightLogR/reference/cut_Datetime.html">cut_Datetime()</a>&nbsp;with a 5-second interval), provided the deviations are small and this rounding won’t create duplicate timestamps.</p></li>
<li><p><em>Aggregate to a coarser time interval.</em>&nbsp;For instance, grouping and averaging data into 1-minute bins with&nbsp;<a href="https://tscnlab.github.io/LightLogR/reference/aggregate_Datetime.html">aggregate_Datetime()</a> can mask irregularities at finer scales, at the cost of some temporal resolution.</p></li>
</ul>
<p>In this case, the deviations from the 5-second schedule are relatively minor. We choose to&nbsp;<strong>round the timestamps to the nearest 5 seconds</strong>&nbsp;to enforce a uniform sampling grid, which simplifies downstream gap handling. We further add a separate date column for convenience:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Regularize timestamps by rounding to nearest 5-second interval</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-7"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-7-1"><a href="#annotated-cell-7-1"></a>dataCC <span class="ot">&lt;-</span> dataCC <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-7-2" class="code-annotation-target"><a href="#annotated-cell-7-2"></a>  <span class="fu">cut_Datetime</span>(<span class="st">"5 secs"</span>, <span class="at">New.colname =</span> Datetime) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-7-3" class="code-annotation-target"><a href="#annotated-cell-7-3"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-7" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="2" data-code-annotation="1">Round times to 5-second bins</span>
</dd>
<dt data-target-cell="annotated-cell-7" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="3" data-code-annotation="2">Add a Date column for grouping by day</span>
</dd>
</dl>
</div>
</div>
<p>After this operation, all&nbsp;<code>Datetime</code>&nbsp;entries in&nbsp;<code>dataCC</code>&nbsp;align perfectly on 5-second boundaries (e.g.&nbsp;<strong>08:00:00</strong>,&nbsp;<strong>08:00:05</strong>,&nbsp;<strong>08:00:10</strong>, etc.). We can verify that no irregular intervals remain by re-running&nbsp;<code>has_irregulars()</code>:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Re-check if data are on a perfectly regular 5-second schedule</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">has_irregulars</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<p>Next, we want to quantify the missing data. LightLogR distinguishes between&nbsp;<strong>explicit missing values</strong>&nbsp;(actual&nbsp;<code>NA</code>s in the data, possibly from sentinel replacements or gaps we have filled in) and&nbsp;<strong>implicit missing intervals</strong>&nbsp;(time points where the device&nbsp;<em>should</em>&nbsp;have a reading but none was recorded, and we have not yet filled them in). See <a href="#fig-gaps" class="quarto-xref">Figure&nbsp;4</a> for a visual aid to these terms. Initially, many gaps are still implicit (between the first and last timestamp of each day).</p>
<div id="fig-gaps" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-gaps-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="assets/gap_terminology.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Terminology of gaps and irregular data in LightLogR"><img src="assets/gap_terminology.png" class="img-fluid figure-img"></a></p>
<figcaption>Terminology of gaps and irregular data in LightLogR</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="assets/gap_handler.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="Handling of gaps with LightLogR’s gap_handler() function"><img src="assets/gap_handler.png" class="img-fluid figure-img"></a></p>
<figcaption>Handling of gaps with LightLogR’s <code>gap_handler()</code> function</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-gaps-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Gaps and irregular data
</figcaption>
</figure>
</div>
<p>We can generate a gap summary table:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Summarize observed vs missing data by day for distance</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-9"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-9-1"><a href="#annotated-cell-9-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">gap_table</span>(Dis, <span class="at">Variable.label =</span> <span class="st">"Distance (cm)"</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-9-2" class="code-annotation-target"><a href="#annotated-cell-9-2"></a>  <span class="fu">cols_hide</span>(<span class="fu">contains</span>(<span class="st">"_n"</span>))</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-9" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="2" data-code-annotation="1">Hide absolute counts for brevity in output</span>
</dd>
</dl>
</div>
<div id="tbl-gaps" class="cell quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-gaps-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;1: Summary of missing and observed data for the Clouclip device
</figcaption>
<div aria-describedby="tbl-gaps-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div id="aoamsqfhtb" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#aoamsqfhtb table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#aoamsqfhtb thead, #aoamsqfhtb tbody, #aoamsqfhtb tfoot, #aoamsqfhtb tr, #aoamsqfhtb td, #aoamsqfhtb th {
  border-style: none;
}

#aoamsqfhtb p {
  margin: 0;
  padding: 0;
}

#aoamsqfhtb .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#aoamsqfhtb .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#aoamsqfhtb .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#aoamsqfhtb .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#aoamsqfhtb .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#aoamsqfhtb .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#aoamsqfhtb .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#aoamsqfhtb .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#aoamsqfhtb .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#aoamsqfhtb .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#aoamsqfhtb .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#aoamsqfhtb .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#aoamsqfhtb .gt_spanner_row {
  border-bottom-style: hidden;
}

#aoamsqfhtb .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#aoamsqfhtb .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#aoamsqfhtb .gt_from_md > :first-child {
  margin-top: 0;
}

#aoamsqfhtb .gt_from_md > :last-child {
  margin-bottom: 0;
}

#aoamsqfhtb .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#aoamsqfhtb .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#aoamsqfhtb .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#aoamsqfhtb .gt_row_group_first td {
  border-top-width: 2px;
}

#aoamsqfhtb .gt_row_group_first th {
  border-top-width: 2px;
}

#aoamsqfhtb .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#aoamsqfhtb .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#aoamsqfhtb .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#aoamsqfhtb .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#aoamsqfhtb .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#aoamsqfhtb .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#aoamsqfhtb .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#aoamsqfhtb .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#aoamsqfhtb .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#aoamsqfhtb .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#aoamsqfhtb .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#aoamsqfhtb .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#aoamsqfhtb .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#aoamsqfhtb .gt_left {
  text-align: left;
}

#aoamsqfhtb .gt_center {
  text-align: center;
}

#aoamsqfhtb .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#aoamsqfhtb .gt_font_normal {
  font-weight: normal;
}

#aoamsqfhtb .gt_font_bold {
  font-weight: bold;
}

#aoamsqfhtb .gt_font_italic {
  font-style: italic;
}

#aoamsqfhtb .gt_super {
  font-size: 65%;
}

#aoamsqfhtb .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#aoamsqfhtb .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#aoamsqfhtb .gt_indent_1 {
  text-indent: 5px;
}

#aoamsqfhtb .gt_indent_2 {
  text-indent: 10px;
}

#aoamsqfhtb .gt_indent_3 {
  text-indent: 15px;
}

#aoamsqfhtb .gt_indent_4 {
  text-indent: 20px;
}

#aoamsqfhtb .gt_indent_5 {
  text-indent: 25px;
}

#aoamsqfhtb .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#aoamsqfhtb div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table do-not-create-environment cell caption-top table table-sm table-striped small" data-quarto-bootstrap="false">
<colgroup>
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
</colgroup>
<thead>
<tr class="gt_heading header">
<td colspan="14" class="gt_heading gt_title gt_font_normal">Summary of available and missing data</td>
</tr>
<tr class="gt_heading even">
<td colspan="14" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border">Variable: Distance (cm)</td>
</tr>
<tr class="gt_col_headings gt_spanner_row header">
<th data-quarto-table-cell-role="th" scope="col"></th>
<th colspan="5" id="data_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold; font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Data
</div></th>
<th colspan="8" id="missing_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold; font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Missing
</div></th>
</tr>
<tr class="gt_col_headings gt_spanner_row even">
<th rowspan="2" id="a::stub" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col"></th>
<th colspan="2" id="Regular" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Regular
</div></th>
<th id="irregular_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Irregular
</div></th>
<th id="Range" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Range
</div></th>
<th id="interval_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Interval
</div></th>
<th colspan="4" id="gap_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Gaps
</div></th>
<th colspan="2" id="implicit_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Implicit
</div></th>
<th colspan="2" id="explicit_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Explicit
</div></th>
</tr>
<tr class="gt_col_headings header">
<th id="available_data" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="available_data_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="number_irregulars" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">n<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1,2</sup></span></th>
<th id="total_duration" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="interval" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="number_gaps" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">N</th>
<th id="mean_gap_duration" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ø</th>
<th id="missing_data" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_data_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="missing_implicit" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_implicit_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="missing_explicit" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_explicit_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<th id="grand_summary_stub_1" class="gt_row gt_left gt_stub gt_grand_summary_row gt_last_grand_summary_row_top" data-quarto-table-cell-role="th" scope="row" style="background-color: #E5E5E5"><strong>Overall</strong></th>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 available_data" style="background-color: #E5E5E5">2d 14h 43m 10s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 available_data_pct" style="background-color: #E5E5E5">29.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 number_irregulars" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 total_duration" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1w 2d</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 interval" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 number_gaps" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2,690</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 mean_gap_duration" style="background-color: #E5E5E5">1h 35m 57s</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_data" style="background-color: #E5E5E5">6d 9h 16m 50s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_data_pct" style="background-color: #E5E5E5">71.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_implicit" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5d 15h 19m 55s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_implicit_pct" style="background-color: #E5E5E5">62.7%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_explicit" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">17h 56m 55s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_explicit_pct" style="background-color: #E5E5E5">8.3%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="Clouclip - 2021-02-06" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-06</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_1" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 available_data">43m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 available_data_pct">3.0%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">26</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 mean_gap_duration">53m 43s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 missing_data">23h 16m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 missing_data_pct">97.0%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">22h 53m 20s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 missing_implicit_pct">95.4%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">23m 10s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 missing_explicit_pct">1.6%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="Clouclip - 2021-02-07" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-07</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_2" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 available_data">2h 45m</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 available_data_pct">11.5%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">139</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 mean_gap_duration">9m 10s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 missing_data">21h 15m</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 missing_data_pct">88.5%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">19h 42m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 missing_implicit_pct">82.1%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1h 32m 25s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 missing_explicit_pct">6.4%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="Clouclip - 2021-02-08" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-08</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_3" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 available_data">11h 13m 55s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 available_data_pct">46.8%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">443</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 mean_gap_duration">1m 44s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 missing_data">12h 46m 5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 missing_data_pct">53.2%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">10h 47m 50s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 missing_implicit_pct">45.0%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1h 58m 15s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 missing_explicit_pct">8.2%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="Clouclip - 2021-02-09" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-09</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_4" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 available_data">8h 46m 25s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 available_data_pct">36.6%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">278</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 mean_gap_duration">3m 17s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 missing_data">15h 13m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 missing_data_pct">63.4%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">13h 50m</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 missing_implicit_pct">57.6%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1h 23m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 missing_explicit_pct">5.8%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="Clouclip - 2021-02-10" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-10</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_5" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 available_data">7h 1m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 available_data_pct">29.3%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">367</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 mean_gap_duration">2m 47s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 missing_data">16h 58m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 missing_data_pct">70.7%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">14h 18m 40s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 missing_implicit_pct">59.6%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2h 39m 50s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 missing_explicit_pct">11.1%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="Clouclip - 2021-02-11" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-11</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_6" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 available_data">8h 31m 55s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 available_data_pct">35.5%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">423</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 mean_gap_duration">2m 12s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 missing_data">15h 28m 5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 missing_data_pct">64.5%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">11h 43m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 missing_implicit_pct">48.9%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">3h 44m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 missing_explicit_pct">15.6%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="Clouclip - 2021-02-12" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-12</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_7" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 available_data">12h 17m 55s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 available_data_pct">51.2%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">417</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 mean_gap_duration">1m 41s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 missing_data">11h 42m 5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 missing_data_pct">48.8%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">9h 17m 45s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 missing_implicit_pct">38.7%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2h 24m 20s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 missing_explicit_pct">10.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="Clouclip - 2021-02-13" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-13</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_8" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 available_data">10h 32m 15s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 available_data_pct">43.9%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">527</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 mean_gap_duration">1m 32s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 missing_data">13h 27m 45s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 missing_data_pct">56.1%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">10h 42m 10s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 missing_implicit_pct">44.6%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2h 45m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 missing_explicit_pct">11.5%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="Clouclip - 2021-02-14" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-14</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_9" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 available_data">50m 45s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 available_data_pct">3.5%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">70</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 mean_gap_duration">19m 51s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 missing_data">23h 9m 15s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 missing_data_pct">96.5%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">22h 4m 5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 missing_implicit_pct">92.0%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1h 5m 10s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 missing_explicit_pct">4.5%</td>
</tr>
</tbody><tfoot>
<tr class="gt_footnotes odd">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span> If n &gt; 0: it is possible that the other summary statistics are affected, as they are calculated based on the most prominent interval.</td>
</tr>
<tr class="gt_footnotes even">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>2</sup></span> Number of (missing or actual) observations</td>
</tr>
<tr class="gt_footnotes odd">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span> Based on times, not necessarily number of observations</td>
</tr>
</tfoot>

</table>

</div>
</div>
</div>
</figure>
</div>
</div>
<p>This summary (<a href="#tbl-gaps" class="quarto-xref">Table&nbsp;1</a>) breaks down, for each day, how much data is present vs.&nbsp;missing. It reports the total duration of recorded data and the duration of gaps. After rounding the times, there are no irregular timestamp issues, but we see substantial <strong>implicit gaps</strong> — periods where the device was not recording (e.g., overnight when the device was likely not worn or was in sleep mode). Notably, the first and last days of the week have very little data (less than 1 hour each), since they probably represent partial recording days (the trial started and ended on those days).</p>
<p>To prepare the dataset for analysis, we will convert all those implicit gaps into explicit missing entries, and remove days that are mostly incomplete. Converting implicit gaps means inserting rows with <code>NA</code> for each missing 5-second slot, so that the time series becomes continuous and explicit about missingness (see <a href="#fig-gaps" class="quarto-xref">Figure&nbsp;4</a>). We use <code>gap_handler()</code> for this, and then drop the nearly-empty days:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Convert implicit gaps to explicit NA gaps, and drop days with &lt;1 hour of data</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-10"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-10-1"><a href="#annotated-cell-10-1"></a>dataCC <span class="ot">&lt;-</span> dataCC <span class="sc">|&gt;</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-10-2" class="code-annotation-target"><a href="#annotated-cell-10-2"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(Lux_status, Dis_status), <span class="sc">~</span> <span class="fu">replace_na</span>(.x, <span class="st">"operational"</span>))) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-10-3" class="code-annotation-target"><a href="#annotated-cell-10-3"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-10" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-10-4" class="code-annotation-target"><a href="#annotated-cell-10-4"></a>  <span class="fu">remove_partial_data</span>(Dis, <span class="at">threshold.missing =</span> <span class="st">"-1 hour"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-10" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="2" data-code-annotation="1">First ensure that status columns have an “operational” tag for non-missing periods</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="3" data-code-annotation="2">Fill in all implicit gaps with explicit NA rows (for full days range)</span>
</dd>
<dt data-target-cell="annotated-cell-10" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-10" data-code-lines="4" data-code-annotation="3">Remove any day that has less than 1 hour of recorded data</span>
</dd>
</dl>
</div>
</div>
<p>After these steps,&nbsp;<code>dataCC</code>&nbsp;contains continuous 5-second timestamps for each day that remains. We chose a threshold of “-1 hours” to remove days with less than one hour of data, which in this dataset removes the first and last (partial) days. The cleaned <code>Clouclip</code> data now covers six full days with bouts of continuous wear.</p>
<p>It is often helpful to double-check how the sentinel values and missing data are distributed over time. We can visualize the distance time-series with status annotations and day/night periods:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Visualize observations and sentinel states</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-11"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-11-1" class="code-annotation-target"><a href="#annotated-cell-11-1"></a>coordinates <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">29.75</span>, <span class="sc">-</span><span class="fl">95.36</span>)</span>
<span id="annotated-cell-11-2"><a href="#annotated-cell-11-2"></a>dataCC <span class="sc">|&gt;</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-11-3" class="code-annotation-target"><a href="#annotated-cell-11-3"></a>  <span class="fu">fill</span>(<span class="fu">c</span>(Lux_status, Dis_status), <span class="at">.direction =</span> <span class="st">"downup"</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-11-4" class="code-annotation-target"><a href="#annotated-cell-11-4"></a>  <span class="fu">gg_day</span>(<span class="at">y.axis =</span> Dis, <span class="at">geom =</span> <span class="st">"line"</span>, <span class="at">y.axis.label =</span> y.label) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-11-5" class="code-annotation-target"><a href="#annotated-cell-11-5"></a>  <span class="fu">gg_state</span>(Dis_status, <span class="at">aes_fill =</span> Dis_status, <span class="at">ymin =</span> <span class="dv">0</span>, <span class="at">ymax =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-11" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-11-6" class="code-annotation-target"><a href="#annotated-cell-11-6"></a>  <span class="fu">gg_photoperiod</span>(coordinates, <span class="at">alpha =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="annotated-cell-11-7"><a href="#annotated-cell-11-7"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-11" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="1" data-code-annotation="1">Setting coordinates for Houston, Texas (recording location)</span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="3" data-code-annotation="2">Retain status markers until a new marker arrives</span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="4" data-code-annotation="3">Create the basic plot</span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="5" data-code-annotation="4">Add the status times</span>
</dd>
<dt data-target-cell="annotated-cell-11" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-11" data-code-lines="6" data-code-annotation="5">Add the photoperiod (day/night)</span>
</dd>
</dl>
</div>
<div class="cell-output-display">
<div id="fig-state" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-state-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-state-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6" title="Figure&nbsp;5: Distance measurements across days. Blue, grey and yellow-colored bars at the bottom of each day show sentinel states of the device. Blue indicates an operational status, grey sleep mode (not recording), and yellow an out of range measurement. Shaded areas in the main plot show nighttime from civil dusk until dawn, which are calculated based on the recording date and geographic coordinates"><img src="supplement_files/figure-html/fig-state-1.png" class="img-fluid figure-img" width="672"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-state-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Distance measurements across days. Blue, grey and yellow-colored bars at the bottom of each day show sentinel states of the device. Blue indicates an operational status, grey sleep mode (not recording), and yellow an out of range measurement. Shaded areas in the main plot show nighttime from civil dusk until dawn, which are calculated based on the recording date and geographic coordinates
</figcaption>
</figure>
</div>
</div>
</div>
<p>In this plot, blue segments indicate times when the Clouclip was&nbsp;<em>operational</em>&nbsp;(actively measuring), grey segments indicate the device in&nbsp;<em>sleep mode</em>&nbsp;(no recording, typically at night), and yellow segments indicate&nbsp;<em>out-of-range</em>&nbsp;distance readings. The red outlined regions show nighttime (from civil dusk to dawn) based on the given latitude/longitude and dates. As expected, most of the grey “sleep” periods align with night hours, and we see a few yellow spans when the user’s viewing distance exceeded the device’s range (e.g.&nbsp;presumably when no object was within 100&nbsp;cm, such as when the user looked into the distance). At this stage, the Clouclip dataset&nbsp;<code>dataCC</code>&nbsp;is fully preprocessed: all timestamps are regular 5-second intervals, missing data are explicitly marked, extraneous partial days are removed, and sentinel codes are handled via the status columns. The data are ready for calculating daily distance and light exposure metrics (as done in the <a href="./index.html">main tutorial</a>’s Results).</p>
</section>
<section id="veet-data-ambient-light-illuminance-processing" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="veet-data-ambient-light-illuminance-processing"><span class="header-section-number">3</span> <code>VEET</code> Data: Ambient Light (Illuminance) Processing</h2>
<p>The&nbsp;<strong>VEET</strong>&nbsp;device <span class="citation" data-cites="Sah2025OphtalmicPhysiolOpt">Sullivan et al. (<a href="#ref-Sullivan2024" role="doc-biblioref">2024</a>)</span> is a more complex logger that records multiple data modalities in one combined file. Its raw data file contains interleaved records for different sensor types, distinguished by a “modality” field. We focus first on the ambient light sensor modality (abbreviated&nbsp;<strong>ALS</strong>), which provides broad-spectrum illuminance readings (lux) and related information like sensor gains and flicker, recorded every 2 seconds. Later we will import the spectral sensor modality (<strong>PHO</strong>) for spectral irradiance data, and the time-of-flight modality (<strong>TOF</strong>) for distance data.</p>
<p>In the <code>VEET</code>’s export file, each line includes a timestamp and a modality code, followed by fields specific to that modality. Importantly, this means that the <code>VEET</code> export is not rectangular, i.e., tabular (see <a href="#fig-VEET_file" class="quarto-xref">Figure&nbsp;6</a>). This makes it challenging for many import functions that expect the equal number of columns in every row, which is not the case in this instance. For the&nbsp;<strong>ALS modality</strong>, the relevant columns include a high-resolution timestamp (in Unix epoch format), integration time, UV/VIS/IR sensor gain settings, raw UV/VIS/IR sensor counts, a flicker measurement, and the computed illuminance in lux. For example, the ALS data columns are named:&nbsp;<code>time_stamp</code>,&nbsp;<code>integration_time</code>,&nbsp;<code>uvGain</code>,&nbsp;<code>visGain</code>,&nbsp;<code>irGain</code>,&nbsp;<code>uvValue</code>,&nbsp;<code>visValue</code>,&nbsp;<code>irValue</code>,&nbsp;<code>Flicker</code>, and&nbsp;<code>Lux</code>.</p>
<div id="fig-VEET_file" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET_file-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/VEET_fileformat.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7" title="Figure&nbsp;6: Screenshot of the first rows of the VEET export file format as seen in a text editor"><img src="assets/VEET_fileformat.png" class="img-fluid figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET_file-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Screenshot of the first rows of the <code>VEET</code> export file format as seen in a text editor
</figcaption>
</figure>
</div>
<p>For the&nbsp;<strong>PHO (spectral) modality</strong>, the columns include a timestamp, integration time, a general&nbsp;<code>Gain</code>&nbsp;factor, and nine sensor channel readings covering different wavelengths (with names like&nbsp;<code>s415, s445, ..., s680, s940</code>) as well as a&nbsp;<code>Dark</code>&nbsp;channel and a broadband channel&nbsp;<code>Clear</code>. In essence, the <code>VEET</code>’s spectral sensor captures light in several wavelength bands (from ~415&nbsp;nm up to 940&nbsp;nm, plus an infrared and “clear” channel) rather than outputting a single lux value like the ambient light sensor does (<em>PHO</em>).</p>
<p>To import the <code>VEET</code> ambient light data, we again use the LightLogR import function, specifying the <code>ALS</code> modality. The raw VEET data in our example is provided as a zip file (<code>01_VEET_L.csv.zip</code>) containing the logged data for one week. We do the following:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Import VEET Ambient Light Sensor (ALS) data</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-12"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-12-1"><a href="#annotated-cell-12-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="annotated-cell-12-2"><a href="#annotated-cell-12-2"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-12" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-12-3" class="code-annotation-target"><a href="#annotated-cell-12-3"></a>dataVEET <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"ALS"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-12" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-12" data-code-lines="3" data-code-annotation="1">In difference to the <code>Clouclip</code> file, we simply respecify the device type with <code>import$VEET(...)</code>, but must also provide a <code>modality</code> argument.</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Successfully read in 304'193 observations across 1 Ids from 1 VEET-file(s).
Timezone set is US/Central.
The system timezone is UTC. Please correct if necessary!
1 observations were dropped due to a missing or non-parseable Datetime value (e.g., non-valid timestamps during DST jumps). 

First Observation: 2024-06-04 15:00:37
Last Observation: 2024-06-12 08:29:43
Timespan: 7.7 days

Observation intervals: 
  Id    interval.time              n pct      
1 VEET  0s                         1 0.00033% 
2 VEET  1s                      1957 0.64334% 
3 VEET  2s                    300147 98.67025%
4 VEET  3s                      2074 0.68181% 
5 VEET  4s                         3 0.00099% 
6 VEET  9s                         5 0.00164% 
7 VEET  10s                        3 0.00099% 
8 VEET  109s (~1.82 minutes)       1 0.00033% 
9 VEET  59077s (~16.41 hours)      1 0.00033% </code></pre>
</div>
<div class="cell-output-display">
<div id="fig-VEET-overview" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-overview-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-VEET-overview-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8" title="Figure&nbsp;7: Overview plot of imported VEET data"><img src="supplement_files/figure-html/fig-VEET-overview-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-overview-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: Overview plot of imported VEET data
</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>We get one warning as a single time stamp could not be parsed into a datetime - with ~300k observations, this is not an issue.</p>
</div>
</div>
<p>This call reads in only the lines corresponding to the <code>ALS</code> modality from the <code>VEET</code> file. The result&nbsp;<code>dataVEET</code>&nbsp;is a tibble<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> with columns such as&nbsp;<code>Datetime</code>&nbsp;(parsed from the&nbsp;<code>time_stamp</code>&nbsp; to POSIXct in US/Central time),&nbsp;<code>Lux</code>&nbsp;(illuminance in lux),&nbsp;<code>Flicker</code>, and the various sensor gains/values. Columns we don’t need for our analysis, like the modality code or file name, are also included but can be ignored or removed. From the import summary, we learn that the <code>VEET</code> light data, like the <code>Clouclip</code>, also exhibits irregularities and gaps. (The device nominally records every 2 seconds, but timing may drift or pause when not worn.)</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Print the first 6 rows of the VEETs light dataset</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>dataVEET <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 14
# Groups:   Id [1]
  Id    Datetime            file.name     time_stamp modality integration_time
  &lt;fct&gt; &lt;dttm&gt;              &lt;chr&gt;              &lt;dbl&gt; &lt;chr&gt;               &lt;dbl&gt;
1 VEET  2024-06-04 15:00:37 01_VEET_L.csv 1717531237 ALS                   100
2 VEET  2024-06-04 15:00:39 01_VEET_L.csv 1717531239 ALS                   100
3 VEET  2024-06-04 15:00:41 01_VEET_L.csv 1717531241 ALS                   100
4 VEET  2024-06-04 15:00:43 01_VEET_L.csv 1717531243 ALS                   100
5 VEET  2024-06-04 15:00:45 01_VEET_L.csv 1717531245 ALS                   100
6 VEET  2024-06-04 15:00:47 01_VEET_L.csv 1717531247 ALS                   100
# ℹ 8 more variables: uvGain &lt;dbl&gt;, visGain &lt;dbl&gt;, irGain &lt;dbl&gt;, uvValue &lt;dbl&gt;,
#   visValue &lt;dbl&gt;, irValue &lt;dbl&gt;, Flicker &lt;dbl&gt;, Lux &lt;dbl&gt;</code></pre>
</div>
</div>
<p>To make the VEET light data comparable to the Clouclip’s and to simplify analysis, we choose to&nbsp;<strong>aggregate the VEET illuminance data to 5-second intervals</strong>. This slight downsampling will both reduce data volume and remove irregularities.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Aggregate VEET light data to 5-second intervals and mark gaps</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-14"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-14-1"><a href="#annotated-cell-14-1"></a>dataVEET <span class="ot">&lt;-</span> dataVEET <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-14-2" class="code-annotation-target"><a href="#annotated-cell-14-2"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 seconds"</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-14-3" class="code-annotation-target"><a href="#annotated-cell-14-3"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-14-4" class="code-annotation-target"><a href="#annotated-cell-14-4"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-14-5" class="code-annotation-target"><a href="#annotated-cell-14-5"></a>  <span class="fu">remove_partial_data</span>(Lux, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-14" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="2" data-code-annotation="1">Resample to 5-sec bins (e.g.&nbsp;average Lux over 2-sec readings)</span>
</dd>
<dt data-target-cell="annotated-cell-14" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="3" data-code-annotation="2">Fill in implicit gaps with NA rows</span>
</dd>
<dt data-target-cell="annotated-cell-14" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="4" data-code-annotation="3">Add Date column for daily grouping</span>
</dd>
<dt data-target-cell="annotated-cell-14" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="5" data-code-annotation="4">Remove participant days with more than one hour of missing data</span>
</dd>
</dl>
</div>
</div>
<p>First, <code>aggregate_Datetime(unit = "5 seconds")</code> combines the high-frequency 2-second observations into 5-second slots. By default, this function will average numeric columns like&nbsp;<code>Lux</code>&nbsp;over each 5-second period (and have sensible defaults for strings or categorical data). All of these <em>data type handlers</em> can be changed with the function call. The result is that&nbsp;<code>dataVEET</code>&nbsp;now has a reading every 5 seconds (or an NA if none were present in that window). Next,&nbsp;<code>gap_handler(full.days = TRUE)</code>&nbsp;inserts explicit NA entries for any 5-second timestamp that had no data within the continuous span of the recording. Then we add a&nbsp;<code>Date</code>&nbsp;column for grouping, and finally we remove days with more than 1 hour of missing data (using a more strict criterion as we did for <code>Clouclip</code>). According to the gap summary (<a href="#tbl-gaps2" class="quarto-xref">Table&nbsp;2</a>), this leaves six full days of <code>VEET</code> light data with good coverage, after dropping the very incomplete start/end days.</p>
<p>We can inspect the missing-data summary for the <code>VEET</code> illuminance data:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>dataVEET <span class="sc">|&gt;</span> <span class="fu">gap_table</span>(Lux, <span class="st">"Illuminance (lx)"</span>) <span class="sc">|&gt;</span> <span class="fu">cols_hide</span>(<span class="fu">contains</span>(<span class="st">"_n"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div id="tbl-gaps2" class="cell quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-gaps2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;2: Summary of missing and observed data for the VEET device, light modality
</figcaption>
<div aria-describedby="tbl-gaps2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div id="qkqxejjdyg" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#qkqxejjdyg table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#qkqxejjdyg thead, #qkqxejjdyg tbody, #qkqxejjdyg tfoot, #qkqxejjdyg tr, #qkqxejjdyg td, #qkqxejjdyg th {
  border-style: none;
}

#qkqxejjdyg p {
  margin: 0;
  padding: 0;
}

#qkqxejjdyg .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#qkqxejjdyg .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#qkqxejjdyg .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#qkqxejjdyg .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#qkqxejjdyg .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#qkqxejjdyg .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#qkqxejjdyg .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#qkqxejjdyg .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#qkqxejjdyg .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#qkqxejjdyg .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#qkqxejjdyg .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#qkqxejjdyg .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#qkqxejjdyg .gt_spanner_row {
  border-bottom-style: hidden;
}

#qkqxejjdyg .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#qkqxejjdyg .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#qkqxejjdyg .gt_from_md > :first-child {
  margin-top: 0;
}

#qkqxejjdyg .gt_from_md > :last-child {
  margin-bottom: 0;
}

#qkqxejjdyg .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#qkqxejjdyg .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#qkqxejjdyg .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#qkqxejjdyg .gt_row_group_first td {
  border-top-width: 2px;
}

#qkqxejjdyg .gt_row_group_first th {
  border-top-width: 2px;
}

#qkqxejjdyg .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#qkqxejjdyg .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#qkqxejjdyg .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#qkqxejjdyg .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#qkqxejjdyg .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#qkqxejjdyg .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#qkqxejjdyg .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#qkqxejjdyg .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#qkqxejjdyg .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#qkqxejjdyg .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#qkqxejjdyg .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#qkqxejjdyg .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#qkqxejjdyg .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#qkqxejjdyg .gt_left {
  text-align: left;
}

#qkqxejjdyg .gt_center {
  text-align: center;
}

#qkqxejjdyg .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#qkqxejjdyg .gt_font_normal {
  font-weight: normal;
}

#qkqxejjdyg .gt_font_bold {
  font-weight: bold;
}

#qkqxejjdyg .gt_font_italic {
  font-style: italic;
}

#qkqxejjdyg .gt_super {
  font-size: 65%;
}

#qkqxejjdyg .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#qkqxejjdyg .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#qkqxejjdyg .gt_indent_1 {
  text-indent: 5px;
}

#qkqxejjdyg .gt_indent_2 {
  text-indent: 10px;
}

#qkqxejjdyg .gt_indent_3 {
  text-indent: 15px;
}

#qkqxejjdyg .gt_indent_4 {
  text-indent: 20px;
}

#qkqxejjdyg .gt_indent_5 {
  text-indent: 25px;
}

#qkqxejjdyg .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#qkqxejjdyg div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table do-not-create-environment cell caption-top table table-sm table-striped small" data-quarto-bootstrap="false">
<colgroup>
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
</colgroup>
<thead>
<tr class="gt_heading header">
<td colspan="14" class="gt_heading gt_title gt_font_normal">Summary of available and missing data</td>
</tr>
<tr class="gt_heading even">
<td colspan="14" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border">Variable: Illuminance (lx)</td>
</tr>
<tr class="gt_col_headings gt_spanner_row header">
<th data-quarto-table-cell-role="th" scope="col"></th>
<th colspan="5" id="data_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold; font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Data
</div></th>
<th colspan="8" id="missing_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold; font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Missing
</div></th>
</tr>
<tr class="gt_col_headings gt_spanner_row even">
<th rowspan="2" id="a::stub" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col"></th>
<th colspan="2" id="Regular" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Regular
</div></th>
<th id="irregular_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Irregular
</div></th>
<th id="Range" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Range
</div></th>
<th id="interval_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Interval
</div></th>
<th colspan="4" id="gap_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Gaps
</div></th>
<th colspan="2" id="implicit_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Implicit
</div></th>
<th colspan="2" id="explicit_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Explicit
</div></th>
</tr>
<tr class="gt_col_headings header">
<th id="available_data" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="available_data_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="number_irregulars" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">n<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1,2</sup></span></th>
<th id="total_duration" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="interval" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="number_gaps" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">N</th>
<th id="mean_gap_duration" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ø</th>
<th id="missing_data" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_data_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="missing_implicit" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_implicit_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="missing_explicit" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_explicit_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<th id="grand_summary_stub_1" class="gt_row gt_left gt_stub gt_grand_summary_row gt_last_grand_summary_row_top" data-quarto-table-cell-role="th" scope="row" style="background-color: #E5E5E5"><strong>Overall</strong></th>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 available_data" style="background-color: #E5E5E5">5d 23h 57m 40s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 available_data_pct" style="background-color: #E5E5E5">100.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 number_irregulars" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 total_duration" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">6d</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 interval" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 number_gaps" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">8</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 mean_gap_duration" style="background-color: #E5E5E5">58s</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_data" style="background-color: #E5E5E5">2m 20s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_data_pct" style="background-color: #E5E5E5">0.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_implicit" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_implicit_pct" style="background-color: #E5E5E5">0.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_explicit" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2m 20s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_explicit_pct" style="background-color: #E5E5E5">0.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="VEET - 2024-06-06" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-06</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_1" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 available_data">23h 58m 5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 available_data_pct">99.9%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">3</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 mean_gap_duration">38s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 missing_data">1m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 missing_data_pct">0.1%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 missing_explicit_pct">0.1%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="VEET - 2024-06-07" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-07</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_2" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 available_data">1d</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 mean_gap_duration">0s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 missing_data">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 missing_explicit_pct">0.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="VEET - 2024-06-08" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-08</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_3" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 available_data">23h 59m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 mean_gap_duration">5s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 missing_data">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 missing_explicit_pct">0.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="VEET - 2024-06-09" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-09</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_4" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 available_data">23h 59m 50s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 mean_gap_duration">5s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 missing_data">10s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">10s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 missing_explicit_pct">0.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="VEET - 2024-06-10" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-10</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_5" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 available_data">23h 59m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 mean_gap_duration">5s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 missing_data">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 missing_explicit_pct">0.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<th colspan="14" id="VEET - 2024-06-11" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-11</th>
</tr>
<tr class="gt_row_group_first odd">
<th id="stub_1_6" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></th>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 available_data">23h 59m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 mean_gap_duration">5s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 missing_data">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 missing_explicit_pct">0.0%</td>
</tr>
</tbody><tfoot>
<tr class="gt_footnotes odd">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span> If n &gt; 0: it is possible that the other summary statistics are affected, as they are calculated based on the most prominent interval.</td>
</tr>
<tr class="gt_footnotes even">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>2</sup></span> Number of (missing or actual) observations</td>
</tr>
<tr class="gt_footnotes odd">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span> Based on times, not necessarily number of observations</td>
</tr>
</tfoot>

</table>

</div>
</div>
</div>
</figure>
</div>
</div>
<p><a href="#tbl-gaps2" class="quarto-xref">Table&nbsp;2</a> shows, for each retained day, the total recorded duration and the duration of gaps. The <code>VEET</code> device, like the <code>Clouclip</code>, was not worn continuously 24&nbsp;hours per day, so there are nightly gaps of wear time (when the device was likely off the participant), but not of recordings. After our preprocessing, any implicit gaps are represented as explicit missing intervals. The <code>VEET</code>’s time sampling was originally more frequent, but by aggregating to 5&nbsp;s we have ensured a uniform timeline akin to the <code>Clouclip</code>’s.</p>
<p>At this point, the&nbsp;<code>dataVEET</code>&nbsp;object (for illuminance) is cleaned and ready for computing light exposure metrics. For example, one could calculate daily mean illuminance or the duration spent above certain light thresholds (e.g.&nbsp;“outdoor light exposure” defined as &gt;1000&nbsp;lx) using this dataset. Indeed, basic summary tables in the <a href="./index.html">main tutorial</a> illustrate the highly skewed nature of light exposure data and the calculation of outdoor light metrics. We will not repeat those metric calculations here in the supplement, as our focus is on data preprocessing; however, having a cleaned, gap-marked time series is crucial for those metrics to be accurate.</p>
<section id="identifying-non-wear-times" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="identifying-non-wear-times"><span class="header-section-number">3.1</span> Identifying non-wear times</h3>
<p>Not all recorded time points of the <code>VEET</code> can be considered as <code>wear time</code>. <code>Non-wear</code> times should generally be removed from time series, if possible. There are several ways how <code>Non-wear</code> can be classified <span class="citation" data-cites="Zauner2025npj">(<a href="#ref-Zauner2025npj" role="doc-biblioref">Zauner et al. 2025</a>)</span>:</p>
<ol type="1">
<li><p>A device detects <code>non-wear</code> and does not record any measurements during non-wear times, or has a variable in the exported data to denote non-wear times. The <code>Clouclip</code> is an example here with the <code>sleep mode</code> sentinel state</p></li>
<li><p>Record non-wear separately, e.g.&nbsp;with a wear-log or diary. This is easy to implement, but puts a higher burden on participants and research staff, and can be error prone: forgotten, or incorrect data entries can mis-classify wear and non-wear times. For an implementation on how to remove non-wear times based on such a log, see, e.g., this <a href="https://tscnlab.github.io/LightLogR_webinar/advanced_01_a_day_in_daylight-live.html">tutorial</a>.</p></li>
<li><p>Instruct participants with a certain behavior when removing the device to ease automated detection of non-wear times in the data. These include putting a device in a black opaque bag, so that sensor readings for light are zero, or pressing an event button (if one exists). Up- and downsides are similar to the previous option, but the measures are usually harder to decipher in the data: when a buttonpress was forgotten, does that mess up the automated detection moving forward, which expects two presses per non-wear period (at beginning and at start). When there is zero lux during the day - was it in a black bag, or was this person just in a dark environment?</p></li>
<li><p>Automated detection by an algorithm. There are several possibilities to try to classify non-wear times based on the available data from a wearable. E.g., if there is an activity tracker, that can be used to determine periods of inactivity, which could indicate non-wear. If only light data is available, the standard deviation (or coefficient of variance) of light in a moving window of a few minutes can be used to determine low variance, which also indicates inactivity. Usually a histogram of values shows a spike in those periods of inactivity.</p></li>
</ol>
<p>If datasets are large enough, i.e., collect data from a long enought period, <code>non-wear</code> has a diminishing effect, at least as daily metrics of light are concerned. E.g., in a monthlong study of 39 participants from Switzerland and Malaysia, <span class="citation" data-cites="Biller2025">(<a href="#ref-Biller2025" role="doc-biblioref">Biller et al. 2025</a>)</span> found that based on a sensitivity anaylsis up to 6 hours of data could be missing <strong>every day</strong>, and still the daily metrics across that month would not significantly change. This assumption only holds when non-wear times are not systematic, i.e., at the same time each day.</p>
<p>We will perform a simple variant of the third option, and look at the <code>VEET</code>’s <code>IMU</code> modality</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Import VEET activity (IMU) data</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-16"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-16-1"><a href="#annotated-cell-16-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="annotated-cell-16-2"><a href="#annotated-cell-16-2"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span></span>
<span id="annotated-cell-16-3"><a href="#annotated-cell-16-3"></a>dataIMU <span class="ot">&lt;-</span> </span>
<span id="annotated-cell-16-4"><a href="#annotated-cell-16-4"></a>  import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"IMU"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-16-5" class="code-annotation-target"><a href="#annotated-cell-16-5"></a>              <span class="at">remove_duplicates =</span> <span class="cn">TRUE</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-16" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-16-6" class="code-annotation-target"><a href="#annotated-cell-16-6"></a>              <span class="at">silent =</span> <span class="cn">TRUE</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-16" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="5" data-code-annotation="1">Some rows in the data are duplicated. These will be removed during import</span>
</dd>
<dt data-target-cell="annotated-cell-16" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-16" data-code-lines="6" data-code-annotation="2">In this instance, we skip the import summary</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>15 duplicate rows were removed during import.</code></pre>
</div>
</div>
<p>To get a quick feeling for the data, we visualize the activity variable for the x axis in <a href="#fig-IMU" class="quarto-xref">Figure&nbsp;8</a></p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Plotting the IMU timeline</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>dataIMU <span class="sc">|&gt;</span> </span>
<span id="cb14-2"><a href="#cb14-2"></a>  <span class="fu">gg_days</span>(ax, </span>
<span id="cb14-3"><a href="#cb14-3"></a>          <span class="at">aes_col =</span> <span class="fu">abs</span>(ax) <span class="sc">&lt;</span> <span class="dv">1</span>, </span>
<span id="cb14-4"><a href="#cb14-4"></a>          <span class="at">group =</span> <span class="fu">consecutive_id</span>(<span class="fu">abs</span>(ax) <span class="sc">&lt;</span> <span class="dv">1</span>),</span>
<span id="cb14-5"><a href="#cb14-5"></a>          <span class="at">y.axis.label =</span> <span class="st">"ax"</span></span>
<span id="cb14-6"><a href="#cb14-6"></a>          )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output-display">
<div id="fig-IMU" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-IMU-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-IMU-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9" title="Figure&nbsp;8: Timeline of the activity sensor (x direction)"><img src="supplement_files/figure-html/fig-IMU-1.png" class="img-fluid figure-img" width="1344"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-IMU-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8: Timeline of the activity sensor (x direction)
</figcaption>
</figure>
</div>
</div>
</div>
<p><code>ax</code> seems to vary between ±10. In general, the periods of inactivity seem to lie within a band of ±1. However, simply choosing this band does not eliminate false detections during times of high movement. Some data transformations can make these more clear:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Plotting a transformed IMU timeline</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>dataIMU <span class="sc">|&gt;</span> </span>
<span id="cb15-2"><a href="#cb15-2"></a>  <span class="fu">aggregate_Datetime</span>(</span>
<span id="cb15-3"><a href="#cb15-3"></a>    <span class="st">"5 mins"</span>,</span>
<span id="cb15-4"><a href="#cb15-4"></a>    <span class="at">numeric.handler =</span> sd</span>
<span id="cb15-5"><a href="#cb15-5"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb15-6"><a href="#cb15-6"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(ax, ay, az)) <span class="sc">|&gt;</span> </span>
<span id="cb15-7"><a href="#cb15-7"></a>  <span class="fu">group_by</span>(name) <span class="sc">|&gt;</span> </span>
<span id="cb15-8"><a href="#cb15-8"></a>  <span class="fu">gg_days</span>(value, </span>
<span id="cb15-9"><a href="#cb15-9"></a>          <span class="at">aes_col =</span> value <span class="sc">&lt;</span> <span class="fl">0.05</span>,</span>
<span id="cb15-10"><a href="#cb15-10"></a>          <span class="at">group =</span> <span class="fu">consecutive_id</span>(value <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb15-11"><a href="#cb15-11"></a>          <span class="at">y.axis.label =</span> <span class="st">"activity axes"</span></span>
<span id="cb15-12"><a href="#cb15-12"></a>          )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output-display">
<div id="fig-IMU2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-IMU2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-IMU2-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10" title="Figure&nbsp;9: Timeline of the activity sensor (x direction) - transformed to distinguish times of low activity"><img src="supplement_files/figure-html/fig-IMU2-1.png" class="img-fluid figure-img" width="1344"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-IMU2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9: Timeline of the activity sensor (x direction) - transformed to distinguish times of low activity
</figcaption>
</figure>
</div>
</div>
</div>
<p>What <a href="#fig-IMU2" class="quarto-xref">Figure&nbsp;9</a> shows is that the 5-minute standard deviation of the activity channels (x, y, and z direction) allows a pretty stable distinction between periods of high and low activity, with a threshold of <strong>0.05</strong>. We can us this to create a wear column. As the choice of channel does not seem to matter much, we will use <code>ax</code>.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Calculating times of wear and non-wear</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-19"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-19-1"><a href="#annotated-cell-19-1"></a>wear_data <span class="ot">&lt;-</span> </span>
<span id="annotated-cell-19-2"><a href="#annotated-cell-19-2"></a>dataIMU <span class="sc">|&gt;</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-19-3" class="code-annotation-target"><a href="#annotated-cell-19-3"></a>  <span class="fu">aggregate_Datetime</span>(<span class="st">"5 mins"</span>,<span class="at">numeric.handler =</span> sd) <span class="sc">|&gt;</span></span>
<span id="annotated-cell-19-4"><a href="#annotated-cell-19-4"></a>  <span class="fu">mutate</span>(<span class="at">wear =</span> ax <span class="sc">&gt;</span> <span class="fl">0.05</span>) <span class="sc">|&gt;</span></span>
<span id="annotated-cell-19-5"><a href="#annotated-cell-19-5"></a>  <span class="fu">select</span>(Id, Datetime, wear) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-19-6" class="code-annotation-target"><a href="#annotated-cell-19-6"></a>  <span class="fu">extract_states</span>(wear)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-19" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-19-7" class="code-annotation-target"><a href="#annotated-cell-19-7"></a>wear_data <span class="sc">|&gt;</span></span>
<span id="annotated-cell-19-8"><a href="#annotated-cell-19-8"></a>  <span class="fu">ungroup</span>(Id) <span class="sc">|&gt;</span></span>
<span id="annotated-cell-19-9"><a href="#annotated-cell-19-9"></a>  <span class="fu">summarize_numeric</span>(<span class="at">remove =</span> <span class="fu">c</span>(<span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"epoch"</span>)) <span class="sc">|&gt;</span></span>
<span id="annotated-cell-19-10"><a href="#annotated-cell-19-10"></a>  <span class="fu">gt</span>()</span>
<span id="annotated-cell-19-11"><a href="#annotated-cell-19-11"></a>wear_data <span class="ot">&lt;-</span> wear_data <span class="sc">|&gt;</span> <span class="fu">select</span>(Id, wear, start, end)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-19" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="3,4,5" data-code-annotation="1">Calculation of the wear-variable</span>
</dd>
<dt data-target-cell="annotated-cell-19" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="6,11" data-code-annotation="2">Extracting the start and end times of wear and non-wear</span>
</dd>
<dt data-target-cell="annotated-cell-19" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-19" data-code-lines="7,8,9,10" data-code-annotation="3">Summarizing wear and non-wear times in a table</span>
</dd>
</dl>
</div>
<div class="cell-output-display">
<div id="xyqagpjbjv" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#xyqagpjbjv table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#xyqagpjbjv thead, #xyqagpjbjv tbody, #xyqagpjbjv tfoot, #xyqagpjbjv tr, #xyqagpjbjv td, #xyqagpjbjv th {
  border-style: none;
}

#xyqagpjbjv p {
  margin: 0;
  padding: 0;
}

#xyqagpjbjv .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#xyqagpjbjv .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#xyqagpjbjv .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#xyqagpjbjv .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#xyqagpjbjv .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#xyqagpjbjv .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#xyqagpjbjv .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#xyqagpjbjv .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#xyqagpjbjv .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#xyqagpjbjv .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#xyqagpjbjv .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#xyqagpjbjv .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#xyqagpjbjv .gt_spanner_row {
  border-bottom-style: hidden;
}

#xyqagpjbjv .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#xyqagpjbjv .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#xyqagpjbjv .gt_from_md > :first-child {
  margin-top: 0;
}

#xyqagpjbjv .gt_from_md > :last-child {
  margin-bottom: 0;
}

#xyqagpjbjv .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#xyqagpjbjv .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#xyqagpjbjv .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#xyqagpjbjv .gt_row_group_first td {
  border-top-width: 2px;
}

#xyqagpjbjv .gt_row_group_first th {
  border-top-width: 2px;
}

#xyqagpjbjv .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#xyqagpjbjv .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#xyqagpjbjv .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#xyqagpjbjv .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#xyqagpjbjv .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#xyqagpjbjv .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#xyqagpjbjv .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#xyqagpjbjv .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#xyqagpjbjv .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#xyqagpjbjv .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#xyqagpjbjv .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#xyqagpjbjv .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#xyqagpjbjv .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#xyqagpjbjv .gt_left {
  text-align: left;
}

#xyqagpjbjv .gt_center {
  text-align: center;
}

#xyqagpjbjv .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#xyqagpjbjv .gt_font_normal {
  font-weight: normal;
}

#xyqagpjbjv .gt_font_bold {
  font-weight: bold;
}

#xyqagpjbjv .gt_font_italic {
  font-style: italic;
}

#xyqagpjbjv .gt_super {
  font-size: 65%;
}

#xyqagpjbjv .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#xyqagpjbjv .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#xyqagpjbjv .gt_indent_1 {
  text-indent: 5px;
}

#xyqagpjbjv .gt_indent_2 {
  text-indent: 10px;
}

#xyqagpjbjv .gt_indent_3 {
  text-indent: 15px;
}

#xyqagpjbjv .gt_indent_4 {
  text-indent: 20px;
}

#xyqagpjbjv .gt_indent_5 {
  text-indent: 25px;
}

#xyqagpjbjv .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#xyqagpjbjv div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table caption-top table table-sm table-striped small" data-quarto-bootstrap="false">
<thead>
<tr class="gt_col_headings header">
<th id="wear" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">wear</th>
<th id="mean_duration" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">mean_duration</th>
<th id="total_duration" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">total_duration</th>
<th id="episodes" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">episodes</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_center" headers="wear">FALSE</td>
<td class="gt_row gt_center" headers="mean_duration">11447s (~3.18 hours)</td>
<td class="gt_row gt_center" headers="total_duration">366300s (~4.24 days)</td>
<td class="gt_row gt_right" headers="episodes">32</td>
</tr>
<tr class="even">
<td class="gt_row gt_center" headers="wear">TRUE</td>
<td class="gt_row gt_center" headers="mean_duration">9431s (~2.62 hours)</td>
<td class="gt_row gt_center" headers="total_duration">301800s (~3.49 days)</td>
<td class="gt_row gt_right" headers="episodes">32</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>We see that most of the recorded timespan is actually non-wear (including sleep), with an average wear time of 2.6 hours.</p>
<p>Let’s add this data to the <code>ALS</code> dataset and visualize it in <a href="#fig-nonwear" class="quarto-xref">Figure&nbsp;10</a></p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Plotting non-wear for light</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>dataVEET <span class="ot">&lt;-</span> </span>
<span id="cb16-2"><a href="#cb16-2"></a>dataVEET <span class="sc">|&gt;</span> </span>
<span id="cb16-3"><a href="#cb16-3"></a>  <span class="fu">group_by</span>(Id) <span class="sc">|&gt;</span> </span>
<span id="cb16-4"><a href="#cb16-4"></a>  <span class="fu">add_states</span>(wear_data)</span>
<span id="cb16-5"><a href="#cb16-5"></a></span>
<span id="cb16-6"><a href="#cb16-6"></a>dataVEET <span class="sc">|&gt;</span> </span>
<span id="cb16-7"><a href="#cb16-7"></a>  <span class="fu">gg_days</span>(Lux) <span class="sc">|&gt;</span> </span>
<span id="cb16-8"><a href="#cb16-8"></a>  <span class="fu">gg_states</span>(wear, <span class="at">ymax =</span> <span class="dv">0</span>, <span class="at">alpha =</span> <span class="dv">1</span>, <span class="at">fill =</span> <span class="st">"red"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output-display">
<div id="fig-nonwear" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-nonwear-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-nonwear-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11" title="Figure&nbsp;10: Wear times for light data, shown as red bars"><img src="supplement_files/figure-html/fig-nonwear-1.png" class="img-fluid figure-img" width="1344"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-nonwear-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10: Wear times for light data, shown as red bars
</figcaption>
</figure>
</div>
</div>
</div>
<p>This looks sensible, e.g., when we look at noon of 7 June 2024. We can remove these observations based on the <code>wear</code> column to make our summaries more valid.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Removing non-wear observations from light modality</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>dataVEET <span class="ot">&lt;-</span> dataVEET <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">Lux =</span> <span class="fu">ifelse</span>(wear, Lux, <span class="cn">NA</span>)) <span class="sc">|&gt;</span> <span class="fu">group_by</span>(Id, Date)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
</section>
</section>
<section id="veet-data-spectral-data-processing" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="veet-data-spectral-data-processing"><span class="header-section-number">4</span> <code>VEET</code> Data: Spectral Data Processing</h2>
<section id="import" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="import"><span class="header-section-number">4.1</span> Import</h3>
<p>In addition to broad-band illuminance and distance, the <code>VEET</code> provides&nbsp;<strong>spectral sensor data</strong>&nbsp;through its <code>PHO</code> modality. Unlike illuminance, the spectral data are not given as directly interpretable radiometric metrics but rather as raw sensor counts across multiple wavelength channels, which require conversion to reconstruct a spectral power distribution. In our analysis, spectral data allow us to compute metrics like the relative contribution of short-wavelength (blue) light versus long-wavelength light in the participant’s environment. Processing this spectral data involves several necessary steps.</p>
<p>First, we import the spectral modality from a second <code>VEET</code> file. This time we need to extract the lines marked as <code>PHO</code>. We will store the spectral dataset in a separate object&nbsp;<code>dataVEET2</code>&nbsp;so as not to overwrite the&nbsp;<code>dataVEET</code>&nbsp;illuminance data in our R session:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Import VEET Spectral Sensor (PHO) data</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/02_VEET_L.csv.zip"</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>dataVEET2 <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"PHO"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Successfully read in 173'013 observations across 1 Ids from 1 VEET-file(s).
Timezone set is US/Central.
The system timezone is UTC. Please correct if necessary!

First Observation: 2025-06-17 12:25:13
Last Observation: 2025-06-21 22:47:01
Timespan: 4.4 days

Observation intervals: 
   Id    interval.time      n pct      
 1 VEET  1s               417 0.24102% 
 2 VEET  2s            171837 99.32086%
 3 VEET  3s               738 0.42656% 
 4 VEET  4s                 7 0.00405% 
 5 VEET  9s                 1 0.00058% 
 6 VEET  11s                1 0.00058% 
 7 VEET  12s                2 0.00116% 
 8 VEET  13s                4 0.00231% 
 9 VEET  15s                1 0.00058% 
10 VEET  16s                1 0.00058% 
# ℹ 3 more rows</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-VEET-overview2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-overview2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-VEET-overview2-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12" title="Figure&nbsp;11: Overview plot of imported VEET data"><img src="supplement_files/figure-html/fig-VEET-overview2-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-overview2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11: Overview plot of imported VEET data
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Print the first 6 rows of the VEETs spectral dataset</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>dataVEET2 <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 18
# Groups:   Id [1]
  Id    Datetime            file.name time_stamp modality integration_time  Gain
  &lt;fct&gt; &lt;dttm&gt;              &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt;               &lt;dbl&gt; &lt;dbl&gt;
1 VEET  2025-06-17 12:25:13 02_VEET_… 1750181113 PHO                   100   512
2 VEET  2025-06-17 12:25:17 02_VEET_… 1750181117 PHO                   100   512
3 VEET  2025-06-17 12:25:19 02_VEET_… 1750181119 PHO                   100   512
4 VEET  2025-06-17 12:25:21 02_VEET_… 1750181121 PHO                   100   512
5 VEET  2025-06-17 12:25:23 02_VEET_… 1750181123 PHO                   100   512
6 VEET  2025-06-17 12:25:25 02_VEET_… 1750181125 PHO                   100   512
# ℹ 11 more variables: s415 &lt;dbl&gt;, s445 &lt;dbl&gt;, s480 &lt;dbl&gt;, s515 &lt;dbl&gt;,
#   s555 &lt;dbl&gt;, s590 &lt;dbl&gt;, s630 &lt;dbl&gt;, s680 &lt;dbl&gt;, s910 &lt;dbl&gt;, Dark &lt;dbl&gt;,
#   Clear &lt;dbl&gt;</code></pre>
</div>
</div>
<p>After import,&nbsp;<code>data</code>&nbsp;contains columns for the timestamp (<code>Datetime</code>),&nbsp;<code>Gain</code>&nbsp;(the sensor gain setting), and the nine spectral sensor channels plus a clear channel. These appear as numeric columns named&nbsp;<code>s415, s445, ..., s940, Dark, Clear</code>. Other columns are also present but not needed for now. The spectral sensor was logging at a 2-second rate. It is informative to look at a snippet of the imported spectral data before further processing. <a href="#tbl-PHO" class="quarto-xref">Table&nbsp;3</a> shows three rows of&nbsp;<code>data</code>&nbsp;after import (before calibration), with some technical columns omitted for brevity:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Table overview of spectral sensor data</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="cb22-2"><a href="#cb22-2"></a>  <span class="fu">slice</span>(<span class="dv">6000</span><span class="sc">:</span><span class="dv">6003</span>) <span class="sc">|&gt;</span> </span>
<span id="cb22-3"><a href="#cb22-3"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(modality, file.name, time_stamp)) <span class="sc">|&gt;</span> </span>
<span id="cb22-4"><a href="#cb22-4"></a>  <span class="fu">gt</span>() <span class="sc">|&gt;</span> </span>
<span id="cb22-5"><a href="#cb22-5"></a>  <span class="fu">fmt_number</span>(s415<span class="sc">:</span>Clear) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="tbl-PHO" class="cell quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-PHO-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;3: Overview of the spectral sensor import from the VEET device (3 observations). Each row corresponds to a 2-second timestamp (Datetime) and shows the raw sensor readings for the spectral channels (s415–s940, Dark, Clear). All values are in arbitrary sensor units (counts). Gain values and integration_time are also relevant for each interval, depending on the downstream computation.
</figcaption>
<div aria-describedby="tbl-PHO-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div id="knhsibmlkf" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#knhsibmlkf table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#knhsibmlkf thead, #knhsibmlkf tbody, #knhsibmlkf tfoot, #knhsibmlkf tr, #knhsibmlkf td, #knhsibmlkf th {
  border-style: none;
}

#knhsibmlkf p {
  margin: 0;
  padding: 0;
}

#knhsibmlkf .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#knhsibmlkf .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#knhsibmlkf .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#knhsibmlkf .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#knhsibmlkf .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#knhsibmlkf .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#knhsibmlkf .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#knhsibmlkf .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#knhsibmlkf .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#knhsibmlkf .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#knhsibmlkf .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#knhsibmlkf .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#knhsibmlkf .gt_spanner_row {
  border-bottom-style: hidden;
}

#knhsibmlkf .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#knhsibmlkf .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#knhsibmlkf .gt_from_md > :first-child {
  margin-top: 0;
}

#knhsibmlkf .gt_from_md > :last-child {
  margin-bottom: 0;
}

#knhsibmlkf .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#knhsibmlkf .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#knhsibmlkf .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#knhsibmlkf .gt_row_group_first td {
  border-top-width: 2px;
}

#knhsibmlkf .gt_row_group_first th {
  border-top-width: 2px;
}

#knhsibmlkf .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#knhsibmlkf .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#knhsibmlkf .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#knhsibmlkf .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#knhsibmlkf .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#knhsibmlkf .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#knhsibmlkf .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#knhsibmlkf .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#knhsibmlkf .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#knhsibmlkf .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#knhsibmlkf .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#knhsibmlkf .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#knhsibmlkf .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#knhsibmlkf .gt_left {
  text-align: left;
}

#knhsibmlkf .gt_center {
  text-align: center;
}

#knhsibmlkf .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#knhsibmlkf .gt_font_normal {
  font-weight: normal;
}

#knhsibmlkf .gt_font_bold {
  font-weight: bold;
}

#knhsibmlkf .gt_font_italic {
  font-style: italic;
}

#knhsibmlkf .gt_super {
  font-size: 65%;
}

#knhsibmlkf .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#knhsibmlkf .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#knhsibmlkf .gt_indent_1 {
  text-indent: 5px;
}

#knhsibmlkf .gt_indent_2 {
  text-indent: 10px;
}

#knhsibmlkf .gt_indent_3 {
  text-indent: 15px;
}

#knhsibmlkf .gt_indent_4 {
  text-indent: 20px;
}

#knhsibmlkf .gt_indent_5 {
  text-indent: 25px;
}

#knhsibmlkf .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#knhsibmlkf div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table do-not-create-environment cell caption-top table table-sm table-striped small" data-quarto-bootstrap="false">
<thead>
<tr class="gt_col_headings header">
<th id="Datetime" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Datetime</th>
<th id="integration_time" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">integration_time</th>
<th id="Gain" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Gain</th>
<th id="s415" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s415</th>
<th id="s445" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s445</th>
<th id="s480" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s480</th>
<th id="s515" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s515</th>
<th id="s555" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s555</th>
<th id="s590" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s590</th>
<th id="s630" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s630</th>
<th id="s680" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s680</th>
<th id="s910" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s910</th>
<th id="Dark" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Dark</th>
<th id="Clear" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Clear</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="gt_group_heading_row odd">
<th colspan="14" id="VEET" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET</th>
</tr>
<tr class="gt_row_group_first even">
<td class="gt_row gt_right" headers="VEET  Datetime">2025-06-18 00:11:22</td>
<td class="gt_row gt_right" headers="VEET  integration_time">100</td>
<td class="gt_row gt_right" headers="VEET  Gain">512</td>
<td class="gt_row gt_right" headers="VEET  s415">20.00</td>
<td class="gt_row gt_right" headers="VEET  s445">27.00</td>
<td class="gt_row gt_right" headers="VEET  s480">31.00</td>
<td class="gt_row gt_right" headers="VEET  s515">42.00</td>
<td class="gt_row gt_right" headers="VEET  s555">47.00</td>
<td class="gt_row gt_right" headers="VEET  s590">63.00</td>
<td class="gt_row gt_right" headers="VEET  s630">90.00</td>
<td class="gt_row gt_right" headers="VEET  s680">139.00</td>
<td class="gt_row gt_right" headers="VEET  s910">756.00</td>
<td class="gt_row gt_right" headers="VEET  Dark">0.00</td>
<td class="gt_row gt_right" headers="VEET  Clear">534.00</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="VEET  Datetime">2025-06-18 00:11:24</td>
<td class="gt_row gt_right" headers="VEET  integration_time">100</td>
<td class="gt_row gt_right" headers="VEET  Gain">512</td>
<td class="gt_row gt_right" headers="VEET  s415">21.00</td>
<td class="gt_row gt_right" headers="VEET  s445">28.00</td>
<td class="gt_row gt_right" headers="VEET  s480">31.00</td>
<td class="gt_row gt_right" headers="VEET  s515">41.00</td>
<td class="gt_row gt_right" headers="VEET  s555">47.00</td>
<td class="gt_row gt_right" headers="VEET  s590">63.00</td>
<td class="gt_row gt_right" headers="VEET  s630">90.00</td>
<td class="gt_row gt_right" headers="VEET  s680">140.00</td>
<td class="gt_row gt_right" headers="VEET  s910">756.00</td>
<td class="gt_row gt_right" headers="VEET  Dark">0.00</td>
<td class="gt_row gt_right" headers="VEET  Clear">534.00</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="VEET  Datetime">2025-06-18 00:11:26</td>
<td class="gt_row gt_right" headers="VEET  integration_time">100</td>
<td class="gt_row gt_right" headers="VEET  Gain">512</td>
<td class="gt_row gt_right" headers="VEET  s415">21.00</td>
<td class="gt_row gt_right" headers="VEET  s445">27.00</td>
<td class="gt_row gt_right" headers="VEET  s480">31.00</td>
<td class="gt_row gt_right" headers="VEET  s515">41.00</td>
<td class="gt_row gt_right" headers="VEET  s555">47.00</td>
<td class="gt_row gt_right" headers="VEET  s590">63.00</td>
<td class="gt_row gt_right" headers="VEET  s630">90.00</td>
<td class="gt_row gt_right" headers="VEET  s680">139.00</td>
<td class="gt_row gt_right" headers="VEET  s910">756.00</td>
<td class="gt_row gt_right" headers="VEET  Dark">0.00</td>
<td class="gt_row gt_right" headers="VEET  Clear">534.00</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="VEET  Datetime">2025-06-18 00:11:28</td>
<td class="gt_row gt_right" headers="VEET  integration_time">100</td>
<td class="gt_row gt_right" headers="VEET  Gain">512</td>
<td class="gt_row gt_right" headers="VEET  s415">21.00</td>
<td class="gt_row gt_right" headers="VEET  s445">27.00</td>
<td class="gt_row gt_right" headers="VEET  s480">31.00</td>
<td class="gt_row gt_right" headers="VEET  s515">42.00</td>
<td class="gt_row gt_right" headers="VEET  s555">47.00</td>
<td class="gt_row gt_right" headers="VEET  s590">63.00</td>
<td class="gt_row gt_right" headers="VEET  s630">90.00</td>
<td class="gt_row gt_right" headers="VEET  s680">140.00</td>
<td class="gt_row gt_right" headers="VEET  s910">755.00</td>
<td class="gt_row gt_right" headers="VEET  Dark">0.00</td>
<td class="gt_row gt_right" headers="VEET  Clear">534.00</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</figure>
</div>
</div>
</section>
<section id="spectral-calibration" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="spectral-calibration"><span class="header-section-number">4.2</span> Spectral calibration</h3>
<p>Now we proceed with&nbsp;<strong>spectral calibration</strong>. The <code>VEET</code>’s spectral sensor counts need to be converted to physical units (spectral irradiance) via a calibration matrix provided by the manufacturer. For this example, we assume we have a calibration matrix that maps all the channel readings to an estimated spectral power distribution (SPD). The LightLogR package provides a function&nbsp;<code>spectral_reconstruction()</code>&nbsp;to perform this conversion. However, before applying it, we must ensure the sensor counts are in a normalized form. This procedure is laid out by the manufacturer. In our version, we refer to the <a href="https://projectveet.com/wp-content/uploads/VEET-Spectral-Power-Distribution-Reconstruction-Guide.pdf" target="_blank"><em>VEET SPD Reconstruction Guide.pdf</em></a>, version <em>06/05/2025</em>. Note that each manufacturer has to specify the method of count normalization (if any) and spectral reconstruction. In our raw data, each observation comes with a&nbsp;<code>Gain</code>&nbsp;setting that indicates how the sensor’s sensitivity was adjusted; we need to divide the raw counts by the gain to get normalized counts. LightLogR offers&nbsp;<code>normalize_counts()</code>&nbsp;for this purpose. We further need to scale by integration time (in milliseconds) and adjust depending on counts in the <code>Dark</code> sensor channel.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Normalize spectral sensor counts</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-25"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-25" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-25-1" class="code-annotation-target"><a href="#annotated-cell-25-1"></a>count.columns <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"s415"</span>, <span class="st">"s445"</span>, <span class="st">"s480"</span>, <span class="st">"s515"</span>, <span class="st">"s555"</span>, <span class="st">"s590"</span>, <span class="st">"s630"</span>,</span>
<span id="annotated-cell-25-2"><a href="#annotated-cell-25-2"></a>                      <span class="st">"s680"</span>, <span class="st">"s910"</span>, <span class="st">"Dark"</span>, <span class="st">"Clear"</span>)</span>
<span id="annotated-cell-25-3"><a href="#annotated-cell-25-3"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-25" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-25-4" class="code-annotation-target"><a href="#annotated-cell-25-4"></a>gain.ratios <span class="ot">&lt;-</span></span>
<span id="annotated-cell-25-5"><a href="#annotated-cell-25-5"></a>  <span class="fu">tibble</span>(</span>
<span id="annotated-cell-25-6"><a href="#annotated-cell-25-6"></a>    <span class="at">gain =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">8</span>, <span class="dv">16</span>, <span class="dv">32</span>, <span class="dv">64</span>, <span class="dv">128</span>, <span class="dv">256</span>, <span class="dv">512</span>),</span>
<span id="annotated-cell-25-7"><a href="#annotated-cell-25-7"></a>    <span class="at">gain.ratio =</span></span>
<span id="annotated-cell-25-8"><a href="#annotated-cell-25-8"></a>      <span class="fu">c</span>(<span class="fl">0.008</span>, <span class="fl">0.016</span>, <span class="fl">0.032</span>, <span class="fl">0.065</span>, <span class="fl">0.125</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="fl">3.95</span>, <span class="fl">7.75</span>)</span>
<span id="annotated-cell-25-9"><a href="#annotated-cell-25-9"></a>  )</span>
<span id="annotated-cell-25-10"><a href="#annotated-cell-25-10"></a></span>
<span id="annotated-cell-25-11"><a href="#annotated-cell-25-11"></a><span class="co">#normalize data:</span></span>
<span id="annotated-cell-25-12"><a href="#annotated-cell-25-12"></a>dataVEET2 <span class="ot">&lt;-</span></span>
<span id="annotated-cell-25-13"><a href="#annotated-cell-25-13"></a>  dataVEET2 <span class="sc">|&gt;</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-25" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-25-14" class="code-annotation-target"><a href="#annotated-cell-25-14"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(s415<span class="sc">:</span>Clear), \(x) (x <span class="sc">-</span> Dark)<span class="sc">/</span>integration_time)) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-25" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-25-15" class="code-annotation-target"><a href="#annotated-cell-25-15"></a>  <span class="fu">normalize_counts</span>(</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-25" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-25-16" class="code-annotation-target"><a href="#annotated-cell-25-16"></a>    <span class="at">gain.columns =</span> <span class="fu">rep</span>(<span class="st">"Gain"</span>, <span class="dv">11</span>),</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-25" data-target-annotation="6" onclick="event.preventDefault();">6</a><span id="annotated-cell-25-17" class="code-annotation-target"><a href="#annotated-cell-25-17"></a>    <span class="at">count.columns =</span> count.columns,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-25" data-target-annotation="7" onclick="event.preventDefault();">7</a><span id="annotated-cell-25-18" class="code-annotation-target"><a href="#annotated-cell-25-18"></a>    gain.ratios</span>
<span id="annotated-cell-25-19"><a href="#annotated-cell-25-19"></a>  ) <span class="sc">|&gt;</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-25" data-target-annotation="8" onclick="event.preventDefault();">8</a><span id="annotated-cell-25-20" class="code-annotation-target"><a href="#annotated-cell-25-20"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(s415<span class="sc">:</span>Clear)) <span class="sc">|&gt;</span></span>
<span id="annotated-cell-25-21"><a href="#annotated-cell-25-21"></a>  <span class="fu">rename_with</span>(<span class="sc">~</span> <span class="fu">str_remove</span>(.x, <span class="st">".normalized"</span>))</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-25" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-25" data-code-lines="1,2" data-code-annotation="1">Column names of variables that need to be normalized</span>
</dd>
<dt data-target-cell="annotated-cell-25" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-25" data-code-lines="4,5,6,7,8,9" data-code-annotation="2">Gain ratios as specified by the manufacturer’s reconstruction guide</span>
</dd>
<dt data-target-cell="annotated-cell-25" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-25" data-code-lines="14" data-code-annotation="3">Remove dark counts &amp; scale by integration time</span>
</dd>
<dt data-target-cell="annotated-cell-25" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-25" data-code-lines="15" data-code-annotation="4">Function to normalize counts</span>
</dd>
<dt data-target-cell="annotated-cell-25" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-25" data-code-lines="16" data-code-annotation="5">All sensor channels share the gain value</span>
</dd>
<dt data-target-cell="annotated-cell-25" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-25" data-code-lines="17" data-code-annotation="6">Sensor channels to normalize (see 1.)</span>
</dd>
<dt data-target-cell="annotated-cell-25" data-target-annotation="7">7</dt>
<dd>
<span data-code-cell="annotated-cell-25" data-code-lines="18" data-code-annotation="7">Gain ratios (see 2.)</span>
</dd>
<dt data-target-cell="annotated-cell-25" data-target-annotation="8">8</dt>
<dd>
<span data-code-cell="annotated-cell-25" data-code-lines="20,21" data-code-annotation="8">Drop original raw count columns</span>
</dd>
</dl>
</div>
</div>
<p>In this call, we specified <code>gain.columns = rep("Gain", 11)</code> because we have 11 sensor columns that all use the same gain factor column (<code>Gain</code>). This step will add new columns (with a suffix, e.g.&nbsp;<code>.normalized</code>) for each spectral channel representing the count normalized by the gain. We then dropped the <code>raw</code> count columns and renamed the normalized ones by dropping <code>.normalized</code> from the names. After this, <code>dataVEET2</code> contains the normalized sensor readings for <code>s415, s445, ..., s940, Dark, Clear</code> for each time point time point.</p>
<p>Because we do not need this high a resolution, we will&nbsp;<strong>aggregate it to a 5-minute interval</strong> for computational efficiency. The assumption is that spectral composition does not need to be examined at every 2-second instant for our purposes, and 5-minute averages can capture the general trends while drastically reducing data size and downstream computational costs.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Aggregate spectral data to 5-minute intervals and mark gaps</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-26"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-26-1"><a href="#annotated-cell-26-1"></a>dataVEET2 <span class="ot">&lt;-</span> dataVEET2 <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-26" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-26-2" class="code-annotation-target"><a href="#annotated-cell-26-2"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 mins"</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-26" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-26-3" class="code-annotation-target"><a href="#annotated-cell-26-3"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-26" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-26-4" class="code-annotation-target"><a href="#annotated-cell-26-4"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-26" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-26-5" class="code-annotation-target"><a href="#annotated-cell-26-5"></a>  <span class="fu">remove_partial_data</span>(Clear, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-26" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-26" data-code-lines="2" data-code-annotation="1">Aggregate to 5-minute bins</span>
</dd>
<dt data-target-cell="annotated-cell-26" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-26" data-code-lines="3" data-code-annotation="2">Explicit NA for any gaps in between</span>
</dd>
<dt data-target-cell="annotated-cell-26" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-26" data-code-lines="4" data-code-annotation="3">Add a date identifier for grouping</span>
</dd>
<dt data-target-cell="annotated-cell-26" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-26" data-code-lines="5" data-code-annotation="4">remove days with more than one hour of data missing</span>
</dd>
</dl>
</div>
</div>
<p>We aggregate over 5-minute windows; within each 5-minute bin, multiple spectral readings (if present) are combined (averaged). We use one of the channels (here&nbsp;<code>Clear</code>) as the reference variable for&nbsp;<code>remove_partial_data</code>&nbsp;to drop incomplete days (the choice of channel is arbitrary as all channels share the same level of completeness).</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Please note that <code>normalize_counts()</code> requires the <code>Gain</code> values according to the gain table. If we had aggregated the data before normalizing it, <code>Gain</code> values would have been averaged within each bin (5 minutes in this case). If the <code>Gain</code> did not change in that time, it is not an issue. Any mix of <code>Gain</code> values will lead to a <code>Gain</code> value that is not represented in the gain table. While outputs for <code>normalize_counts()</code> are not wrong in these cases, they will output <code>NA</code> if a <code>Gain</code> value is not found in the table. Thus we recommend to always normalize counts based on the raw dataset.</p>
</div>
</div>
</section>
<section id="spectral-reconstruction" class="level3" data-number="4.3">
<h3 data-number="4.3" class="anchored" data-anchor-id="spectral-reconstruction"><span class="header-section-number">4.3</span> Spectral reconstruction</h3>
<p>For spectral reconstruction, we require a calibration matrix that corresponds to the <code>VEET</code>’s sensor channels. This matrix would typically be obtained from the device manufacturer or a calibration procedure. It defines how each channel’s normalized count relates to intensity at various wavelengths. For demonstration, the calibration matrix was provided by the manufacturer and is specific to the make and model (see <a href="#fig-calibmtx" class="quarto-xref">Figure&nbsp;12</a>). It should not be used for research purposes without confirming its accuracy with the manufacturer.</p>
<div id="fig-calibmtx" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-calibmtx-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="assets/calibration_matrix.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13" title="Figure&nbsp;12: Calibration matrix"><img src="assets/calibration_matrix.png" class="img-fluid figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-calibmtx-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;12: Calibration matrix
</figcaption>
</figure>
</div>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Calibration matrica and reconstruction of spectral power distribution</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-27"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-27" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-27-1" class="code-annotation-target"><a href="#annotated-cell-27-1"></a>calib_mtx <span class="ot">&lt;-</span></span>
<span id="annotated-cell-27-2"><a href="#annotated-cell-27-2"></a>  <span class="fu">read_csv</span>(<span class="st">"data/VEET_calibration_matrix.csv"</span>,</span>
<span id="annotated-cell-27-3"><a href="#annotated-cell-27-3"></a>           <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">|&gt;</span></span>
<span id="annotated-cell-27-4"><a href="#annotated-cell-27-4"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"wavelength"</span>)</span>
<span id="annotated-cell-27-5"><a href="#annotated-cell-27-5"></a></span>
<span id="annotated-cell-27-6"><a href="#annotated-cell-27-6"></a>dataVEET2 <span class="ot">&lt;-</span></span>
<span id="annotated-cell-27-7"><a href="#annotated-cell-27-7"></a>  dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="annotated-cell-27-8"><a href="#annotated-cell-27-8"></a>  <span class="fu">mutate</span>(</span>
<span id="annotated-cell-27-9"><a href="#annotated-cell-27-9"></a>  <span class="at">Spectrum =</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-27" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-27-10" class="code-annotation-target"><a href="#annotated-cell-27-10"></a>    <span class="fu">spectral_reconstruction</span>(</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-27" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-27-11" class="code-annotation-target"><a href="#annotated-cell-27-11"></a>      <span class="fu">pick</span>(s415<span class="sc">:</span>s910),</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-27" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-27-12" class="code-annotation-target"><a href="#annotated-cell-27-12"></a>      <span class="at">calibration_matrix =</span> calib_mtx,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-27" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-27-13" class="code-annotation-target"><a href="#annotated-cell-27-13"></a>      <span class="at">format =</span> <span class="st">"long"</span></span>
<span id="annotated-cell-27-14"><a href="#annotated-cell-27-14"></a>  )</span>
<span id="annotated-cell-27-15"><a href="#annotated-cell-27-15"></a>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-27" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-27" data-code-lines="1,2,3,4" data-code-annotation="1">Import the calibration matrix and make certain <code>wavelength</code> is set a rownames</span>
</dd>
<dt data-target-cell="annotated-cell-27" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-27" data-code-lines="10" data-code-annotation="2">The function <code>spectral_reconstruction()</code> does not work on the level of the dataset, but has to be called within <code>mutate()</code>(or provided the data directly)</span>
</dd>
<dt data-target-cell="annotated-cell-27" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-27" data-code-lines="11" data-code-annotation="3">Pick the normalized sensor columns</span>
</dd>
<dt data-target-cell="annotated-cell-27" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-27" data-code-lines="12" data-code-annotation="4">Provide the calibration matrix</span>
</dd>
<dt data-target-cell="annotated-cell-27" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-27" data-code-lines="13" data-code-annotation="5">Return a long-form list column (wavelength, intensity)</span>
</dd>
</dl>
</div>
</div>
<p>Here, we use&nbsp;<code>format = "long"</code>&nbsp;so that the result for each observation is a&nbsp;<strong>list-column</strong>&nbsp;<code>Spectrum</code>, where each entry is a tibble<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a> containing two columns:&nbsp;<code>wavelength</code>&nbsp;and&nbsp;<code>irradiance</code>&nbsp;(one row per wavelength in the calibration matrix). In other words, each row of&nbsp;<code>dataVEET2</code>&nbsp;now holds a full reconstructed spectrum in the&nbsp;<code>Spectrum</code>&nbsp;column. The long format is convenient for further calculations and plotting. (An alternative&nbsp;<code>format = "wide"</code>&nbsp;would add each wavelength as a separate column, but that is less practical when there are many wavelengths.)</p>
<p>To visualize the data we will calculate the photopic illuminance based on the spectra and plot each spectrum color-scaled by their illuminance. For clarity, we reduce the data to observations within the day that has the most observations (non-<code>NA</code>).</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Calculate photopic illuminance</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-28"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-28-1"><a href="#annotated-cell-28-1"></a>data_spectra <span class="ot">&lt;-</span> </span>
<span id="annotated-cell-28-2"><a href="#annotated-cell-28-2"></a>dataVEET2 <span class="sc">|&gt;</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-28" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-28-3" class="code-annotation-target"><a href="#annotated-cell-28-3"></a>  <span class="fu">sample_groups</span>(<span class="at">order.by =</span> <span class="fu">sum</span>(<span class="sc">!</span>is.implicit)) <span class="sc">|&gt;</span></span>
<span id="annotated-cell-28-4"><a href="#annotated-cell-28-4"></a>  <span class="fu">mutate</span>( </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-28" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-28-5" class="code-annotation-target"><a href="#annotated-cell-28-5"></a>    <span class="at">Illuminance =</span> Spectrum <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-28" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-28-6" class="code-annotation-target"><a href="#annotated-cell-28-6"></a>      <span class="fu">map_dbl</span>(spectral_integration,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-28" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-28-7" class="code-annotation-target"><a href="#annotated-cell-28-7"></a>              <span class="at">action.spectrum =</span> <span class="st">"photopic"</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-28" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-28-8" class="code-annotation-target"><a href="#annotated-cell-28-8"></a>              <span class="at">general.weight =</span> <span class="st">"auto"</span>)</span>
<span id="annotated-cell-28-9"><a href="#annotated-cell-28-9"></a>  ) <span class="sc">|&gt;</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-28" data-target-annotation="6" onclick="event.preventDefault();">6</a><span id="annotated-cell-28-10" class="code-annotation-target"><a href="#annotated-cell-28-10"></a>  <span class="fu">unnest</span>(Spectrum)</span>
<span id="annotated-cell-28-11"><a href="#annotated-cell-28-11"></a>data_spectra <span class="sc">|&gt;</span> <span class="fu">select</span>(Id, Date, Datetime, Illuminance) <span class="sc">|&gt;</span> <span class="fu">distinct</span>()</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-28" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-28" data-code-lines="3" data-code-annotation="1">Keep only observations for one day (with the lowest missing intervals)</span>
</dd>
<dt data-target-cell="annotated-cell-28" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-28" data-code-lines="5" data-code-annotation="2">Use the spectrum,…</span>
</dd>
<dt data-target-cell="annotated-cell-28" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-28" data-code-lines="6" data-code-annotation="3">… call the function <code>spectral_integration()</code> for each,…</span>
</dd>
<dt data-target-cell="annotated-cell-28" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-28" data-code-lines="7" data-code-annotation="4">… use the brightness sensitivity function,…</span>
</dd>
<dt data-target-cell="annotated-cell-28" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-28" data-code-lines="8" data-code-annotation="5">… and apply the appropriate efficacy weight.</span>
</dd>
<dt data-target-cell="annotated-cell-28" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-28" data-code-lines="10" data-code-annotation="6">Create a long format of the data where the spectrum is unnested</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 288 × 4
# Groups:   Id, Date [1]
   Id    Date       Datetime            Illuminance
   &lt;fct&gt; &lt;date&gt;     &lt;dttm&gt;                    &lt;dbl&gt;
 1 VEET  2025-06-18 2025-06-18 00:00:00        3.70
 2 VEET  2025-06-18 2025-06-18 00:05:00        3.79
 3 VEET  2025-06-18 2025-06-18 00:10:00        3.72
 4 VEET  2025-06-18 2025-06-18 00:15:00        6.90
 5 VEET  2025-06-18 2025-06-18 00:20:00        3.53
 6 VEET  2025-06-18 2025-06-18 00:25:00        3.26
 7 VEET  2025-06-18 2025-06-18 00:30:00        3.59
 8 VEET  2025-06-18 2025-06-18 00:35:00        3.49
 9 VEET  2025-06-18 2025-06-18 00:40:00        3.54
10 VEET  2025-06-18 2025-06-18 00:45:00        3.62
# ℹ 278 more rows</code></pre>
</div>
</div>
<p>The following plot visualizes the spectra:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Plot spectra</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>data_spectra <span class="sc">|&gt;</span> </span>
<span id="cb24-2"><a href="#cb24-2"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> wavelength,<span class="at">group =</span> Datetime)) <span class="sc">+</span></span>
<span id="cb24-3"><a href="#cb24-3"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> irradiance<span class="sc">*</span><span class="dv">1000</span>, <span class="at">col =</span> Illuminance)) <span class="sc">+</span></span>
<span id="cb24-4"><a href="#cb24-4"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Irradiance (mW/m²/nm)"</span>, </span>
<span id="cb24-5"><a href="#cb24-5"></a>       <span class="at">x =</span> <span class="st">"Wavelength (nm)"</span>, </span>
<span id="cb24-6"><a href="#cb24-6"></a>       <span class="at">col =</span> <span class="st">"Photopic illuminance (lx)"</span>) <span class="sc">+</span></span>
<span id="cb24-7"><a href="#cb24-7"></a>  <span class="fu">scale_color_viridis_b</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span><span class="sc">^</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">3</span>))) <span class="sc">+</span></span>
<span id="cb24-8"><a href="#cb24-8"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">trans =</span> <span class="st">"symlog"</span>, <span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">50</span>)) <span class="sc">+</span></span>
<span id="cb24-9"><a href="#cb24-9"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="cn">NA</span>), <span class="at">expand =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb24-10"><a href="#cb24-10"></a>  <span class="fu">theme_minimal</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output-display">
<div id="fig-VEET-spectra-intensity" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-spectra-intensity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-VEET-spectra-intensity-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14" title="Figure&nbsp;13: Overview of the reconstructed spectra, color-scaled by photopic illuminance (lx)"><img src="supplement_files/figure-html/fig-VEET-spectra-intensity-1.png" class="img-fluid figure-img" width="768"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-spectra-intensity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;13: Overview of the reconstructed spectra, color-scaled by photopic illuminance (lx)
</figcaption>
</figure>
</div>
</div>
</div>
<p>The following ridgeline plot can be used to assess <strong>when</strong> in the day certain spectral wavelenghts are dominant:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Plot spectra across the time of day</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>data_spectra <span class="sc">|&gt;</span> </span>
<span id="cb25-2"><a href="#cb25-2"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> wavelength, <span class="at">group =</span> Datetime)) <span class="sc">+</span></span>
<span id="cb25-3"><a href="#cb25-3"></a>  <span class="fu">geom_ridgeline</span>(<span class="fu">aes</span>(<span class="at">height =</span> irradiance<span class="sc">*</span><span class="dv">1000</span>, </span>
<span id="cb25-4"><a href="#cb25-4"></a>                     <span class="at">y =</span> Datetime, </span>
<span id="cb25-5"><a href="#cb25-5"></a>                     <span class="at">fill =</span> Illuminance), </span>
<span id="cb25-6"><a href="#cb25-6"></a>                 <span class="at">scale =</span> <span class="dv">400</span>, <span class="at">lwd =</span> <span class="fl">0.1</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb25-7"><a href="#cb25-7"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Local time &amp; Irradiance (mW/m²/nm)"</span>, </span>
<span id="cb25-8"><a href="#cb25-8"></a>       <span class="at">x =</span> <span class="st">"Wavelength (nm)"</span>, </span>
<span id="cb25-9"><a href="#cb25-9"></a>       <span class="at">fill =</span> <span class="st">"Photopic illuminance (lx)"</span>)<span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10"></a>  <span class="fu">scale_fill_viridis_b</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span><span class="sc">^</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">3</span>))) <span class="sc">+</span></span>
<span id="cb25-11"><a href="#cb25-11"></a>  <span class="fu">theme_minimal</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output-display">
<div id="fig-VEET-spectra-time" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-spectra-time-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-VEET-spectra-time-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15" title="Figure&nbsp;14: Overview of the reconstructed spectra by time of day, color-scaled by photopic illuminance (lx)"><img src="supplement_files/figure-html/fig-VEET-spectra-time-1.png" class="img-fluid figure-img" width="768"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-spectra-time-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;14: Overview of the reconstructed spectra by time of day, color-scaled by photopic illuminance (lx)
</figcaption>
</figure>
</div>
</div>
</div>
<p>At this stage, the&nbsp;<code>dataVEET2</code>&nbsp;dataset has been processed to yield time-series of spectral power distributions. We can use these to compute biologically relevant light metrics. For instance, one possible metric is the proportion of power in short wavelengths versus long wavelengths.</p>
<p>In the main analysis, we defined&nbsp;<strong>short-wavelength (blue light) content</strong>&nbsp;as the integrated intensity in the 400–500&nbsp;nm range, and&nbsp;<strong>long-wavelength content</strong>&nbsp;as the integrated intensity in a longer range (e.g.&nbsp;600–700&nbsp;nm), then computed the&nbsp;<em>short-to-long ratio</em>&nbsp;(“sl ratio”). Calculating these metrics is the first step of spectrum analysis in the <a href="./index.html#spectrum">main tutorial</a>.</p>
<pre><code>dataVEET2 |&gt; 
  select(Id, Date, Datetime, Spectrum) |&gt;    # focus on ID, date, time, and spectrum
  mutate(
    short = Spectrum |&gt; map_dbl(spectral_integration, wavelength.range = c(400, 500)),
    long  = Spectrum |&gt; map_dbl(spectral_integration, wavelength.range = c(600, 700)),
    `sl ratio` = short / long   # compute short-to-long ratio
  )</code></pre>
<p><em>(The cutoff of 500&nbsp;nm here is hypothetical for demonstration; actual definitions might vary.)</em>&nbsp;We would then have columns&nbsp;<code>short</code>,&nbsp;<code>long</code>, and&nbsp;<code>sl_ratio</code>&nbsp;for each observation, which could be averaged per day or analyzed further. The cleaned spectral data in&nbsp;<code>dataVEET2</code>&nbsp;makes it straightforward to calculate such metrics or apply spectral weighting functions (for melatonin suppression, circadian stimulus, etc., if one has the spectral sensitivity curves).</p>
<p>With the <code>VEET</code> spectral preprocessing complete, we emphasize that these steps – normalizing by gain, applying calibration, and perhaps simplifying channels – are&nbsp;<strong>device-specific</strong>&nbsp;requirements. They ensure that the raw sensor counts are translated into meaningful physical measures (like spectral irradiance). Researchers using other spectral devices would follow a similar procedure, adjusting for their device’s particulars (some may output spectra directly, whereas others, like <code>VEET</code>, require reconstruction.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Some devices may output normalized counts instead of raw counts. For example, the <code>ActLumus</code> device outputs normalized counts, while the <code>VEET</code> device records raw counts and the gain. Manufacturers will be able to speficy exact outputs for a given model and software version.</p>
</div>
</div>
</section>
</section>
<section id="veet-data-time-of-flight-distance" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="veet-data-time-of-flight-distance"><span class="header-section-number">5</span> <code>VEET</code> Data: Time of flight (distance)</h2>
<p>In this last section, the distance data of the <code>VEET</code> device will be imported, analogous to the other modalities. The <code>TOF</code> modality contains information for up to two objects in a 8x8 grid of measurements, spanning a total of about 52° vertically and 41° horizontally. Because the <code>VEET</code> device can detect up to two objects in a given grid point, and there is a confidence value assigned to every measurement, each observation contains <span class="math inline">\(2*2*8*8 = 256\)</span> measurements.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Import VEET Spectral Sensor (TOF) data</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-32"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-32-1"><a href="#annotated-cell-32-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="annotated-cell-32-2"><a href="#annotated-cell-32-2"></a>dataVEET3 <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, </span>
<span id="annotated-cell-32-3"><a href="#annotated-cell-32-3"></a>                        <span class="at">tz =</span> tz, </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-32" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-32-4" class="code-annotation-target"><a href="#annotated-cell-32-4"></a>                        <span class="at">modality =</span> <span class="st">"TOF"</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-32" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-32-5" class="code-annotation-target"><a href="#annotated-cell-32-5"></a>                        <span class="at">manual.id =</span> <span class="st">"VEET"</span></span>
<span id="annotated-cell-32-6"><a href="#annotated-cell-32-6"></a>                        ) </span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-32" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-32" data-code-lines="4" data-code-annotation="1"><code>modality</code> is a parameter only the <code>VEET</code> device requires. If uncertain, which devices require special parameters, have a look a the import help page (<code>?import</code>) under the <code>VEET</code> device. Setting it to <code>TOF</code> gives us the distance modality.</span>
</dd>
<dt data-target-cell="annotated-cell-32" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-32" data-code-lines="5" data-code-annotation="2">As we are only dealing with one individual here, we set a manual <code>Id</code></span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Successfully read in 304'195 observations across 1 Ids from 1 VEET-file(s).
Timezone set is US/Central.
The system timezone is UTC. Please correct if necessary!

First Observation: 2024-06-04 15:00:36
Last Observation: 2024-06-12 08:29:43
Timespan: 7.7 days

Observation intervals: 
  Id    interval.time              n pct      
1 VEET  0s                         3 0.00099% 
2 VEET  1s                      2089 0.68673% 
3 VEET  2s                    299876 98.58051%
4 VEET  3s                      2213 0.72750% 
5 VEET  4s                         3 0.00099% 
6 VEET  6s                         1 0.00033% 
7 VEET  9s                         7 0.00230% 
8 VEET  109s (~1.82 minutes)       1 0.00033% 
9 VEET  59077s (~16.41 hours)      1 0.00033% </code></pre>
</div>
<div class="cell-output-display">
<div id="fig-VEET-overview3" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-overview3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-VEET-overview3-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16" title="Figure&nbsp;15: Overview plot of imported VEET data"><img src="supplement_files/figure-html/fig-VEET-overview3-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-overview3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15: Overview plot of imported VEET data
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Print the first 6 rows of the VEETs distance dataset</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>dataVEET3 <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 261
# Groups:   Id [1]
  Id    Datetime            file.name     time_stamp modality conf1_0 conf1_1
  &lt;fct&gt; &lt;dttm&gt;              &lt;chr&gt;              &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;
1 VEET  2024-06-04 15:00:36 01_VEET_L.csv 1717531236 TOF           46      36
2 VEET  2024-06-04 15:00:38 01_VEET_L.csv 1717531238 TOF           29      22
3 VEET  2024-06-04 15:00:40 01_VEET_L.csv 1717531240 TOF            0       0
4 VEET  2024-06-04 15:00:42 01_VEET_L.csv 1717531242 TOF            0       0
5 VEET  2024-06-04 15:00:44 01_VEET_L.csv 1717531244 TOF            0       0
6 VEET  2024-06-04 15:00:46 01_VEET_L.csv 1717531246 TOF           59      40
# ℹ 254 more variables: conf1_2 &lt;dbl&gt;, conf1_3 &lt;dbl&gt;, conf1_4 &lt;dbl&gt;,
#   conf1_5 &lt;dbl&gt;, conf1_6 &lt;dbl&gt;, conf1_7 &lt;dbl&gt;, conf1_8 &lt;dbl&gt;, conf1_9 &lt;dbl&gt;,
#   conf1_10 &lt;dbl&gt;, conf1_11 &lt;dbl&gt;, conf1_12 &lt;dbl&gt;, conf1_13 &lt;dbl&gt;,
#   conf1_14 &lt;dbl&gt;, conf1_15 &lt;dbl&gt;, conf1_16 &lt;dbl&gt;, conf1_17 &lt;dbl&gt;,
#   conf1_18 &lt;dbl&gt;, conf1_19 &lt;dbl&gt;, conf1_20 &lt;dbl&gt;, conf1_21 &lt;dbl&gt;,
#   conf1_22 &lt;dbl&gt;, conf1_23 &lt;dbl&gt;, conf1_24 &lt;dbl&gt;, conf1_25 &lt;dbl&gt;,
#   conf1_26 &lt;dbl&gt;, conf1_27 &lt;dbl&gt;, conf1_28 &lt;dbl&gt;, conf1_29 &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>In a first step, we condition the data similarly to the other <code>VEET</code> modalities. For computational reasons of the use cases, we remove the second object and set the interval to <code>10 seconds</code>. <em>Note that the next step still takes considerable computation time</em>.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Aggregate distance data to 5-second intervals and mark gaps</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-34"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-34-1"><a href="#annotated-cell-34-1"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="annotated-cell-34-2"><a href="#annotated-cell-34-2"></a>  dataVEET3 <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-34" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-34-3" class="code-annotation-target"><a href="#annotated-cell-34-3"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">contains</span>(<span class="fu">c</span>(<span class="st">"conf2_"</span>, <span class="st">"dist2_"</span>))) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-34" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-34-4" class="code-annotation-target"><a href="#annotated-cell-34-4"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 secs"</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-34" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-34-5" class="code-annotation-target"><a href="#annotated-cell-34-5"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="annotated-cell-34-6"><a href="#annotated-cell-34-6"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="annotated-cell-34-7"><a href="#annotated-cell-34-7"></a>  <span class="fu">remove_partial_data</span>(dist1_0, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span>
<span id="annotated-cell-34-8"><a href="#annotated-cell-34-8"></a>dataVEET3 <span class="sc">|&gt;</span> <span class="fu">summary_overview</span>(dist1_0)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-34" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-34" data-code-lines="3" data-code-annotation="1">Remove the second object (for computational reasons)</span>
</dd>
<dt data-target-cell="annotated-cell-34" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-34" data-code-lines="4" data-code-annotation="2">Aggregate to 10-second bins</span>
</dd>
<dt data-target-cell="annotated-cell-34" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-34" data-code-lines="5" data-code-annotation="3">Explicit NA for any gaps</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 4
  name                mean   min   max
  &lt;chr&gt;              &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 Participants           1    NA    NA
2 Participant-days       6     6     6
3 Days ≥80% complete     6     6     6
4 Missing/Irregular      0     0     0</code></pre>
</div>
</div>
<p>In the next step, we need to transform the <code>wide</code> format of the imported dataset into a <code>long</code> format, where each row contains exactly one observation for one grid-point.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Pivot distance grid from wide to long</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb31-2"><a href="#cb31-2"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb31-3"><a href="#cb31-3"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb31-4"><a href="#cb31-4"></a>    <span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(Datetime, file.name, Id, is.implicit, time_stamp, modality, Date),</span>
<span id="cb31-5"><a href="#cb31-5"></a>    <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">".value"</span>, <span class="st">"position"</span>),</span>
<span id="cb31-6"><a href="#cb31-6"></a>    <span class="at">names_pattern =</span> <span class="st">"(conf1|conf2|dist1|dist2)_(</span><span class="sc">\\</span><span class="st">d+)"</span></span>
<span id="cb31-7"><a href="#cb31-7"></a>  )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
<p>In a final step before we can use the data in the analysis, we need to assign <code>x</code> and <code>y</code> coordinates based on the <code>position</code> column that was created when pivoting longer. Positions are counted from 0 to 63 starting at the top right and increasing towards the left, before continuing on the right in the next row below. <code>y</code> positions thus depend on the row count, i.e., how often a row of 8 values fits into the <code>position</code> column. <code>x</code> positions consequently depend on the <code>position</code> within each 8-value row. We also add an <code>observation</code> variable that increases by <code>+1</code> every time, the <code>position</code> column hits <code>0</code>. We then center both <code>x</code> and <code>y</code> coordinates to receive meaningful values, i.e., 0° indicates the center of the overall measurement cone. Lastly, we convert both confidence columns, which are scaled from 0-255 into percentages by dividing them by <code>255</code>. Empirical data from the manufacturer points to a threshold of about <code>10%</code>, under which the respective distance data is not reliable.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Calculate grid positions of spatial distance measurements</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-36"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><span id="annotated-cell-36-1"><a href="#annotated-cell-36-1"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="annotated-cell-36-2"><a href="#annotated-cell-36-2"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="annotated-cell-36-3"><a href="#annotated-cell-36-3"></a>  <span class="fu">mutate</span>(<span class="at">position =</span> <span class="fu">as.numeric</span>(position),</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-36" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-36-4" class="code-annotation-target"><a href="#annotated-cell-36-4"></a>         <span class="at">y.pos =</span> (position <span class="sc">%/%</span> <span class="dv">8</span>)<span class="sc">+</span><span class="dv">1</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-36" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-36-5" class="code-annotation-target"><a href="#annotated-cell-36-5"></a>         <span class="at">y.pos =</span> <span class="fu">scale</span>(y.pos, <span class="at">scale =</span> <span class="cn">FALSE</span>)<span class="sc">*</span><span class="dv">52</span><span class="sc">/</span><span class="dv">8</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-36" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-36-6" class="code-annotation-target"><a href="#annotated-cell-36-6"></a>         <span class="at">x.pos =</span> <span class="dv">8</span> <span class="sc">-</span> (position <span class="sc">%%</span> <span class="dv">8</span>),</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-36" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-36-7" class="code-annotation-target"><a href="#annotated-cell-36-7"></a>         <span class="at">x.pos =</span> <span class="fu">scale</span>(x.pos, <span class="at">scale =</span> <span class="cn">FALSE</span>)<span class="sc">*</span><span class="dv">41</span><span class="sc">/</span><span class="dv">8</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-36" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-36-8" class="code-annotation-target"><a href="#annotated-cell-36-8"></a>         <span class="at">observation =</span> <span class="fu">cumsum</span>(position <span class="sc">==</span> <span class="dv">0</span>),</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-36" data-target-annotation="6" onclick="event.preventDefault();">6</a><span id="annotated-cell-36-9" class="code-annotation-target"><a href="#annotated-cell-36-9"></a>         <span class="fu">across</span>(<span class="fu">starts_with</span>(<span class="st">"conf"</span>), \(x) x<span class="sc">/</span><span class="dv">255</span>)</span>
<span id="annotated-cell-36-10"><a href="#annotated-cell-36-10"></a>         )</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-36" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-36" data-code-lines="4" data-code-annotation="1">Increment the y position for every 8 steps in <code>position</code></span>
</dd>
<dt data-target-cell="annotated-cell-36" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-36" data-code-lines="5" data-code-annotation="2">Center <code>y.pos</code> and rescale it to cover 52° across 8 steps</span>
</dd>
<dt data-target-cell="annotated-cell-36" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-36" data-code-lines="6" data-code-annotation="3">Increment the x position for every step in <code>position</code>, resetting every 8 steps</span>
</dd>
<dt data-target-cell="annotated-cell-36" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-36" data-code-lines="7" data-code-annotation="4">Center <code>x.pos</code> and rescale it to cover 41° across 8 steps</span>
</dd>
<dt data-target-cell="annotated-cell-36" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-36" data-code-lines="8" data-code-annotation="5">Increase an observation counter every time we restart with <code>position</code> at 0</span>
</dd>
<dt data-target-cell="annotated-cell-36" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-36" data-code-lines="9" data-code-annotation="6">Scale the confidence columns so that 255 = 100%</span>
</dd>
</dl>
</div>
</div>
<p>Now this dataset is ready for further analysis. We finish by visualizing the same observation time on different days. Note that we replace zero distance values with <code>Infinity</code>, as these indicate measurements outside the 5m measurement radius of the device.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Plot spatial distance grid for the same time point on each day</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-37"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-37" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-37-1" class="code-annotation-target"><a href="#annotated-cell-37-1"></a>extras <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-37-2"><a href="#annotated-cell-37-2"></a>  <span class="fu">geom_tile</span>(),</span>
<span id="annotated-cell-37-3"><a href="#annotated-cell-37-3"></a>  <span class="fu">scale_fill_viridis_c</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="annotated-cell-37-4"><a href="#annotated-cell-37-4"></a>                       <span class="at">oob =</span> scales<span class="sc">::</span>oob_squish_any),</span>
<span id="annotated-cell-37-5"><a href="#annotated-cell-37-5"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"white"</span>)),</span>
<span id="annotated-cell-37-6"><a href="#annotated-cell-37-6"></a>  <span class="fu">theme_minimal</span>(),</span>
<span id="annotated-cell-37-7"><a href="#annotated-cell-37-7"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="st">"none"</span>),</span>
<span id="annotated-cell-37-8"><a href="#annotated-cell-37-8"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> (dist1<span class="sc">/</span><span class="dv">10</span>) <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">0</span>), <span class="at">colour =</span> dist1<span class="sc">&gt;</span><span class="dv">1000</span>),</span>
<span id="annotated-cell-37-9"><a href="#annotated-cell-37-9"></a>            <span class="at">size =</span> <span class="fl">2.5</span>),</span>
<span id="annotated-cell-37-10"><a href="#annotated-cell-37-10"></a>  <span class="fu">coord_fixed</span>(),</span>
<span id="annotated-cell-37-11"><a href="#annotated-cell-37-11"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"X position (°)"</span>, <span class="at">y =</span> <span class="st">"Y position (°)"</span>,</span>
<span id="annotated-cell-37-12"><a href="#annotated-cell-37-12"></a>       <span class="at">fill =</span> <span class="st">"Distance (cm)"</span>))</span>
<span id="annotated-cell-37-13"><a href="#annotated-cell-37-13"></a></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-37" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-37-14" class="code-annotation-target"><a href="#annotated-cell-37-14"></a>slicer <span class="ot">&lt;-</span> <span class="cf">function</span>(x){<span class="fu">seq</span>(<span class="fu">min</span>((x<span class="dv">-1</span>)<span class="sc">*</span><span class="dv">64</span><span class="sc">+</span><span class="dv">1</span>), <span class="fu">max</span>(x<span class="sc">*</span><span class="dv">64</span>, <span class="at">by =</span> <span class="dv">1</span>))}</span>
<span id="annotated-cell-37-15"><a href="#annotated-cell-37-15"></a></span>
<span id="annotated-cell-37-16"><a href="#annotated-cell-37-16"></a>dataVEET3 <span class="sc">|&gt;</span> </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-37" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-37-17" class="code-annotation-target"><a href="#annotated-cell-37-17"></a>  <span class="fu">slice</span>(<span class="fu">slicer</span>(<span class="dv">9530</span>)) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-37" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-37-18" class="code-annotation-target"><a href="#annotated-cell-37-18"></a>  <span class="fu">mutate</span>(<span class="at">dist1 =</span> <span class="fu">ifelse</span>(dist1 <span class="sc">==</span> <span class="dv">0</span>, <span class="cn">Inf</span>, dist1)) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-37" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-37-19" class="code-annotation-target"><a href="#annotated-cell-37-19"></a>  <span class="fu">filter</span>(conf1 <span class="sc">&gt;=</span> <span class="fl">0.1</span> <span class="sc">|</span> dist1 <span class="sc">==</span> <span class="cn">Inf</span>) <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-37" data-target-annotation="6" onclick="event.preventDefault();">6</a><span id="annotated-cell-37-20" class="code-annotation-target"><a href="#annotated-cell-37-20"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x.pos, <span class="at">y=</span>y.pos, <span class="at">fill =</span> dist1<span class="sc">/</span><span class="dv">10</span>))<span class="sc">+</span> extras <span class="sc">+</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-37" data-target-annotation="7" onclick="event.preventDefault();">7</a><span id="annotated-cell-37-21" class="code-annotation-target"><a href="#annotated-cell-37-21"></a>  <span class="fu">facet_grid</span>(<span class="sc">~</span>Datetime)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-37" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-37" data-code-lines="1,2,3,4,5,6,7,8,9,10,11,12" data-code-annotation="1">Set visualization parameters</span>
</dd>
<dt data-target-cell="annotated-cell-37" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-37" data-code-lines="14" data-code-annotation="2">Allows to choose an observation</span>
</dd>
<dt data-target-cell="annotated-cell-37" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-37" data-code-lines="17" data-code-annotation="3">Choose a particular observation</span>
</dd>
<dt data-target-cell="annotated-cell-37" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-37" data-code-lines="18" data-code-annotation="4">Replace 0 distances with Infinity</span>
</dd>
<dt data-target-cell="annotated-cell-37" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-37" data-code-lines="19" data-code-annotation="5">Remove data that has less than 10% confidence</span>
</dd>
<dt data-target-cell="annotated-cell-37" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-37" data-code-lines="20" data-code-annotation="6">Plot the data</span>
</dd>
<dt data-target-cell="annotated-cell-37" data-target-annotation="7">7</dt>
<dd>
<span data-code-cell="annotated-cell-37" data-code-lines="21" data-code-annotation="7">Show one plot per day</span>
</dd>
</dl>
</div>
<div class="cell-output-display">
<div id="fig-VEET-distance" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-distance-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="supplement_files/figure-html/fig-VEET-distance-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17" title="Figure&nbsp;16: Example observations of the measurement grid at 1:14 p.m. for each measurement day. Text values show distance in cm. Empty grid points show values with low confidence. Zero-distance values were replaced with infinite distance and plotted despite low confidence."><img src="supplement_files/figure-html/fig-VEET-distance-1.png" class="img-fluid figure-img" width="1152"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-distance-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;16: Example observations of the measurement grid at 1:14 p.m. for each measurement day. Text values show distance in cm. Empty grid points show values with low confidence. Zero-distance values were replaced with infinite distance and plotted despite low confidence.
</figcaption>
</figure>
</div>
</div>
</div>
<p>As we can see from the figure, different days have - at a given time - a vastly different distribution of distance data, and measurement confidence (values with confidence &lt; 10% are removed)</p>
<section id="removing-non-wear-times" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="removing-non-wear-times"><span class="header-section-number">5.1</span> Removing non-wear times</h3>
<p>Same as for the <code>ALS</code> modality, we can remove times of non-wear - we can even use the same <code>wear_data</code> set for it:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Removing non-wear observations from distance modality</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb32-2"><a href="#cb32-2"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb32-3"><a href="#cb32-3"></a>  <span class="fu">group_by</span>(Id) <span class="sc">|&gt;</span> </span>
<span id="cb32-4"><a href="#cb32-4"></a>  <span class="fu">add_states</span>(wear_data)</span>
<span id="cb32-5"><a href="#cb32-5"></a></span>
<span id="cb32-6"><a href="#cb32-6"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb32-7"><a href="#cb32-7"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb32-8"><a href="#cb32-8"></a>  <span class="fu">mutate</span>(<span class="at">dist1 =</span> <span class="fu">ifelse</span>(wear, dist1, <span class="cn">NA</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb32-9"><a href="#cb32-9"></a>  <span class="fu">group_by</span>(Id, Date)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
</section>
</section>
<section id="saving-the-cleaned-data" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="saving-the-cleaned-data"><span class="header-section-number">6</span> Saving the Cleaned Data</h2>
<p>After executing all the above steps, we have three cleaned data frames in our R session:</p>
<ul>
<li><p><strong><code>dataCC</code></strong>&nbsp;– the processed <code>Clouclip</code> dataset (5-second intervals, with distance and lux, including NA for gaps and sentinel statuses).</p></li>
<li><p><strong><code>dataVEET</code></strong>&nbsp;– the processed <code>VEET</code> ambient light dataset (5-second intervals, illuminance in lux, with gaps filled).</p></li>
<li><p><strong><code>dataVEET2</code></strong>&nbsp;– the processed <code>VEET</code> spectral dataset (5-minute intervals, each entry containing a spectrum or derived spectral metrics).</p></li>
<li><p><strong><code>dataVEET3</code></strong>&nbsp;– the processed <code>VEET</code> distance dataset (5-second intervals, each entry containing the distance of up to two objects in the 8x8 grid).</p></li>
</ul>
<p>For convenience and future reproducibility, we will save these combined results to a single R data file. Storing all cleaned data together ensures that any analysis can reload the exact same data state without re-running the import and cleaning (which can be time-consuming for large raw files).</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Save preprocessed files</strong></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="annotated-cell-39"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-39" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-39-1" class="code-annotation-target"><a href="#annotated-cell-39-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">"data/cleaned"</span>)) <span class="fu">dir.create</span>(<span class="st">"data/cleaned"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-39" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-39-2" class="code-annotation-target"><a href="#annotated-cell-39-2"></a><span class="fu">save</span>(dataCC, dataVEET, dataVEET2, dataVEET3, <span class="at">file =</span> <span class="st">"data/cleaned/data.RData"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-39" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-39" data-code-lines="1" data-code-annotation="1">Create directory for cleaned data if it doesn’t exist</span>
</dd>
<dt data-target-cell="annotated-cell-39" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-39" data-code-lines="2" data-code-annotation="2">Save all cleaned datasets into one .RData file</span>
</dd>
</dl>
</div>
</div>
<p>The above code creates (if necessary) a folder&nbsp;<code>data/cleaned/</code>&nbsp;and saves a single RData file (<code>data.RData</code>) containing the three objects. To retrieve them later, one can use&nbsp;<code>&lt;- load("data/cleaned/data.RData")</code>, which will return the objects into the environment. This single-file approach simplifies sharing and keeps the cleaned data together.</p>
<p>In summary, this supplement has walked through the full preprocessing pipeline for two example devices. We began by describing the raw data format for each device and then demonstrated how to import the data with correct time zone settings. We handled device-specific quirks like sentinel codes (for <code>Clouclip</code>) and multiple modalities with gain normalization (for <code>VEET</code>). We showed how to detect and address irregular sampling, how to explicitly mark missing data gaps to avoid analytic pitfalls, and how to reduce data granularity via rounding or aggregation when appropriate. Throughout, we used functions from LightLogR in a tidyverse workflow, aiming to make the steps clear and modular. By saving the final cleaned datasets, we set the stage for the computation of visual experience metrics such as working distance, time in bright light, spectral composition ratios, as presented in the <a href="./index.html">main tutorial</a>. We hope this detailed tutorial empowers researchers to adopt similar pipelines for their own data, facilitating reproducible and accurate analyses of visual experience.</p>
</section>
<section id="references" class="level2" data-number="7">



<!-- -->


</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">7 References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Biller2025" class="csl-entry" role="listitem">
Biller, Anna M., Johannes Zauner, Christian Cajochen, Marisa A. Gerle, Vineetha Kalavally, Anas Mohamed, Lukas Rottländer, Ming-Yi Seah, Oliver Stefani, and Manuel Spitschan. 2025. <span>“Physiologically-Relevant Light Exposure and Light Behaviour in Switzerland and Malaysia.”</span> <em>Journal of Exposure Science &amp; Environmental Epidemiology</em>. <a href="https://doi.org/10.1038/s41370-025-00825-8">https://doi.org/10.1038/s41370-025-00825-8</a>.
</div>
<div id="ref-Sah2025OphtalmicPhysiolOpt" class="csl-entry" role="listitem">
Sah, Raman Prasad, Pavan Kalyan Narra, and Lisa A. Ostrin. 2025. <span>“A Novel Wearable Sensor for Objective Measurement of Distance and Illumination.”</span> <em>Ophthalmic and Physiological Optics</em> 00 (n/a): 1–13. https://doi.org/<a href="https://doi.org/10.1111/opo.13523">https://doi.org/10.1111/opo.13523</a>.
</div>
<div id="ref-Sullivan2024" class="csl-entry" role="listitem">
Sullivan, David, Aaron Nicholls, George Hatoun, Samuel Thompson, Cory Schwarzmiller, Fathollah Memarzanjany, Alyssa Gunderson, et al. 2024. <span>“The Visual Experience Evaluation Tool: A Myopia Research Instrument for Quantifying Visual Experience.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2024.09.20.614212">https://doi.org/10.1101/2024.09.20.614212</a>.
</div>
<div id="ref-Wen2020bjophthalmol" class="csl-entry" role="listitem">
Wen, Longbo, Yingpin Cao, Qian Cheng, Xiaoning Li, Lun Pan, Lei Li, HaoGang Zhu, Weizhong Lan, and Zhikuan Yang. 2020. <span>“Objectively Measured Near Work, Outdoor Exposure and Myopia in Children.”</span> <em>British Journal of Ophthalmology</em> 104 (11): 1542–47. <a href="https://doi.org/10.1136/bjophthalmol-2019-315258">https://doi.org/10.1136/bjophthalmol-2019-315258</a>.
</div>
<div id="ref-Wen2021ActaOphtalmol" class="csl-entry" role="listitem">
Wen, Longbo, Qian Cheng, Yingpin Cao, Xiaoning Li, Lun Pan, Lei Li, Haogang Zhu, Ian Mogran, Weizhong Lan, and Zhikuan Yang. 2021. <span>“The Clouclip, a Wearable Device for Measuring Near-Work and Outdoor Time: Validation and Comparison of Objective Measures with Questionnaire Estimates.”</span> <em>Acta Ophthalmologica</em> 99 (7): e1222–35. https://doi.org/<a href="https://doi.org/10.1111/aos.14785">https://doi.org/10.1111/aos.14785</a>.
</div>
<div id="ref-Zauner2025JOpenSourceSoftw" class="csl-entry" role="listitem">
Zauner, J., S. Hartmeyer, and M. Spitschan. 2025. <span>“LightLogR: Reproducible Analysis of Personal Light Exposure Data.”</span> Journal Article. <em>J Open Source Softw</em> 10 (107): 7601. <a href="https://doi.org/10.21105/joss.07601">https://doi.org/10.21105/joss.07601</a>.
</div>
<div id="ref-Zauner2025npj" class="csl-entry" role="listitem">
Zauner, J., O. Stefani, G. Bocanegra, C. Guidolin, B. Schrader, L. Udovicic, and M. Spitschan. 2025. <span>“Auxiliary Data, Quality Assurance and Quality Control for Wearable Light Loggers and Optical Radiation Dosimeters.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2025.09.11.675633">https://doi.org/10.1101/2025.09.11.675633</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>A <em>sentinel value</em> is a special placeholder value used in data recording to signal a particular condition. It does not represent a valid measured quantity but rather acts as a marker (for example, “device off” or “value out of range”).<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p><strong>tibble</strong> are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <a href="https://tibble.tidyverse.org/index.html">tibbles</a><a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p><strong>tibble</strong> are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <a href="https://tibble.tidyverse.org/index.html">tibbles</a><a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p><strong>tibble</strong> are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <a href="https://tibble.tidyverse.org/index.html">tibbles</a><a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/tscnlab\.github\.io\/ZaunerEtAl_JVis_2026\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb33" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb33-1"><a href="#cb33-1"></a><span class="co">---</span></span>
<span id="cb33-2"><a href="#cb33-2"></a><span class="an">title:</span><span class="co"> "Supplement 1"</span></span>
<span id="cb33-3"><a href="#cb33-3"></a><span class="an">subtitle:</span><span class="co"> "Analysis of human visual experience data"</span></span>
<span id="cb33-4"><a href="#cb33-4"></a><span class="an">author:</span></span>
<span id="cb33-5"><a href="#cb33-5"></a><span class="co">  - name: "Johannes Zauner"</span></span>
<span id="cb33-6"><a href="#cb33-6"></a><span class="co">    id: JZ</span></span>
<span id="cb33-7"><a href="#cb33-7"></a><span class="co">    affiliation: </span></span>
<span id="cb33-8"><a href="#cb33-8"></a><span class="co">      - Technical University of Munich, Germany</span></span>
<span id="cb33-9"><a href="#cb33-9"></a><span class="co">      - Max Planck Institute for Biological Cybernetics, Tübingen, Germany</span></span>
<span id="cb33-10"><a href="#cb33-10"></a><span class="co">    orcid: "0000-0003-2171-4566"</span></span>
<span id="cb33-11"><a href="#cb33-11"></a><span class="co">    corresponding: true</span></span>
<span id="cb33-12"><a href="#cb33-12"></a><span class="co">    email: johannes.zauner@tum.de</span></span>
<span id="cb33-13"><a href="#cb33-13"></a><span class="co">  - name: "Aaron Nicholls"</span></span>
<span id="cb33-14"><a href="#cb33-14"></a><span class="co">    affiliation: "Reality Labs Research, USA"</span></span>
<span id="cb33-15"><a href="#cb33-15"></a><span class="co">    orcid: "0009-0001-6683-6826"</span></span>
<span id="cb33-16"><a href="#cb33-16"></a><span class="co">  - name: "Lisa A. Ostrin"</span></span>
<span id="cb33-17"><a href="#cb33-17"></a><span class="co">    affiliation: "University of Houston College of Optometry, USA"</span></span>
<span id="cb33-18"><a href="#cb33-18"></a><span class="co">    orcid: "0000-0001-8675-1478"</span></span>
<span id="cb33-19"><a href="#cb33-19"></a><span class="co">  - name: "Manuel Spitschan"</span></span>
<span id="cb33-20"><a href="#cb33-20"></a><span class="co">    affiliation: </span></span>
<span id="cb33-21"><a href="#cb33-21"></a><span class="co">      - Technical University of Munich, Germany</span></span>
<span id="cb33-22"><a href="#cb33-22"></a><span class="co">      - Max Planck Institute for Biological Cybernetics, Tübingen, Germany</span></span>
<span id="cb33-23"><a href="#cb33-23"></a><span class="co">      - Technical University of Munich Institute of advanced study (TUM-IAS), Munich, Germany</span></span>
<span id="cb33-24"><a href="#cb33-24"></a><span class="co">    orcid: "0000-0002-8572-9268"</span></span>
<span id="cb33-25"><a href="#cb33-25"></a><span class="an">doi:</span><span class="co"> 10.5281/zenodo.16566014</span></span>
<span id="cb33-26"><a href="#cb33-26"></a><span class="an">format:</span></span>
<span id="cb33-27"><a href="#cb33-27"></a><span class="co">  html:</span></span>
<span id="cb33-28"><a href="#cb33-28"></a><span class="co">    toc: true</span></span>
<span id="cb33-29"><a href="#cb33-29"></a><span class="co">    include-in-header:</span></span>
<span id="cb33-30"><a href="#cb33-30"></a><span class="co">      - text: |</span></span>
<span id="cb33-31"><a href="#cb33-31"></a><span class="co">          &lt;style&gt;</span></span>
<span id="cb33-32"><a href="#cb33-32"></a><span class="co">          .quarto-notebook .cell-container .cell-decorator {</span></span>
<span id="cb33-33"><a href="#cb33-33"></a><span class="co">            display: none;</span></span>
<span id="cb33-34"><a href="#cb33-34"></a><span class="co">          }</span></span>
<span id="cb33-35"><a href="#cb33-35"></a><span class="co">          &lt;/style&gt;</span></span>
<span id="cb33-36"><a href="#cb33-36"></a><span class="co">    number-sections: true</span></span>
<span id="cb33-37"><a href="#cb33-37"></a><span class="co">    code-tools: true</span></span>
<span id="cb33-38"><a href="#cb33-38"></a><span class="co">    html-math-method:</span></span>
<span id="cb33-39"><a href="#cb33-39"></a><span class="co">      method: mathjax</span></span>
<span id="cb33-40"><a href="#cb33-40"></a><span class="co">      url: "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML"</span></span>
<span id="cb33-41"><a href="#cb33-41"></a><span class="an">bibliography:</span><span class="co"> references.bib</span></span>
<span id="cb33-42"><a href="#cb33-42"></a><span class="an">lightbox:</span><span class="co"> true</span></span>
<span id="cb33-43"><a href="#cb33-43"></a><span class="an">execute:</span></span>
<span id="cb33-44"><a href="#cb33-44"></a><span class="co">  echo: true</span></span>
<span id="cb33-45"><a href="#cb33-45"></a><span class="co">  warning: false</span></span>
<span id="cb33-46"><a href="#cb33-46"></a><span class="co">---</span></span>
<span id="cb33-47"><a href="#cb33-47"></a></span>
<span id="cb33-48"><a href="#cb33-48"></a><span class="fu">## Abstract {.unnumbered}</span></span>
<span id="cb33-49"><a href="#cb33-49"></a></span>
<span id="cb33-50"><a href="#cb33-50"></a>This supplementary document provides a detailed, step-by-step tutorial on importing and preprocessing raw data from two wearable devices: <span class="in">`Clouclip`</span> and <span class="in">`VEET`</span>. We describe the structure of the raw datasets recorded by each device and explain how to parse these data, specify time zones, handle special sentinel values, clean missing observations, regularize timestamps to fixed intervals, and aggregate data as needed. All original R code from the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](index.qmd)</span> is shown here for transparency, with additional guidance intended for a broad research audience. We demonstrate how to detect gaps and irregular sampling, convert implicit missing periods into explicit missing values, and address device-specific quirks such as the <span class="in">`Clouclip`</span>’s use of sentinel codes for “sleep mode” and “out of range” readings. Special procedures for processing the <span class="in">`VEET`</span>’s rich spectral data (e.g. normalizing sensor counts and reconstructing full spectra from multiple sensor channels) are also outlined. Finally, we show how to save the cleaned datasets from both devices into a single R data file for downstream analysis. This comprehensive walkthrough is designed to ensure reproducibility and to assist researchers in understanding and adapting the data pipeline for their own visual experience datasets.</span>
<span id="cb33-51"><a href="#cb33-51"></a></span>
<span id="cb33-52"><a href="#cb33-52"></a><span class="fu">## Introduction</span></span>
<span id="cb33-53"><a href="#cb33-53"></a></span>
<span id="cb33-54"><a href="#cb33-54"></a>Wearable sensors like the <span class="in">`Clouclip`</span> and the Visual Environment Evaluation Tool (<span class="in">`VEET`</span>) produce high-dimensional time-series data on viewing distance and light exposure. Proper handling of these raw data is essential before any analysis of visual experience metrics. In the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](index.qmd)</span>, we introduce an analysis pipeline using the open-source R package LightLogR <span class="co">[</span><span class="ot">@Zauner2025JOpenSourceSoftw</span><span class="co">]</span> to calculate various distance and light exposure metrics. Here, we present a full account of the data import and preparation steps as a supplement to the methods, with an emphasis on clarity for researchers who may be less familiar with data processing in R.</span>
<span id="cb33-55"><a href="#cb33-55"></a></span>
<span id="cb33-56"><a href="#cb33-56"></a>We use example datasets from a <span class="in">`Clouclip`</span> device <span class="co">[</span><span class="ot">@Wen2021ActaOphtalmol; @Wen2020bjophthalmol</span><span class="co">]</span> and from a <span class="in">`VEET`</span> device <span class="co">[</span><span class="ot">@Sah2025OphtalmicPhysiolOpt</span><span class="co">]</span> (both provided in the accompanying <span class="co">[</span><span class="ot">repository</span><span class="co">](https://github.com/tscnlab/ZaunerEtAl_JVis_2026)</span>). The <span class="in">`Clouclip`</span> is a glasses-mounted sensor that records working distance (distance from eyes to object, in cm) and ambient illuminance (in lux) at 5-second intervals. The <span class="in">`VEET`</span> is a head-mounted multi-modal sensor that logs ambient light and spectral information (along with other data like motion and distance via a depth sensor) - in this exemplary case at 2-second intervals. A single week of continuous data illustrates the contrast in complexity: approximately 1.6&nbsp;MB for the <span class="in">`Clouclip`</span>’s simple two-column output versus up to 270&nbsp;MB for the <span class="in">`VEET`</span>’s multi-channel output (due to its higher sampling rate and richer sensor modalities).</span>
<span id="cb33-57"><a href="#cb33-57"></a></span>
<span id="cb33-58"><a href="#cb33-58"></a>In the following sections, we detail how these raw data are structured and how to import and preprocess them using LightLogR. We cover device-specific considerations such as file format quirks and sensor range limitations, as well as general best practices like handling missing timestamps and normalizing sensor readings. All code blocks can be executed in R (with the required packages loaded) to reproduce the cleaning steps. The end result will be clean, regularized datasets (<span class="in">`dataCC`</span> for Clouclip, <span class="in">`dataVEET`</span> for VEET light data, <span class="in">`dataVEET2`</span> for VEET spectral data, and <span class="in">`dataVEET3`</span> for VEET distance) ready for calculating visual experience metrics. We conclude by saving these cleaned datasets into a single file for convenient reuse.</span>
<span id="cb33-59"><a href="#cb33-59"></a></span>
<span id="cb33-60"><a href="#cb33-60"></a><span class="fu">## `Clouclip` Data: Raw Structure and Import</span></span>
<span id="cb33-61"><a href="#cb33-61"></a></span>
<span id="cb33-62"><a href="#cb33-62"></a>The <span class="in">`Clouclip`</span> device exports its data as a text file (not a true Excel file despite sometimes using an .xls extension), which is actually tab-separated values. See @fig-Clouclip_file for the raw file.</span>
<span id="cb33-63"><a href="#cb33-63"></a></span>
<span id="cb33-64"><a href="#cb33-64"></a><span class="al">![Screenshot of the first rows of the `Clouclip` export file format as seen in a text editor](assets/Clouclip_fileformat.png)</span>{#fig-Clouclip_file width=35%}</span>
<span id="cb33-65"><a href="#cb33-65"></a></span>
<span id="cb33-66"><a href="#cb33-66"></a>Each record corresponds to one timestamped observation (nominally every 5 seconds) and includes two measured variables: **distance** and **illuminance**. In the sample dataset (<span class="in">`Sample_Clouclip.csv`</span> provided), the columns are:</span>
<span id="cb33-67"><a href="#cb33-67"></a></span>
<span id="cb33-68"><a href="#cb33-68"></a><span class="ss">-   </span><span class="in">`Date`</span> – the date and time of the observation (in the device’s local time, here one week in 2021).</span>
<span id="cb33-69"><a href="#cb33-69"></a></span>
<span id="cb33-70"><a href="#cb33-70"></a><span class="ss">-   </span><span class="in">`Dis`</span> – the viewing distance in centimeters.</span>
<span id="cb33-71"><a href="#cb33-71"></a></span>
<span id="cb33-72"><a href="#cb33-72"></a><span class="ss">-   </span><span class="in">`Lux`</span> – the ambient illuminance in lux.</span>
<span id="cb33-73"><a href="#cb33-73"></a></span>
<span id="cb33-74"><a href="#cb33-74"></a>For example, a raw data line might look like:</span>
<span id="cb33-75"><a href="#cb33-75"></a></span>
<span id="cb33-76"><a href="#cb33-76"></a><span class="in">```         </span></span>
<span id="cb33-77"><a href="#cb33-77"></a><span class="in">2021-07-01 08:00:00 45  320</span></span>
<span id="cb33-78"><a href="#cb33-78"></a><span class="in">```</span></span>
<span id="cb33-79"><a href="#cb33-79"></a></span>
<span id="cb33-80"><a href="#cb33-80"></a>indicating that at *2021-07-01 08:00:00* local time, the device recorded a working distance of 45&nbsp;cm and illuminance of 320&nbsp;lx. The Clouclip uses special **sentinel values**[^1] in these measurement columns to denote certain device states. Specifically, a distance (`Dis`) value of **204** is a code meaning the object is *out of the sensor’s range*, and a value of **-1** in either `Dis` or `Lux` indicates the device was in *sleep mode* (not actively recording). During normal operation, distance measurements are limited by the device’s range, and illuminance readings are positive lux values. Any sentinel codes in the raw file need to be handled appropriately, as described below.</span>
<span id="cb33-81"><a href="#cb33-81"></a></span>
<span id="cb33-82"><a href="#cb33-82"></a><span class="ot">[^1]: </span>A *sentinel value* is a special placeholder value used in data recording to signal a particular condition. It does not represent a valid measured quantity but rather acts as a marker (for example, “device off” or “value out of range”).</span>
<span id="cb33-83"><a href="#cb33-83"></a></span>
<span id="cb33-84"><a href="#cb33-84"></a>We will use <span class="in">`LightLogR`</span>’s built-in import function for <span class="in">`Clouclip`</span>, which automatically reads the file, parses the timestamps, and converts sentinel codes into a separate status annotation. To begin, we load the necessary libraries and import the raw <span class="in">`Clouclip`</span> dataset:</span>
<span id="cb33-85"><a href="#cb33-85"></a></span>
<span id="cb33-88"><a href="#cb33-88"></a><span class="in">```{r}</span></span>
<span id="cb33-89"><a href="#cb33-89"></a><span class="co">#| label: setup</span></span>
<span id="cb33-90"><a href="#cb33-90"></a><span class="co">#| output: false</span></span>
<span id="cb33-91"><a href="#cb33-91"></a><span class="co">#| filename: Load required packages</span></span>
<span id="cb33-92"><a href="#cb33-92"></a><span class="fu">library</span>(tidyverse) <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-93"><a href="#cb33-93"></a><span class="fu">library</span>(LightLogR) <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-94"><a href="#cb33-94"></a><span class="fu">library</span>(gt) <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-95"><a href="#cb33-95"></a><span class="fu">library</span>(ggridges) <span class="co">#&lt;4&gt;</span></span>
<span id="cb33-96"><a href="#cb33-96"></a><span class="fu">library</span>(downlit) <span class="co">#&lt;5&gt;</span></span>
<span id="cb33-97"><a href="#cb33-97"></a><span class="fu">library</span>(magick) <span class="co">#&lt;5&gt;</span></span>
<span id="cb33-98"><a href="#cb33-98"></a><span class="in">```</span></span>
<span id="cb33-99"><a href="#cb33-99"></a></span>
<span id="cb33-100"><a href="#cb33-100"></a><span class="ss">1. </span>For tidy data science</span>
<span id="cb33-101"><a href="#cb33-101"></a><span class="ss">2. </span>Wearable analysis package</span>
<span id="cb33-102"><a href="#cb33-102"></a><span class="ss">3. </span>For great tables</span>
<span id="cb33-103"><a href="#cb33-103"></a><span class="ss">4. </span>For ridgeline plots</span>
<span id="cb33-104"><a href="#cb33-104"></a><span class="ss">5. </span>These packages are not used, but are needed for dependencies</span>
<span id="cb33-105"><a href="#cb33-105"></a></span>
<span id="cb33-108"><a href="#cb33-108"></a><span class="in">```{r}</span></span>
<span id="cb33-109"><a href="#cb33-109"></a><span class="co">#| label: fig-importCC</span></span>
<span id="cb33-110"><a href="#cb33-110"></a><span class="co">#| filename: Import of Clouclip data</span></span>
<span id="cb33-111"><a href="#cb33-111"></a><span class="co">#| fig-cap: "Overview plot of imported Clouclip data"</span></span>
<span id="cb33-112"><a href="#cb33-112"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb33-113"><a href="#cb33-113"></a><span class="co">#| fig-width: 6</span></span>
<span id="cb33-114"><a href="#cb33-114"></a>path <span class="ot">&lt;-</span> <span class="st">"data/Sample_Clouclip.csv"</span> <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-115"><a href="#cb33-115"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span>   <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-116"><a href="#cb33-116"></a>dataCC <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">Clouclip</span>(path, <span class="at">tz =</span> tz, <span class="at">manual.id =</span> <span class="st">"Clouclip"</span>) <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-117"><a href="#cb33-117"></a><span class="in">```</span></span>
<span id="cb33-118"><a href="#cb33-118"></a></span>
<span id="cb33-119"><a href="#cb33-119"></a><span class="ss">1. </span>Define file path</span>
<span id="cb33-120"><a href="#cb33-120"></a><span class="ss">2. </span>Time zone in which device was recording (e.g., US Central Time)</span>
<span id="cb33-121"><a href="#cb33-121"></a><span class="ss">3. </span>Import Clouclip data</span>
<span id="cb33-122"><a href="#cb33-122"></a></span>
<span id="cb33-123"><a href="#cb33-123"></a>In this code, <span class="in">`import$Clouclip()`</span> reads the tab-delimited file and returns a **tibble**<span class="ot">[^2]</span> (saved in the variable <span class="in">`dataCC`</span>) containing the data. We specify&nbsp;<span class="in">`tz = "US/Central"`</span>&nbsp;because the device’s clock was set to U.S. Central time; this ensures that the&nbsp;<span class="in">`Datetime`</span>values are properly interpreted with the correct time zone. The argument&nbsp;<span class="in">`manual.id = "Clouclip"`</span>&nbsp;simply tags the dataset with an identifier (useful if combining data from multiple devices).</span>
<span id="cb33-124"><a href="#cb33-124"></a></span>
<span id="cb33-125"><a href="#cb33-125"></a><span class="ot">[^2]: </span>**tibble** are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <span class="co">[</span><span class="ot">tibbles</span><span class="co">](https://tibble.tidyverse.org/index.html)</span></span>
<span id="cb33-126"><a href="#cb33-126"></a></span>
<span id="cb33-127"><a href="#cb33-127"></a>During import, LightLogR automatically handles the <span class="in">`Clouclip`</span>’s sentinel codes. The&nbsp;<span class="in">`Date`</span>&nbsp;column from the raw file is parsed into a <span class="in">`POSIXct`</span> date-time (<span class="in">`Datetime`</span>) with the specified time zone. The&nbsp;<span class="in">`Lux`</span>&nbsp;and&nbsp;<span class="in">`Dis`</span>&nbsp;columns are read as numeric, but any occurrences of&nbsp;**-1**&nbsp;or&nbsp;**204**&nbsp;are treated specially: these are replaced with&nbsp;<span class="in">`NA`</span>&nbsp;(missing values) in the numeric columns, and corresponding status columns&nbsp;<span class="in">`Lux_status`</span>&nbsp;and&nbsp;<span class="in">`Dis_status`</span>&nbsp;are created to indicate the reason for those <span class="in">`NA`</span> values. For example, if a&nbsp;<span class="in">`Dis`</span>&nbsp;value of *204* was encountered, that row’s&nbsp;<span class="in">`Dis`</span>&nbsp;will be set to NA and&nbsp;<span class="in">`Dis_status`</span>&nbsp;will contain&nbsp;<span class="in">`"out_of_range"`</span>; if&nbsp;<span class="in">`Lux`</span>&nbsp;or&nbsp;<span class="in">`Dis`</span>&nbsp;was -1, the status is&nbsp;<span class="in">`"sleep_mode"`</span>. We will set all other readings to &nbsp;<span class="in">`"operational"`</span> (meaning the device was functioning normally at that time) for visualisation purposes.</span>
<span id="cb33-128"><a href="#cb33-128"></a></span>
<span id="cb33-131"><a href="#cb33-131"></a><span class="in">```{r}</span></span>
<span id="cb33-132"><a href="#cb33-132"></a><span class="co">#| filename: Print the first 6 rows of the Clouclip dataset</span></span>
<span id="cb33-133"><a href="#cb33-133"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">head</span>()</span>
<span id="cb33-134"><a href="#cb33-134"></a><span class="in">```</span></span>
<span id="cb33-135"><a href="#cb33-135"></a></span>
<span id="cb33-136"><a href="#cb33-136"></a></span>
<span id="cb33-137"><a href="#cb33-137"></a>After import, it is good practice to get an overview of the data. The import function by default prints a brief summary (and generates an overview plot of the timeline) showing the number of observations, the time span, and any irregularities or large gaps. In our case, the <span class="in">`Clouclip`</span> summary indicates the data spans one week and reveals that there are&nbsp;**irregular intervals**&nbsp;in the timestamps. This means some observations do not occur exactly every 5 seconds as expected. We can programmatically check for irregular timing:</span>
<span id="cb33-138"><a href="#cb33-138"></a></span>
<span id="cb33-141"><a href="#cb33-141"></a><span class="in">```{r}</span></span>
<span id="cb33-142"><a href="#cb33-142"></a><span class="co">#| filename: Check if data are on a perfectly regular 5-second schedule</span></span>
<span id="cb33-143"><a href="#cb33-143"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">has_irregulars</span>()</span>
<span id="cb33-144"><a href="#cb33-144"></a><span class="in">```</span></span>
<span id="cb33-145"><a href="#cb33-145"></a></span>
<span id="cb33-146"><a href="#cb33-146"></a>If the result is&nbsp;<span class="in">`TRUE`</span>, it confirms that the time sequence has deviations from the regular interval. Indeed, our example dataset has many small timing irregularities and gaps (periods with no data). Understanding the pattern of these missing or irregular readings is important. We can visualize them using a gap plot:</span>
<span id="cb33-147"><a href="#cb33-147"></a></span>
<span id="cb33-150"><a href="#cb33-150"></a><span class="in">```{r}</span></span>
<span id="cb33-151"><a href="#cb33-151"></a><span class="co">#| label: fig-irregular</span></span>
<span id="cb33-152"><a href="#cb33-152"></a><span class="co">#| filename: Plot gaps and irregular timestamps for Clouclip data</span></span>
<span id="cb33-153"><a href="#cb33-153"></a><span class="co">#| fig-height: 12</span></span>
<span id="cb33-154"><a href="#cb33-154"></a><span class="co">#| warning: false</span></span>
<span id="cb33-155"><a href="#cb33-155"></a><span class="co">#| fig-cap: "Visualization of gaps and irregular data. Black traces show available data. Red shaded areas show times of missing data. Red dots show instances where observations occur off the regular interval from start to finish, i.e., irregular data."</span></span>
<span id="cb33-156"><a href="#cb33-156"></a>y.label <span class="ot">&lt;-</span> <span class="st">"Distance (cm)"</span></span>
<span id="cb33-157"><a href="#cb33-157"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">gg_gaps</span>(Dis, <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-158"><a href="#cb33-158"></a>                  <span class="at">include.implicit.gaps =</span> <span class="cn">FALSE</span>,  <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-159"><a href="#cb33-159"></a>                  <span class="at">show.irregulars =</span> <span class="cn">TRUE</span>,         <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-160"><a href="#cb33-160"></a>                  <span class="at">y.axis.label =</span> y.label,</span>
<span id="cb33-161"><a href="#cb33-161"></a>                  <span class="at">group.by.days =</span> <span class="cn">TRUE</span>, <span class="at">col.irregular =</span> <span class="fu">alpha</span>(<span class="st">"red"</span>, <span class="fl">0.03</span>)</span>
<span id="cb33-162"><a href="#cb33-162"></a>                  ) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="cn">NULL</span>)</span>
<span id="cb33-163"><a href="#cb33-163"></a><span class="in">```</span></span>
<span id="cb33-164"><a href="#cb33-164"></a></span>
<span id="cb33-165"><a href="#cb33-165"></a><span class="ss">1. </span>Basing the gap-figure on the <span class="in">`Dis`</span> distance variable</span>
<span id="cb33-166"><a href="#cb33-166"></a><span class="ss">2. </span>Only show missing observations (<span class="in">`NA`</span> values)</span>
<span id="cb33-167"><a href="#cb33-167"></a><span class="ss">3. </span>Highlight irregular timing</span>
<span id="cb33-168"><a href="#cb33-168"></a></span>
<span id="cb33-169"><a href="#cb33-169"></a>In @fig-irregular, time periods where data are missing appear as red-shaded areas, and any off-schedule observation times are marked with red dots. The <span class="in">`Clouclip`</span> example shows extensive gaps (red blocks) on certain days and irregular timing on all days except the first and last. These irregular timestamps likely arise from the device’s logging process (e.g. slight clock drift or buffering when the device was turned on/off). Such issues must be addressed before further analysis.</span>
<span id="cb33-170"><a href="#cb33-170"></a></span>
<span id="cb33-171"><a href="#cb33-171"></a>When faced with irregular or gapped data, we recommend a few strategies:</span>
<span id="cb33-172"><a href="#cb33-172"></a></span>
<span id="cb33-173"><a href="#cb33-173"></a><span class="ss">-   </span>*Remove leading/trailing segments that cause irregularity.*&nbsp;For example, if only the first day is regular and subsequent days drift, one might exclude the problematic portion using date filters (see&nbsp;<span class="co">[</span><span class="ot">filter_Date() / filter_Datetime()</span><span class="co">](https://tscnlab.github.io/LightLogR/reference/filter_Datetime.html)</span>&nbsp;in LightLogR).</span>
<span id="cb33-174"><a href="#cb33-174"></a></span>
<span id="cb33-175"><a href="#cb33-175"></a><span class="ss">-   </span>*Round timestamps to the nearest regular interval.*&nbsp;This relabels slightly off-schedule times back to the 5-second grid (using&nbsp;<span class="co">[</span><span class="ot">cut_Datetime()</span><span class="co">](https://tscnlab.github.io/LightLogR/reference/cut_Datetime.html)</span>&nbsp;with a 5-second interval), provided the deviations are small and this rounding won’t create duplicate timestamps.</span>
<span id="cb33-176"><a href="#cb33-176"></a></span>
<span id="cb33-177"><a href="#cb33-177"></a><span class="ss">-   </span>*Aggregate to a coarser time interval.*&nbsp;For instance, grouping and averaging data into 1-minute bins with&nbsp;<span class="co">[</span><span class="ot">aggregate_Datetime()</span><span class="co">](https://tscnlab.github.io/LightLogR/reference/aggregate_Datetime.html)</span> can mask irregularities at finer scales, at the cost of some temporal resolution.</span>
<span id="cb33-178"><a href="#cb33-178"></a></span>
<span id="cb33-179"><a href="#cb33-179"></a>In this case, the deviations from the 5-second schedule are relatively minor. We choose to&nbsp;**round the timestamps to the nearest 5 seconds**&nbsp;to enforce a uniform sampling grid, which simplifies downstream gap handling. We further add a separate date column for convenience:</span>
<span id="cb33-180"><a href="#cb33-180"></a></span>
<span id="cb33-183"><a href="#cb33-183"></a><span class="in">```{r}</span></span>
<span id="cb33-184"><a href="#cb33-184"></a><span class="co">#| filename: Regularize timestamps by rounding to nearest 5-second interval</span></span>
<span id="cb33-185"><a href="#cb33-185"></a>dataCC <span class="ot">&lt;-</span> dataCC <span class="sc">|&gt;</span></span>
<span id="cb33-186"><a href="#cb33-186"></a>  <span class="fu">cut_Datetime</span>(<span class="st">"5 secs"</span>, <span class="at">New.colname =</span> Datetime) <span class="sc">|&gt;</span>  <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-187"><a href="#cb33-187"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>)                     <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-188"><a href="#cb33-188"></a><span class="in">```</span></span>
<span id="cb33-189"><a href="#cb33-189"></a></span>
<span id="cb33-190"><a href="#cb33-190"></a><span class="ss">1. </span>Round times to 5-second bins</span>
<span id="cb33-191"><a href="#cb33-191"></a><span class="ss">2. </span>Add a Date column for grouping by day</span>
<span id="cb33-192"><a href="#cb33-192"></a></span>
<span id="cb33-193"><a href="#cb33-193"></a>After this operation, all&nbsp;<span class="in">`Datetime`</span>&nbsp;entries in&nbsp;<span class="in">`dataCC`</span>&nbsp;align perfectly on 5-second boundaries (e.g.&nbsp;**08:00:00**,&nbsp;**08:00:05**,&nbsp;**08:00:10**, etc.). We can verify that no irregular intervals remain by re-running&nbsp;<span class="in">`has_irregulars()`</span>:</span>
<span id="cb33-194"><a href="#cb33-194"></a></span>
<span id="cb33-197"><a href="#cb33-197"></a><span class="in">```{r}</span></span>
<span id="cb33-198"><a href="#cb33-198"></a><span class="co">#| filename: Re-check if data are on a perfectly regular 5-second schedule</span></span>
<span id="cb33-199"><a href="#cb33-199"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">has_irregulars</span>()</span>
<span id="cb33-200"><a href="#cb33-200"></a><span class="in">```</span></span>
<span id="cb33-201"><a href="#cb33-201"></a></span>
<span id="cb33-202"><a href="#cb33-202"></a>Next, we want to quantify the missing data. LightLogR distinguishes between&nbsp;**explicit missing values**&nbsp;(actual&nbsp;`NA`s in the data, possibly from sentinel replacements or gaps we have filled in) and&nbsp;**implicit missing intervals**&nbsp;(time points where the device&nbsp;*should*&nbsp;have a reading but none was recorded, and we have not yet filled them in). See @fig-gaps for a visual aid to these terms. Initially, many gaps are still implicit (between the first and last timestamp of each day). </span>
<span id="cb33-203"><a href="#cb33-203"></a></span>
<span id="cb33-204"><a href="#cb33-204"></a>::: {#fig-gaps layout-ncol=2}</span>
<span id="cb33-205"><a href="#cb33-205"></a></span>
<span id="cb33-206"><a href="#cb33-206"></a><span class="al">![Terminology of gaps and irregular data in LightLogR](assets/gap_terminology.png)</span></span>
<span id="cb33-207"><a href="#cb33-207"></a></span>
<span id="cb33-208"><a href="#cb33-208"></a><span class="al">![Handling of gaps with LightLogR's `gap_handler()` function](assets/gap_handler.png)</span></span>
<span id="cb33-209"><a href="#cb33-209"></a></span>
<span id="cb33-210"><a href="#cb33-210"></a>Gaps and irregular data</span>
<span id="cb33-211"><a href="#cb33-211"></a></span>
<span id="cb33-212"><a href="#cb33-212"></a>:::</span>
<span id="cb33-213"><a href="#cb33-213"></a></span>
<span id="cb33-214"><a href="#cb33-214"></a>We can generate a gap summary table:</span>
<span id="cb33-215"><a href="#cb33-215"></a></span>
<span id="cb33-218"><a href="#cb33-218"></a><span class="in">```{r}</span></span>
<span id="cb33-219"><a href="#cb33-219"></a><span class="co">#| label: tbl-gaps</span></span>
<span id="cb33-220"><a href="#cb33-220"></a><span class="co">#| tbl-cap: "Summary of missing and observed data for the Clouclip device"</span></span>
<span id="cb33-221"><a href="#cb33-221"></a><span class="co">#| filename: Summarize observed vs missing data by day for distance</span></span>
<span id="cb33-222"><a href="#cb33-222"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">gap_table</span>(Dis, <span class="at">Variable.label =</span> <span class="st">"Distance (cm)"</span>) <span class="sc">|&gt;</span></span>
<span id="cb33-223"><a href="#cb33-223"></a>  <span class="fu">cols_hide</span>(<span class="fu">contains</span>(<span class="st">"_n"</span>))   <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-224"><a href="#cb33-224"></a><span class="in">```</span></span>
<span id="cb33-225"><a href="#cb33-225"></a></span>
<span id="cb33-226"><a href="#cb33-226"></a><span class="ss">1. </span>Hide absolute counts for brevity in output</span>
<span id="cb33-227"><a href="#cb33-227"></a></span>
<span id="cb33-228"><a href="#cb33-228"></a>This summary (@tbl-gaps) breaks down, for each day, how much data is present vs. missing. It reports the total duration of recorded data and the duration of gaps. After rounding the times, there are no irregular timestamp issues, but we see substantial **implicit gaps** — periods where the device was not recording (e.g., overnight when the device was likely not worn or was in sleep mode). Notably, the first and last days of the week have very little data (less than 1 hour each), since they probably represent partial recording days (the trial started and ended on those days).</span>
<span id="cb33-229"><a href="#cb33-229"></a></span>
<span id="cb33-230"><a href="#cb33-230"></a>To prepare the dataset for analysis, we will convert all those implicit gaps into explicit missing entries, and remove days that are mostly incomplete. Converting implicit gaps means inserting rows with <span class="in">`NA`</span> for each missing 5-second slot, so that the time series becomes continuous and explicit about missingness (see @fig-gaps). We use <span class="in">`gap_handler()`</span> for this, and then drop the nearly-empty days:</span>
<span id="cb33-231"><a href="#cb33-231"></a></span>
<span id="cb33-234"><a href="#cb33-234"></a><span class="in">```{r}</span></span>
<span id="cb33-235"><a href="#cb33-235"></a><span class="co">#| filename: Convert implicit gaps to explicit NA gaps, and drop days with &lt;1 hour of data</span></span>
<span id="cb33-236"><a href="#cb33-236"></a>dataCC <span class="ot">&lt;-</span> dataCC <span class="sc">|&gt;</span> </span>
<span id="cb33-237"><a href="#cb33-237"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(Lux_status, Dis_status), <span class="sc">~</span> <span class="fu">replace_na</span>(.x, <span class="st">"operational"</span>))) <span class="sc">|&gt;</span> <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-238"><a href="#cb33-238"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> <span class="co">#&lt;2&gt; </span></span>
<span id="cb33-239"><a href="#cb33-239"></a>  <span class="fu">remove_partial_data</span>(Dis, <span class="at">threshold.missing =</span> <span class="st">"-1 hour"</span>) <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-240"><a href="#cb33-240"></a><span class="in">```</span></span>
<span id="cb33-241"><a href="#cb33-241"></a></span>
<span id="cb33-242"><a href="#cb33-242"></a><span class="ss">1. </span>First ensure that status columns have an "operational" tag for non-missing periods</span>
<span id="cb33-243"><a href="#cb33-243"></a><span class="ss">2. </span>Fill in all implicit gaps with explicit NA rows (for full days range)</span>
<span id="cb33-244"><a href="#cb33-244"></a><span class="ss">3. </span>Remove any day that has less than 1 hour of recorded data</span>
<span id="cb33-245"><a href="#cb33-245"></a></span>
<span id="cb33-246"><a href="#cb33-246"></a>After these steps,&nbsp;<span class="in">`dataCC`</span>&nbsp;contains continuous 5-second timestamps for each day that remains. We chose a threshold of “-1 hours” to remove days with less than one hour of data, which in this dataset removes the first and last (partial) days. The cleaned <span class="in">`Clouclip`</span> data now covers six full days with bouts of continuous wear.</span>
<span id="cb33-247"><a href="#cb33-247"></a></span>
<span id="cb33-248"><a href="#cb33-248"></a>It is often helpful to double-check how the sentinel values and missing data are distributed over time. We can visualize the distance time-series with status annotations and day/night periods:</span>
<span id="cb33-249"><a href="#cb33-249"></a></span>
<span id="cb33-252"><a href="#cb33-252"></a><span class="in">```{r}</span></span>
<span id="cb33-253"><a href="#cb33-253"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb33-254"><a href="#cb33-254"></a><span class="co">#| warning: false</span></span>
<span id="cb33-255"><a href="#cb33-255"></a><span class="co">#| filename: Visualize observations and sentinel states</span></span>
<span id="cb33-256"><a href="#cb33-256"></a><span class="co">#| label: fig-state</span></span>
<span id="cb33-257"><a href="#cb33-257"></a><span class="co">#| fig-cap: "Distance measurements across days. Blue, grey and yellow-colored bars at the bottom of each day show sentinel states of the device. Blue indicates an operational status, grey sleep mode (not recording), and yellow an out of range measurement. Shaded areas in the main plot show nighttime from civil dusk until dawn, which are calculated based on the recording date and geographic coordinates"</span></span>
<span id="cb33-258"><a href="#cb33-258"></a>coordinates <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">29.75</span>, <span class="sc">-</span><span class="fl">95.36</span>) <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-259"><a href="#cb33-259"></a>dataCC <span class="sc">|&gt;</span> </span>
<span id="cb33-260"><a href="#cb33-260"></a>  <span class="fu">fill</span>(<span class="fu">c</span>(Lux_status, Dis_status), <span class="at">.direction =</span> <span class="st">"downup"</span>) <span class="sc">|&gt;</span> <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-261"><a href="#cb33-261"></a>  <span class="fu">gg_day</span>(<span class="at">y.axis =</span> Dis, <span class="at">geom =</span> <span class="st">"line"</span>, <span class="at">y.axis.label =</span> y.label) <span class="sc">|&gt;</span> <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-262"><a href="#cb33-262"></a>  <span class="fu">gg_state</span>(Dis_status, <span class="at">aes_fill =</span> Dis_status, <span class="at">ymin =</span> <span class="dv">0</span>, <span class="at">ymax =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="dv">1</span>) <span class="sc">|&gt;</span> <span class="co">#&lt;4&gt;</span></span>
<span id="cb33-263"><a href="#cb33-263"></a>  <span class="fu">gg_photoperiod</span>(coordinates, <span class="at">alpha =</span> <span class="fl">0.1</span>) <span class="sc">+</span> <span class="co">#&lt;5&gt;</span></span>
<span id="cb33-264"><a href="#cb33-264"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb33-265"><a href="#cb33-265"></a><span class="in">```</span></span>
<span id="cb33-266"><a href="#cb33-266"></a></span>
<span id="cb33-267"><a href="#cb33-267"></a><span class="ss">1. </span>Setting coordinates for Houston, Texas (recording location)</span>
<span id="cb33-268"><a href="#cb33-268"></a><span class="ss">2. </span>Retain status markers until a new marker arrives</span>
<span id="cb33-269"><a href="#cb33-269"></a><span class="ss">3. </span>Create the basic plot</span>
<span id="cb33-270"><a href="#cb33-270"></a><span class="ss">4. </span>Add the status times</span>
<span id="cb33-271"><a href="#cb33-271"></a><span class="ss">5. </span>Add the photoperiod (day/night)</span>
<span id="cb33-272"><a href="#cb33-272"></a></span>
<span id="cb33-273"><a href="#cb33-273"></a>In this plot, blue segments indicate times when the Clouclip was&nbsp;*operational*&nbsp;(actively measuring), grey segments indicate the device in&nbsp;*sleep mode*&nbsp;(no recording, typically at night), and yellow segments indicate&nbsp;*out-of-range*&nbsp;distance readings. The red outlined regions show nighttime (from civil dusk to dawn) based on the given latitude/longitude and dates. As expected, most of the grey “sleep” periods align with night hours, and we see a few yellow spans when the user’s viewing distance exceeded the device’s range (e.g. presumably when no object was within 100&nbsp;cm, such as when the user looked into the distance). At this stage, the Clouclip dataset&nbsp;<span class="in">`dataCC`</span>&nbsp;is fully preprocessed: all timestamps are regular 5-second intervals, missing data are explicitly marked, extraneous partial days are removed, and sentinel codes are handled via the status columns. The data are ready for calculating daily distance and light exposure metrics (as done in the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](index.qmd)</span>’s Results).</span>
<span id="cb33-274"><a href="#cb33-274"></a></span>
<span id="cb33-275"><a href="#cb33-275"></a><span class="fu">## `VEET` Data: Ambient Light (Illuminance) Processing</span></span>
<span id="cb33-276"><a href="#cb33-276"></a></span>
<span id="cb33-277"><a href="#cb33-277"></a>The&nbsp;**VEET**&nbsp;device [@Sah2025OphtalmicPhysiolOpt, @Sullivan2024] is a more complex logger that records multiple data modalities in one combined file. Its raw data file contains interleaved records for different sensor types, distinguished by a “modality” field. We focus first on the ambient light sensor modality (abbreviated&nbsp;**ALS**), which provides broad-spectrum illuminance readings (lux) and related information like sensor gains and flicker, recorded every 2 seconds. Later we will import the spectral sensor modality (**PHO**) for spectral irradiance data, and the time-of-flight modality (**TOF**) for distance data.</span>
<span id="cb33-278"><a href="#cb33-278"></a></span>
<span id="cb33-279"><a href="#cb33-279"></a>In the <span class="in">`VEET`</span>’s export file, each line includes a timestamp and a modality code, followed by fields specific to that modality. Importantly, this means that the <span class="in">`VEET`</span> export is not rectangular, i.e., tabular (see @fig-VEET_file). This makes it challenging for many import functions that expect the equal number of columns in every row, which is not the case in this instance. For the&nbsp;**ALS modality**, the relevant columns include a high-resolution timestamp (in Unix epoch format), integration time, UV/VIS/IR sensor gain settings, raw UV/VIS/IR sensor counts, a flicker measurement, and the computed illuminance in lux. For example, the ALS data columns are named:&nbsp;<span class="in">`time_stamp`</span>,&nbsp;<span class="in">`integration_time`</span>,&nbsp;<span class="in">`uvGain`</span>,&nbsp;<span class="in">`visGain`</span>,&nbsp;<span class="in">`irGain`</span>,&nbsp;<span class="in">`uvValue`</span>,&nbsp;<span class="in">`visValue`</span>,&nbsp;<span class="in">`irValue`</span>,&nbsp;<span class="in">`Flicker`</span>, and&nbsp;<span class="in">`Lux`</span>.</span>
<span id="cb33-280"><a href="#cb33-280"></a></span>
<span id="cb33-281"><a href="#cb33-281"></a><span class="al">![Screenshot of the first rows of the `VEET` export file format as seen in a text editor](assets/VEET_fileformat.png)</span>{#fig-VEET_file}</span>
<span id="cb33-282"><a href="#cb33-282"></a></span>
<span id="cb33-283"><a href="#cb33-283"></a>For the&nbsp;**PHO (spectral) modality**, the columns include a timestamp, integration time, a general&nbsp;<span class="in">`Gain`</span>&nbsp;factor, and nine sensor channel readings covering different wavelengths (with names like&nbsp;<span class="in">`s415, s445, ..., s680, s940`</span>) as well as a&nbsp;<span class="in">`Dark`</span>&nbsp;channel and a broadband channel&nbsp;<span class="in">`Clear`</span>. In essence, the <span class="in">`VEET`</span>’s spectral sensor captures light in several wavelength bands (from \~415&nbsp;nm up to 940&nbsp;nm, plus an infrared and “clear” channel) rather than outputting a single lux value like the ambient light sensor does (*PHO*).</span>
<span id="cb33-284"><a href="#cb33-284"></a></span>
<span id="cb33-285"><a href="#cb33-285"></a>To import the <span class="in">`VEET`</span> ambient light data, we again use the LightLogR import function, specifying the <span class="in">`ALS`</span> modality. The raw VEET data in our example is provided as a zip file (<span class="in">`01_VEET_L.csv.zip`</span>) containing the logged data for one week. We do the following:</span>
<span id="cb33-286"><a href="#cb33-286"></a></span>
<span id="cb33-289"><a href="#cb33-289"></a><span class="in">```{r}</span></span>
<span id="cb33-290"><a href="#cb33-290"></a><span class="co">#| label: fig-VEET-overview</span></span>
<span id="cb33-291"><a href="#cb33-291"></a><span class="co">#| fig-cap: "Overview plot of imported VEET data"</span></span>
<span id="cb33-292"><a href="#cb33-292"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb33-293"><a href="#cb33-293"></a><span class="co">#| fig-width: 6</span></span>
<span id="cb33-294"><a href="#cb33-294"></a><span class="co">#| filename: Import VEET Ambient Light Sensor (ALS) data</span></span>
<span id="cb33-295"><a href="#cb33-295"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="cb33-296"><a href="#cb33-296"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span></span>
<span id="cb33-297"><a href="#cb33-297"></a>dataVEET <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"ALS"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>) <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-298"><a href="#cb33-298"></a><span class="in">```</span></span>
<span id="cb33-299"><a href="#cb33-299"></a></span>
<span id="cb33-300"><a href="#cb33-300"></a><span class="ss">1. </span>In difference to the <span class="in">`Clouclip`</span> file, we simply respecify the device type with <span class="in">`import$VEET(...)`</span>, but must also provide a <span class="in">`modality`</span> argument.</span>
<span id="cb33-301"><a href="#cb33-301"></a></span>
<span id="cb33-302"><a href="#cb33-302"></a>::: {.callout-note}</span>
<span id="cb33-303"><a href="#cb33-303"></a></span>
<span id="cb33-304"><a href="#cb33-304"></a>We get one warning as a single time stamp could not be parsed into a datetime - with \~300k observations, this is not an issue.</span>
<span id="cb33-305"><a href="#cb33-305"></a></span>
<span id="cb33-306"><a href="#cb33-306"></a>:::</span>
<span id="cb33-307"><a href="#cb33-307"></a></span>
<span id="cb33-308"><a href="#cb33-308"></a>This call reads in only the lines corresponding to the <span class="in">`ALS`</span> modality from the <span class="in">`VEET`</span> file. The result&nbsp;<span class="in">`dataVEET`</span>&nbsp;is a tibble<span class="ot">[^2]</span> with columns such as&nbsp;<span class="in">`Datetime`</span>&nbsp;(parsed from the&nbsp;<span class="in">`time_stamp`</span>&nbsp; to POSIXct in US/Central time),&nbsp;<span class="in">`Lux`</span>&nbsp;(illuminance in lux),&nbsp;<span class="in">`Flicker`</span>, and the various sensor gains/values. Columns we don't need for our analysis, like the modality code or file name, are also included but can be ignored or removed. From the import summary, we learn that the <span class="in">`VEET`</span> light data, like the <span class="in">`Clouclip`</span>, also exhibits irregularities and gaps. (The device nominally records every 2 seconds, but timing may drift or pause when not worn.)</span>
<span id="cb33-309"><a href="#cb33-309"></a></span>
<span id="cb33-312"><a href="#cb33-312"></a><span class="in">```{r}</span></span>
<span id="cb33-313"><a href="#cb33-313"></a><span class="co">#| filename: Print the first 6 rows of the VEETs light dataset</span></span>
<span id="cb33-314"><a href="#cb33-314"></a>dataVEET <span class="sc">|&gt;</span> <span class="fu">head</span>()</span>
<span id="cb33-315"><a href="#cb33-315"></a><span class="in">```</span></span>
<span id="cb33-316"><a href="#cb33-316"></a></span>
<span id="cb33-317"><a href="#cb33-317"></a></span>
<span id="cb33-318"><a href="#cb33-318"></a>To make the VEET light data comparable to the Clouclip’s and to simplify analysis, we choose to&nbsp;**aggregate the VEET illuminance data to 5-second intervals**. This slight downsampling will both reduce data volume and remove irregularities.</span>
<span id="cb33-319"><a href="#cb33-319"></a></span>
<span id="cb33-322"><a href="#cb33-322"></a><span class="in">```{r}</span></span>
<span id="cb33-323"><a href="#cb33-323"></a><span class="co">#| filename: Aggregate VEET light data to 5-second intervals and mark gaps</span></span>
<span id="cb33-324"><a href="#cb33-324"></a>dataVEET <span class="ot">&lt;-</span> dataVEET <span class="sc">|&gt;</span></span>
<span id="cb33-325"><a href="#cb33-325"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 seconds"</span>) <span class="sc">|&gt;</span>  <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-326"><a href="#cb33-326"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>          <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-327"><a href="#cb33-327"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>          <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-328"><a href="#cb33-328"></a>  <span class="fu">remove_partial_data</span>(Lux, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>) <span class="co">#&lt;4&gt;</span></span>
<span id="cb33-329"><a href="#cb33-329"></a><span class="in">```</span></span>
<span id="cb33-330"><a href="#cb33-330"></a></span>
<span id="cb33-331"><a href="#cb33-331"></a><span class="ss">1. </span>Resample to 5-sec bins (e.g. average Lux over 2-sec readings)</span>
<span id="cb33-332"><a href="#cb33-332"></a><span class="ss">2. </span>Fill in implicit gaps with NA rows</span>
<span id="cb33-333"><a href="#cb33-333"></a><span class="ss">3. </span>Add Date column for daily grouping</span>
<span id="cb33-334"><a href="#cb33-334"></a><span class="ss">4. </span>Remove participant days with more than one hour of missing data</span>
<span id="cb33-335"><a href="#cb33-335"></a></span>
<span id="cb33-336"><a href="#cb33-336"></a>First, <span class="in">`aggregate_Datetime(unit = "5 seconds")`</span> combines the high-frequency 2-second observations into 5-second slots. By default, this function will average numeric columns like&nbsp;<span class="in">`Lux`</span>&nbsp;over each 5-second period (and have sensible defaults for strings or categorical data). All of these *data type handlers* can be changed with the function call. The result is that&nbsp;<span class="in">`dataVEET`</span>&nbsp;now has a reading every 5 seconds (or an NA if none were present in that window). Next,&nbsp;<span class="in">`gap_handler(full.days = TRUE)`</span>&nbsp;inserts explicit NA entries for any 5-second timestamp that had no data within the continuous span of the recording. Then we add a&nbsp;<span class="in">`Date`</span>&nbsp;column for grouping, and finally we remove days with more than 1 hour of missing data (using a more strict criterion as we did for <span class="in">`Clouclip`</span>). According to the gap summary (@tbl-gaps2), this leaves six full days of <span class="in">`VEET`</span> light data with good coverage, after dropping the very incomplete start/end days.</span>
<span id="cb33-337"><a href="#cb33-337"></a></span>
<span id="cb33-338"><a href="#cb33-338"></a>We can inspect the missing-data summary for the <span class="in">`VEET`</span> illuminance data:</span>
<span id="cb33-339"><a href="#cb33-339"></a></span>
<span id="cb33-342"><a href="#cb33-342"></a><span class="in">```{r}</span></span>
<span id="cb33-343"><a href="#cb33-343"></a><span class="co">#| label: tbl-gaps2</span></span>
<span id="cb33-344"><a href="#cb33-344"></a><span class="co">#| tbl-cap: "Summary of missing and observed data for the VEET device, light modality"</span></span>
<span id="cb33-345"><a href="#cb33-345"></a>dataVEET <span class="sc">|&gt;</span> <span class="fu">gap_table</span>(Lux, <span class="st">"Illuminance (lx)"</span>) <span class="sc">|&gt;</span> <span class="fu">cols_hide</span>(<span class="fu">contains</span>(<span class="st">"_n"</span>))</span>
<span id="cb33-346"><a href="#cb33-346"></a><span class="in">```</span></span>
<span id="cb33-347"><a href="#cb33-347"></a></span>
<span id="cb33-348"><a href="#cb33-348"></a>@tbl-gaps2 shows, for each retained day, the total recorded duration and the duration of gaps. The <span class="in">`VEET`</span> device, like the <span class="in">`Clouclip`</span>, was not worn continuously 24&nbsp;hours per day, so there are nightly gaps of wear time (when the device was likely off the participant), but not of recordings. After our preprocessing, any implicit gaps are represented as explicit missing intervals. The <span class="in">`VEET`</span>’s time sampling was originally more frequent, but by aggregating to 5&nbsp;s we have ensured a uniform timeline akin to the <span class="in">`Clouclip`</span>’s.</span>
<span id="cb33-349"><a href="#cb33-349"></a></span>
<span id="cb33-350"><a href="#cb33-350"></a>At this point, the&nbsp;<span class="in">`dataVEET`</span>&nbsp;object (for illuminance) is cleaned and ready for computing light exposure metrics. For example, one could calculate daily mean illuminance or the duration spent above certain light thresholds (e.g. “outdoor light exposure” defined as <span class="sc">\&gt;</span>1000&nbsp;lx) using this dataset. Indeed, basic summary tables in the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](index.qmd)</span> illustrate the highly skewed nature of light exposure data and the calculation of outdoor light metrics. We will not repeat those metric calculations here in the supplement, as our focus is on data preprocessing; however, having a cleaned, gap-marked time series is crucial for those metrics to be accurate.</span>
<span id="cb33-351"><a href="#cb33-351"></a></span>
<span id="cb33-352"><a href="#cb33-352"></a><span class="fu">### Identifying non-wear times</span></span>
<span id="cb33-353"><a href="#cb33-353"></a></span>
<span id="cb33-354"><a href="#cb33-354"></a>Not all recorded time points of the <span class="in">`VEET`</span> can be considered as <span class="in">`wear time`</span>. <span class="in">`Non-wear`</span> times should generally be removed from time series, if possible. There are several ways how <span class="in">`Non-wear`</span> can be classified <span class="co">[</span><span class="ot">@Zauner2025npj</span><span class="co">]</span>:</span>
<span id="cb33-355"><a href="#cb33-355"></a></span>
<span id="cb33-356"><a href="#cb33-356"></a><span class="ss">1. </span>A device detects <span class="in">`non-wear`</span> and does not record any measurements during non-wear times, or has a variable in the exported data to denote non-wear times. The <span class="in">`Clouclip`</span> is an example here with the <span class="in">`sleep mode`</span> sentinel state</span>
<span id="cb33-357"><a href="#cb33-357"></a></span>
<span id="cb33-358"><a href="#cb33-358"></a><span class="ss">2. </span>Record non-wear separately, e.g. with a wear-log or diary. This is easy to implement, but puts a higher burden on participants and research staff, and can be error prone: forgotten, or incorrect data entries can mis-classify wear and non-wear times. For an implementation on how to remove non-wear times based on such a log, see, e.g., this <span class="co">[</span><span class="ot">tutorial</span><span class="co">](https://tscnlab.github.io/LightLogR_webinar/advanced_01_a_day_in_daylight-live.html)</span>.</span>
<span id="cb33-359"><a href="#cb33-359"></a></span>
<span id="cb33-360"><a href="#cb33-360"></a><span class="ss">3. </span>Instruct participants with a certain behavior when removing the device to ease automated detection of non-wear times in the data. These include putting a device in a black opaque bag, so that sensor readings for light are zero, or pressing an event button (if one exists). Up- and downsides are similar to the previous option, but the measures are usually harder to decipher in the data: when a buttonpress was forgotten, does that mess up the automated detection moving forward, which expects two presses per non-wear period (at beginning and at start). When there is zero lux during the day - was it in a black bag, or was this person just in a dark environment?</span>
<span id="cb33-361"><a href="#cb33-361"></a></span>
<span id="cb33-362"><a href="#cb33-362"></a><span class="ss">4. </span>Automated detection by an algorithm. There are several possibilities to try to classify non-wear times based on the available data from a wearable. E.g., if there is an activity tracker, that can be used to determine periods of inactivity, which could indicate non-wear. If only light data is available, the standard deviation (or coefficient of variance) of light in a moving window of a few minutes can be used to determine low variance, which also indicates inactivity. Usually a histogram of values shows a spike in those periods of inactivity.</span>
<span id="cb33-363"><a href="#cb33-363"></a></span>
<span id="cb33-364"><a href="#cb33-364"></a>If datasets are large enough, i.e., collect data from a long enought period, <span class="in">`non-wear`</span> has a diminishing effect, at least as daily metrics of light are concerned. E.g., in a monthlong study of 39 participants from Switzerland and Malaysia, <span class="co">[</span><span class="ot">@Biller2025</span><span class="co">]</span> found that based on a sensitivity anaylsis up to 6 hours of data could be missing **every day**, and still the daily metrics across that month would not significantly change. This assumption only holds when non-wear times are not systematic, i.e., at the same time each day.</span>
<span id="cb33-365"><a href="#cb33-365"></a></span>
<span id="cb33-366"><a href="#cb33-366"></a>We will perform a simple variant of the third option, and look at the <span class="in">`VEET`</span>'s <span class="in">`IMU`</span> modality</span>
<span id="cb33-367"><a href="#cb33-367"></a></span>
<span id="cb33-370"><a href="#cb33-370"></a><span class="in">```{r}</span></span>
<span id="cb33-371"><a href="#cb33-371"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb33-372"><a href="#cb33-372"></a><span class="co">#| fig-width: 6</span></span>
<span id="cb33-373"><a href="#cb33-373"></a><span class="co">#| filename: Import VEET activity (IMU) data</span></span>
<span id="cb33-374"><a href="#cb33-374"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="cb33-375"><a href="#cb33-375"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span></span>
<span id="cb33-376"><a href="#cb33-376"></a>dataIMU <span class="ot">&lt;-</span> </span>
<span id="cb33-377"><a href="#cb33-377"></a>  import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"IMU"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>,</span>
<span id="cb33-378"><a href="#cb33-378"></a>              <span class="at">remove_duplicates =</span> <span class="cn">TRUE</span>, <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-379"><a href="#cb33-379"></a>              <span class="at">silent =</span> <span class="cn">TRUE</span>) <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-380"><a href="#cb33-380"></a><span class="in">```</span></span>
<span id="cb33-381"><a href="#cb33-381"></a></span>
<span id="cb33-382"><a href="#cb33-382"></a><span class="ss">1. </span>Some rows in the data are duplicated. These will be removed during import</span>
<span id="cb33-383"><a href="#cb33-383"></a><span class="ss">2. </span>In this instance, we skip the import summary</span>
<span id="cb33-384"><a href="#cb33-384"></a></span>
<span id="cb33-385"><a href="#cb33-385"></a>To get a quick feeling for the data, we visualize the activity variable for the x axis in @fig-IMU</span>
<span id="cb33-386"><a href="#cb33-386"></a></span>
<span id="cb33-389"><a href="#cb33-389"></a><span class="in">```{r}</span></span>
<span id="cb33-390"><a href="#cb33-390"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb33-391"><a href="#cb33-391"></a><span class="co">#| label: fig-IMU</span></span>
<span id="cb33-392"><a href="#cb33-392"></a><span class="co">#| fig-width: 14</span></span>
<span id="cb33-393"><a href="#cb33-393"></a><span class="co">#| filename: Plotting the IMU timeline</span></span>
<span id="cb33-394"><a href="#cb33-394"></a><span class="co">#| fig-cap: "Timeline of the activity sensor (x direction)"</span></span>
<span id="cb33-395"><a href="#cb33-395"></a>dataIMU <span class="sc">|&gt;</span> </span>
<span id="cb33-396"><a href="#cb33-396"></a>  <span class="fu">gg_days</span>(ax, </span>
<span id="cb33-397"><a href="#cb33-397"></a>          <span class="at">aes_col =</span> <span class="fu">abs</span>(ax) <span class="sc">&lt;</span> <span class="dv">1</span>, </span>
<span id="cb33-398"><a href="#cb33-398"></a>          <span class="at">group =</span> <span class="fu">consecutive_id</span>(<span class="fu">abs</span>(ax) <span class="sc">&lt;</span> <span class="dv">1</span>),</span>
<span id="cb33-399"><a href="#cb33-399"></a>          <span class="at">y.axis.label =</span> <span class="st">"ax"</span></span>
<span id="cb33-400"><a href="#cb33-400"></a>          )</span>
<span id="cb33-401"><a href="#cb33-401"></a><span class="in">```</span></span>
<span id="cb33-402"><a href="#cb33-402"></a></span>
<span id="cb33-403"><a href="#cb33-403"></a><span class="in">`ax`</span> seems to vary between ±10. In general, the periods of inactivity seem to lie within a band of ±1. However, simply choosing this band does not eliminate false detections during times of high movement. Some data transformations can make these more clear:</span>
<span id="cb33-404"><a href="#cb33-404"></a></span>
<span id="cb33-407"><a href="#cb33-407"></a><span class="in">```{r}</span></span>
<span id="cb33-408"><a href="#cb33-408"></a><span class="co">#| label: fig-IMU2</span></span>
<span id="cb33-409"><a href="#cb33-409"></a><span class="co">#| fig-width: 14</span></span>
<span id="cb33-410"><a href="#cb33-410"></a><span class="co">#| filename: Plotting a transformed IMU timeline</span></span>
<span id="cb33-411"><a href="#cb33-411"></a><span class="co">#| fig-cap: "Timeline of the activity sensor (x direction) - transformed to distinguish times of low activity"</span></span>
<span id="cb33-412"><a href="#cb33-412"></a>dataIMU <span class="sc">|&gt;</span> </span>
<span id="cb33-413"><a href="#cb33-413"></a>  <span class="fu">aggregate_Datetime</span>(</span>
<span id="cb33-414"><a href="#cb33-414"></a>    <span class="st">"5 mins"</span>,</span>
<span id="cb33-415"><a href="#cb33-415"></a>    <span class="at">numeric.handler =</span> sd</span>
<span id="cb33-416"><a href="#cb33-416"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb33-417"><a href="#cb33-417"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(ax, ay, az)) <span class="sc">|&gt;</span> </span>
<span id="cb33-418"><a href="#cb33-418"></a>  <span class="fu">group_by</span>(name) <span class="sc">|&gt;</span> </span>
<span id="cb33-419"><a href="#cb33-419"></a>  <span class="fu">gg_days</span>(value, </span>
<span id="cb33-420"><a href="#cb33-420"></a>          <span class="at">aes_col =</span> value <span class="sc">&lt;</span> <span class="fl">0.05</span>,</span>
<span id="cb33-421"><a href="#cb33-421"></a>          <span class="at">group =</span> <span class="fu">consecutive_id</span>(value <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb33-422"><a href="#cb33-422"></a>          <span class="at">y.axis.label =</span> <span class="st">"activity axes"</span></span>
<span id="cb33-423"><a href="#cb33-423"></a>          )</span>
<span id="cb33-424"><a href="#cb33-424"></a><span class="in">```</span></span>
<span id="cb33-425"><a href="#cb33-425"></a></span>
<span id="cb33-426"><a href="#cb33-426"></a>What @fig-IMU2 shows is that the 5-minute standard deviation of the activity channels (x, y, and z direction) allows a pretty stable distinction between periods of high and low activity, with a threshold of **0.05**. We can us this to create a wear column. As the choice of channel does not seem to matter much, we will use <span class="in">`ax`</span>.</span>
<span id="cb33-427"><a href="#cb33-427"></a></span>
<span id="cb33-430"><a href="#cb33-430"></a><span class="in">```{r}</span></span>
<span id="cb33-431"><a href="#cb33-431"></a><span class="co">#| filename: Calculating times of wear and non-wear</span></span>
<span id="cb33-432"><a href="#cb33-432"></a>wear_data <span class="ot">&lt;-</span> </span>
<span id="cb33-433"><a href="#cb33-433"></a>dataIMU <span class="sc">|&gt;</span> </span>
<span id="cb33-434"><a href="#cb33-434"></a>  <span class="fu">aggregate_Datetime</span>(<span class="st">"5 mins"</span>,<span class="at">numeric.handler =</span> sd) <span class="sc">|&gt;</span> <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-435"><a href="#cb33-435"></a>  <span class="fu">mutate</span>(<span class="at">wear =</span> ax <span class="sc">&gt;</span> <span class="fl">0.05</span>) <span class="sc">|&gt;</span> <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-436"><a href="#cb33-436"></a>  <span class="fu">select</span>(Id, Datetime, wear) <span class="sc">|&gt;</span> <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-437"><a href="#cb33-437"></a>  <span class="fu">extract_states</span>(wear) <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-438"><a href="#cb33-438"></a>wear_data <span class="sc">|&gt;</span> <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-439"><a href="#cb33-439"></a>  <span class="fu">ungroup</span>(Id) <span class="sc">|&gt;</span> <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-440"><a href="#cb33-440"></a>  <span class="fu">summarize_numeric</span>(<span class="at">remove =</span> <span class="fu">c</span>(<span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"epoch"</span>)) <span class="sc">|&gt;</span> <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-441"><a href="#cb33-441"></a>  <span class="fu">gt</span>() <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-442"><a href="#cb33-442"></a>wear_data <span class="ot">&lt;-</span> wear_data <span class="sc">|&gt;</span> <span class="fu">select</span>(Id, wear, start, end) <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-443"><a href="#cb33-443"></a><span class="in">```</span></span>
<span id="cb33-444"><a href="#cb33-444"></a></span>
<span id="cb33-445"><a href="#cb33-445"></a><span class="ss">1. </span>Calculation of the wear-variable</span>
<span id="cb33-446"><a href="#cb33-446"></a><span class="ss">2. </span>Extracting the start and end times of wear and non-wear</span>
<span id="cb33-447"><a href="#cb33-447"></a><span class="ss">3. </span>Summarizing wear and non-wear times in a table</span>
<span id="cb33-448"><a href="#cb33-448"></a></span>
<span id="cb33-449"><a href="#cb33-449"></a>We see that most of the recorded timespan is actually non-wear (including sleep), with an average wear time of 2.6 hours.</span>
<span id="cb33-450"><a href="#cb33-450"></a></span>
<span id="cb33-451"><a href="#cb33-451"></a>Let's add this data to the <span class="in">`ALS`</span> dataset and visualize it in @fig-nonwear</span>
<span id="cb33-452"><a href="#cb33-452"></a></span>
<span id="cb33-455"><a href="#cb33-455"></a><span class="in">```{r}</span></span>
<span id="cb33-456"><a href="#cb33-456"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb33-457"><a href="#cb33-457"></a><span class="co">#| label: fig-nonwear</span></span>
<span id="cb33-458"><a href="#cb33-458"></a><span class="co">#| fig-width: 14</span></span>
<span id="cb33-459"><a href="#cb33-459"></a><span class="co">#| filename: Plotting non-wear for light</span></span>
<span id="cb33-460"><a href="#cb33-460"></a><span class="co">#| fig-cap: "Wear times for light data, shown as red bars"</span></span>
<span id="cb33-461"><a href="#cb33-461"></a>dataVEET <span class="ot">&lt;-</span> </span>
<span id="cb33-462"><a href="#cb33-462"></a>dataVEET <span class="sc">|&gt;</span> </span>
<span id="cb33-463"><a href="#cb33-463"></a>  <span class="fu">group_by</span>(Id) <span class="sc">|&gt;</span> </span>
<span id="cb33-464"><a href="#cb33-464"></a>  <span class="fu">add_states</span>(wear_data)</span>
<span id="cb33-465"><a href="#cb33-465"></a></span>
<span id="cb33-466"><a href="#cb33-466"></a>dataVEET <span class="sc">|&gt;</span> </span>
<span id="cb33-467"><a href="#cb33-467"></a>  <span class="fu">gg_days</span>(Lux) <span class="sc">|&gt;</span> </span>
<span id="cb33-468"><a href="#cb33-468"></a>  <span class="fu">gg_states</span>(wear, <span class="at">ymax =</span> <span class="dv">0</span>, <span class="at">alpha =</span> <span class="dv">1</span>, <span class="at">fill =</span> <span class="st">"red"</span>)</span>
<span id="cb33-469"><a href="#cb33-469"></a><span class="in">```</span></span>
<span id="cb33-470"><a href="#cb33-470"></a></span>
<span id="cb33-471"><a href="#cb33-471"></a>This looks sensible, e.g., when we look at noon of 7 June 2024. We can remove these observations based on the <span class="in">`wear`</span> column to make our summaries more valid.</span>
<span id="cb33-472"><a href="#cb33-472"></a></span>
<span id="cb33-475"><a href="#cb33-475"></a><span class="in">```{r}</span></span>
<span id="cb33-476"><a href="#cb33-476"></a><span class="co">#| filename: Removing non-wear observations from light modality</span></span>
<span id="cb33-477"><a href="#cb33-477"></a><span class="co">#| warning: false</span></span>
<span id="cb33-478"><a href="#cb33-478"></a>dataVEET <span class="ot">&lt;-</span> dataVEET <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">Lux =</span> <span class="fu">ifelse</span>(wear, Lux, <span class="cn">NA</span>)) <span class="sc">|&gt;</span> <span class="fu">group_by</span>(Id, Date)</span>
<span id="cb33-479"><a href="#cb33-479"></a><span class="in">```</span></span>
<span id="cb33-480"><a href="#cb33-480"></a></span>
<span id="cb33-481"><a href="#cb33-481"></a><span class="fu">## `VEET` Data: Spectral Data Processing</span></span>
<span id="cb33-482"><a href="#cb33-482"></a></span>
<span id="cb33-483"><a href="#cb33-483"></a><span class="fu">### Import</span></span>
<span id="cb33-484"><a href="#cb33-484"></a></span>
<span id="cb33-485"><a href="#cb33-485"></a>In addition to broad-band illuminance and distance, the <span class="in">`VEET`</span> provides&nbsp;**spectral sensor data**&nbsp;through its <span class="in">`PHO`</span> modality. Unlike illuminance, the spectral data are not given as directly interpretable radiometric metrics but rather as raw sensor counts across multiple wavelength channels, which require conversion to reconstruct a spectral power distribution. In our analysis, spectral data allow us to compute metrics like the relative contribution of short-wavelength (blue) light versus long-wavelength light in the participant’s environment. Processing this spectral data involves several necessary steps.</span>
<span id="cb33-486"><a href="#cb33-486"></a></span>
<span id="cb33-487"><a href="#cb33-487"></a>First, we import the spectral modality from a second <span class="in">`VEET`</span> file. This time we need to extract the lines marked as <span class="in">`PHO`</span>. We will store the spectral dataset in a separate object&nbsp;<span class="in">`dataVEET2`</span>&nbsp;so as not to overwrite the&nbsp;<span class="in">`dataVEET`</span>&nbsp;illuminance data in our R session:</span>
<span id="cb33-488"><a href="#cb33-488"></a></span>
<span id="cb33-491"><a href="#cb33-491"></a><span class="in">```{r}</span></span>
<span id="cb33-492"><a href="#cb33-492"></a><span class="co">#| label: fig-VEET-overview2</span></span>
<span id="cb33-493"><a href="#cb33-493"></a><span class="co">#| fig-cap: "Overview plot of imported VEET data"</span></span>
<span id="cb33-494"><a href="#cb33-494"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb33-495"><a href="#cb33-495"></a><span class="co">#| fig-width: 6</span></span>
<span id="cb33-496"><a href="#cb33-496"></a><span class="co">#| filename: Import VEET Spectral Sensor (PHO) data</span></span>
<span id="cb33-497"><a href="#cb33-497"></a>path <span class="ot">&lt;-</span> <span class="st">"data/02_VEET_L.csv.zip"</span></span>
<span id="cb33-498"><a href="#cb33-498"></a>dataVEET2 <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"PHO"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>)</span>
<span id="cb33-499"><a href="#cb33-499"></a><span class="in">```</span></span>
<span id="cb33-500"><a href="#cb33-500"></a></span>
<span id="cb33-503"><a href="#cb33-503"></a><span class="in">```{r}</span></span>
<span id="cb33-504"><a href="#cb33-504"></a><span class="co">#| filename: Print the first 6 rows of the VEETs spectral dataset</span></span>
<span id="cb33-505"><a href="#cb33-505"></a>dataVEET2 <span class="sc">|&gt;</span> <span class="fu">head</span>()</span>
<span id="cb33-506"><a href="#cb33-506"></a><span class="in">```</span></span>
<span id="cb33-507"><a href="#cb33-507"></a></span>
<span id="cb33-508"><a href="#cb33-508"></a></span>
<span id="cb33-509"><a href="#cb33-509"></a>After import,&nbsp;<span class="in">`data`</span>&nbsp;contains columns for the timestamp (<span class="in">`Datetime`</span>),&nbsp;<span class="in">`Gain`</span>&nbsp;(the sensor gain setting), and the nine spectral sensor channels plus a clear channel. These appear as numeric columns named&nbsp;<span class="in">`s415, s445, ..., s940, Dark, Clear`</span>. Other columns are also present but not needed for now. The spectral sensor was logging at a 2-second rate. It is informative to look at a snippet of the imported spectral data before further processing. @tbl-PHO shows three rows of&nbsp;<span class="in">`data`</span>&nbsp;after import (before calibration), with some technical columns omitted for brevity:</span>
<span id="cb33-510"><a href="#cb33-510"></a></span>
<span id="cb33-513"><a href="#cb33-513"></a><span class="in">```{r}</span></span>
<span id="cb33-514"><a href="#cb33-514"></a><span class="co">#| label: tbl-PHO</span></span>
<span id="cb33-515"><a href="#cb33-515"></a><span class="co">#| filename: Table overview of spectral sensor data</span></span>
<span id="cb33-516"><a href="#cb33-516"></a><span class="co">#| tbl-cap: "Overview of the spectral sensor import from the VEET device (3 observations). Each row corresponds to a 2-second timestamp (Datetime) and shows the raw sensor readings for the spectral channels (s415–s940, Dark, Clear). All values are in arbitrary sensor units (counts). Gain values and integration_time are also relevant for each interval, depending on the downstream computation."</span></span>
<span id="cb33-517"><a href="#cb33-517"></a>dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="cb33-518"><a href="#cb33-518"></a>  <span class="fu">slice</span>(<span class="dv">6000</span><span class="sc">:</span><span class="dv">6003</span>) <span class="sc">|&gt;</span> </span>
<span id="cb33-519"><a href="#cb33-519"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(modality, file.name, time_stamp)) <span class="sc">|&gt;</span> </span>
<span id="cb33-520"><a href="#cb33-520"></a>  <span class="fu">gt</span>() <span class="sc">|&gt;</span> </span>
<span id="cb33-521"><a href="#cb33-521"></a>  <span class="fu">fmt_number</span>(s415<span class="sc">:</span>Clear) </span>
<span id="cb33-522"><a href="#cb33-522"></a><span class="in">```</span></span>
<span id="cb33-523"><a href="#cb33-523"></a></span>
<span id="cb33-524"><a href="#cb33-524"></a><span class="fu">### Spectral calibration</span></span>
<span id="cb33-525"><a href="#cb33-525"></a></span>
<span id="cb33-526"><a href="#cb33-526"></a>Now we proceed with&nbsp;**spectral calibration**. The <span class="in">`VEET`</span>’s spectral sensor counts need to be converted to physical units (spectral irradiance) via a calibration matrix provided by the manufacturer. For this example, we assume we have a calibration matrix that maps all the channel readings to an estimated spectral power distribution (SPD). The LightLogR package provides a function&nbsp;<span class="in">`spectral_reconstruction()`</span>&nbsp;to perform this conversion. However, before applying it, we must ensure the sensor counts are in a normalized form. This procedure is laid out by the manufacturer. In our version, we refer to the <span class="co">[</span><span class="ot">*VEET SPD Reconstruction Guide.pdf*</span><span class="co">](https://projectveet.com/wp-content/uploads/VEET-Spectral-Power-Distribution-Reconstruction-Guide.pdf)</span>{target="_blank"}, version *06/05/2025*. Note that each manufacturer has to specify the method of count normalization (if any) and spectral reconstruction. In our raw data, each observation comes with a&nbsp;<span class="in">`Gain`</span>&nbsp;setting that indicates how the sensor’s sensitivity was adjusted; we need to divide the raw counts by the gain to get normalized counts. LightLogR offers&nbsp;<span class="in">`normalize_counts()`</span>&nbsp;for this purpose. We further need to scale by integration time (in milliseconds) and adjust depending on counts in the <span class="in">`Dark`</span> sensor channel.</span>
<span id="cb33-527"><a href="#cb33-527"></a></span>
<span id="cb33-530"><a href="#cb33-530"></a><span class="in">```{r}</span></span>
<span id="cb33-531"><a href="#cb33-531"></a><span class="co">#| filename: Normalize spectral sensor counts</span></span>
<span id="cb33-532"><a href="#cb33-532"></a>count.columns <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"s415"</span>, <span class="st">"s445"</span>, <span class="st">"s480"</span>, <span class="st">"s515"</span>, <span class="st">"s555"</span>, <span class="st">"s590"</span>, <span class="st">"s630"</span>, <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-533"><a href="#cb33-533"></a>                      <span class="st">"s680"</span>, <span class="st">"s910"</span>, <span class="st">"Dark"</span>, <span class="st">"Clear"</span>) <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-534"><a href="#cb33-534"></a></span>
<span id="cb33-535"><a href="#cb33-535"></a>gain.ratios <span class="ot">&lt;-</span> <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-536"><a href="#cb33-536"></a>  <span class="fu">tibble</span>( <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-537"><a href="#cb33-537"></a>    <span class="at">gain =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">8</span>, <span class="dv">16</span>, <span class="dv">32</span>, <span class="dv">64</span>, <span class="dv">128</span>, <span class="dv">256</span>, <span class="dv">512</span>), <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-538"><a href="#cb33-538"></a>    <span class="at">gain.ratio =</span> <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-539"><a href="#cb33-539"></a>      <span class="fu">c</span>(<span class="fl">0.008</span>, <span class="fl">0.016</span>, <span class="fl">0.032</span>, <span class="fl">0.065</span>, <span class="fl">0.125</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="fl">3.95</span>, <span class="fl">7.75</span>) <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-540"><a href="#cb33-540"></a>  ) <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-541"><a href="#cb33-541"></a></span>
<span id="cb33-542"><a href="#cb33-542"></a><span class="co">#normalize data:</span></span>
<span id="cb33-543"><a href="#cb33-543"></a>dataVEET2 <span class="ot">&lt;-</span></span>
<span id="cb33-544"><a href="#cb33-544"></a>  dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="cb33-545"><a href="#cb33-545"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(s415<span class="sc">:</span>Clear), \(x) (x <span class="sc">-</span> Dark)<span class="sc">/</span>integration_time)) <span class="sc">|&gt;</span>  <span class="co">#&lt;3&gt; </span></span>
<span id="cb33-546"><a href="#cb33-546"></a>  <span class="fu">normalize_counts</span>( <span class="co">#&lt;4&gt;</span></span>
<span id="cb33-547"><a href="#cb33-547"></a>    <span class="at">gain.columns =</span> <span class="fu">rep</span>(<span class="st">"Gain"</span>, <span class="dv">11</span>), <span class="co">#&lt;5&gt; </span></span>
<span id="cb33-548"><a href="#cb33-548"></a>    <span class="at">count.columns =</span> count.columns, <span class="co">#&lt;6&gt;</span></span>
<span id="cb33-549"><a href="#cb33-549"></a>    gain.ratios <span class="co">#&lt;7&gt;</span></span>
<span id="cb33-550"><a href="#cb33-550"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb33-551"><a href="#cb33-551"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(s415<span class="sc">:</span>Clear)) <span class="sc">|&gt;</span> <span class="co">#&lt;8&gt;</span></span>
<span id="cb33-552"><a href="#cb33-552"></a>  <span class="fu">rename_with</span>(<span class="sc">~</span> <span class="fu">str_remove</span>(.x, <span class="st">".normalized"</span>)) <span class="co">#&lt;8&gt;</span></span>
<span id="cb33-553"><a href="#cb33-553"></a><span class="in">```</span></span>
<span id="cb33-554"><a href="#cb33-554"></a></span>
<span id="cb33-555"><a href="#cb33-555"></a><span class="ss">1.  </span>Column names of variables that need to be normalized</span>
<span id="cb33-556"><a href="#cb33-556"></a><span class="ss">2.  </span>Gain ratios as specified by the manufacturer's reconstruction guide</span>
<span id="cb33-557"><a href="#cb33-557"></a><span class="ss">3.  </span>Remove dark counts &amp; scale by integration time</span>
<span id="cb33-558"><a href="#cb33-558"></a><span class="ss">4.  </span>Function to normalize counts</span>
<span id="cb33-559"><a href="#cb33-559"></a><span class="ss">5.  </span>All sensor channels share the gain value</span>
<span id="cb33-560"><a href="#cb33-560"></a><span class="ss">6.  </span>Sensor channels to normalize (see 1.)</span>
<span id="cb33-561"><a href="#cb33-561"></a><span class="ss">7.  </span>Gain ratios (see 2.)</span>
<span id="cb33-562"><a href="#cb33-562"></a><span class="ss">8.  </span>Drop original raw count columns</span>
<span id="cb33-563"><a href="#cb33-563"></a></span>
<span id="cb33-564"><a href="#cb33-564"></a>In this call, we specified <span class="in">`gain.columns = rep("Gain", 11)`</span> because we have 11 sensor columns that all use the same gain factor column (<span class="in">`Gain`</span>). This step will add new columns (with a suffix, e.g. <span class="in">`.normalized`</span>) for each spectral channel representing the count normalized by the gain. We then dropped the <span class="in">`raw`</span> count columns and renamed the normalized ones by dropping <span class="in">`.normalized`</span> from the names. After this, <span class="in">`dataVEET2`</span> contains the normalized sensor readings for <span class="in">`s415, s445, ..., s940, Dark, Clear`</span> for each time point time point.</span>
<span id="cb33-565"><a href="#cb33-565"></a></span>
<span id="cb33-566"><a href="#cb33-566"></a>Because we do not need this high a resolution, we will&nbsp;**aggregate it to a 5-minute interval** for computational efficiency. The assumption is that spectral composition does not need to be examined at every 2-second instant for our purposes, and 5-minute averages can capture the general trends while drastically reducing data size and downstream computational costs.</span>
<span id="cb33-567"><a href="#cb33-567"></a></span>
<span id="cb33-570"><a href="#cb33-570"></a><span class="in">```{r}</span></span>
<span id="cb33-571"><a href="#cb33-571"></a><span class="co">#| filename: Aggregate spectral data to 5-minute intervals and mark gaps</span></span>
<span id="cb33-572"><a href="#cb33-572"></a>dataVEET2 <span class="ot">&lt;-</span> dataVEET2 <span class="sc">|&gt;</span></span>
<span id="cb33-573"><a href="#cb33-573"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 mins"</span>) <span class="sc">|&gt;</span>     <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-574"><a href="#cb33-574"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>           <span class="co">#&lt;2&gt; </span></span>
<span id="cb33-575"><a href="#cb33-575"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>           <span class="co">#&lt;3&gt; </span></span>
<span id="cb33-576"><a href="#cb33-576"></a>  <span class="fu">remove_partial_data</span>(Clear, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>) <span class="co">#&lt;4&gt; </span></span>
<span id="cb33-577"><a href="#cb33-577"></a><span class="in">```</span></span>
<span id="cb33-578"><a href="#cb33-578"></a></span>
<span id="cb33-579"><a href="#cb33-579"></a><span class="ss">1. </span>Aggregate to 5-minute bins</span>
<span id="cb33-580"><a href="#cb33-580"></a><span class="ss">2. </span>Explicit NA for any gaps in between</span>
<span id="cb33-581"><a href="#cb33-581"></a><span class="ss">3. </span>Add a date identifier for grouping</span>
<span id="cb33-582"><a href="#cb33-582"></a><span class="ss">4. </span>remove days with more than one hour of data missing</span>
<span id="cb33-583"><a href="#cb33-583"></a></span>
<span id="cb33-584"><a href="#cb33-584"></a>We aggregate over 5-minute windows; within each 5-minute bin, multiple spectral readings (if present) are combined (averaged). We use one of the channels (here&nbsp;<span class="in">`Clear`</span>) as the reference variable for&nbsp;<span class="in">`remove_partial_data`</span>&nbsp;to drop incomplete days (the choice of channel is arbitrary as all channels share the same level of completeness).</span>
<span id="cb33-585"><a href="#cb33-585"></a></span>
<span id="cb33-586"><a href="#cb33-586"></a>::: callout-warning</span>
<span id="cb33-587"><a href="#cb33-587"></a>Please note that <span class="in">`normalize_counts()`</span> requires the <span class="in">`Gain`</span> values according to the gain table. If we had aggregated the data before normalizing it, <span class="in">`Gain`</span> values would have been averaged within each bin (5 minutes in this case). If the <span class="in">`Gain`</span> did not change in that time, it is not an issue. Any mix of <span class="in">`Gain`</span> values will lead to a <span class="in">`Gain`</span> value that is not represented in the gain table. While outputs for <span class="in">`normalize_counts()`</span> are not wrong in these cases, they will output <span class="in">`NA`</span> if a <span class="in">`Gain`</span> value is not found in the table. Thus we recommend to always normalize counts based on the raw dataset.</span>
<span id="cb33-588"><a href="#cb33-588"></a>:::</span>
<span id="cb33-589"><a href="#cb33-589"></a></span>
<span id="cb33-590"><a href="#cb33-590"></a><span class="fu">### Spectral reconstruction</span></span>
<span id="cb33-591"><a href="#cb33-591"></a></span>
<span id="cb33-592"><a href="#cb33-592"></a>For spectral reconstruction, we require a calibration matrix that corresponds to the <span class="in">`VEET`</span>’s sensor channels. This matrix would typically be obtained from the device manufacturer or a calibration procedure. It defines how each channel’s normalized count relates to intensity at various wavelengths. For demonstration, the calibration matrix was provided by the manufacturer and is specific to the make and model (see @fig-calibmtx). It should not be used for research purposes without confirming its accuracy with the manufacturer.</span>
<span id="cb33-593"><a href="#cb33-593"></a></span>
<span id="cb33-594"><a href="#cb33-594"></a><span class="al">![Calibration matrix](assets/calibration_matrix.png)</span>{#fig-calibmtx}</span>
<span id="cb33-595"><a href="#cb33-595"></a></span>
<span id="cb33-598"><a href="#cb33-598"></a><span class="in">```{r}</span></span>
<span id="cb33-599"><a href="#cb33-599"></a><span class="co">#| filename: Calibration matrica and reconstruction of spectral power distribution</span></span>
<span id="cb33-600"><a href="#cb33-600"></a>calib_mtx <span class="ot">&lt;-</span> <span class="co"># &lt;1&gt;</span></span>
<span id="cb33-601"><a href="#cb33-601"></a>  <span class="fu">read_csv</span>(<span class="st">"data/VEET_calibration_matrix.csv"</span>, <span class="co"># &lt;1&gt;</span></span>
<span id="cb33-602"><a href="#cb33-602"></a>           <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">|&gt;</span> <span class="co"># &lt;1&gt;</span></span>
<span id="cb33-603"><a href="#cb33-603"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"wavelength"</span>) <span class="co"># &lt;1&gt;</span></span>
<span id="cb33-604"><a href="#cb33-604"></a></span>
<span id="cb33-605"><a href="#cb33-605"></a>dataVEET2 <span class="ot">&lt;-</span></span>
<span id="cb33-606"><a href="#cb33-606"></a>  dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="cb33-607"><a href="#cb33-607"></a>  <span class="fu">mutate</span>(</span>
<span id="cb33-608"><a href="#cb33-608"></a>  <span class="at">Spectrum =</span> </span>
<span id="cb33-609"><a href="#cb33-609"></a>    <span class="fu">spectral_reconstruction</span>( <span class="co"># &lt;2&gt;</span></span>
<span id="cb33-610"><a href="#cb33-610"></a>      <span class="fu">pick</span>(s415<span class="sc">:</span>s910),   <span class="co"># &lt;3&gt;</span></span>
<span id="cb33-611"><a href="#cb33-611"></a>      <span class="at">calibration_matrix =</span> calib_mtx, <span class="co">#&lt;4&gt;</span></span>
<span id="cb33-612"><a href="#cb33-612"></a>      <span class="at">format =</span> <span class="st">"long"</span> <span class="co"># &lt;5&gt;</span></span>
<span id="cb33-613"><a href="#cb33-613"></a>  )</span>
<span id="cb33-614"><a href="#cb33-614"></a>)</span>
<span id="cb33-615"><a href="#cb33-615"></a><span class="in">```</span></span>
<span id="cb33-616"><a href="#cb33-616"></a></span>
<span id="cb33-617"><a href="#cb33-617"></a><span class="ss">1. </span>Import the calibration matrix and make certain <span class="in">`wavelength`</span> is set a rownames</span>
<span id="cb33-618"><a href="#cb33-618"></a><span class="ss">2.  </span>The function <span class="in">`spectral_reconstruction()`</span> does not work on the level of the dataset, but has to be called within <span class="in">`mutate()`</span>(or provided the data directly)</span>
<span id="cb33-619"><a href="#cb33-619"></a><span class="ss">3.  </span>Pick the normalized sensor columns</span>
<span id="cb33-620"><a href="#cb33-620"></a><span class="ss">4.  </span>Provide the calibration matrix</span>
<span id="cb33-621"><a href="#cb33-621"></a><span class="ss">5.  </span>Return a long-form list column (wavelength, intensity)</span>
<span id="cb33-622"><a href="#cb33-622"></a></span>
<span id="cb33-623"><a href="#cb33-623"></a>Here, we use&nbsp;<span class="in">`format = "long"`</span>&nbsp;so that the result for each observation is a&nbsp;**list-column**&nbsp;<span class="in">`Spectrum`</span>, where each entry is a tibble<span class="ot">[^2]</span> containing two columns:&nbsp;<span class="in">`wavelength`</span>&nbsp;and&nbsp;<span class="in">`irradiance`</span>&nbsp;(one row per wavelength in the calibration matrix). In other words, each row of&nbsp;<span class="in">`dataVEET2`</span>&nbsp;now holds a full reconstructed spectrum in the&nbsp;<span class="in">`Spectrum`</span>&nbsp;column. The long format is convenient for further calculations and plotting. (An alternative&nbsp;<span class="in">`format = "wide"`</span>&nbsp;would add each wavelength as a separate column, but that is less practical when there are many wavelengths.)</span>
<span id="cb33-624"><a href="#cb33-624"></a></span>
<span id="cb33-625"><a href="#cb33-625"></a>To visualize the data we will calculate the photopic illuminance based on the spectra and plot each spectrum color-scaled by their illuminance. For clarity, we reduce the data to observations within the day that has the most observations (non-<span class="in">`NA`</span>).</span>
<span id="cb33-626"><a href="#cb33-626"></a></span>
<span id="cb33-629"><a href="#cb33-629"></a><span class="in">```{r}</span></span>
<span id="cb33-630"><a href="#cb33-630"></a><span class="co">#| filename: Calculate photopic illuminance</span></span>
<span id="cb33-631"><a href="#cb33-631"></a>data_spectra <span class="ot">&lt;-</span> </span>
<span id="cb33-632"><a href="#cb33-632"></a>dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="cb33-633"><a href="#cb33-633"></a>  <span class="fu">sample_groups</span>(<span class="at">order.by =</span> <span class="fu">sum</span>(<span class="sc">!</span>is.implicit)) <span class="sc">|&gt;</span> <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-634"><a href="#cb33-634"></a>  <span class="fu">mutate</span>( </span>
<span id="cb33-635"><a href="#cb33-635"></a>    <span class="at">Illuminance =</span> Spectrum <span class="sc">|&gt;</span> <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-636"><a href="#cb33-636"></a>      <span class="fu">map_dbl</span>(spectral_integration, <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-637"><a href="#cb33-637"></a>              <span class="at">action.spectrum =</span> <span class="st">"photopic"</span>, <span class="co">#&lt;4&gt;</span></span>
<span id="cb33-638"><a href="#cb33-638"></a>              <span class="at">general.weight =</span> <span class="st">"auto"</span>) <span class="co">#&lt;5&gt;</span></span>
<span id="cb33-639"><a href="#cb33-639"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb33-640"><a href="#cb33-640"></a>  <span class="fu">unnest</span>(Spectrum) <span class="co">#&lt;6&gt;</span></span>
<span id="cb33-641"><a href="#cb33-641"></a>data_spectra <span class="sc">|&gt;</span> <span class="fu">select</span>(Id, Date, Datetime, Illuminance) <span class="sc">|&gt;</span> <span class="fu">distinct</span>()</span>
<span id="cb33-642"><a href="#cb33-642"></a><span class="in">```</span></span>
<span id="cb33-643"><a href="#cb33-643"></a></span>
<span id="cb33-644"><a href="#cb33-644"></a><span class="ss">1.  </span>Keep only observations for one day (with the lowest missing intervals)</span>
<span id="cb33-645"><a href="#cb33-645"></a><span class="ss">2.  </span>Use the spectrum,...</span>
<span id="cb33-646"><a href="#cb33-646"></a><span class="ss">3.  </span>... call the function <span class="in">`spectral_integration()`</span> for each,...</span>
<span id="cb33-647"><a href="#cb33-647"></a><span class="ss">4.  </span>... use the brightness sensitivity function,...</span>
<span id="cb33-648"><a href="#cb33-648"></a><span class="ss">5.  </span>... and apply the appropriate efficacy weight.</span>
<span id="cb33-649"><a href="#cb33-649"></a><span class="ss">6.  </span>Create a long format of the data where the spectrum is unnested</span>
<span id="cb33-650"><a href="#cb33-650"></a></span>
<span id="cb33-651"><a href="#cb33-651"></a>The following plot visualizes the spectra:</span>
<span id="cb33-652"><a href="#cb33-652"></a></span>
<span id="cb33-655"><a href="#cb33-655"></a><span class="in">```{r}</span></span>
<span id="cb33-656"><a href="#cb33-656"></a><span class="co">#| label: fig-VEET-spectra-intensity</span></span>
<span id="cb33-657"><a href="#cb33-657"></a><span class="co">#| fig-cap: "Overview of the reconstructed spectra, color-scaled by photopic illuminance (lx)"</span></span>
<span id="cb33-658"><a href="#cb33-658"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb33-659"><a href="#cb33-659"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb33-660"><a href="#cb33-660"></a><span class="co">#| filename: Plot spectra</span></span>
<span id="cb33-661"><a href="#cb33-661"></a>data_spectra <span class="sc">|&gt;</span> </span>
<span id="cb33-662"><a href="#cb33-662"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> wavelength,<span class="at">group =</span> Datetime)) <span class="sc">+</span></span>
<span id="cb33-663"><a href="#cb33-663"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> irradiance<span class="sc">*</span><span class="dv">1000</span>, <span class="at">col =</span> Illuminance)) <span class="sc">+</span></span>
<span id="cb33-664"><a href="#cb33-664"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Irradiance (mW/m²/nm)"</span>, </span>
<span id="cb33-665"><a href="#cb33-665"></a>       <span class="at">x =</span> <span class="st">"Wavelength (nm)"</span>, </span>
<span id="cb33-666"><a href="#cb33-666"></a>       <span class="at">col =</span> <span class="st">"Photopic illuminance (lx)"</span>) <span class="sc">+</span></span>
<span id="cb33-667"><a href="#cb33-667"></a>  <span class="fu">scale_color_viridis_b</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span><span class="sc">^</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">3</span>))) <span class="sc">+</span></span>
<span id="cb33-668"><a href="#cb33-668"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">trans =</span> <span class="st">"symlog"</span>, <span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">50</span>)) <span class="sc">+</span></span>
<span id="cb33-669"><a href="#cb33-669"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="cn">NA</span>), <span class="at">expand =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb33-670"><a href="#cb33-670"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb33-671"><a href="#cb33-671"></a><span class="in">```</span></span>
<span id="cb33-672"><a href="#cb33-672"></a></span>
<span id="cb33-673"><a href="#cb33-673"></a>The following ridgeline plot can be used to assess **when** in the day certain spectral wavelenghts are dominant:</span>
<span id="cb33-674"><a href="#cb33-674"></a></span>
<span id="cb33-677"><a href="#cb33-677"></a><span class="in">```{r}</span></span>
<span id="cb33-678"><a href="#cb33-678"></a><span class="co">#| label: fig-VEET-spectra-time</span></span>
<span id="cb33-679"><a href="#cb33-679"></a><span class="co">#| fig-cap: "Overview of the reconstructed spectra by time of day, color-scaled by photopic illuminance (lx)"</span></span>
<span id="cb33-680"><a href="#cb33-680"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb33-681"><a href="#cb33-681"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb33-682"><a href="#cb33-682"></a><span class="co">#| filename: Plot spectra across the time of day</span></span>
<span id="cb33-683"><a href="#cb33-683"></a>data_spectra <span class="sc">|&gt;</span> </span>
<span id="cb33-684"><a href="#cb33-684"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> wavelength, <span class="at">group =</span> Datetime)) <span class="sc">+</span></span>
<span id="cb33-685"><a href="#cb33-685"></a>  <span class="fu">geom_ridgeline</span>(<span class="fu">aes</span>(<span class="at">height =</span> irradiance<span class="sc">*</span><span class="dv">1000</span>, </span>
<span id="cb33-686"><a href="#cb33-686"></a>                     <span class="at">y =</span> Datetime, </span>
<span id="cb33-687"><a href="#cb33-687"></a>                     <span class="at">fill =</span> Illuminance), </span>
<span id="cb33-688"><a href="#cb33-688"></a>                 <span class="at">scale =</span> <span class="dv">400</span>, <span class="at">lwd =</span> <span class="fl">0.1</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb33-689"><a href="#cb33-689"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Local time &amp; Irradiance (mW/m²/nm)"</span>, </span>
<span id="cb33-690"><a href="#cb33-690"></a>       <span class="at">x =</span> <span class="st">"Wavelength (nm)"</span>, </span>
<span id="cb33-691"><a href="#cb33-691"></a>       <span class="at">fill =</span> <span class="st">"Photopic illuminance (lx)"</span>)<span class="sc">+</span></span>
<span id="cb33-692"><a href="#cb33-692"></a>  <span class="fu">scale_fill_viridis_b</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span><span class="sc">^</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">3</span>))) <span class="sc">+</span></span>
<span id="cb33-693"><a href="#cb33-693"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb33-694"><a href="#cb33-694"></a><span class="in">```</span></span>
<span id="cb33-695"><a href="#cb33-695"></a></span>
<span id="cb33-696"><a href="#cb33-696"></a></span>
<span id="cb33-697"><a href="#cb33-697"></a>At this stage, the&nbsp;<span class="in">`dataVEET2`</span>&nbsp;dataset has been processed to yield time-series of spectral power distributions. We can use these to compute biologically relevant light metrics. For instance, one possible metric is the proportion of power in short wavelengths versus long wavelengths. </span>
<span id="cb33-698"><a href="#cb33-698"></a></span>
<span id="cb33-699"><a href="#cb33-699"></a>In the main analysis, we defined&nbsp;**short-wavelength (blue light) content**&nbsp;as the integrated intensity in the 400–500&nbsp;nm range, and&nbsp;**long-wavelength content**&nbsp;as the integrated intensity in a longer range (e.g. 600–700&nbsp;nm), then computed the&nbsp;*short-to-long ratio*&nbsp;(“sl ratio”). Calculating these metrics is the first step of spectrum analysis in the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](index.qmd#spectrum)</span>.</span>
<span id="cb33-700"><a href="#cb33-700"></a></span>
<span id="cb33-701"><a href="#cb33-701"></a><span class="in">```</span></span>
<span id="cb33-702"><a href="#cb33-702"></a><span class="in">dataVEET2 |&gt; </span></span>
<span id="cb33-703"><a href="#cb33-703"></a><span class="in">  select(Id, Date, Datetime, Spectrum) |&gt;    # focus on ID, date, time, and spectrum</span></span>
<span id="cb33-704"><a href="#cb33-704"></a><span class="in">  mutate(</span></span>
<span id="cb33-705"><a href="#cb33-705"></a><span class="in">    short = Spectrum |&gt; map_dbl(spectral_integration, wavelength.range = c(400, 500)),</span></span>
<span id="cb33-706"><a href="#cb33-706"></a><span class="in">    long  = Spectrum |&gt; map_dbl(spectral_integration, wavelength.range = c(600, 700)),</span></span>
<span id="cb33-707"><a href="#cb33-707"></a><span class="in">    `sl ratio` = short / long   # compute short-to-long ratio</span></span>
<span id="cb33-708"><a href="#cb33-708"></a><span class="in">  )</span></span>
<span id="cb33-709"><a href="#cb33-709"></a><span class="in">```</span></span>
<span id="cb33-710"><a href="#cb33-710"></a></span>
<span id="cb33-711"><a href="#cb33-711"></a>*(The cutoff of 500&nbsp;nm here is hypothetical for demonstration; actual definitions might vary.)*&nbsp;We would then have columns&nbsp;<span class="in">`short`</span>,&nbsp;<span class="in">`long`</span>, and&nbsp;<span class="in">`sl_ratio`</span>&nbsp;for each observation, which could be averaged per day or analyzed further. The cleaned spectral data in&nbsp;<span class="in">`dataVEET2`</span>&nbsp;makes it straightforward to calculate such metrics or apply spectral weighting functions (for melatonin suppression, circadian stimulus, etc., if one has the spectral sensitivity curves).</span>
<span id="cb33-712"><a href="#cb33-712"></a></span>
<span id="cb33-713"><a href="#cb33-713"></a>With the <span class="in">`VEET`</span> spectral preprocessing complete, we emphasize that these steps – normalizing by gain, applying calibration, and perhaps simplifying channels – are&nbsp;**device-specific**&nbsp;requirements. They ensure that the raw sensor counts are translated into meaningful physical measures (like spectral irradiance). Researchers using other spectral devices would follow a similar procedure, adjusting for their device’s particulars (some may output spectra directly, whereas others, like <span class="in">`VEET`</span>, require reconstruction.</span>
<span id="cb33-714"><a href="#cb33-714"></a></span>
<span id="cb33-715"><a href="#cb33-715"></a>::: {.callout-note}</span>
<span id="cb33-716"><a href="#cb33-716"></a></span>
<span id="cb33-717"><a href="#cb33-717"></a>Some devices may output normalized counts instead of raw counts. For example, the <span class="in">`ActLumus`</span> device outputs normalized counts, while the <span class="in">`VEET`</span> device records raw counts and the gain. Manufacturers will be able to speficy exact outputs for a given model and software version.</span>
<span id="cb33-718"><a href="#cb33-718"></a></span>
<span id="cb33-719"><a href="#cb33-719"></a>:::</span>
<span id="cb33-720"><a href="#cb33-720"></a></span>
<span id="cb33-721"><a href="#cb33-721"></a><span class="fu">## `VEET` Data: Time of flight (distance)</span></span>
<span id="cb33-722"><a href="#cb33-722"></a></span>
<span id="cb33-723"><a href="#cb33-723"></a>In this last section, the distance data of the <span class="in">`VEET`</span> device will be imported, analogous to the other modalities. The <span class="in">`TOF`</span> modality contains information for up to two objects in a 8x8 grid of measurements, spanning a total of about 52° vertically and 41° horizontally. Because the <span class="in">`VEET`</span> device can detect up to two objects in a given grid point, and there is a confidence value assigned to every measurement, each observation contains $2*2*8*8 = 256$ measurements.</span>
<span id="cb33-724"><a href="#cb33-724"></a></span>
<span id="cb33-727"><a href="#cb33-727"></a><span class="in">```{r}</span></span>
<span id="cb33-728"><a href="#cb33-728"></a><span class="co">#| label: fig-VEET-overview3</span></span>
<span id="cb33-729"><a href="#cb33-729"></a><span class="co">#| fig-cap: "Overview plot of imported VEET data"</span></span>
<span id="cb33-730"><a href="#cb33-730"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb33-731"><a href="#cb33-731"></a><span class="co">#| fig-width: 6</span></span>
<span id="cb33-732"><a href="#cb33-732"></a><span class="co">#| filename: Import VEET Spectral Sensor (TOF) data</span></span>
<span id="cb33-733"><a href="#cb33-733"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="cb33-734"><a href="#cb33-734"></a>dataVEET3 <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, </span>
<span id="cb33-735"><a href="#cb33-735"></a>                        <span class="at">tz =</span> tz, </span>
<span id="cb33-736"><a href="#cb33-736"></a>                        <span class="at">modality =</span> <span class="st">"TOF"</span>, <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-737"><a href="#cb33-737"></a>                        <span class="at">manual.id =</span> <span class="st">"VEET"</span> <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-738"><a href="#cb33-738"></a>                        ) </span>
<span id="cb33-739"><a href="#cb33-739"></a><span class="in">```</span></span>
<span id="cb33-740"><a href="#cb33-740"></a></span>
<span id="cb33-741"><a href="#cb33-741"></a><span class="ss">1.  </span><span class="in">`modality`</span> is a parameter only the <span class="in">`VEET`</span> device requires. If uncertain, which devices require special parameters, have a look a the import help page (<span class="in">`?import`</span>) under the <span class="in">`VEET`</span> device. Setting it to <span class="in">`TOF`</span> gives us the distance modality.</span>
<span id="cb33-742"><a href="#cb33-742"></a><span class="ss">2.  </span>As we are only dealing with one individual here, we set a manual <span class="in">`Id`</span></span>
<span id="cb33-743"><a href="#cb33-743"></a></span>
<span id="cb33-746"><a href="#cb33-746"></a><span class="in">```{r}</span></span>
<span id="cb33-747"><a href="#cb33-747"></a><span class="co">#| filename: Print the first 6 rows of the VEETs distance dataset</span></span>
<span id="cb33-748"><a href="#cb33-748"></a>dataVEET3 <span class="sc">|&gt;</span> <span class="fu">head</span>()</span>
<span id="cb33-749"><a href="#cb33-749"></a><span class="in">```</span></span>
<span id="cb33-750"><a href="#cb33-750"></a></span>
<span id="cb33-751"><a href="#cb33-751"></a></span>
<span id="cb33-752"><a href="#cb33-752"></a>In a first step, we condition the data similarly to the other <span class="in">`VEET`</span> modalities. For computational reasons of the use cases, we remove the second object and set the interval to <span class="in">`10 seconds`</span>. *Note that the next step still takes considerable computation time*.</span>
<span id="cb33-753"><a href="#cb33-753"></a></span>
<span id="cb33-756"><a href="#cb33-756"></a><span class="in">```{r}</span></span>
<span id="cb33-757"><a href="#cb33-757"></a><span class="co">#| label: VEET-cleaning</span></span>
<span id="cb33-758"><a href="#cb33-758"></a><span class="co">#| filename: Aggregate distance data to 5-second intervals and mark gaps</span></span>
<span id="cb33-759"><a href="#cb33-759"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb33-760"><a href="#cb33-760"></a>  dataVEET3 <span class="sc">|&gt;</span></span>
<span id="cb33-761"><a href="#cb33-761"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">contains</span>(<span class="fu">c</span>(<span class="st">"conf2_"</span>, <span class="st">"dist2_"</span>))) <span class="sc">|&gt;</span> <span class="co"># &lt;1&gt;</span></span>
<span id="cb33-762"><a href="#cb33-762"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 secs"</span>) <span class="sc">|&gt;</span>     <span class="co"># &lt;2&gt;</span></span>
<span id="cb33-763"><a href="#cb33-763"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>           <span class="co"># &lt;3&gt;</span></span>
<span id="cb33-764"><a href="#cb33-764"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb33-765"><a href="#cb33-765"></a>  <span class="fu">remove_partial_data</span>(dist1_0, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span>
<span id="cb33-766"><a href="#cb33-766"></a>dataVEET3 <span class="sc">|&gt;</span> <span class="fu">summary_overview</span>(dist1_0)</span>
<span id="cb33-767"><a href="#cb33-767"></a><span class="in">```</span></span>
<span id="cb33-768"><a href="#cb33-768"></a></span>
<span id="cb33-769"><a href="#cb33-769"></a><span class="ss">1.  </span>Remove the second object (for computational reasons)</span>
<span id="cb33-770"><a href="#cb33-770"></a><span class="ss">2.  </span>Aggregate to 10-second bins</span>
<span id="cb33-771"><a href="#cb33-771"></a><span class="ss">3.  </span>Explicit NA for any gaps</span>
<span id="cb33-772"><a href="#cb33-772"></a></span>
<span id="cb33-773"><a href="#cb33-773"></a>In the next step, we need to transform the <span class="in">`wide`</span> format of the imported dataset into a <span class="in">`long`</span> format, where each row contains exactly one observation for one grid-point. </span>
<span id="cb33-774"><a href="#cb33-774"></a></span>
<span id="cb33-777"><a href="#cb33-777"></a><span class="in">```{r}</span></span>
<span id="cb33-778"><a href="#cb33-778"></a><span class="co">#| label: VEET-long</span></span>
<span id="cb33-779"><a href="#cb33-779"></a><span class="co">#| filename: Pivot distance grid from wide to long</span></span>
<span id="cb33-780"><a href="#cb33-780"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb33-781"><a href="#cb33-781"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb33-782"><a href="#cb33-782"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb33-783"><a href="#cb33-783"></a>    <span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(Datetime, file.name, Id, is.implicit, time_stamp, modality, Date),</span>
<span id="cb33-784"><a href="#cb33-784"></a>    <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">".value"</span>, <span class="st">"position"</span>),</span>
<span id="cb33-785"><a href="#cb33-785"></a>    <span class="at">names_pattern =</span> <span class="st">"(conf1|conf2|dist1|dist2)_(</span><span class="sc">\\</span><span class="st">d+)"</span></span>
<span id="cb33-786"><a href="#cb33-786"></a>  )</span>
<span id="cb33-787"><a href="#cb33-787"></a><span class="in">```</span></span>
<span id="cb33-788"><a href="#cb33-788"></a></span>
<span id="cb33-789"><a href="#cb33-789"></a>In a final step before we can use the data in the analysis, we need to assign <span class="in">`x`</span> and <span class="in">`y`</span> coordinates based on the <span class="in">`position`</span> column that was created when pivoting longer. Positions are counted from 0 to 63 starting at the top right and increasing towards the left, before continuing on the right in the next row below. <span class="in">`y`</span> positions thus depend on the row count, i.e., how often a row of 8 values fits into the <span class="in">`position`</span> column. <span class="in">`x`</span> positions consequently depend on the <span class="in">`position`</span> within each 8-value row. We also add an <span class="in">`observation`</span> variable that increases by <span class="in">`+1`</span> every time, the <span class="in">`position`</span> column hits <span class="in">`0`</span>. We then center both <span class="in">`x`</span> and <span class="in">`y`</span> coordinates to receive meaningful values, i.e., 0° indicates the center of the overall measurement cone. Lastly, we convert both confidence columns, which are scaled from 0-255 into percentages by dividing them by <span class="in">`255`</span>. Empirical data from the manufacturer points to a threshold of about <span class="in">`10%`</span>, under which the respective distance data is not reliable.</span>
<span id="cb33-790"><a href="#cb33-790"></a></span>
<span id="cb33-793"><a href="#cb33-793"></a><span class="in">```{r}</span></span>
<span id="cb33-794"><a href="#cb33-794"></a><span class="co">#| label: VEET-distance preparation</span></span>
<span id="cb33-795"><a href="#cb33-795"></a><span class="co">#| filename: Calculate grid positions of spatial distance measurements</span></span>
<span id="cb33-796"><a href="#cb33-796"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb33-797"><a href="#cb33-797"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb33-798"><a href="#cb33-798"></a>  <span class="fu">mutate</span>(<span class="at">position =</span> <span class="fu">as.numeric</span>(position),</span>
<span id="cb33-799"><a href="#cb33-799"></a>         <span class="at">y.pos =</span> (position <span class="sc">%/%</span> <span class="dv">8</span>)<span class="sc">+</span><span class="dv">1</span>, <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-800"><a href="#cb33-800"></a>         <span class="at">y.pos =</span> <span class="fu">scale</span>(y.pos, <span class="at">scale =</span> <span class="cn">FALSE</span>)<span class="sc">*</span><span class="dv">52</span><span class="sc">/</span><span class="dv">8</span>, <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-801"><a href="#cb33-801"></a>         <span class="at">x.pos =</span> <span class="dv">8</span> <span class="sc">-</span> (position <span class="sc">%%</span> <span class="dv">8</span>), <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-802"><a href="#cb33-802"></a>         <span class="at">x.pos =</span> <span class="fu">scale</span>(x.pos, <span class="at">scale =</span> <span class="cn">FALSE</span>)<span class="sc">*</span><span class="dv">41</span><span class="sc">/</span><span class="dv">8</span>, <span class="co">#&lt;4&gt;</span></span>
<span id="cb33-803"><a href="#cb33-803"></a>         <span class="at">observation =</span> <span class="fu">cumsum</span>(position <span class="sc">==</span> <span class="dv">0</span>), <span class="co">#&lt;5&gt;</span></span>
<span id="cb33-804"><a href="#cb33-804"></a>         <span class="fu">across</span>(<span class="fu">starts_with</span>(<span class="st">"conf"</span>), \(x) x<span class="sc">/</span><span class="dv">255</span>) <span class="co">#&lt;6&gt;</span></span>
<span id="cb33-805"><a href="#cb33-805"></a>         )</span>
<span id="cb33-806"><a href="#cb33-806"></a><span class="in">```</span></span>
<span id="cb33-807"><a href="#cb33-807"></a></span>
<span id="cb33-808"><a href="#cb33-808"></a><span class="ss">1.  </span>Increment the y position for every 8 steps in <span class="in">`position`</span></span>
<span id="cb33-809"><a href="#cb33-809"></a><span class="ss">2.  </span>Center <span class="in">`y.pos`</span> and rescale it to cover 52° across 8 steps</span>
<span id="cb33-810"><a href="#cb33-810"></a><span class="ss">3.  </span>Increment the x position for every step in <span class="in">`position`</span>, resetting every 8 steps</span>
<span id="cb33-811"><a href="#cb33-811"></a><span class="ss">4.  </span>Center <span class="in">`x.pos`</span> and rescale it to cover 41° across 8 steps</span>
<span id="cb33-812"><a href="#cb33-812"></a><span class="ss">5.  </span>Increase an observation counter every time we restart with <span class="in">`position`</span> at 0</span>
<span id="cb33-813"><a href="#cb33-813"></a><span class="ss">6.  </span>Scale the confidence columns so that 255 = 100%</span>
<span id="cb33-814"><a href="#cb33-814"></a></span>
<span id="cb33-815"><a href="#cb33-815"></a>Now this dataset is ready for further analysis. We finish by visualizing the same observation time on different days. Note that we replace zero distance values with <span class="in">`Infinity`</span>, as these indicate measurements outside the 5m measurement radius of the device.</span>
<span id="cb33-816"><a href="#cb33-816"></a></span>
<span id="cb33-819"><a href="#cb33-819"></a><span class="in">```{r}</span></span>
<span id="cb33-820"><a href="#cb33-820"></a><span class="co">#| label: fig-VEET-distance</span></span>
<span id="cb33-821"><a href="#cb33-821"></a><span class="co">#| fig-cap: "Example observations of the measurement grid at 1:14 p.m. for each measurement day. Text values show distance in cm. Empty grid points show values with low confidence. Zero-distance values were replaced with infinite distance and plotted despite low confidence."</span></span>
<span id="cb33-822"><a href="#cb33-822"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb33-823"><a href="#cb33-823"></a><span class="co">#| message: false</span></span>
<span id="cb33-824"><a href="#cb33-824"></a><span class="co">#| fig-width: 12</span></span>
<span id="cb33-825"><a href="#cb33-825"></a><span class="co">#| filename: Plot spatial distance grid for the same time point on each day</span></span>
<span id="cb33-826"><a href="#cb33-826"></a>extras <span class="ot">&lt;-</span> <span class="fu">list</span>( <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-827"><a href="#cb33-827"></a>  <span class="fu">geom_tile</span>(), <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-828"><a href="#cb33-828"></a>  <span class="fu">scale_fill_viridis_c</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">200</span>), <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-829"><a href="#cb33-829"></a>                       <span class="at">oob =</span> scales<span class="sc">::</span>oob_squish_any), <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-830"><a href="#cb33-830"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"white"</span>)), <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-831"><a href="#cb33-831"></a>  <span class="fu">theme_minimal</span>(), <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-832"><a href="#cb33-832"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="st">"none"</span>), <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-833"><a href="#cb33-833"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> (dist1<span class="sc">/</span><span class="dv">10</span>) <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">0</span>), <span class="at">colour =</span> dist1<span class="sc">&gt;</span><span class="dv">1000</span>), <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-834"><a href="#cb33-834"></a>            <span class="at">size =</span> <span class="fl">2.5</span>), <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-835"><a href="#cb33-835"></a>  <span class="fu">coord_fixed</span>(), <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-836"><a href="#cb33-836"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"X position (°)"</span>, <span class="at">y =</span> <span class="st">"Y position (°)"</span>, <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-837"><a href="#cb33-837"></a>       <span class="at">fill =</span> <span class="st">"Distance (cm)"</span>)) <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-838"><a href="#cb33-838"></a></span>
<span id="cb33-839"><a href="#cb33-839"></a>slicer <span class="ot">&lt;-</span> <span class="cf">function</span>(x){<span class="fu">seq</span>(<span class="fu">min</span>((x<span class="dv">-1</span>)<span class="sc">*</span><span class="dv">64</span><span class="sc">+</span><span class="dv">1</span>), <span class="fu">max</span>(x<span class="sc">*</span><span class="dv">64</span>, <span class="at">by =</span> <span class="dv">1</span>))} <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-840"><a href="#cb33-840"></a></span>
<span id="cb33-841"><a href="#cb33-841"></a>dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb33-842"><a href="#cb33-842"></a>  <span class="fu">slice</span>(<span class="fu">slicer</span>(<span class="dv">9530</span>)) <span class="sc">|&gt;</span>  <span class="co">#&lt;3&gt;</span></span>
<span id="cb33-843"><a href="#cb33-843"></a>  <span class="fu">mutate</span>(<span class="at">dist1 =</span> <span class="fu">ifelse</span>(dist1 <span class="sc">==</span> <span class="dv">0</span>, <span class="cn">Inf</span>, dist1)) <span class="sc">|&gt;</span> <span class="co">#&lt;4&gt;</span></span>
<span id="cb33-844"><a href="#cb33-844"></a>  <span class="fu">filter</span>(conf1 <span class="sc">&gt;=</span> <span class="fl">0.1</span> <span class="sc">|</span> dist1 <span class="sc">==</span> <span class="cn">Inf</span>) <span class="sc">|&gt;</span> <span class="co">#&lt;5&gt;</span></span>
<span id="cb33-845"><a href="#cb33-845"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x.pos, <span class="at">y=</span>y.pos, <span class="at">fill =</span> dist1<span class="sc">/</span><span class="dv">10</span>))<span class="sc">+</span> extras <span class="sc">+</span> <span class="co">#&lt;6&gt;</span></span>
<span id="cb33-846"><a href="#cb33-846"></a>  <span class="fu">facet_grid</span>(<span class="sc">~</span>Datetime) <span class="co">#&lt;7&gt;</span></span>
<span id="cb33-847"><a href="#cb33-847"></a><span class="in">```</span></span>
<span id="cb33-848"><a href="#cb33-848"></a></span>
<span id="cb33-849"><a href="#cb33-849"></a><span class="ss">1.  </span>Set visualization parameters</span>
<span id="cb33-850"><a href="#cb33-850"></a><span class="ss">2.  </span>Allows to choose an observation</span>
<span id="cb33-851"><a href="#cb33-851"></a><span class="ss">3.  </span>Choose a particular observation</span>
<span id="cb33-852"><a href="#cb33-852"></a><span class="ss">4.  </span>Replace 0 distances with Infinity</span>
<span id="cb33-853"><a href="#cb33-853"></a><span class="ss">5.  </span>Remove data that has less than 10% confidence</span>
<span id="cb33-854"><a href="#cb33-854"></a><span class="ss">6.  </span>Plot the data</span>
<span id="cb33-855"><a href="#cb33-855"></a><span class="ss">7.  </span>Show one plot per day</span>
<span id="cb33-856"><a href="#cb33-856"></a></span>
<span id="cb33-857"><a href="#cb33-857"></a>As we can see from the figure, different days have - at a given time - a vastly different distribution of distance data, and measurement confidence (values with confidence <span class="sc">\&lt;</span> 10% are removed)</span>
<span id="cb33-858"><a href="#cb33-858"></a></span>
<span id="cb33-859"><a href="#cb33-859"></a><span class="fu">### Removing non-wear times</span></span>
<span id="cb33-860"><a href="#cb33-860"></a></span>
<span id="cb33-861"><a href="#cb33-861"></a>Same as for the <span class="in">`ALS`</span> modality, we can remove times of non-wear - we can even use the same <span class="in">`wear_data`</span> set for it:</span>
<span id="cb33-862"><a href="#cb33-862"></a></span>
<span id="cb33-865"><a href="#cb33-865"></a><span class="in">```{r}</span></span>
<span id="cb33-866"><a href="#cb33-866"></a><span class="co">#| filename: Removing non-wear observations from distance modality</span></span>
<span id="cb33-867"><a href="#cb33-867"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb33-868"><a href="#cb33-868"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb33-869"><a href="#cb33-869"></a>  <span class="fu">group_by</span>(Id) <span class="sc">|&gt;</span> </span>
<span id="cb33-870"><a href="#cb33-870"></a>  <span class="fu">add_states</span>(wear_data)</span>
<span id="cb33-871"><a href="#cb33-871"></a></span>
<span id="cb33-872"><a href="#cb33-872"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb33-873"><a href="#cb33-873"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb33-874"><a href="#cb33-874"></a>  <span class="fu">mutate</span>(<span class="at">dist1 =</span> <span class="fu">ifelse</span>(wear, dist1, <span class="cn">NA</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb33-875"><a href="#cb33-875"></a>  <span class="fu">group_by</span>(Id, Date)</span>
<span id="cb33-876"><a href="#cb33-876"></a><span class="in">```</span></span>
<span id="cb33-877"><a href="#cb33-877"></a></span>
<span id="cb33-878"><a href="#cb33-878"></a></span>
<span id="cb33-879"><a href="#cb33-879"></a><span class="fu">## Saving the Cleaned Data</span></span>
<span id="cb33-880"><a href="#cb33-880"></a></span>
<span id="cb33-881"><a href="#cb33-881"></a>After executing all the above steps, we have three cleaned data frames in our R session:</span>
<span id="cb33-882"><a href="#cb33-882"></a></span>
<span id="cb33-883"><a href="#cb33-883"></a><span class="ss">-   </span>**`dataCC`**&nbsp;– the processed <span class="in">`Clouclip`</span> dataset (5-second intervals, with distance and lux, including NA for gaps and sentinel statuses).</span>
<span id="cb33-884"><a href="#cb33-884"></a></span>
<span id="cb33-885"><a href="#cb33-885"></a><span class="ss">-   </span>**`dataVEET`**&nbsp;– the processed <span class="in">`VEET`</span> ambient light dataset (5-second intervals, illuminance in lux, with gaps filled).</span>
<span id="cb33-886"><a href="#cb33-886"></a></span>
<span id="cb33-887"><a href="#cb33-887"></a><span class="ss">-   </span>**`dataVEET2`**&nbsp;– the processed <span class="in">`VEET`</span> spectral dataset (5-minute intervals, each entry containing a spectrum or derived spectral metrics).</span>
<span id="cb33-888"><a href="#cb33-888"></a></span>
<span id="cb33-889"><a href="#cb33-889"></a><span class="ss">-   </span>**`dataVEET3`**&nbsp;– the processed <span class="in">`VEET`</span> distance dataset (5-second intervals, each entry containing the distance of up to two objects in the 8x8 grid).</span>
<span id="cb33-890"><a href="#cb33-890"></a></span>
<span id="cb33-891"><a href="#cb33-891"></a>For convenience and future reproducibility, we will save these combined results to a single R data file. Storing all cleaned data together ensures that any analysis can reload the exact same data state without re-running the import and cleaning (which can be time-consuming for large raw files).</span>
<span id="cb33-892"><a href="#cb33-892"></a></span>
<span id="cb33-895"><a href="#cb33-895"></a><span class="in">```{r}</span></span>
<span id="cb33-896"><a href="#cb33-896"></a><span class="co">#| filename: Save preprocessed files</span></span>
<span id="cb33-897"><a href="#cb33-897"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">"data/cleaned"</span>)) <span class="fu">dir.create</span>(<span class="st">"data/cleaned"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>) <span class="co">#&lt;1&gt;</span></span>
<span id="cb33-898"><a href="#cb33-898"></a><span class="fu">save</span>(dataCC, dataVEET, dataVEET2, dataVEET3, <span class="at">file =</span> <span class="st">"data/cleaned/data.RData"</span>) <span class="co">#&lt;2&gt;</span></span>
<span id="cb33-899"><a href="#cb33-899"></a><span class="in">```</span></span>
<span id="cb33-900"><a href="#cb33-900"></a></span>
<span id="cb33-901"><a href="#cb33-901"></a><span class="ss">1. </span>Create directory for cleaned data if it doesn't exist</span>
<span id="cb33-902"><a href="#cb33-902"></a><span class="ss">2. </span>Save all cleaned datasets into one .RData file</span>
<span id="cb33-903"><a href="#cb33-903"></a></span>
<span id="cb33-904"><a href="#cb33-904"></a>The above code creates (if necessary) a folder&nbsp;<span class="in">`data/cleaned/`</span>&nbsp;and saves a single RData file (<span class="in">`data.RData`</span>) containing the three objects. To retrieve them later, one can use&nbsp;<span class="in">`&lt;- load("data/cleaned/data.RData")`</span>, which will return the objects into the environment. This single-file approach simplifies sharing and keeps the cleaned data together.</span>
<span id="cb33-905"><a href="#cb33-905"></a></span>
<span id="cb33-906"><a href="#cb33-906"></a>In summary, this supplement has walked through the full preprocessing pipeline for two example devices. We began by describing the raw data format for each device and then demonstrated how to import the data with correct time zone settings. We handled device-specific quirks like sentinel codes (for <span class="in">`Clouclip`</span>) and multiple modalities with gain normalization (for <span class="in">`VEET`</span>). We showed how to detect and address irregular sampling, how to explicitly mark missing data gaps to avoid analytic pitfalls, and how to reduce data granularity via rounding or aggregation when appropriate. Throughout, we used functions from LightLogR in a tidyverse workflow, aiming to make the steps clear and modular. By saving the final cleaned datasets, we set the stage for the computation of visual experience metrics such as working distance, time in bright light, spectral composition ratios, as presented in the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](index.qmd)</span>. We hope this detailed tutorial empowers researchers to adopt similar pipelines for their own data, facilitating reproducible and accurate analyses of visual experience.</span>
<span id="cb33-907"><a href="#cb33-907"></a></span>
<span id="cb33-908"><a href="#cb33-908"></a><span class="fu">## References</span></span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/tscnlab/ZaunerEtAl_JVis_2026/edit/main/supplement.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/tscnlab/ZaunerEtAl_JVis_2026/issues" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div></div></footer><script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>