<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
    <meta charset="utf-8">
    <meta name="generator" content="quarto-1.7.33">

    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

    <meta name="author" content="Johannes Zauner">
    <meta name="author" content="Aaron Nicholls">
    <meta name="author" content="Lisa A. Ostrin">
    <meta name="author" content="Manuel Spitschan">
    <meta name="dcterms.date" content="2025-08-11">

    <title>Supplement 1</title>
    <style>
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      div.columns{display: flex; gap: min(4vw, 1.5em);}
      div.column{flex: auto; overflow-x: auto;}
      div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
      ul.task-list{list-style: none;}
      ul.task-list li input[type="checkbox"] {
        width: 0.8em;
        margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
        vertical-align: middle;
      }
      /* CSS for syntax highlighting */
      html { -webkit-text-size-adjust: 100%; }
      pre > code.sourceCode { white-space: pre; position: relative; }
      pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
      pre > code.sourceCode > span:empty { height: 1.2em; }
      .sourceCode { overflow: visible; }
      code.sourceCode > span { color: inherit; text-decoration: inherit; }
      div.sourceCode { margin: 1em 0; }
      pre.sourceCode { margin: 0; }
      @media screen {
      div.sourceCode { overflow: auto; }
      }
      @media print {
      pre > code.sourceCode { white-space: pre-wrap; }
      pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
      }
      pre.numberSource code
        { counter-reset: source-line 0; }
      pre.numberSource code > span
        { position: relative; left: -4em; counter-increment: source-line; }
      pre.numberSource code > span > a:first-child::before
        { content: counter(source-line);
          position: relative; left: -1em; text-align: right; vertical-align: baseline;
          border: none; display: inline-block;
          -webkit-touch-callout: none; -webkit-user-select: none;
          -khtml-user-select: none; -moz-user-select: none;
          -ms-user-select: none; user-select: none;
          padding: 0 4px; width: 4em;
        }
      pre.numberSource { margin-left: 3em;  padding-left: 4px; }
      div.sourceCode
        {   }
      @media screen {
      pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
      }
      /* CSS for citations */
      div.csl-bib-body { }
      div.csl-entry {
        clear: both;
        margin-bottom: 0em;
      }
      .hanging-indent div.csl-entry {
        margin-left:2em;
        text-indent:-2em;
      }
      div.csl-left-margin {
        min-width:2em;
        float:left;
      }
      div.csl-right-inline {
        margin-left:2em;
        padding-left:1em;
      }
      div.csl-indent {
        margin-left: 2em;
      }    </style>

    <style>
      body.hypothesis-enabled #quarto-embed-header {
        padding-right: 36px;
      }

      #quarto-embed-header {
        height: 3em;
        width: 100%;
        display: flex;
        justify-content: space-between;
        align-items: center;
        border-bottom: solid 1px;
      }

      #quarto-embed-header h6 {
        font-size: 1.1em;
        padding-top: 0.6em;
        margin-left: 1em;
        margin-right: 1em;
        font-weight: 400;
      }

      #quarto-embed-header a.quarto-back-link,
      #quarto-embed-header a.quarto-download-embed {
        font-size: 0.8em;
        margin-top: 1em;
        margin-bottom: 1em;
        margin-left: 1em;
        margin-right: 1em;
      }

      .quarto-back-container {
        padding-left: 0.5em;
        display: flex;
      }

      .headroom {
          will-change: transform;
          transition: transform 200ms linear;
      }

      .headroom--pinned {
          transform: translateY(0%);
      }

      .headroom--unpinned {
          transform: translateY(-100%);
      }      
    </style>

    <script>
    window.document.addEventListener("DOMContentLoaded", function () {

      var header = window.document.querySelector("#quarto-embed-header");
      const titleBannerEl = window.document.querySelector("body > #title-block-header");
      if (titleBannerEl) {
        titleBannerEl.style.paddingTop = header.clientHeight + "px";
      }
      const contentEl = window.document.getElementById('quarto-content');
      for (const child of contentEl.children) {
        child.style.paddingTop = header.clientHeight + "px";
        child.style.marginTop = "1em";
      }

      // Use the article root if the `back` call doesn't work. This isn't perfect
      // but should typically work
      window.quartoBackToArticle = () => {
        var currentUrl = window.location.href;
        window.history.back();
        setTimeout(() => {
            // if location was not changed in 100 ms, then there is no history back
            if(currentUrl === window.location.href){              
                // redirect to site root
                window.location.href = "index.html";
            }
        }, 100);
      }

      const headroom = new window.Headroom(header, {
        tolerance: 5,
        onPin: function () {
        },
        onUnpin: function () {
        },
      });
      headroom.init();
    });
    </script>

    
<script src="site_libs/manuscript-notebook/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d24ab4a4f01adf7b6391c077e236a78f.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
     <script async="" src="https://hypothes.is/embed.js"></script>
<script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script>  <style>
.quarto-notebook .cell-container .cell-decorator {
  display: none;
}
</style>   <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML" type="text/javascript"></script> 
        <link rel="stylesheet" href="styles.css">
      </head>

  <body class="quarto-notebook quarto-light">
    <div id="quarto-embed-header" class="headroom fixed-top bg-primary">
      
      <a onclick="window.quartoBackToArticle(); return false;" class="btn btn-primary quarto-back-link" href=""><i class="bi bi-caret-left"></i> Back to Article</a>
      <h6><i class="bi bi-journal-code"></i> Supplement 1</h6>

            <a href="./supplement.qmd" class="btn btn-primary quarto-download-embed" download="supplement.qmd">Download Source</a>
          </div>

     <header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Supplement 1</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
            <p class="subtitle lead">Analysis of human visual experience data</p>
          </div>

    
    <div class="quarto-title-meta-container">
      <div class="quarto-title-meta-column-start">
            <div class="quarto-title-meta-author">
          <div class="quarto-title-meta-heading">Authors</div>
          <div class="quarto-title-meta-heading">Affiliations</div>
          
                <div class="quarto-title-meta-contents">
            <p class="author">Johannes Zauner <a href="mailto:johannes.zauner@tum.de" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0003-2171-4566" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
                <div class="quarto-title-meta-contents">
                    <p class="affiliation">
                        Technical University of Munich, Germany
                      </p>
                    <p class="affiliation">
                        Max Planck Institute for Biological Cybernetics, Tübingen, Germany
                      </p>
                  </div>
                      <div class="quarto-title-meta-contents">
            <p class="author">Aaron Nicholls <a href="https://orcid.org/0009-0001-6683-6826" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
                <div class="quarto-title-meta-contents">
                    <p class="affiliation">
                        Meta Reality Labs Research
                      </p>
                  </div>
                      <div class="quarto-title-meta-contents">
            <p class="author">Lisa A. Ostrin <a href="https://orcid.org/0000-0001-8675-1478" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
                <div class="quarto-title-meta-contents">
                    <p class="affiliation">
                        University of Houston, Texas, USA
                      </p>
                  </div>
                      <div class="quarto-title-meta-contents">
            <p class="author">Manuel Spitschan <a href="https://orcid.org/0000-0002-8572-9268" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
                <div class="quarto-title-meta-contents">
                    <p class="affiliation">
                        Technical University of Munich, Germany
                      </p>
                    <p class="affiliation">
                        Max Planck Institute for Biological Cybernetics, Tübingen, Germany
                      </p>
                    <p class="affiliation">
                        Technical University of Munich Institute of advanced study (TUM-IAS), Munich, Germany
                      </p>
                  </div>
                    </div>
        
        <div class="quarto-title-meta">

                      
                <div>
            <div class="quarto-title-meta-heading">Last modified:</div>
            <div class="quarto-title-meta-contents">
              <p class="date">August 11, 2025</p>
            </div>
          </div>
          
                
                <div>
            <div class="quarto-title-meta-heading">Doi</div>
            <div class="quarto-title-meta-contents">
              <p class="doi">
                <a href="https://doi.org/10.5281/zenodo.16566014">10.5281/zenodo.16566014</a>
              </p>
            </div>
          </div>
              </div>
      </div>
      <div class="quarto-title-meta-column-end quarto-other-formats-target">
      </div>
    </div>



    <div class="quarto-other-links-text-target">
    </div>  </div>
</header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#abstract" id="toc-abstract" class="nav-link active" data-scroll-target="#abstract">Abstract</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#clouclip-data-raw-structure-and-import" id="toc-clouclip-data-raw-structure-and-import" class="nav-link" data-scroll-target="#clouclip-data-raw-structure-and-import"><code>Clouclip</code> Data: Raw Structure and Import</a></li>
  <li><a href="#veet-data-ambient-light-illuminance-processing" id="toc-veet-data-ambient-light-illuminance-processing" class="nav-link" data-scroll-target="#veet-data-ambient-light-illuminance-processing"><code>VEET</code> Data: Ambient Light (Illuminance) Processing</a></li>
  <li><a href="#veet-data-spectral-data-processing" id="toc-veet-data-spectral-data-processing" class="nav-link" data-scroll-target="#veet-data-spectral-data-processing"><code>VEET</code> Data: Spectral Data Processing</a>
  <ul class="collapse">
  <li><a href="#import" id="toc-import" class="nav-link" data-scroll-target="#import">Import</a></li>
  <li><a href="#spectral-calibration" id="toc-spectral-calibration" class="nav-link" data-scroll-target="#spectral-calibration">Spectral calibration</a></li>
  <li><a href="#spectral-reconstruction" id="toc-spectral-reconstruction" class="nav-link" data-scroll-target="#spectral-reconstruction">Spectral reconstruction</a></li>
  </ul></li>
  <li><a href="#veet-data-time-of-flight-distance" id="toc-veet-data-time-of-flight-distance" class="nav-link" data-scroll-target="#veet-data-time-of-flight-distance"><code>VEET</code> Data: Time of flight (distance)</a></li>
  <li><a href="#saving-the-cleaned-data" id="toc-saving-the-cleaned-data" class="nav-link" data-scroll-target="#saving-the-cleaned-data">Saving the Cleaned Data</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">      

       <div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Please note that this supplement is not in its final styling form, as it necessarily has to be formatted as a <code>notebook</code> to enable commenting. Current issues include:</p>
<ul>
<li>Cross-references not working (e.g.&nbsp;Figure labeling and referencing figures)</li>
</ul>
<p>These issues will be resolved when all comments have been gathered and implemented</p>
</div>
</div>
<section id="abstract" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="abstract">Abstract</h2>
<p>This supplementary document provides a detailed, step-by-step tutorial on importing and preprocessing raw data from two wearable devices: <code>Clouclip</code> and <code>VEET</code>. We describe the structure of the raw datasets recorded by each device and explain how to parse these data, specify time zones, handle special sentinel values, clean missing observations, regularize timestamps to fixed intervals, and aggregate data as needed. All original R code from the <a href="https://tscnlab.github.io/ZaunerEtAl_JVis_2025/">main tutorial</a> is preserved here for transparency, with additional guidance intended for a broad research audience. We demonstrate how to detect gaps and irregular sampling, convert implicit missing periods into explicit missing values, and address device-specific quirks such as the <code>Clouclip</code>’s use of sentinel codes for “sleep mode” and “out of range” readings. Special procedures for processing the <code>VEET</code>’s rich spectral data (e.g.&nbsp;normalizing sensor counts and reconstructing full spectra from multiple sensor channels) are also outlined. Finally, we show how to save the cleaned datasets from both devices into a single R data file for downstream analysis. This comprehensive walkthrough is designed to ensure reproducibility and to assist researchers in understanding and adapting the data pipeline for their own visual experience datasets.</p>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Wearable sensors like the Clouclip and the Visual Environment Evaluation Tool (VEET) produce high-dimensional time-series data on viewing distance and light exposure. Proper handling of these raw data is essential before any analysis of visual experience metrics. In the <a href="https://tscnlab.github.io/ZaunerEtAl_JVis_2025/">main tutorial</a>, we introduced an analysis pipeline using the open-source R package LightLogR <span class="citation" data-cites="Zauner2025JOpenSourceSoftw">(<a href="#ref-Zauner2025JOpenSourceSoftw" role="doc-biblioref">Zauner, Hartmeyer, and Spitschan 2025</a>)</span> to calculate various distance and light exposure metrics. Here, we present a full account of the data import and preparation steps as a supplement to the methods, with an emphasis on clarity for researchers who may be less familiar with data processing in R.</p>
<p>We use example datasets from a <code>Clouclip</code> device <span class="citation" data-cites="Wen2021ActaOphtalmol Wen2020bjophthalmol">(<a href="#ref-Wen2021ActaOphtalmol" role="doc-biblioref">Wen et al. 2021</a>, <a href="#ref-Wen2020bjophthalmol" role="doc-biblioref">2020</a>)</span> and from a <code>VEET</code> device <span class="citation" data-cites="Sah2025OphtalmicPhysiolOpt">(<a href="#ref-Sah2025OphtalmicPhysiolOpt" role="doc-biblioref">Sah, Narra, and Ostrin 2025</a>)</span> (both provided in the accompanying <a href="https://github.com/tscnlab/ZaunerEtAl_2025_JOV">repository</a>). The <code>Clouclip</code> is a glasses-mounted sensor that records working distance (distance from eyes to object, in cm) and ambient illuminance (in lux) at 5-second intervals. The <code>VEET</code> is a head-mounted multi-modal sensor that logs ambient light and spectral information (along with other data like motion and distance via a depth sensor) - in this exemplary case at 2-second intervals. A single week of continuous data illustrates the contrast in complexity: approximately 1.6&nbsp;MB for the <code>Clouclip</code>’s simple two-column output versus up to 270&nbsp;MB for the <code>VEET</code>’s multi-channel output (due to its higher sampling rate and richer sensor modalities).</p>
<p>In the following sections, we detail how these raw data are structured and how to import and preprocess them using LightLogR. We cover device-specific considerations such as file format quirks and sensor range limitations, as well as general best practices like handling missing timestamps and normalizing sensor readings. All code blocks can be executed in R (with the required packages loaded) to reproduce the cleaning steps. The end result will be clean, regularized datasets (<code>dataCC</code> for Clouclip, <code>dataVEET</code> for VEET light data, and <code>dataVEET2</code> for VEET spectral data) ready for calculating visual experience metrics. We conclude by saving these cleaned datasets into a single file for convenient reuse.</p>
</section>
<section id="clouclip-data-raw-structure-and-import" class="level2">
<h2 class="anchored" data-anchor-id="clouclip-data-raw-structure-and-import"><code>Clouclip</code> Data: Raw Structure and Import</h2>
<p>The <code>Clouclip</code> device exports its data as a text file (not a true Excel file despite sometimes using an .xls extension), which is actually tab-separated values. Each record corresponds to one timestamped observation (nominally every 5 seconds) and includes two measured variables: <strong>distance</strong> and <strong>illuminance</strong>. In the sample dataset (<code>Sample_Clouclip.csv</code> provided), the columns are:</p>
<ul>
<li><p><code>Date</code> – the date and time of the observation (in the device’s local time, here one week in 2021).</p></li>
<li><p><code>Dis</code> – the viewing distance in centimeters.</p></li>
<li><p><code>Lux</code> – the ambient illuminance in lux.</p></li>
</ul>
<p>For example, a raw data line might look like:</p>
<pre><code>2021-07-01 08:00:00 45  320</code></pre>
<p>indicating that at <em>2021-07-01 08:00:00</em> local time, the device recorded a working distance of 45&nbsp;cm and illuminance of 320&nbsp;lx. The Clouclip uses special <strong>sentinel values</strong><a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> in these measurement columns to denote certain device states. Specifically, a distance (<code>Dis</code>) value of <strong>204</strong> is a code meaning the object is <em>out of the sensor’s range</em>, and a value of <strong>-1</strong> in either <code>Dis</code> or <code>Lux</code> indicates the device was in <em>sleep mode</em> (not actively recording). During normal operation, distance measurements are limited by the device’s range, and illuminance readings are positive lux values. Any sentinel codes in the raw file need to be handled appropriately, as described below.</p>
<p>We will use LightLogR’s built-in import function for <code>Clouclip</code>, which automatically reads the file, parses the timestamps, and converts sentinel codes into a separate status annotation. To begin, we load the necessary libraries and import the raw <code>Clouclip</code> dataset:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [1]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load required packages</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.2     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.0.4     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(LightLogR)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gt)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(downlit) <span class="co">#not used, but important for code-linking feature</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [2]:</pre></div><div id="cell-fig-importCC" class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define file path and time zone</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/Sample_Clouclip.csv"</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span>   <span class="co"># Time zone in which device was recording (e.g., US Central Time)</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Import Clouclip data</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>dataCC <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">Clouclip</span>(path, <span class="at">tz =</span> tz, <span class="at">manual.id =</span> <span class="st">"Clouclip"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Successfully read in 58'081 observations across 1 Ids from 1 Clouclip-file(s).
Timezone set is US/Central.
The system timezone is Europe/Berlin. Please correct if necessary!

First Observation: 2021-02-06 17:12:47
Last Observation: 2021-02-14 17:12:36
Timespan: 8 days

Observation intervals: 
  Id       interval.time            n pct     
1 Clouclip 5s                   54572 93.9601%
2 Clouclip 17s                     12 0.0207% 
3 Clouclip 18s                     14 0.0241% 
4 Clouclip 120s (~2 minutes)     3479 5.9900% 
5 Clouclip 128s (~2.13 minutes)     1 0.0017% 
6 Clouclip 132s (~2.2 minutes)      1 0.0017% 
7 Clouclip 133s (~2.22 minutes)     1 0.0017% </code></pre>
</div>
<div class="cell-output-display">
<div id="fig-importCC" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-importCC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="supplement_files/figure-html/fig-importCC-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="Overview plot of imported Clouclip data"><img src="supplement_files/figure-html/fig-importCC-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-importCC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Overview plot of imported Clouclip data
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>In this code, <code>import$Clouclip()</code> reads the tab-delimited file and returns a <strong>tibble</strong><a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> (saved in the variable <code>dataCC</code>) containing the data. We specify&nbsp;<code>tz = "US/Central"</code>&nbsp;because the device’s clock was set to U.S. Central time; this ensures that the&nbsp;<code>Datetime</code>values are properly interpreted with the correct time zone. The argument&nbsp;<code>manual.id = "Clouclip"</code>&nbsp;simply tags the dataset with an identifier (useful if combining data from multiple devices).</p>
<p>During import, LightLogR automatically handles the <code>Clouclip</code>’s sentinel codes. The&nbsp;<code>Date</code>&nbsp;column from the raw file is parsed into a POSIX date-time (<code>Datetime</code>) with the specified time zone. The&nbsp;<code>Lux</code>&nbsp;and&nbsp;<code>Dis</code>&nbsp;columns are read as numeric, but any occurrences of&nbsp;<strong>-1</strong>&nbsp;or&nbsp;<strong>204</strong>&nbsp;are treated specially: these are replaced with&nbsp;<code>NA</code>&nbsp;(missing values) in the numeric columns, and corresponding status columns&nbsp;<code>Lux_status</code>&nbsp;and&nbsp;<code>Dis_status</code>&nbsp;are created to indicate the reason for those <code>NA</code> values. For example, if a&nbsp;<code>Dis</code>&nbsp;value of 204 was encountered, that row’s&nbsp;<code>Dis</code>&nbsp;will be set to NA and&nbsp;<code>Dis_status</code>&nbsp;will contain&nbsp;<code>"out_of_range"</code>; if&nbsp;<code>Lux</code>&nbsp;or&nbsp;<code>Dis</code>&nbsp;was -1, the status is&nbsp;<code>"sleep_mode"</code>. We will set all other readings to &nbsp;<code>"operational"</code> (meaning the device was functioning normally at that time) for visualisation purposes.</p>
<p>After import, it is good practice to get an overview of the data. The import function by default prints a brief summary (and generates an overview plot of the timeline) showing the number of observations, the time span, and any irregularities or large gaps. In our case, the <code>Clouclip</code> summary indicates the data spans one week and reveals that there are&nbsp;<strong>irregular intervals</strong>&nbsp;in the timestamps. This means some observations do not occur exactly every 5 seconds as expected. We can programmatically check for irregular timing:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [3]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if data are on a perfectly regular 5-second schedule</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">has_irregulars</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div></div>
<p>If the result is&nbsp;<code>TRUE</code>, it confirms that the time sequence has deviations from the 5-second interval. Indeed, our example dataset has many small timing irregularities and gaps (periods with no data). Understanding the pattern of these missing or irregular readings is important. We can visualize them using a gap plot:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [4]:</pre></div><div id="cell-fig-irregular" class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot gaps and irregular timestamps for Clouclip data</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>y.label <span class="ot">&lt;-</span> <span class="st">"Distance (cm)"</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">gg_gaps</span>(Dis, </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">include.implicit.gaps =</span> <span class="cn">FALSE</span>,  <span class="co"># only show recorded gaps, not every missing point</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">show.irregulars =</span> <span class="cn">TRUE</span>,         <span class="co"># highlight irregular timing</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">y.axis.label =</span> y.label,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">group.by.days =</span> <span class="cn">TRUE</span>, <span class="at">col.irregular =</span> <span class="fu">alpha</span>(<span class="st">"red"</span>, <span class="fl">0.03</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>                  ) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="cn">NULL</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 1 row containing missing values or values outside the scale range
(`geom_line()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5476 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-irregular" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-irregular-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="supplement_files/figure-html/fig-irregular-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Visualization of gaps and irregular data. Black traces show available data. Red shaded areas show times of missing data. Red dots show instances where observations occur off the regular interval from start to finish, i.e., irregular data."><img src="supplement_files/figure-html/fig-irregular-1.png" class="img-fluid figure-img" width="672"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-irregular-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Visualization of gaps and irregular data. Black traces show available data. Red shaded areas show times of missing data. Red dots show instances where observations occur off the regular interval from start to finish, i.e., irregular data.
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>In <span class="citation" data-cites="fig-irregular">(<a href="#ref-fig-irregular" role="doc-biblioref"><strong>fig-irregular?</strong></a>)</span>, time periods where data are missing appear as red-shaded areas, and any off-schedule observation times are marked with red dots. The <code>Clouclip</code> example shows extensive gaps (red blocks) on certain days and irregular timing on all days except the first and last. These irregular timestamps likely arise from the device’s logging process (e.g.&nbsp;slight clock drift or buffering when the device was turned on/off). Such issues must be addressed before further analysis.</p>
<p>When faced with irregular or gapped data, LightLogR recommends a few strategies:</p>
<ul>
<li><p><em>Remove leading/trailing segments that cause irregularity.</em>&nbsp;For example, if only the first day is regular and subsequent days drift, one might exclude the problematic portion using date filters (see&nbsp;<a href="https://tscnlab.github.io/LightLogR/reference/filter_Datetime.html">filter_Date() / filter_Datetime()</a>&nbsp;in LightLogR).</p></li>
<li><p><em>Round timestamps to the nearest regular interval.</em>&nbsp;This can snap slightly off-schedule times back to the 5-second grid (using&nbsp;<a href="https://tscnlab.github.io/LightLogR/reference/cut_Datetime.html">cut_Datetime()</a>&nbsp;with a 5-second interval), provided the deviations are small and this rounding won’t create duplicate timestamps.</p></li>
<li><p><em>Aggregate to a coarser time interval.</em>&nbsp;For instance, grouping data into 1-minute bins with&nbsp;<a href="https://tscnlab.github.io/LightLogR/reference/aggregate_Datetime.html">aggregate_Datetime()</a> can mask irregularities at finer scales, at the cost of some temporal resolution.</p></li>
</ul>
<p>In this case, the deviations from the 5-second schedule are relatively minor. We choose to&nbsp;<strong>round the timestamps to the nearest 5 seconds</strong>&nbsp;to enforce a uniform sampling grid, which simplifies downstream gap handling. We further add a separate date column for convenience:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [5]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Regularize timestamps by rounding to nearest 5-second interval</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>dataCC <span class="ot">&lt;-</span> dataCC <span class="sc">|&gt;</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cut_Datetime</span>(<span class="st">"5 secs"</span>, <span class="at">New.colname =</span> Datetime) <span class="sc">|&gt;</span>  <span class="co"># round times to 5-second bins</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>)                     <span class="co"># add a Date column for grouping by day</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>After this operation, all&nbsp;<code>Datetime</code>&nbsp;entries in&nbsp;<code>dataCC</code>&nbsp;align perfectly on 5-second boundaries (e.g.&nbsp;<strong>08:00:00</strong>,&nbsp;<strong>08:00:05</strong>,&nbsp;<strong>08:00:10</strong>, etc.). We can verify that no irregular intervals remain by re-running&nbsp;<code>has_irregulars()</code>&nbsp;(it now returns&nbsp;<code>FALSE</code>). Next, we want to quantify the missing data. LightLogR distinguishes between&nbsp;<strong>explicit missing values</strong>&nbsp;(actual&nbsp;<code>NA</code>s in the data, possibly from sentinel replacements or gaps we have filled in) and&nbsp;<strong>implicit missing intervals</strong>&nbsp;(time points where the device&nbsp;<em>should</em>&nbsp;have a reading but none was recorded, and we have not yet filled them in). Initially, many gaps are still implicit (between the first and last timestamp of each day). We can generate a gap summary table:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [6]:</pre></div><div class="cell">

<div class="cell-container"><div class="cell-decorator"><pre>In [7]:</pre></div><div><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize observed vs missing data by day for distance</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">gap_table</span>(Dis, <span class="at">Variable.label =</span> <span class="st">"Distance (cm)"</span>) <span class="sc">|&gt;</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cols_hide</span>(<span class="fu">contains</span>(<span class="st">"_n"</span>))   <span class="co"># hide absolute counts for brevity in output</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div><div id="tbl-gaps" class="cell quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-gaps-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Summary of missing and observed data for the Clouclip device
</figcaption>
<div aria-describedby="tbl-gaps-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div id="enbfpgvfab" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#enbfpgvfab table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#enbfpgvfab thead, #enbfpgvfab tbody, #enbfpgvfab tfoot, #enbfpgvfab tr, #enbfpgvfab td, #enbfpgvfab th {
  border-style: none;
}

#enbfpgvfab p {
  margin: 0;
  padding: 0;
}

#enbfpgvfab .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#enbfpgvfab .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#enbfpgvfab .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#enbfpgvfab .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#enbfpgvfab .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#enbfpgvfab .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#enbfpgvfab .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#enbfpgvfab .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#enbfpgvfab .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#enbfpgvfab .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#enbfpgvfab .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#enbfpgvfab .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#enbfpgvfab .gt_spanner_row {
  border-bottom-style: hidden;
}

#enbfpgvfab .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#enbfpgvfab .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#enbfpgvfab .gt_from_md > :first-child {
  margin-top: 0;
}

#enbfpgvfab .gt_from_md > :last-child {
  margin-bottom: 0;
}

#enbfpgvfab .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#enbfpgvfab .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#enbfpgvfab .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#enbfpgvfab .gt_row_group_first td {
  border-top-width: 2px;
}

#enbfpgvfab .gt_row_group_first th {
  border-top-width: 2px;
}

#enbfpgvfab .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#enbfpgvfab .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#enbfpgvfab .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#enbfpgvfab .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#enbfpgvfab .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#enbfpgvfab .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#enbfpgvfab .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#enbfpgvfab .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#enbfpgvfab .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#enbfpgvfab .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#enbfpgvfab .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#enbfpgvfab .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#enbfpgvfab .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#enbfpgvfab .gt_left {
  text-align: left;
}

#enbfpgvfab .gt_center {
  text-align: center;
}

#enbfpgvfab .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#enbfpgvfab .gt_font_normal {
  font-weight: normal;
}

#enbfpgvfab .gt_font_bold {
  font-weight: bold;
}

#enbfpgvfab .gt_font_italic {
  font-style: italic;
}

#enbfpgvfab .gt_super {
  font-size: 65%;
}

#enbfpgvfab .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#enbfpgvfab .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#enbfpgvfab .gt_indent_1 {
  text-indent: 5px;
}

#enbfpgvfab .gt_indent_2 {
  text-indent: 10px;
}

#enbfpgvfab .gt_indent_3 {
  text-indent: 15px;
}

#enbfpgvfab .gt_indent_4 {
  text-indent: 20px;
}

#enbfpgvfab .gt_indent_5 {
  text-indent: 25px;
}

#enbfpgvfab .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#enbfpgvfab div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table do-not-create-environment cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<colgroup>
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
</colgroup>
<thead>
<tr class="gt_heading header">
<th colspan="14" class="gt_heading gt_title gt_font_normal">Summary of available and missing data</th>
</tr>
<tr class="gt_heading even">
<th colspan="14" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border">Variable: Distance (cm)</th>
</tr>
<tr class="gt_col_headings gt_spanner_row header">
<th data-quarto-table-cell-role="th" scope="col"></th>
<th colspan="5" id="data_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold; font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Data
</div></th>
<th colspan="8" id="missing_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold; font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Missing
</div></th>
</tr>
<tr class="gt_col_headings gt_spanner_row even">
<th rowspan="2" id="a::stub" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col"></th>
<th colspan="2" id="Regular" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Regular
</div></th>
<th id="irregular_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Irregular
</div></th>
<th id="Range" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Range
</div></th>
<th id="interval_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Interval
</div></th>
<th colspan="4" id="gap_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Gaps
</div></th>
<th colspan="2" id="implicit_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Implicit
</div></th>
<th colspan="2" id="explicit_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Explicit
</div></th>
</tr>
<tr class="gt_col_headings header">
<th id="available_data" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="available_data_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="number_irregulars" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">n<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1,2</sup></span></th>
<th id="total_duration" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="interval" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="number_gaps" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">N</th>
<th id="mean_gap_duration" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ø</th>
<th id="missing_data" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_data_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="missing_implicit" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_implicit_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="missing_explicit" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_explicit_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td id="grand_summary_stub_1" class="gt_row gt_left gt_stub gt_grand_summary_row gt_last_grand_summary_row_top" data-quarto-table-cell-role="th" scope="row" style="background-color: #E5E5E5"><strong>Overall</strong></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 available_data" style="background-color: #E5E5E5">2d 14h 43m 10s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 available_data_pct" style="background-color: #E5E5E5">29.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 number_irregulars" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 total_duration" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1w 2d</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 interval" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 number_gaps" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2,690</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 mean_gap_duration" style="background-color: #E5E5E5">1h 35m 57s</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_data" style="background-color: #E5E5E5">6d 9h 16m 50s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_data_pct" style="background-color: #E5E5E5">71.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_implicit" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5d 15h 19m 55s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_implicit_pct" style="background-color: #E5E5E5">62.7%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_explicit" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">17h 56m 55s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_explicit_pct" style="background-color: #E5E5E5">8.3%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="Clouclip - 2021-02-06" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-06</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_1" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 available_data">43m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 available_data_pct">3.0%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">26</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 mean_gap_duration">53m 43s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 missing_data">23h 16m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 missing_data_pct">97.0%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">22h 53m 20s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 missing_implicit_pct">95.4%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-06 stub_1_1 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">23m 10s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-06 stub_1_1 missing_explicit_pct">1.6%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="Clouclip - 2021-02-07" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-07</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_2" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 available_data">2h 45m</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 available_data_pct">11.5%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">139</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 mean_gap_duration">9m 10s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 missing_data">21h 15m</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 missing_data_pct">88.5%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">19h 42m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 missing_implicit_pct">82.1%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-07 stub_1_2 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1h 32m 25s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-07 stub_1_2 missing_explicit_pct">6.4%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="Clouclip - 2021-02-08" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-08</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_3" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 available_data">11h 13m 55s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 available_data_pct">46.8%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">443</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 mean_gap_duration">1m 44s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 missing_data">12h 46m 5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 missing_data_pct">53.2%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">10h 47m 50s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 missing_implicit_pct">45.0%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-08 stub_1_3 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1h 58m 15s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-08 stub_1_3 missing_explicit_pct">8.2%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="Clouclip - 2021-02-09" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-09</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_4" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 available_data">8h 46m 25s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 available_data_pct">36.6%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">278</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 mean_gap_duration">3m 17s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 missing_data">15h 13m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 missing_data_pct">63.4%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">13h 50m</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 missing_implicit_pct">57.6%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-09 stub_1_4 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1h 23m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-09 stub_1_4 missing_explicit_pct">5.8%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="Clouclip - 2021-02-10" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-10</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_5" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 available_data">7h 1m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 available_data_pct">29.3%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">367</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 mean_gap_duration">2m 47s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 missing_data">16h 58m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 missing_data_pct">70.7%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">14h 18m 40s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 missing_implicit_pct">59.6%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-10 stub_1_5 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2h 39m 50s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-10 stub_1_5 missing_explicit_pct">11.1%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="Clouclip - 2021-02-11" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-11</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_6" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 available_data">8h 31m 55s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 available_data_pct">35.5%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">423</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 mean_gap_duration">2m 12s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 missing_data">15h 28m 5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 missing_data_pct">64.5%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">11h 43m 30s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 missing_implicit_pct">48.9%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-11 stub_1_6 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">3h 44m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-11 stub_1_6 missing_explicit_pct">15.6%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="Clouclip - 2021-02-12" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-12</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_7" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 available_data">12h 17m 55s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 available_data_pct">51.2%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">417</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 mean_gap_duration">1m 41s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 missing_data">11h 42m 5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 missing_data_pct">48.8%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">9h 17m 45s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 missing_implicit_pct">38.7%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-12 stub_1_7 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2h 24m 20s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-12 stub_1_7 missing_explicit_pct">10.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="Clouclip - 2021-02-13" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-13</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_8" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 available_data">10h 32m 15s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 available_data_pct">43.9%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">527</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 mean_gap_duration">1m 32s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 missing_data">13h 27m 45s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 missing_data_pct">56.1%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">10h 42m 10s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 missing_implicit_pct">44.6%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-13 stub_1_8 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2h 45m 35s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-13 stub_1_8 missing_explicit_pct">11.5%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="Clouclip - 2021-02-14" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">Clouclip - 2021-02-14</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_9" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 available_data">50m 45s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 available_data_pct">3.5%</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">70</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 mean_gap_duration">19m 51s</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 missing_data">23h 9m 15s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 missing_data_pct">96.5%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">22h 4m 5s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 missing_implicit_pct">92.0%</td>
<td class="gt_row gt_center" headers="Clouclip - 2021-02-14 stub_1_9 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1h 5m 10s</td>
<td class="gt_row gt_right" headers="Clouclip - 2021-02-14 stub_1_9 missing_explicit_pct">4.5%</td>
</tr>
</tbody><tfoot class="gt_footnotes">
<tr class="odd">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span> If n &gt; 0: it is possible that the other summary statistics are affected, as they are calculated based on the most prominent interval.</td>
</tr>
<tr class="even">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>2</sup></span> Number of (missing or actual) observations</td>
</tr>
<tr class="odd">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span> Based on times, not necessarily number of observations</td>
</tr>
</tfoot>

</table>

</div>
</div>
</div>
</figure>
</div></div></div>
</div></div>
<p>This summary (<span class="citation" data-cites="tbl-gaps">(<a href="#ref-tbl-gaps" role="doc-biblioref"><strong>tbl-gaps?</strong></a>)</span>) breaks down, for each day, how much data is present vs.&nbsp;missing. It reports the total duration of recorded data and the duration of gaps. After rounding the times, there are no irregular timestamp issues, but we see substantial <strong>implicit gaps</strong> — periods where the device was not recording (e.g., overnight when the device was likely not worn or was in sleep mode). Notably, the first and last days of the week have very little data (less than 1 hour each), since they probably represent partial recording days (the device was put on and removed on those days).</p>
<p>To prepare the dataset for analysis, we will convert all those implicit gaps into explicit missing entries, and remove days that are mostly incomplete. Converting implicit gaps means inserting rows with <code>NA</code> for each missing 5-second slot, so that the time series becomes continuous and explicit about missingness. We use <code>gap_handler()</code> for this, and then drop the nearly-empty days:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [8]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert implicit gaps to explicit NA gaps, and drop days with &lt;1 hour of data</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>dataCC <span class="ot">&lt;-</span> dataCC <span class="sc">|&gt;</span> </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># First ensure that status columns have an "operational" tag for non-missing periods:</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(Lux_status, Dis_status), <span class="sc">~</span> <span class="fu">replace_na</span>(.x, <span class="st">"operational"</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Fill in all implicit gaps with explicit NA rows (for full days range):</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>  </span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove any day that has less than 1 hour of recorded data</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">remove_partial_data</span>(Dis, <span class="at">threshold.missing =</span> <span class="st">"23 hours"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>After these steps,&nbsp;<code>dataCC</code>&nbsp;contains continuous 5-second timestamps for each day that remains. We chose a threshold of “23 hours” missing to remove days with &lt;1 hour of data, which in this dataset removes the first and last (partial) days. The cleaned <code>Clouclip</code> data now covers six full days with bouts of continuous wear.</p>
<p>It is often helpful to double-check how the sentinel values and missing data are distributed over time. We can visualize the distance time-series with status annotations and day/night periods:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [9]:</pre></div><div id="cell-fig-state" class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#setting coordinates for Houston, Texas</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>coordinates <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">29.75</span>, <span class="sc">-</span><span class="fl">95.36</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize observations</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>dataCC <span class="sc">|&gt;</span> </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fill</span>(<span class="fu">c</span>(Lux_status, Dis_status), <span class="at">.direction =</span> <span class="st">"downup"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gg_day</span>(<span class="at">y.axis =</span> Dis, <span class="at">geom =</span> <span class="st">"line"</span>, <span class="at">y.axis.label =</span> y.label) <span class="sc">|&gt;</span> <span class="co">#create a basic plot</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gg_state</span>(Dis_status, <span class="at">aes_fill =</span> Dis_status) <span class="sc">|&gt;</span> <span class="co">#add the status times</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gg_photoperiod</span>(coordinates, <span class="at">alpha =</span> <span class="fl">0.1</span>, <span class="at">col =</span> <span class="st">"red"</span>) <span class="sc">+</span> <span class="co">#add the photoperiod (day/night)</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 6889 rows containing missing values or values outside the scale range
(`geom_line()`).</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-state" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-state-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="supplement_files/figure-html/fig-state-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Distance measurements across days. Blue, grey and yellow-colored areas show sentinel states of the device. Blue indicates an operational status, grey sleep mode (not recording), and yellow an out of range measurement. Red boxed areas show nighttime from civil dusk until dawn, which are calculated based on the recording date and geographic coordinates"><img src="supplement_files/figure-html/fig-state-1.png" class="img-fluid figure-img" width="672"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-state-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Distance measurements across days. Blue, grey and yellow-colored areas show sentinel states of the device. Blue indicates an operational status, grey sleep mode (not recording), and yellow an out of range measurement. Red boxed areas show nighttime from civil dusk until dawn, which are calculated based on the recording date and geographic coordinates
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>In this plot, blue segments indicate times when the Clouclip was&nbsp;<em>operational</em>&nbsp;(actively measuring), grey segments indicate the device in&nbsp;<em>sleep mode</em>&nbsp;(no recording, typically at night), and yellow segments indicate&nbsp;<em>out-of-range</em>&nbsp;distance readings. The red outlined regions show nighttime (from civil dusk to dawn) based on the given latitude/longitude and dates. As expected, most of the grey “sleep” periods align with night hours, and we see a few yellow spans when the user’s viewing distance exceeded the device’s range (e.g.&nbsp;presumably when no object was within 100&nbsp;cm, such as when the user looked into the distance). At this stage, the Clouclip dataset&nbsp;<code>dataCC</code>&nbsp;is fully preprocessed: all timestamps are regular 5-second intervals, missing data are explicitly marked, extraneous partial days are removed, and sentinel codes are handled via the status columns. The data are ready for calculating daily distance and light exposure metrics (as done in the <a href="https://tscnlab.github.io/ZaunerEtAl_JVis_2025/">main tutorial</a>’s Results).</p>
</section>
<section id="veet-data-ambient-light-illuminance-processing" class="level2">
<h2 class="anchored" data-anchor-id="veet-data-ambient-light-illuminance-processing"><code>VEET</code> Data: Ambient Light (Illuminance) Processing</h2>
<p>The&nbsp;<strong>VEET</strong>&nbsp;device <span class="citation" data-cites="Sah2025OphtalmicPhysiolOpt">Sullivan et al. (<a href="#ref-Sullivan2024" role="doc-biblioref">2024</a>)</span> is a more complex logger that records multiple data modalities in one combined file. Its raw data file contains interleaved records for different sensor types, distinguished by a “modality” field. We focus first on the ambient light sensor modality (abbreviated&nbsp;<strong>ALS</strong>), which provides broad-spectrum illuminance readings (lux) and related information like sensor gains and flicker, recorded every 2 seconds. Later we will import the spectral sensor modality (<strong>PHO</strong>) for spectral irradiance data, and the time-of-flight modality (<strong>TOF</strong>) for distance data.</p>
<p>In the <code>VEET</code>’s export file, each line includes a timestamp and a modality code, followed by fields specific to that modality. Importantly, this means that the <code>VEET</code> export is not rectangular, i.e., tabular. This makes it challenging for many import functions that expect the equal number of columns in every row, which is not the case in this instance. For the&nbsp;<strong>ALS modality</strong>, the relevant columns include a high-resolution timestamp (in Unix epoch format), integration time, UV/VIS/IR sensor gain settings, raw UV/VIS/IR sensor counts, a flicker measurement, and the computed illuminance in lux. For example, the ALS data columns are named:&nbsp;<code>time_stamp</code>,&nbsp;<code>integration_time</code>,&nbsp;<code>uvGain</code>,&nbsp;<code>visGain</code>,&nbsp;<code>irGain</code>,&nbsp;<code>uvValue</code>,&nbsp;<code>visValue</code>,&nbsp;<code>irValue</code>,&nbsp;<code>Flicker</code>, and&nbsp;<code>Lux</code>.</p>
<p>For the&nbsp;<strong>PHO (spectral) modality</strong>, the columns include a timestamp, integration time, a general&nbsp;<code>Gain</code>&nbsp;factor, and nine sensor channel readings covering different wavelengths (with names like&nbsp;<code>s415, s445, ..., s680, s940</code>) as well as a&nbsp;<code>Dark</code>&nbsp;channel and two broadband channels&nbsp;<code>ClearL</code>&nbsp;and&nbsp;<code>ClearR</code>. In essence, the <code>VEET</code>’s spectral sensor captures light in several wavelength bands (from ~415&nbsp;nm up to 940&nbsp;nm, plus an infrared and two “clear” channels) rather than outputting a single lux value like the ambient light sensor does (<em>PHO</em>).</p>
<p>To import the <code>VEET</code> ambient light data, we again use the LightLogR import function, specifying the <code>ALS</code> modality. The raw VEET data in our example is provided as a zip file (<code>01_VEET_L.csv.zip</code>) containing the logged data for one week. We do the following:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [10]:</pre></div><div id="cell-fig-VEET-overview" class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import VEET Ambient Light Sensor (ALS) data</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span>    <span class="co"># assuming device clock was set to US Central, for consistency</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>dataVEET <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"ALS"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Multiple files in zip: reading '01_VEET_L.csv'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: There was 1 warning in `dplyr::mutate()`.
ℹ In argument: `dplyr::across(...)`.
Caused by warning:
! NAs introduced by coercion</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Successfully read in 304'193 observations across 1 Ids from 1 VEET-file(s).
Timezone set is US/Central.
The system timezone is Europe/Berlin. Please correct if necessary!
1 observations were dropped due to a missing or non-parseable Datetime value (e.g., non-valid timestamps during DST jumps). 

First Observation: 2024-06-04 15:00:37
Last Observation: 2024-06-12 08:29:43
Timespan: 7.7 days

Observation intervals: 
  Id    interval.time              n pct      
1 VEET  0s                         1 0.00033% 
2 VEET  1s                      1957 0.64334% 
3 VEET  2s                    300147 98.67025%
4 VEET  3s                      2074 0.68181% 
5 VEET  4s                         3 0.00099% 
6 VEET  9s                         5 0.00164% 
7 VEET  10s                        3 0.00099% 
8 VEET  109s (~1.82 minutes)       1 0.00033% 
9 VEET  59077s (~16.41 hours)      1 0.00033% </code></pre>
</div>
<div class="cell-output-display">
<div id="fig-VEET-overview" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-overview-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="supplement_files/figure-html/fig-VEET-overview-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Overview plot of imported VEET data"><img src="supplement_files/figure-html/fig-VEET-overview-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-overview-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Overview plot of imported VEET data
</figcaption>
</figure>
</div>
</div>
</div></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>We get one warning as a single time stamp could not be parsed into a datetime - with ~300k observations, this is not an issue.</p>
</div>
</div>
<p>This call reads in only the lines corresponding to the <code>ALS</code> modality from the <code>VEET</code> file. The result&nbsp;<code>dataVEET</code>&nbsp;is a tibble<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> with columns such as&nbsp;<code>Datetime</code>&nbsp;(parsed from the&nbsp;<code>time_stamp</code>&nbsp;epoch to POSIXct in US/Central time),&nbsp;<code>Lux</code>&nbsp;(illuminance in lux),&nbsp;<code>Flicker</code>, and the various sensor gains/values. Unneeded columns like the modality code or file name are also included but can be ignored or removed. After import, LightLogR again provides an overview of the data. We learn that the <code>VEET</code> light data, like the <code>Clouclip</code>, also exhibits irregularities and gaps. (The device nominally records every 2 seconds, but timing may drift or pause when not worn.)</p>
<p>To make the VEET light data comparable to the Clouclip’s and to simplify analysis, we choose to&nbsp;<strong>aggregate the VEET illuminance data to 5-second intervals</strong>. This slight downsampling will both reduce data volume and help align with the Clouclip’s timeline for any combined analysis. Before aggregation, we will also explicitly mark gaps in the VEET data so that missing intervals are not overlooked.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [11]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Aggregate VEET light data to 5-second intervals and mark gaps</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>dataVEET <span class="ot">&lt;-</span> dataVEET <span class="sc">|&gt;</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 seconds"</span>) <span class="sc">|&gt;</span>  <span class="co"># resample to 5-sec bins (e.g. average Lux over 2-sec readings)</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>          <span class="co"># fill in implicit gaps with NA rows</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>          <span class="co"># add Date column for daily grouping</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">remove_partial_data</span>(Lux, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>First, <code>aggregate_Datetime(unit = "5 seconds")</code> combines the high-frequency 2-second observations into 5-second slots. By default, this function will average numeric columns like&nbsp;<code>Lux</code>&nbsp;over each 5-second period (and handle counts or categorical appropriately). All of these <em>data type handlers</em> can be changed with the function call. The result is that&nbsp;<code>dataVEET</code>&nbsp;now has a reading every 5 seconds (or an NA if none were present in that window). Next,&nbsp;<code>gap_handler(full.days = TRUE)</code>&nbsp;inserts explicit NA entries for any 5-second timestamp that had no data within the continuous span of the recording. Then we add a&nbsp;<code>Date</code>&nbsp;column for grouping, and finally we remove days with more than 1 hour of missing data (using a more strict criterion as we did for <code>Clouclip</code>). According to the gap summary (<span class="citation" data-cites="tbl-gaps2">(<a href="#ref-tbl-gaps2" role="doc-biblioref"><strong>tbl-gaps2?</strong></a>)</span>), this leaves six full days of <code>VEET</code> light data with good coverage, after dropping the very incomplete start/end days.</p>
<p>We can inspect the missing-data summary for the <code>VEET</code> illuminance data:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [12]:</pre></div><div class="cell">

<div class="cell-container"><div class="cell-decorator"><pre>In [13]:</pre></div><div><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>dataVEET <span class="sc">|&gt;</span> <span class="fu">gap_table</span>(Lux, <span class="st">"Illuminance (lx)"</span>) <span class="sc">|&gt;</span>   </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cols_hide</span>(<span class="fu">contains</span>(<span class="st">"_n"</span>)) <span class="co">#remove the absolute number of data points</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div><div id="tbl-gaps2" class="cell quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-gaps2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Summary of missing and observed data for the VEET device
</figcaption>
<div aria-describedby="tbl-gaps2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div id="hslfxitads" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#hslfxitads table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#hslfxitads thead, #hslfxitads tbody, #hslfxitads tfoot, #hslfxitads tr, #hslfxitads td, #hslfxitads th {
  border-style: none;
}

#hslfxitads p {
  margin: 0;
  padding: 0;
}

#hslfxitads .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#hslfxitads .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#hslfxitads .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#hslfxitads .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#hslfxitads .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#hslfxitads .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#hslfxitads .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#hslfxitads .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#hslfxitads .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#hslfxitads .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#hslfxitads .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#hslfxitads .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#hslfxitads .gt_spanner_row {
  border-bottom-style: hidden;
}

#hslfxitads .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#hslfxitads .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#hslfxitads .gt_from_md > :first-child {
  margin-top: 0;
}

#hslfxitads .gt_from_md > :last-child {
  margin-bottom: 0;
}

#hslfxitads .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#hslfxitads .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#hslfxitads .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#hslfxitads .gt_row_group_first td {
  border-top-width: 2px;
}

#hslfxitads .gt_row_group_first th {
  border-top-width: 2px;
}

#hslfxitads .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#hslfxitads .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#hslfxitads .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#hslfxitads .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#hslfxitads .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#hslfxitads .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#hslfxitads .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#hslfxitads .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#hslfxitads .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#hslfxitads .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#hslfxitads .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#hslfxitads .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#hslfxitads .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#hslfxitads .gt_left {
  text-align: left;
}

#hslfxitads .gt_center {
  text-align: center;
}

#hslfxitads .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#hslfxitads .gt_font_normal {
  font-weight: normal;
}

#hslfxitads .gt_font_bold {
  font-weight: bold;
}

#hslfxitads .gt_font_italic {
  font-style: italic;
}

#hslfxitads .gt_super {
  font-size: 65%;
}

#hslfxitads .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#hslfxitads .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#hslfxitads .gt_indent_1 {
  text-indent: 5px;
}

#hslfxitads .gt_indent_2 {
  text-indent: 10px;
}

#hslfxitads .gt_indent_3 {
  text-indent: 15px;
}

#hslfxitads .gt_indent_4 {
  text-indent: 20px;
}

#hslfxitads .gt_indent_5 {
  text-indent: 25px;
}

#hslfxitads .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#hslfxitads div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table do-not-create-environment cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<colgroup>
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
</colgroup>
<thead>
<tr class="gt_heading header">
<th colspan="14" class="gt_heading gt_title gt_font_normal">Summary of available and missing data</th>
</tr>
<tr class="gt_heading even">
<th colspan="14" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border">Variable: Illuminance (lx)</th>
</tr>
<tr class="gt_col_headings gt_spanner_row header">
<th data-quarto-table-cell-role="th" scope="col"></th>
<th colspan="5" id="data_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold; font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Data
</div></th>
<th colspan="8" id="missing_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold; font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Missing
</div></th>
</tr>
<tr class="gt_col_headings gt_spanner_row even">
<th rowspan="2" id="a::stub" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col"></th>
<th colspan="2" id="Regular" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Regular
</div></th>
<th id="irregular_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Irregular
</div></th>
<th id="Range" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Range
</div></th>
<th id="interval_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="col"><div class="gt_column_spanner">
Interval
</div></th>
<th colspan="4" id="gap_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Gaps
</div></th>
<th colspan="2" id="implicit_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Implicit
</div></th>
<th colspan="2" id="explicit_spanner" class="gt_center gt_columns_top_border gt_column_spanner_outer" data-quarto-table-cell-role="th" style="font-weight: bold" scope="colgroup"><div class="gt_column_spanner">
Explicit
</div></th>
</tr>
<tr class="gt_col_headings header">
<th id="available_data" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="available_data_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="number_irregulars" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">n<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1,2</sup></span></th>
<th id="total_duration" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="interval" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="number_gaps" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">N</th>
<th id="mean_gap_duration" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ø</th>
<th id="missing_data" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_data_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="missing_implicit" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_implicit_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
<th id="missing_explicit" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Time</th>
<th id="missing_explicit_pct" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">%</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td id="grand_summary_stub_1" class="gt_row gt_left gt_stub gt_grand_summary_row gt_last_grand_summary_row_top" data-quarto-table-cell-role="th" scope="row" style="background-color: #E5E5E5"><strong>Overall</strong></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 available_data" style="background-color: #E5E5E5">5d 23h 57m 40s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 available_data_pct" style="background-color: #E5E5E5">100.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 number_irregulars" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 total_duration" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">6d</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 interval" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 number_gaps" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">8</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 mean_gap_duration" style="background-color: #E5E5E5">58s</td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_data" style="background-color: #E5E5E5">2m 20s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_data_pct" style="background-color: #E5E5E5">0.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_implicit" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_implicit_pct" style="background-color: #E5E5E5">0.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
<td class="gt_row gt_center gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_explicit" style="background-color: #E5E5E5; border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2m 20s</td>
<td class="gt_row gt_right gt_grand_summary_row gt_last_grand_summary_row_top" headers="grand_summary_stub_1 missing_explicit_pct" style="background-color: #E5E5E5">0.0%<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span></td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="VEET - 2024-06-06" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-06</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_1" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 available_data">23h 58m 5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 available_data_pct">99.9%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">3</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 mean_gap_duration">38s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 missing_data">1m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 missing_data_pct">0.1%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-06 stub_1_1 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-06 stub_1_1 missing_explicit_pct">0.1%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="VEET - 2024-06-07" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-07</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_2" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 available_data">1d</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 mean_gap_duration">0s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 missing_data">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-07 stub_1_2 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-07 stub_1_2 missing_explicit_pct">0.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="VEET - 2024-06-08" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-08</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_3" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 available_data">23h 59m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 mean_gap_duration">5s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 missing_data">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-08 stub_1_3 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-08 stub_1_3 missing_explicit_pct">0.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="VEET - 2024-06-09" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-09</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_4" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 available_data">23h 59m 50s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">2</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 mean_gap_duration">5s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 missing_data">10s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-09 stub_1_4 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">10s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-09 stub_1_4 missing_explicit_pct">0.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="VEET - 2024-06-10" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-10</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_5" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 available_data">23h 59m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 mean_gap_duration">5s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 missing_data">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-10 stub_1_5 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-10 stub_1_5 missing_explicit_pct">0.0%</td>
</tr>
<tr class="gt_group_heading_row even">
<td colspan="14" id="VEET - 2024-06-11" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2024-06-11</td>
</tr>
<tr class="gt_row_group_first odd">
<td id="stub_1_6" class="gt_row gt_left gt_stub" data-quarto-table-cell-role="th" scope="row"></td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 available_data">23h 59m 55s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 available_data_pct">100.0%</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 number_irregulars" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 total_duration" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1d</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 interval" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 number_gaps" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">1</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 mean_gap_duration">5s</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 missing_data">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 missing_data_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 missing_implicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">0s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 missing_implicit_pct">0.0%</td>
<td class="gt_row gt_center" headers="VEET - 2024-06-11 stub_1_6 missing_explicit" style="border-left-width: 1px; border-left-style: solid; border-left-color: grey80">5s</td>
<td class="gt_row gt_right" headers="VEET - 2024-06-11 stub_1_6 missing_explicit_pct">0.0%</td>
</tr>
</tbody><tfoot class="gt_footnotes">
<tr class="odd">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span> If n &gt; 0: it is possible that the other summary statistics are affected, as they are calculated based on the most prominent interval.</td>
</tr>
<tr class="even">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>2</sup></span> Number of (missing or actual) observations</td>
</tr>
<tr class="odd">
<td colspan="14" class="gt_footnote"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>3</sup></span> Based on times, not necessarily number of observations</td>
</tr>
</tfoot>

</table>

</div>
</div>
</div>
</figure>
</div></div></div>
</div></div>
<p><span class="citation" data-cites="tbl-gaps2">(<a href="#ref-tbl-gaps2" role="doc-biblioref"><strong>tbl-gaps2?</strong></a>)</span> shows, for each retained day, the total recorded duration and the duration of gaps. The <code>VEET</code> device, like the <code>Clouclip</code>, was not worn continuously 24&nbsp;hours per day, so there are nightly gaps of wear time (when the device was likely off the participant), but not of recordings. After our preprocessing, any implicit gaps are represented as explicit missing intervals. The <code>VEET</code>’s time sampling was originally more frequent, but by aggregating to 5&nbsp;s we have ensured a uniform timeline akin to the <code>Clouclip</code>’s.</p>
<p>At this point, the&nbsp;<code>dataVEET</code>&nbsp;object (for illuminance) is cleaned and ready for computing light exposure metrics. For example, one could calculate daily mean illuminance or the duration spent above certain light thresholds (e.g.&nbsp;“outdoor light exposure” defined as &gt;1000&nbsp;lx) using this dataset. Indeed, basic summary tables in the <a href="https://tscnlab.github.io/ZaunerEtAl_JVis_2025/">main tutorial</a> illustrate the highly skewed nature of light exposure data and the calculation of outdoor light metrics. We will not repeat those metric calculations here in the supplement, as our focus is on data preprocessing; however, having a cleaned, gap-marked time series is crucial for those metrics to be accurate.</p>
</section>
<section id="veet-data-spectral-data-processing" class="level2">
<h2 class="anchored" data-anchor-id="veet-data-spectral-data-processing"><code>VEET</code> Data: Spectral Data Processing</h2>
<section id="import" class="level3">
<h3 class="anchored" data-anchor-id="import">Import</h3>
<p>In addition to broad-band illuminance and distance, the <code>VEET</code> provides&nbsp;<strong>spectral sensor data</strong>&nbsp;through its <code>PHO</code> modality. Unlike illuminance, the spectral data are not directly given as directly interpretable radiometric metrics but rather as raw sensor counts across multiple wavelength channels, which require conversion to reconstruct a spectral power distribution. In our analysis, spectral data allow us to compute metrics like the relative contribution of short-wavelength (blue) light versus long-wavelength light in the participant’s environment. Processing this spectral data involves several necessary steps.</p>
<p>First, we import the spectral modality from a second <code>VEET</code> file. This device used the latest software version to allow for the best spectral reconstruction. This time we need to extract the lines marked as <code>PHO</code>. We will store the spectral dataset in a separate object&nbsp;<code>dataVEET2</code>&nbsp;so as not to overwrite the&nbsp;<code>dataVEET</code>&nbsp;illuminance data in our R session:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [14]:</pre></div><div id="cell-fig-VEET-overview2" class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import VEET Spectral Sensor (PHO) data</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/02_VEET_L.csv.zip"</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co">#if a version of LightLogR ≤ 0.9.2 is used, this script needs to be imported, as the device data was collected with a newer firmware version that changes the output format.</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"scripts/VEET_import.R"</span>)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"PHO"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Multiple files in zip: reading '02_VEET_L.csv'</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Successfully read in 173'013 observations across 1 Ids from 1 VEET-file(s).
Timezone set is US/Central.
The system timezone is Europe/Berlin. Please correct if necessary!

First Observation: 2025-06-17 12:25:13
Last Observation: 2025-06-21 22:47:01
Timespan: 4.4 days

Observation intervals: 
   Id    interval.time      n pct      
 1 VEET  1s               417 0.24102% 
 2 VEET  2s            171837 99.32086%
 3 VEET  3s               738 0.42656% 
 4 VEET  4s                 7 0.00405% 
 5 VEET  9s                 1 0.00058% 
 6 VEET  11s                1 0.00058% 
 7 VEET  12s                2 0.00116% 
 8 VEET  13s                4 0.00231% 
 9 VEET  15s                1 0.00058% 
10 VEET  16s                1 0.00058% 
# ℹ 3 more rows</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-VEET-overview2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-overview2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="supplement_files/figure-html/fig-VEET-overview2-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="Overview plot of imported VEET data"><img src="supplement_files/figure-html/fig-VEET-overview2-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-overview2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Overview plot of imported VEET data
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>After import,&nbsp;<code>dataVEET2</code>&nbsp;contains columns for the timestamp (<code>Datetime</code>),&nbsp;<code>Gain</code>&nbsp;(the sensor gain setting), and the nine spectral sensor channels plus a clear channel. These appear as numeric columns named&nbsp;<code>s415, s445, ..., s940, Dark, Clear</code>. The spectral sensor was logging at a 2-second rate. Because this dataset is even denser and more high-dimensional, we will&nbsp;<strong>aggregate it to a 5-minute interval</strong> for computational efficiency. The assumption is that spectral composition does not need to be examined at every 2-second instant for our purposes, and 5-minute averages can capture the general trends while drastically reducing data size and downstream computational costs.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [15]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Aggregate spectral data to 5-minute intervals and mark gaps</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="ot">&lt;-</span> dataVEET2 <span class="sc">|&gt;</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 mins"</span>) <span class="sc">|&gt;</span>     <span class="co"># aggregate to 5-minute bins</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>           <span class="co"># explicit NA for any gaps in between</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">remove_partial_data</span>(Clear, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>We aggregate over 5-minute windows; within each 5-minute bin, multiple spectral readings (if present) are combined (averaged). We use one of the channels (here&nbsp;<code>Clear</code>) as the reference variable for&nbsp;<code>remove_partial_data</code>&nbsp;to drop incomplete days (the choice of channel is arbitrary as all channels share the same level of completeness).</p>
<p>It is informative to look at a snippet of the imported spectral data before further processing. <span class="citation" data-cites="tbl-PHO">(<a href="#ref-tbl-PHO" role="doc-biblioref"><strong>tbl-PHO?</strong></a>)</span> shows the first three rows of&nbsp;<code>dataVEET2</code>&nbsp;after import (before calibration), with some technical columns omitted for brevity:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [16]:</pre></div><div class="cell">

<div class="cell-container"><div class="cell-decorator"><pre>In [17]:</pre></div><div><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="dv">3</span>) <span class="sc">|&gt;</span> <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(modality, file.name, is.implicit, time_stamp)) <span class="sc">|&gt;</span> </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gt</span>() <span class="sc">|&gt;</span> <span class="fu">fmt_number</span>(s415<span class="sc">:</span>Clear) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div><div id="tbl-PHO" class="cell quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-PHO-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Overview of the spectral sensor import from the VEET device (first 3 observations). Each row corresponds to a 5-minute timestamp (Datetime) and shows the raw sensor readings for the spectral channels (s415–s940, Dark, Clear). All values are in arbitrary sensor units (counts). Gain values and integration_time are also relevant for each interval, depending on the downstream computation.
</figcaption>
<div aria-describedby="tbl-PHO-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div id="bjzbszcryp" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#bjzbszcryp table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#bjzbszcryp thead, #bjzbszcryp tbody, #bjzbszcryp tfoot, #bjzbszcryp tr, #bjzbszcryp td, #bjzbszcryp th {
  border-style: none;
}

#bjzbszcryp p {
  margin: 0;
  padding: 0;
}

#bjzbszcryp .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#bjzbszcryp .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#bjzbszcryp .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#bjzbszcryp .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#bjzbszcryp .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#bjzbszcryp .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bjzbszcryp .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#bjzbszcryp .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#bjzbszcryp .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#bjzbszcryp .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#bjzbszcryp .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#bjzbszcryp .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#bjzbszcryp .gt_spanner_row {
  border-bottom-style: hidden;
}

#bjzbszcryp .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#bjzbszcryp .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#bjzbszcryp .gt_from_md > :first-child {
  margin-top: 0;
}

#bjzbszcryp .gt_from_md > :last-child {
  margin-bottom: 0;
}

#bjzbszcryp .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#bjzbszcryp .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#bjzbszcryp .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#bjzbszcryp .gt_row_group_first td {
  border-top-width: 2px;
}

#bjzbszcryp .gt_row_group_first th {
  border-top-width: 2px;
}

#bjzbszcryp .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#bjzbszcryp .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#bjzbszcryp .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#bjzbszcryp .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bjzbszcryp .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#bjzbszcryp .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#bjzbszcryp .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#bjzbszcryp .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#bjzbszcryp .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bjzbszcryp .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#bjzbszcryp .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#bjzbszcryp .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#bjzbszcryp .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#bjzbszcryp .gt_left {
  text-align: left;
}

#bjzbszcryp .gt_center {
  text-align: center;
}

#bjzbszcryp .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#bjzbszcryp .gt_font_normal {
  font-weight: normal;
}

#bjzbszcryp .gt_font_bold {
  font-weight: bold;
}

#bjzbszcryp .gt_font_italic {
  font-style: italic;
}

#bjzbszcryp .gt_super {
  font-size: 65%;
}

#bjzbszcryp .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#bjzbszcryp .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#bjzbszcryp .gt_indent_1 {
  text-indent: 5px;
}

#bjzbszcryp .gt_indent_2 {
  text-indent: 10px;
}

#bjzbszcryp .gt_indent_3 {
  text-indent: 15px;
}

#bjzbszcryp .gt_indent_4 {
  text-indent: 20px;
}

#bjzbszcryp .gt_indent_5 {
  text-indent: 25px;
}

#bjzbszcryp .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#bjzbszcryp div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table do-not-create-environment cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="gt_col_headings header">
<th id="Datetime" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Datetime</th>
<th id="integration_time" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">integration_time</th>
<th id="Gain" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Gain</th>
<th id="s415" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s415</th>
<th id="s445" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s445</th>
<th id="s480" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s480</th>
<th id="s515" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s515</th>
<th id="s555" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s555</th>
<th id="s590" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s590</th>
<th id="s630" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s630</th>
<th id="s680" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s680</th>
<th id="s910" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">s910</th>
<th id="Dark" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Dark</th>
<th id="Clear" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Clear</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="gt_group_heading_row odd">
<td colspan="14" id="VEET - 2025-06-18" class="gt_group_heading" data-quarto-table-cell-role="th" scope="colgroup">VEET - 2025-06-18</td>
</tr>
<tr class="gt_row_group_first even">
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Datetime">2025-06-18</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  integration_time">100</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Gain">512</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s415">20.24</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s445">26.95</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s480">30.02</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s515">40.85</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s555">46.83</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s590">63.02</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s630">89.98</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s680">138.35</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s910">736.04</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Dark">0.00</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Clear">525.21</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Datetime">2025-06-18 00:05:00</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  integration_time">100</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Gain">512</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s415">20.93</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s445">27.63</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s480">30.95</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s515">41.99</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s555">47.81</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s590">64.58</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s630">91.27</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s680">140.51</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s910">774.85</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Dark">0.00</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Clear">537.75</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Datetime">2025-06-18 00:10:00</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  integration_time">100</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Gain">512</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s415">20.86</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s445">27.47</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s480">30.60</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s515">41.31</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s555">46.90</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s590">63.20</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s630">89.92</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s680">139.66</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  s910">757.36</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Dark">0.00</td>
<td class="gt_row gt_right" headers="VEET - 2025-06-18  Clear">534.43</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</figure>
</div></div></div>
</div></div>
</section>
<section id="spectral-calibration" class="level3">
<h3 class="anchored" data-anchor-id="spectral-calibration">Spectral calibration</h3>
<p>Now we proceed with&nbsp;<strong>spectral calibration</strong>. The <code>VEET</code>’s spectral sensor counts need to be converted to physical units (spectral irradiance) via a calibration matrix provided by the manufacturer. For this example, we assume we have a calibration matrix that maps all the channel readings to an estimated spectral power distribution (SPD). The LightLogR package provides a function&nbsp;<code>spectral_reconstruction()</code>&nbsp;to perform this conversion. However, before applying it, we must ensure the sensor counts are in a normalized form. This procedure is laid out by the manufacturer. In our version, we refer to the <em>VEET SPD Reconstruction Guide.pdf</em>, version <em>06/05/2025</em>. Note that each manufacturer has to specify the method of count normalization (if any) and spectral reconstruction. In our raw data, each observation comes with a&nbsp;<code>Gain</code>&nbsp;setting that indicates how the sensor’s sensitivity was adjusted; we need to divide the raw counts by the gain to get normalized counts. LightLogR offers&nbsp;<code>normalize_counts()</code>&nbsp;for this purpose. We further need to scale by integration time (in milliseconds) and adjust depending on counts in the <code>Dark</code> sensor channel.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [18]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>count.columns <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"s415"</span>, <span class="st">"s445"</span>, <span class="st">"s480"</span>, <span class="st">"s515"</span>, <span class="st">"s555"</span>, <span class="st">"s590"</span>, <span class="st">"s630"</span>, </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"s680"</span>, <span class="st">"s910"</span>, <span class="st">"Dark"</span>, <span class="st">"Clear"</span>) <span class="co">#column names</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>gain.ratios <span class="ot">&lt;-</span> <span class="co">#gain ratios as specified by the manufacturers reconstruction guide</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">gain =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">8</span>, <span class="dv">16</span>, <span class="dv">32</span>, <span class="dv">64</span>, <span class="dv">128</span>, <span class="dv">256</span>, <span class="dv">512</span>),</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">gain.ratio =</span> <span class="fu">c</span>(<span class="fl">0.008</span>, <span class="fl">0.016</span>, <span class="fl">0.032</span>, <span class="fl">0.065</span>, <span class="fl">0.125</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="fl">3.95</span>, <span class="fl">7.75</span>)</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  ) </span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="co">#normalize data</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="ot">&lt;-</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>  dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(s415<span class="sc">:</span>Clear), \(x) (x <span class="sc">-</span> Dark)<span class="sc">/</span>integration_time)) <span class="sc">|&gt;</span>  <span class="co">#remove dark counts &amp; scale by integration time</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">normalize_counts</span>( <span class="co">#function to normalize counts</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">gain.columns =</span> <span class="fu">rep</span>(<span class="st">"Gain"</span>, <span class="dv">11</span>), <span class="co">#all sensor channels share the gain value</span></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">count.columns =</span> count.columns, <span class="co">#sensor channels to normalize</span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>    gain.ratios <span class="co">#gain ratios</span></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(s415<span class="sc">:</span>Clear)) <span class="sc">|&gt;</span> <span class="co"># drop original raw count columns</span></span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename_with</span>(<span class="sc">~</span> <span class="fu">str_remove</span>(.x, <span class="st">".normalized"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>In this call, we specified <code>gain.columns = rep("Gain", 11)</code> because we have 11 sensor columns that all use the same gain factor column (<code>Gain</code>). This step will add new columns (with a suffix, e.g.&nbsp;<code>.normalized</code>) for each spectral channel representing the count normalized by the gain. We then dropped the <code>raw</code> count columns and renamed the normalized ones by dropping <code>.normalized</code> from the names. After this, <code>dataVEET2</code> contains the normalized sensor readings for <code>s415, s445, ..., s940, Dark, Clear</code> for each 5-minute time point. The dataset is now ready for spectral reconstruction.</p>
</section>
<section id="spectral-reconstruction" class="level3">
<h3 class="anchored" data-anchor-id="spectral-reconstruction">Spectral reconstruction</h3>
<p>For spectral reconstruction, we require a calibration matrix that corresponds to the VEET’s sensor channels. This matrix would typically be obtained from the device manufacturer or a calibration procedure. It defines how each channel’s count relates to intensity at various wavelengths. For demonstration, the calibration matrix was provided by the manufacturer and is specific to the make and model. It should not be used for research purposes without confirming its accuracy with the manufacturer.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [19]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co">#import calibration matrix</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>calib_mtx <span class="ot">&lt;-</span> </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read_csv</span>(<span class="st">"data/VEET_calibration_matrix.csv"</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"wavelength"</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Reconstruct spectral power distribution (SPD) for each observation</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="ot">&lt;-</span> dataVEET2 <span class="sc">|&gt;</span> <span class="fu">mutate</span>(</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">Spectrum =</span> <span class="fu">spectral_reconstruction</span>(</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pick</span>(s415<span class="sc">:</span>s910),   <span class="co"># pick the normalized sensor columns</span></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">calibration_matrix =</span> calib_mtx, </span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"long"</span> <span class="co"># return a long-form list column (wavelength, intensity)</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>Here, we use&nbsp;<code>format = "long"</code>&nbsp;so that the result for each observation is a&nbsp;<strong>list-column</strong>&nbsp;<code>Spectrum</code>, where each entry is a tibble<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a> containing two columns:&nbsp;<code>wavelength</code>&nbsp;and&nbsp;<code>irradiance</code>&nbsp;(one row per wavelength in the calibration matrix). In other words, each row of&nbsp;<code>dataVEET2</code>&nbsp;now holds a full reconstructed spectrum in the&nbsp;<code>Spectrum</code>&nbsp;column. The long format is convenient for further calculations and plotting. (An alternative&nbsp;<code>format = "wide"</code>&nbsp;would add each wavelength as a separate column, but that is less practical when there are many wavelengths.)</p>
<p>To visualize the data we will calculate the photopic illuminance based on the spectra and plot each spectrum color-scaled by their illuminance. For clarity, we reduce the data to observations within the first day.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [20]:</pre></div><div id="cell-fig-VEET-spectra" class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_Date</span>(<span class="at">length =</span> <span class="st">"1 day"</span>) <span class="sc">|&gt;</span> <span class="co">#keep only observations for one day (from start)</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>( </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">Illuminance =</span> Spectrum <span class="sc">|&gt;</span> <span class="co">#Use the spectrum,...</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">map_dbl</span>(spectral_integration, <span class="co">#... call the function spectral_integration for each,...</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">action.spectrum =</span> <span class="st">"photopic"</span>, <span class="co">#... use the brightness sensitivity function,...</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">general.weight =</span> <span class="st">"auto"</span>) <span class="co">#... and apply the appropriate efficacy weight.</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(Spectrum) <span class="sc">|&gt;</span> <span class="co">#create a long format of the data where the spectrum is unnested</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> wavelength, <span class="at">y =</span> irradiance<span class="sc">*</span><span class="dv">1000</span>, <span class="at">group =</span> Datetime)) <span class="sc">+</span></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">col =</span> Illuminance)) <span class="sc">+</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Irradiance (mW/m²/nm)"</span>, <span class="at">x =</span> <span class="st">"Wavelength (nm)"</span>, <span class="at">col =</span> <span class="st">"Photopic illuminance (lx)"</span>)<span class="sc">+</span></span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_b</span>() <span class="sc">+</span></span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">5.3</span>)) <span class="sc">+</span></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 29289 rows containing missing values or values outside the scale range
(`geom_line()`).</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-VEET-spectra" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-spectra-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="supplement_files/figure-html/fig-VEET-spectra-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6" title="Overview of the reconstructed spectra, color-scaled by photopic illuminance (lx)"><img src="supplement_files/figure-html/fig-VEET-spectra-1.png" class="img-fluid figure-img" width="768"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-spectra-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Overview of the reconstructed spectra, color-scaled by photopic illuminance (lx)
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>At this stage, the&nbsp;<code>dataVEET2</code>&nbsp;dataset has been processed to yield time-series of spectral power distributions. We can use these to compute biologically relevant light metrics. For instance, one possible metric is the proportion of power in short wavelengths versus long wavelengths. In the main analysis, we defined&nbsp;<strong>short-wavelength (blue light) content</strong>&nbsp;as the integrated intensity in the 400–500&nbsp;nm range, and&nbsp;<strong>long-wavelength content</strong>&nbsp;as the integrated intensity in a longer range (e.g.&nbsp;600–700&nbsp;nm), then computed the&nbsp;<em>short-to-long ratio</em>&nbsp;(“sl ratio”). Calculating these metrics is the first step of spectrum analysis in the <a href="https://tscnlab.github.io/ZaunerEtAl_JVis_2025/#spectrum">main tutorial</a>.</p>
<pre><code>dataVEET2 |&gt; 
  select(Id, Date, Datetime, Spectrum) |&gt;    # focus on ID, date, time, and spectrum
  mutate(
    short = Spectrum |&gt; map_dbl(spectral_integration, wavelength.range = c(400, 500)),
    long  = Spectrum |&gt; map_dbl(spectral_integration, wavelength.range = c(600, 700)),
    `sl ratio` = short / long   # compute short-to-long ratio
  )</code></pre>
<p><em>(The cutoff of 500&nbsp;nm here is hypothetical for demonstration; actual definitions might vary.)</em>&nbsp;We would then have columns&nbsp;<code>short</code>,&nbsp;<code>long</code>, and&nbsp;<code>sl_ratio</code>&nbsp;for each observation, which could be averaged per day or analyzed further. The cleaned spectral data in&nbsp;<code>dataVEET2</code>&nbsp;makes it straightforward to calculate such metrics or apply spectral weighting functions (for melatonin suppression, circadian stimulus, etc., if one has the spectral sensitivity curves).</p>
<p>With the <code>VEET</code> spectral preprocessing complete, we emphasize that these steps – normaliziing by gain, applying calibration, and perhaps simplifying channels – are&nbsp;<strong>device-specific</strong>&nbsp;requirements. They ensure that the raw sensor counts are translated into meaningful physical measures (like spectral irradiance). Researchers using other spectral devices would follow a similar procedure, adjusting for their device’s particulars (some may output spectra directly, whereas others, like <code>VEET</code>, require reconstruction.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Some devices may output normalized counts instead of raw counts. For example, the <code>ActLumus</code> device outputs normalized counts, while the <code>VEET</code> device records raw counts and the gain. Manufacturers will be able to speficy exact outputs for a given model and software version.</p>
</div>
</div>
</section>
</section>
<section id="veet-data-time-of-flight-distance" class="level2">
<h2 class="anchored" data-anchor-id="veet-data-time-of-flight-distance"><code>VEET</code> Data: Time of flight (distance)</h2>
<p>In this last section, the distance data of the <code>VEET</code> device will be imported, analogous to the other modalities. The <code>TOF</code> modality contains information for up to two objects in a 8x8 grid of measurements, spanning a total of about 52° vertically and 41° horizontally. Because the <code>VEET</code> device can detect up to two objects in a given grid point, and there is a confidence value assigned to every measurement, each observation contains <span class="math inline">\(2*2*8*8 = 256\)</span> measurements.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [21]:</pre></div><div id="cell-fig-VEET-overview3" class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import VEET Spectral Sensor (TOF) data</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="ot">&lt;-</span> LightLogR<span class="sc">::</span>import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"TOF"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Multiple files in zip: reading '01_VEET_L.csv'</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Successfully read in 304'195 observations across 1 Ids from 1 VEET-file(s).
Timezone set is US/Central.
The system timezone is Europe/Berlin. Please correct if necessary!

First Observation: 2024-06-04 15:00:36
Last Observation: 2024-06-12 08:29:43
Timespan: 7.7 days

Observation intervals: 
  Id    interval.time              n pct      
1 VEET  0s                         3 0.00099% 
2 VEET  1s                      2089 0.68673% 
3 VEET  2s                    299876 98.58051%
4 VEET  3s                      2213 0.72750% 
5 VEET  4s                         3 0.00099% 
6 VEET  6s                         1 0.00033% 
7 VEET  9s                         7 0.00230% 
8 VEET  109s (~1.82 minutes)       1 0.00033% 
9 VEET  59077s (~16.41 hours)      1 0.00033% </code></pre>
</div>
<div class="cell-output-display">
<div id="fig-VEET-overview3" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-overview3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="supplement_files/figure-html/fig-VEET-overview3-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7" title="Overview plot of imported VEET data"><img src="supplement_files/figure-html/fig-VEET-overview3-1.png" class="img-fluid figure-img" width="576"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-overview3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Overview plot of imported VEET data
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>In a first step, we condition the data similarly to the other <code>VEET</code> modalities.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [22]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Aggregate spectral data to 5-minute intervals and mark gaps</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="ot">&lt;-</span> dataVEET3 <span class="sc">|&gt;</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 secs"</span>) <span class="sc">|&gt;</span>     <span class="co"># aggregate to 5-second bins</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>           <span class="co"># explicit NA for any gaps in between</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">remove_partial_data</span>(dist1_0, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>In the next step, we need to transform the <code>wide</code> format of the imported dataset into a <code>long</code> format, where each row contains exactly one observation for one grid-point.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [23]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(Datetime, file.name, Id, is.implicit, time_stamp, modality, Date),</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">".value"</span>, <span class="st">"position"</span>),</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_pattern =</span> <span class="st">"(conf1|conf2|dist1|dist2)_(</span><span class="sc">\\</span><span class="st">d+)"</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>In a final step before we can use the data in the analysis, we need to assign <code>x</code> and <code>y</code> coordinates based on the <code>position</code> column that was created when pivoting longer. Positions are counted from 0 to 63 starting at the top left and increasing towards the right, before continuing on the left in the next row below. <code>y</code> positions thus depend on the row count, i.e., how often a row of 8 values fits into the <code>position</code> column. <code>x</code> positions consequently depend on the <code>position</code> within each 8-value row. We also add an <code>observation</code> variable that increases by <code>+1</code> every time, the <code>position</code> column hits <code>0</code>. We then center both <code>x</code> and <code>y</code> coordinates to receive meaningful values, i.e., 0° indicates the center of the overall measurement cone. Lastly, we convert both confidence columns, which are scaled from 0-255 into percentages by dividing them by <code>255</code>. Empirical data from the manufacturer points to a threshold of about <code>10%</code>, under which the respective distance data is not reliable.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [24]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ungroup() |&gt; </span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">position =</span> <span class="fu">as.numeric</span>(position),</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">y.pos =</span> (position <span class="sc">%/%</span> <span class="dv">8</span>)<span class="sc">+</span><span class="dv">1</span>,</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">y.pos =</span> <span class="fu">scale</span>(y.pos, <span class="at">scale =</span> <span class="cn">FALSE</span>)<span class="sc">*</span><span class="dv">52</span><span class="sc">/</span><span class="dv">8</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">x.pos =</span> (position <span class="sc">%%</span> <span class="dv">8</span>)<span class="sc">+</span><span class="dv">1</span>,</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">x.pos =</span> <span class="fu">scale</span>(x.pos, <span class="at">scale =</span> <span class="cn">FALSE</span>)<span class="sc">*</span><span class="dv">41</span><span class="sc">/</span><span class="dv">8</span>,</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">observation =</span> <span class="fu">cumsum</span>(position <span class="sc">==</span> <span class="dv">0</span>),</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>         <span class="fu">across</span>(<span class="fu">starts_with</span>(<span class="st">"conf"</span>), \(x) x<span class="sc">/</span><span class="dv">255</span>)</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>         )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>Now this dataset is ready for further analysis. We finish by visualizing the same observation time on different days. Note that we replace zero distance values with <code>5 meters</code>, which is the maximum distance the <code>VEET</code> can measure.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [25]:</pre></div><div id="cell-fig-VEET-distance" class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="co">#set visualization parameters</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>extras <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(),</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">oob =</span> scales<span class="sc">::</span>oob_squish_any),</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"white"</span>)),</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>(),</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="st">"none"</span>),</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> (dist1<span class="sc">/</span><span class="dv">10</span>) <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">0</span>), <span class="at">colour =</span> dist1<span class="sc">&gt;</span><span class="dv">1000</span>), <span class="at">size =</span> <span class="fl">2.5</span>),</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(),</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"X position (°)"</span>, <span class="at">y =</span> <span class="st">"Y position (°)"</span>, </span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="st">"Distance (cm)"</span>, <span class="at">alpha =</span> <span class="st">"Confidence (0-255)"</span>))</span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>slicer <span class="ot">&lt;-</span> <span class="cf">function</span>(x){<span class="fu">seq</span>(<span class="fu">min</span>((x<span class="dv">-1</span>)<span class="sc">*</span><span class="dv">64</span><span class="sc">+</span><span class="dv">1</span>), <span class="fu">max</span>(x<span class="sc">*</span><span class="dv">64</span>, <span class="at">by =</span> <span class="dv">1</span>))} <span class="co">#allows to choose an observation</span></span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="fu">slicer</span>(<span class="dv">9530</span>)) <span class="sc">|&gt;</span>  <span class="co">#choose a particular observation</span></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">dist1 =</span> <span class="fu">ifelse</span>(dist1 <span class="sc">==</span> <span class="dv">0</span>, <span class="cn">Inf</span>, dist1)) <span class="sc">|&gt;</span> <span class="co">#replace 0 distances with 5m</span></span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(conf1 <span class="sc">&gt;=</span> <span class="fl">0.1</span> <span class="sc">|</span> dist1 <span class="sc">==</span> <span class="cn">Inf</span>) <span class="sc">|&gt;</span> <span class="co">#remove data that has less than 10% confidence</span></span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x.pos, <span class="at">y=</span>y.pos, <span class="at">fill =</span> dist1<span class="sc">/</span><span class="dv">10</span>))<span class="sc">+</span> extras <span class="sc">+</span> <span class="co">#plot the data</span></span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="sc">~</span>Datetime) <span class="co">#show one plot per day</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-VEET-distance" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-VEET-distance-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="">
<a href="supplement_files/figure-html/fig-VEET-distance-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8" title="Example observations of the measurement grid"><img src="supplement_files/figure-html/fig-VEET-distance-1.png" class="img-fluid figure-img" width="1152"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-VEET-distance-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Example observations of the measurement grid
</figcaption>
</figure>
</div>
</div>
</div></div>
<p>As we can see from the figure, different days have a vastly different distribution of distance data, and measurement confidence (values with confidence &lt; 10% are removed)</p>
</section>
<section id="saving-the-cleaned-data" class="level2">
<h2 class="anchored" data-anchor-id="saving-the-cleaned-data">Saving the Cleaned Data</h2>
<p>After executing all the above steps, we have three cleaned data frames in our R session:</p>
<ul>
<li><p><strong><code>dataCC</code></strong>&nbsp;– the processed <code>Clouclip</code> dataset (5-second intervals, with distance and lux, including NA for gaps and sentinel statuses).</p></li>
<li><p><strong><code>dataVEET</code></strong>&nbsp;– the processed <code>VEET</code> ambient light dataset (5-second intervals, illuminance in lux, with gaps filled).</p></li>
<li><p><strong><code>dataVEET2</code></strong>&nbsp;– the processed <code>VEET</code> spectral dataset (5-minute intervals, each entry containing a spectrum or derived spectral metrics).</p></li>
<li><p><strong><code>dataVEET3</code></strong>&nbsp;– the processed <code>VEET</code> distance dataset (5-second intervals, each entry containing the distance of up to two objects in the 8x8 grid).</p></li>
</ul>
<p>For convenience and future reproducibility, we will save these combined results to a single R data file. Storing all cleaned data together ensures that any analysis can reload the exact same data state without re-running the import and cleaning (which can be time-consuming for large raw files).</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [26]:</pre></div><div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create directory for cleaned data if it doesn't exist</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">"data/cleaned"</span>)) <span class="fu">dir.create</span>(<span class="st">"data/cleaned"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Save all cleaned datasets into one .RData file</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(dataCC, dataVEET, dataVEET2, dataVEET3,</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="st">"data/cleaned/data.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div>
<p>The above code creates (if necessary) a folder&nbsp;<code>data/cleaned/</code>&nbsp;and saves a single RData file (<code>data.RData</code>) containing the three objects. To retrieve them later, one can use&nbsp;<code>&lt;- load("data/cleaned/data.RData")</code>, which will return the objects into the environment. This single-file approach simplifies sharing and keeps the cleaned data together.</p>
<p>In summary, this supplement has walked through the full preprocessing pipeline for two example devices. We began by describing the raw data format for each device and then demonstrated how to import the data with correct time zone settings. We handled device-specific quirks like sentinel codes (for <code>Clouclip</code>) and multiple modalities with gain normalization (for <code>VEET</code>). We showed how to detect and address irregular sampling, how to explicitly mark missing data gaps to avoid analytic pitfalls, and how to reduce data granularity via rounding or aggregation when appropriate. Throughout, we used functions from LightLogR in a tidyverse workflow, aiming to make the steps clear and modular. By saving the final cleaned datasets, we set the stage for the computation of visual experience metrics such as working distance, time in bright light, spectral composition ratios, as presented in the <a href="https://tscnlab.github.io/ZaunerEtAl_JVis_2025/">main tutorial</a>. We hope this detailed tutorial empowers researchers to adopt similar pipelines for their own data, facilitating reproducible and accurate analyses of visual experience.</p>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Sah2025OphtalmicPhysiolOpt" class="csl-entry" role="listitem">
Sah, Raman Prasad, Pavan Kalyan Narra, and Lisa A. Ostrin. 2025. <span>“A Novel Wearable Sensor for Objective Measurement of Distance and Illumination.”</span> <em>Ophthalmic and Physiological Optics</em> 00 (n/a): 1–13. https://doi.org/<a href="https://doi.org/10.1111/opo.13523">https://doi.org/10.1111/opo.13523</a>.
</div>
<div id="ref-Sullivan2024" class="csl-entry" role="listitem">
Sullivan, David, Aaron Nicholls, George Hatoun, Samuel Thompson, Cory Schwarzmiller, Fathollah Memarzanjany, Alyssa Gunderson, et al. 2024. <span>“The Visual Experience Evaluation Tool: A Myopia Research Instrument for Quantifying Visual Experience.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2024.09.20.614212">https://doi.org/10.1101/2024.09.20.614212</a>.
</div>
<div id="ref-Wen2020bjophthalmol" class="csl-entry" role="listitem">
Wen, Longbo, Yingpin Cao, Qian Cheng, Xiaoning Li, Lun Pan, Lei Li, HaoGang Zhu, Weizhong Lan, and Zhikuan Yang. 2020. <span>“Objectively Measured Near Work, Outdoor Exposure and Myopia in Children.”</span> <em>British Journal of Ophthalmology</em> 104 (11): 1542–47. <a href="https://doi.org/10.1136/bjophthalmol-2019-315258">https://doi.org/10.1136/bjophthalmol-2019-315258</a>.
</div>
<div id="ref-Wen2021ActaOphtalmol" class="csl-entry" role="listitem">
Wen, Longbo, Qian Cheng, Yingpin Cao, Xiaoning Li, Lun Pan, Lei Li, Haogang Zhu, Ian Mogran, Weizhong Lan, and Zhikuan Yang. 2021. <span>“The Clouclip, a Wearable Device for Measuring Near-Work and Outdoor Time: Validation and Comparison of Objective Measures with Questionnaire Estimates.”</span> <em>Acta Ophthalmologica</em> 99 (7): e1222–35. https://doi.org/<a href="https://doi.org/10.1111/aos.14785">https://doi.org/10.1111/aos.14785</a>.
</div>
<div id="ref-Zauner2025JOpenSourceSoftw" class="csl-entry" role="listitem">
Zauner, J., S. Hartmeyer, and M. Spitschan. 2025. <span>“LightLogR: Reproducible Analysis of Personal Light Exposure Data.”</span> Journal Article. <em>J Open Source Softw</em> 10 (107): 7601. <a href="https://doi.org/10.21105/joss.07601">https://doi.org/10.21105/joss.07601</a>.
</div>
</div>
</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p>A <em>sentinel value</em> is a special placeholder value used in data recording to signal a particular condition. It does not represent a valid measured quantity but rather acts as a marker (for example, “device off” or “value out of range”).<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p><strong>tibble</strong> are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <a href="https://tibble.tidyverse.org/index.html">tibbles</a><a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p><strong>tibble</strong> are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <a href="https://tibble.tidyverse.org/index.html">tibbles</a><a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p><strong>tibble</strong> are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <a href="https://tibble.tidyverse.org/index.html">tibbles</a><a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>
     </main>
<!-- /main column -->  <script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/tscnlab\.github\.io\/ZaunerEtAl_JVis_2025\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb41" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Supplement 1"</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> "Analysis of human visual experience data"</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co">  - name: "Johannes Zauner"</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co">    id: JZ</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="co">    affiliation: </span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a><span class="co">      - Technical University of Munich, Germany</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a><span class="co">      - Max Planck Institute for Biological Cybernetics, Tübingen, Germany</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="co">    orcid: "0000-0003-2171-4566"</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="co">    corresponding: true</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a><span class="co">    email: johannes.zauner@tum.de</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a><span class="co">  - name: "Aaron Nicholls"</span></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a><span class="co">    affiliation: "Meta Reality Labs Research"</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a><span class="co">    orcid: "0009-0001-6683-6826"</span></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a><span class="co">  - name: "Lisa A. Ostrin"</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a><span class="co">    affiliation: "University of Houston, Texas, USA"</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a><span class="co">    orcid: "0000-0001-8675-1478"</span></span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a><span class="co">  - name: "Manuel Spitschan"</span></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a><span class="co">    affiliation: </span></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a><span class="co">      - Technical University of Munich, Germany</span></span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a><span class="co">      - Max Planck Institute for Biological Cybernetics, Tübingen, Germany</span></span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a><span class="co">      - Technical University of Munich Institute of advanced study (TUM-IAS), Munich, Germany</span></span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a><span class="co">    orcid: "0000-0002-8572-9268"</span></span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a><span class="an">doi:</span><span class="co"> 10.5281/zenodo.16566014</span></span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb41-29"><a href="#cb41-29" aria-hidden="true" tabindex="-1"></a><span class="co">    include-in-header:</span></span>
<span id="cb41-30"><a href="#cb41-30" aria-hidden="true" tabindex="-1"></a><span class="co">      - text: |</span></span>
<span id="cb41-31"><a href="#cb41-31" aria-hidden="true" tabindex="-1"></a><span class="co">          &lt;style&gt;</span></span>
<span id="cb41-32"><a href="#cb41-32" aria-hidden="true" tabindex="-1"></a><span class="co">          .quarto-notebook .cell-container .cell-decorator {</span></span>
<span id="cb41-33"><a href="#cb41-33" aria-hidden="true" tabindex="-1"></a><span class="co">            display: none;</span></span>
<span id="cb41-34"><a href="#cb41-34" aria-hidden="true" tabindex="-1"></a><span class="co">          }</span></span>
<span id="cb41-35"><a href="#cb41-35" aria-hidden="true" tabindex="-1"></a><span class="co">          &lt;/style&gt;</span></span>
<span id="cb41-36"><a href="#cb41-36" aria-hidden="true" tabindex="-1"></a><span class="co">    number-sections: true</span></span>
<span id="cb41-37"><a href="#cb41-37" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb41-38"><a href="#cb41-38" aria-hidden="true" tabindex="-1"></a><span class="co">    html-math-method:</span></span>
<span id="cb41-39"><a href="#cb41-39" aria-hidden="true" tabindex="-1"></a><span class="co">      method: mathjax</span></span>
<span id="cb41-40"><a href="#cb41-40" aria-hidden="true" tabindex="-1"></a><span class="co">      url: "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML"</span></span>
<span id="cb41-41"><a href="#cb41-41" aria-hidden="true" tabindex="-1"></a><span class="an">bibliography:</span><span class="co"> references.bib</span></span>
<span id="cb41-42"><a href="#cb41-42" aria-hidden="true" tabindex="-1"></a><span class="an">lightbox:</span><span class="co"> true</span></span>
<span id="cb41-43"><a href="#cb41-43" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span></span>
<span id="cb41-44"><a href="#cb41-44" aria-hidden="true" tabindex="-1"></a><span class="co">  echo: true</span></span>
<span id="cb41-45"><a href="#cb41-45" aria-hidden="true" tabindex="-1"></a><span class="co">  warning: false</span></span>
<span id="cb41-46"><a href="#cb41-46" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb41-47"><a href="#cb41-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-48"><a href="#cb41-48" aria-hidden="true" tabindex="-1"></a>::: {.callout-warning}</span>
<span id="cb41-49"><a href="#cb41-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-50"><a href="#cb41-50" aria-hidden="true" tabindex="-1"></a>Please note that this supplement is not in its final styling form, as it necessarily has to be formatted as a <span class="in">`notebook`</span> to enable commenting. Current issues include:</span>
<span id="cb41-51"><a href="#cb41-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-52"><a href="#cb41-52" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Cross-references not working (e.g. Figure labeling and referencing figures)</span>
<span id="cb41-53"><a href="#cb41-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-54"><a href="#cb41-54" aria-hidden="true" tabindex="-1"></a>These issues will be resolved when all comments have been gathered and implemented</span>
<span id="cb41-55"><a href="#cb41-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-56"><a href="#cb41-56" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb41-57"><a href="#cb41-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-58"><a href="#cb41-58" aria-hidden="true" tabindex="-1"></a><span class="fu">## Abstract {.unnumbered}</span></span>
<span id="cb41-59"><a href="#cb41-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-60"><a href="#cb41-60" aria-hidden="true" tabindex="-1"></a>This supplementary document provides a detailed, step-by-step tutorial on importing and preprocessing raw data from two wearable devices: <span class="in">`Clouclip`</span> and <span class="in">`VEET`</span>. We describe the structure of the raw datasets recorded by each device and explain how to parse these data, specify time zones, handle special sentinel values, clean missing observations, regularize timestamps to fixed intervals, and aggregate data as needed. All original R code from the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](https://tscnlab.github.io/ZaunerEtAl_JVis_2025/)</span> is preserved here for transparency, with additional guidance intended for a broad research audience. We demonstrate how to detect gaps and irregular sampling, convert implicit missing periods into explicit missing values, and address device-specific quirks such as the <span class="in">`Clouclip`</span>’s use of sentinel codes for “sleep mode” and “out of range” readings. Special procedures for processing the <span class="in">`VEET`</span>’s rich spectral data (e.g. normalizing sensor counts and reconstructing full spectra from multiple sensor channels) are also outlined. Finally, we show how to save the cleaned datasets from both devices into a single R data file for downstream analysis. This comprehensive walkthrough is designed to ensure reproducibility and to assist researchers in understanding and adapting the data pipeline for their own visual experience datasets.</span>
<span id="cb41-61"><a href="#cb41-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-62"><a href="#cb41-62" aria-hidden="true" tabindex="-1"></a><span class="fu">## Introduction</span></span>
<span id="cb41-63"><a href="#cb41-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-64"><a href="#cb41-64" aria-hidden="true" tabindex="-1"></a>Wearable sensors like the Clouclip and the Visual Environment Evaluation Tool (VEET) produce high-dimensional time-series data on viewing distance and light exposure. Proper handling of these raw data is essential before any analysis of visual experience metrics. In the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](https://tscnlab.github.io/ZaunerEtAl_JVis_2025/)</span>, we introduced an analysis pipeline using the open-source R package LightLogR <span class="co">[</span><span class="ot">@Zauner2025JOpenSourceSoftw</span><span class="co">]</span> to calculate various distance and light exposure metrics. Here, we present a full account of the data import and preparation steps as a supplement to the methods, with an emphasis on clarity for researchers who may be less familiar with data processing in R.</span>
<span id="cb41-65"><a href="#cb41-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-66"><a href="#cb41-66" aria-hidden="true" tabindex="-1"></a>We use example datasets from a <span class="in">`Clouclip`</span> device <span class="co">[</span><span class="ot">@Wen2021ActaOphtalmol; @Wen2020bjophthalmol</span><span class="co">]</span> and from a <span class="in">`VEET`</span> device <span class="co">[</span><span class="ot">@Sah2025OphtalmicPhysiolOpt</span><span class="co">]</span> (both provided in the accompanying <span class="co">[</span><span class="ot">repository</span><span class="co">](https://github.com/tscnlab/ZaunerEtAl_2025_JOV)</span>). The <span class="in">`Clouclip`</span> is a glasses-mounted sensor that records working distance (distance from eyes to object, in cm) and ambient illuminance (in lux) at 5-second intervals. The <span class="in">`VEET`</span> is a head-mounted multi-modal sensor that logs ambient light and spectral information (along with other data like motion and distance via a depth sensor) - in this exemplary case at 2-second intervals. A single week of continuous data illustrates the contrast in complexity: approximately 1.6&nbsp;MB for the <span class="in">`Clouclip`</span>’s simple two-column output versus up to 270&nbsp;MB for the <span class="in">`VEET`</span>’s multi-channel output (due to its higher sampling rate and richer sensor modalities).</span>
<span id="cb41-67"><a href="#cb41-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-68"><a href="#cb41-68" aria-hidden="true" tabindex="-1"></a>In the following sections, we detail how these raw data are structured and how to import and preprocess them using LightLogR. We cover device-specific considerations such as file format quirks and sensor range limitations, as well as general best practices like handling missing timestamps and normalizing sensor readings. All code blocks can be executed in R (with the required packages loaded) to reproduce the cleaning steps. The end result will be clean, regularized datasets (<span class="in">`dataCC`</span> for Clouclip, <span class="in">`dataVEET`</span> for VEET light data, and <span class="in">`dataVEET2`</span> for VEET spectral data) ready for calculating visual experience metrics. We conclude by saving these cleaned datasets into a single file for convenient reuse.</span>
<span id="cb41-69"><a href="#cb41-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-70"><a href="#cb41-70" aria-hidden="true" tabindex="-1"></a><span class="fu">## `Clouclip` Data: Raw Structure and Import</span></span>
<span id="cb41-71"><a href="#cb41-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-72"><a href="#cb41-72" aria-hidden="true" tabindex="-1"></a>The <span class="in">`Clouclip`</span> device exports its data as a text file (not a true Excel file despite sometimes using an .xls extension), which is actually tab-separated values. Each record corresponds to one timestamped observation (nominally every 5 seconds) and includes two measured variables: **distance** and **illuminance**. In the sample dataset (<span class="in">`Sample_Clouclip.csv`</span> provided), the columns are:</span>
<span id="cb41-73"><a href="#cb41-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-74"><a href="#cb41-74" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`Date`</span> – the date and time of the observation (in the device’s local time, here one week in 2021).</span>
<span id="cb41-75"><a href="#cb41-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-76"><a href="#cb41-76" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`Dis`</span> – the viewing distance in centimeters.</span>
<span id="cb41-77"><a href="#cb41-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-78"><a href="#cb41-78" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`Lux`</span> – the ambient illuminance in lux.</span>
<span id="cb41-79"><a href="#cb41-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-80"><a href="#cb41-80" aria-hidden="true" tabindex="-1"></a>For example, a raw data line might look like:</span>
<span id="cb41-81"><a href="#cb41-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-82"><a href="#cb41-82" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb41-83"><a href="#cb41-83" aria-hidden="true" tabindex="-1"></a><span class="in">2021-07-01 08:00:00 45  320</span></span>
<span id="cb41-84"><a href="#cb41-84" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-85"><a href="#cb41-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-86"><a href="#cb41-86" aria-hidden="true" tabindex="-1"></a>indicating that at *2021-07-01 08:00:00* local time, the device recorded a working distance of 45&nbsp;cm and illuminance of 320&nbsp;lx. The Clouclip uses special **sentinel values**[^1] in these measurement columns to denote certain device states. Specifically, a distance (`Dis`) value of **204** is a code meaning the object is *out of the sensor’s range*, and a value of **-1** in either `Dis` or `Lux` indicates the device was in *sleep mode* (not actively recording). During normal operation, distance measurements are limited by the device’s range, and illuminance readings are positive lux values. Any sentinel codes in the raw file need to be handled appropriately, as described below.</span>
<span id="cb41-87"><a href="#cb41-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-88"><a href="#cb41-88" aria-hidden="true" tabindex="-1"></a><span class="ot">[^1]: </span>A *sentinel value* is a special placeholder value used in data recording to signal a particular condition. It does not represent a valid measured quantity but rather acts as a marker (for example, “device off” or “value out of range”).</span>
<span id="cb41-89"><a href="#cb41-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-90"><a href="#cb41-90" aria-hidden="true" tabindex="-1"></a>We will use LightLogR’s built-in import function for <span class="in">`Clouclip`</span>, which automatically reads the file, parses the timestamps, and converts sentinel codes into a separate status annotation. To begin, we load the necessary libraries and import the raw <span class="in">`Clouclip`</span> dataset:</span>
<span id="cb41-91"><a href="#cb41-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-94"><a href="#cb41-94" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-95"><a href="#cb41-95" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: setup</span></span>
<span id="cb41-96"><a href="#cb41-96" aria-hidden="true" tabindex="-1"></a><span class="co">#| output: false</span></span>
<span id="cb41-97"><a href="#cb41-97" aria-hidden="true" tabindex="-1"></a><span class="co"># Load required packages</span></span>
<span id="cb41-98"><a href="#cb41-98" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb41-99"><a href="#cb41-99" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(LightLogR)</span>
<span id="cb41-100"><a href="#cb41-100" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gt)</span>
<span id="cb41-101"><a href="#cb41-101" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(downlit) <span class="co">#not used, but important for code-linking feature</span></span>
<span id="cb41-102"><a href="#cb41-102" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-103"><a href="#cb41-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-106"><a href="#cb41-106" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-107"><a href="#cb41-107" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-importCC</span></span>
<span id="cb41-108"><a href="#cb41-108" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Overview plot of imported Clouclip data"</span></span>
<span id="cb41-109"><a href="#cb41-109" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb41-110"><a href="#cb41-110" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 6</span></span>
<span id="cb41-111"><a href="#cb41-111" aria-hidden="true" tabindex="-1"></a><span class="co"># Define file path and time zone</span></span>
<span id="cb41-112"><a href="#cb41-112" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/Sample_Clouclip.csv"</span></span>
<span id="cb41-113"><a href="#cb41-113" aria-hidden="true" tabindex="-1"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span>   <span class="co"># Time zone in which device was recording (e.g., US Central Time)</span></span>
<span id="cb41-114"><a href="#cb41-114" aria-hidden="true" tabindex="-1"></a><span class="co"># Import Clouclip data</span></span>
<span id="cb41-115"><a href="#cb41-115" aria-hidden="true" tabindex="-1"></a>dataCC <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">Clouclip</span>(path, <span class="at">tz =</span> tz, <span class="at">manual.id =</span> <span class="st">"Clouclip"</span>)</span>
<span id="cb41-116"><a href="#cb41-116" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-117"><a href="#cb41-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-118"><a href="#cb41-118" aria-hidden="true" tabindex="-1"></a>In this code, <span class="in">`import$Clouclip()`</span> reads the tab-delimited file and returns a **tibble**<span class="ot">[^2]</span> (saved in the variable <span class="in">`dataCC`</span>) containing the data. We specify&nbsp;<span class="in">`tz = "US/Central"`</span>&nbsp;because the device’s clock was set to U.S. Central time; this ensures that the&nbsp;<span class="in">`Datetime`</span>values are properly interpreted with the correct time zone. The argument&nbsp;<span class="in">`manual.id = "Clouclip"`</span>&nbsp;simply tags the dataset with an identifier (useful if combining data from multiple devices).</span>
<span id="cb41-119"><a href="#cb41-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-120"><a href="#cb41-120" aria-hidden="true" tabindex="-1"></a><span class="ot">[^2]: </span>**tibble** are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <span class="co">[</span><span class="ot">tibbles</span><span class="co">](https://tibble.tidyverse.org/index.html)</span></span>
<span id="cb41-121"><a href="#cb41-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-122"><a href="#cb41-122" aria-hidden="true" tabindex="-1"></a>During import, LightLogR automatically handles the <span class="in">`Clouclip`</span>’s sentinel codes. The&nbsp;<span class="in">`Date`</span>&nbsp;column from the raw file is parsed into a POSIX date-time (<span class="in">`Datetime`</span>) with the specified time zone. The&nbsp;<span class="in">`Lux`</span>&nbsp;and&nbsp;<span class="in">`Dis`</span>&nbsp;columns are read as numeric, but any occurrences of&nbsp;**-1**&nbsp;or&nbsp;**204**&nbsp;are treated specially: these are replaced with&nbsp;<span class="in">`NA`</span>&nbsp;(missing values) in the numeric columns, and corresponding status columns&nbsp;<span class="in">`Lux_status`</span>&nbsp;and&nbsp;<span class="in">`Dis_status`</span>&nbsp;are created to indicate the reason for those <span class="in">`NA`</span> values. For example, if a&nbsp;<span class="in">`Dis`</span>&nbsp;value of 204 was encountered, that row’s&nbsp;<span class="in">`Dis`</span>&nbsp;will be set to NA and&nbsp;<span class="in">`Dis_status`</span>&nbsp;will contain&nbsp;<span class="in">`"out_of_range"`</span>; if&nbsp;<span class="in">`Lux`</span>&nbsp;or&nbsp;<span class="in">`Dis`</span>&nbsp;was -1, the status is&nbsp;<span class="in">`"sleep_mode"`</span>. We will set all other readings to &nbsp;<span class="in">`"operational"`</span> (meaning the device was functioning normally at that time) for visualisation purposes.</span>
<span id="cb41-123"><a href="#cb41-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-124"><a href="#cb41-124" aria-hidden="true" tabindex="-1"></a>After import, it is good practice to get an overview of the data. The import function by default prints a brief summary (and generates an overview plot of the timeline) showing the number of observations, the time span, and any irregularities or large gaps. In our case, the <span class="in">`Clouclip`</span> summary indicates the data spans one week and reveals that there are&nbsp;**irregular intervals**&nbsp;in the timestamps. This means some observations do not occur exactly every 5 seconds as expected. We can programmatically check for irregular timing:</span>
<span id="cb41-125"><a href="#cb41-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-128"><a href="#cb41-128" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-129"><a href="#cb41-129" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if data are on a perfectly regular 5-second schedule</span></span>
<span id="cb41-130"><a href="#cb41-130" aria-hidden="true" tabindex="-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">has_irregulars</span>()</span>
<span id="cb41-131"><a href="#cb41-131" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-132"><a href="#cb41-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-133"><a href="#cb41-133" aria-hidden="true" tabindex="-1"></a>If the result is&nbsp;<span class="in">`TRUE`</span>, it confirms that the time sequence has deviations from the 5-second interval. Indeed, our example dataset has many small timing irregularities and gaps (periods with no data). Understanding the pattern of these missing or irregular readings is important. We can visualize them using a gap plot:</span>
<span id="cb41-134"><a href="#cb41-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-137"><a href="#cb41-137" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-138"><a href="#cb41-138" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-irregular</span></span>
<span id="cb41-139"><a href="#cb41-139" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 12</span></span>
<span id="cb41-140"><a href="#cb41-140" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb41-141"><a href="#cb41-141" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Visualization of gaps and irregular data. Black traces show available data. Red shaded areas show times of missing data. Red dots show instances where observations occur off the regular interval from start to finish, i.e., irregular data."</span></span>
<span id="cb41-142"><a href="#cb41-142" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot gaps and irregular timestamps for Clouclip data</span></span>
<span id="cb41-143"><a href="#cb41-143" aria-hidden="true" tabindex="-1"></a>y.label <span class="ot">&lt;-</span> <span class="st">"Distance (cm)"</span></span>
<span id="cb41-144"><a href="#cb41-144" aria-hidden="true" tabindex="-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">gg_gaps</span>(Dis, </span>
<span id="cb41-145"><a href="#cb41-145" aria-hidden="true" tabindex="-1"></a>                  <span class="at">include.implicit.gaps =</span> <span class="cn">FALSE</span>,  <span class="co"># only show recorded gaps, not every missing point</span></span>
<span id="cb41-146"><a href="#cb41-146" aria-hidden="true" tabindex="-1"></a>                  <span class="at">show.irregulars =</span> <span class="cn">TRUE</span>,         <span class="co"># highlight irregular timing</span></span>
<span id="cb41-147"><a href="#cb41-147" aria-hidden="true" tabindex="-1"></a>                  <span class="at">y.axis.label =</span> y.label,</span>
<span id="cb41-148"><a href="#cb41-148" aria-hidden="true" tabindex="-1"></a>                  <span class="at">group.by.days =</span> <span class="cn">TRUE</span>, <span class="at">col.irregular =</span> <span class="fu">alpha</span>(<span class="st">"red"</span>, <span class="fl">0.03</span>)</span>
<span id="cb41-149"><a href="#cb41-149" aria-hidden="true" tabindex="-1"></a>                  ) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="cn">NULL</span>)</span>
<span id="cb41-150"><a href="#cb41-150" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-151"><a href="#cb41-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-152"><a href="#cb41-152" aria-hidden="true" tabindex="-1"></a>In @fig-irregular, time periods where data are missing appear as red-shaded areas, and any off-schedule observation times are marked with red dots. The <span class="in">`Clouclip`</span> example shows extensive gaps (red blocks) on certain days and irregular timing on all days except the first and last. These irregular timestamps likely arise from the device’s logging process (e.g. slight clock drift or buffering when the device was turned on/off). Such issues must be addressed before further analysis.</span>
<span id="cb41-153"><a href="#cb41-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-154"><a href="#cb41-154" aria-hidden="true" tabindex="-1"></a>When faced with irregular or gapped data, LightLogR recommends a few strategies:</span>
<span id="cb41-155"><a href="#cb41-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-156"><a href="#cb41-156" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*Remove leading/trailing segments that cause irregularity.*&nbsp;For example, if only the first day is regular and subsequent days drift, one might exclude the problematic portion using date filters (see&nbsp;<span class="co">[</span><span class="ot">filter_Date() / filter_Datetime()</span><span class="co">](https://tscnlab.github.io/LightLogR/reference/filter_Datetime.html)</span>&nbsp;in LightLogR).</span>
<span id="cb41-157"><a href="#cb41-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-158"><a href="#cb41-158" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*Round timestamps to the nearest regular interval.*&nbsp;This can snap slightly off-schedule times back to the 5-second grid (using&nbsp;<span class="co">[</span><span class="ot">cut_Datetime()</span><span class="co">](https://tscnlab.github.io/LightLogR/reference/cut_Datetime.html)</span>&nbsp;with a 5-second interval), provided the deviations are small and this rounding won’t create duplicate timestamps.</span>
<span id="cb41-159"><a href="#cb41-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-160"><a href="#cb41-160" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>*Aggregate to a coarser time interval.*&nbsp;For instance, grouping data into 1-minute bins with&nbsp;<span class="co">[</span><span class="ot">aggregate_Datetime()</span><span class="co">](https://tscnlab.github.io/LightLogR/reference/aggregate_Datetime.html)</span> can mask irregularities at finer scales, at the cost of some temporal resolution.</span>
<span id="cb41-161"><a href="#cb41-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-162"><a href="#cb41-162" aria-hidden="true" tabindex="-1"></a>In this case, the deviations from the 5-second schedule are relatively minor. We choose to&nbsp;**round the timestamps to the nearest 5 seconds**&nbsp;to enforce a uniform sampling grid, which simplifies downstream gap handling. We further add a separate date column for convenience:</span>
<span id="cb41-163"><a href="#cb41-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-166"><a href="#cb41-166" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-167"><a href="#cb41-167" aria-hidden="true" tabindex="-1"></a><span class="co"># Regularize timestamps by rounding to nearest 5-second interval</span></span>
<span id="cb41-168"><a href="#cb41-168" aria-hidden="true" tabindex="-1"></a>dataCC <span class="ot">&lt;-</span> dataCC <span class="sc">|&gt;</span></span>
<span id="cb41-169"><a href="#cb41-169" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cut_Datetime</span>(<span class="st">"5 secs"</span>, <span class="at">New.colname =</span> Datetime) <span class="sc">|&gt;</span>  <span class="co"># round times to 5-second bins</span></span>
<span id="cb41-170"><a href="#cb41-170" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>)                     <span class="co"># add a Date column for grouping by day</span></span>
<span id="cb41-171"><a href="#cb41-171" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-172"><a href="#cb41-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-173"><a href="#cb41-173" aria-hidden="true" tabindex="-1"></a>After this operation, all&nbsp;<span class="in">`Datetime`</span>&nbsp;entries in&nbsp;<span class="in">`dataCC`</span>&nbsp;align perfectly on 5-second boundaries (e.g.&nbsp;**08:00:00**,&nbsp;**08:00:05**,&nbsp;**08:00:10**, etc.). We can verify that no irregular intervals remain by re-running&nbsp;`has_irregulars()`&nbsp;(it now returns&nbsp;`FALSE`). Next, we want to quantify the missing data. LightLogR distinguishes between&nbsp;**explicit missing values**&nbsp;(actual&nbsp;`NA`s in the data, possibly from sentinel replacements or gaps we have filled in) and&nbsp;**implicit missing intervals**&nbsp;(time points where the device&nbsp;*should*&nbsp;have a reading but none was recorded, and we have not yet filled them in). Initially, many gaps are still implicit (between the first and last timestamp of each day). We can generate a gap summary table:</span>
<span id="cb41-174"><a href="#cb41-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-177"><a href="#cb41-177" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-178"><a href="#cb41-178" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: tbl-gaps</span></span>
<span id="cb41-179"><a href="#cb41-179" aria-hidden="true" tabindex="-1"></a><span class="co">#| tbl-cap: "Summary of missing and observed data for the Clouclip device"</span></span>
<span id="cb41-180"><a href="#cb41-180" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize observed vs missing data by day for distance</span></span>
<span id="cb41-181"><a href="#cb41-181" aria-hidden="true" tabindex="-1"></a>dataCC <span class="sc">|&gt;</span> <span class="fu">gap_table</span>(Dis, <span class="at">Variable.label =</span> <span class="st">"Distance (cm)"</span>) <span class="sc">|&gt;</span></span>
<span id="cb41-182"><a href="#cb41-182" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cols_hide</span>(<span class="fu">contains</span>(<span class="st">"_n"</span>))   <span class="co"># hide absolute counts for brevity in output</span></span>
<span id="cb41-183"><a href="#cb41-183" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-184"><a href="#cb41-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-185"><a href="#cb41-185" aria-hidden="true" tabindex="-1"></a>This summary (@tbl-gaps) breaks down, for each day, how much data is present vs. missing. It reports the total duration of recorded data and the duration of gaps. After rounding the times, there are no irregular timestamp issues, but we see substantial **implicit gaps** — periods where the device was not recording (e.g., overnight when the device was likely not worn or was in sleep mode). Notably, the first and last days of the week have very little data (less than 1 hour each), since they probably represent partial recording days (the device was put on and removed on those days).</span>
<span id="cb41-186"><a href="#cb41-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-187"><a href="#cb41-187" aria-hidden="true" tabindex="-1"></a>To prepare the dataset for analysis, we will convert all those implicit gaps into explicit missing entries, and remove days that are mostly incomplete. Converting implicit gaps means inserting rows with <span class="in">`NA`</span> for each missing 5-second slot, so that the time series becomes continuous and explicit about missingness. We use <span class="in">`gap_handler()`</span> for this, and then drop the nearly-empty days:</span>
<span id="cb41-188"><a href="#cb41-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-191"><a href="#cb41-191" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-192"><a href="#cb41-192" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert implicit gaps to explicit NA gaps, and drop days with &lt;1 hour of data</span></span>
<span id="cb41-193"><a href="#cb41-193" aria-hidden="true" tabindex="-1"></a>dataCC <span class="ot">&lt;-</span> dataCC <span class="sc">|&gt;</span> </span>
<span id="cb41-194"><a href="#cb41-194" aria-hidden="true" tabindex="-1"></a>  <span class="co"># First ensure that status columns have an "operational" tag for non-missing periods:</span></span>
<span id="cb41-195"><a href="#cb41-195" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(Lux_status, Dis_status), <span class="sc">~</span> <span class="fu">replace_na</span>(.x, <span class="st">"operational"</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb41-196"><a href="#cb41-196" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Fill in all implicit gaps with explicit NA rows (for full days range):</span></span>
<span id="cb41-197"><a href="#cb41-197" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>  </span>
<span id="cb41-198"><a href="#cb41-198" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove any day that has less than 1 hour of recorded data</span></span>
<span id="cb41-199"><a href="#cb41-199" aria-hidden="true" tabindex="-1"></a>  <span class="fu">remove_partial_data</span>(Dis, <span class="at">threshold.missing =</span> <span class="st">"23 hours"</span>)</span>
<span id="cb41-200"><a href="#cb41-200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-201"><a href="#cb41-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-202"><a href="#cb41-202" aria-hidden="true" tabindex="-1"></a>After these steps,&nbsp;<span class="in">`dataCC`</span>&nbsp;contains continuous 5-second timestamps for each day that remains. We chose a threshold of “23 hours” missing to remove days with <span class="sc">\&lt;</span>1 hour of data, which in this dataset removes the first and last (partial) days. The cleaned <span class="in">`Clouclip`</span> data now covers six full days with bouts of continuous wear.</span>
<span id="cb41-203"><a href="#cb41-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-204"><a href="#cb41-204" aria-hidden="true" tabindex="-1"></a>It is often helpful to double-check how the sentinel values and missing data are distributed over time. We can visualize the distance time-series with status annotations and day/night periods:</span>
<span id="cb41-205"><a href="#cb41-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-208"><a href="#cb41-208" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-209"><a href="#cb41-209" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb41-210"><a href="#cb41-210" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb41-211"><a href="#cb41-211" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-state</span></span>
<span id="cb41-212"><a href="#cb41-212" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Distance measurements across days. Blue, grey and yellow-colored areas show sentinel states of the device. Blue indicates an operational status, grey sleep mode (not recording), and yellow an out of range measurement. Red boxed areas show nighttime from civil dusk until dawn, which are calculated based on the recording date and geographic coordinates"</span></span>
<span id="cb41-213"><a href="#cb41-213" aria-hidden="true" tabindex="-1"></a><span class="co">#setting coordinates for Houston, Texas</span></span>
<span id="cb41-214"><a href="#cb41-214" aria-hidden="true" tabindex="-1"></a>coordinates <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">29.75</span>, <span class="sc">-</span><span class="fl">95.36</span>)</span>
<span id="cb41-215"><a href="#cb41-215" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize observations</span></span>
<span id="cb41-216"><a href="#cb41-216" aria-hidden="true" tabindex="-1"></a>dataCC <span class="sc">|&gt;</span> </span>
<span id="cb41-217"><a href="#cb41-217" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fill</span>(<span class="fu">c</span>(Lux_status, Dis_status), <span class="at">.direction =</span> <span class="st">"downup"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb41-218"><a href="#cb41-218" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gg_day</span>(<span class="at">y.axis =</span> Dis, <span class="at">geom =</span> <span class="st">"line"</span>, <span class="at">y.axis.label =</span> y.label) <span class="sc">|&gt;</span> <span class="co">#create a basic plot</span></span>
<span id="cb41-219"><a href="#cb41-219" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gg_state</span>(Dis_status, <span class="at">aes_fill =</span> Dis_status) <span class="sc">|&gt;</span> <span class="co">#add the status times</span></span>
<span id="cb41-220"><a href="#cb41-220" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gg_photoperiod</span>(coordinates, <span class="at">alpha =</span> <span class="fl">0.1</span>, <span class="at">col =</span> <span class="st">"red"</span>) <span class="sc">+</span> <span class="co">#add the photoperiod (day/night)</span></span>
<span id="cb41-221"><a href="#cb41-221" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb41-222"><a href="#cb41-222" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-223"><a href="#cb41-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-224"><a href="#cb41-224" aria-hidden="true" tabindex="-1"></a>In this plot, blue segments indicate times when the Clouclip was&nbsp;*operational*&nbsp;(actively measuring), grey segments indicate the device in&nbsp;*sleep mode*&nbsp;(no recording, typically at night), and yellow segments indicate&nbsp;*out-of-range*&nbsp;distance readings. The red outlined regions show nighttime (from civil dusk to dawn) based on the given latitude/longitude and dates. As expected, most of the grey “sleep” periods align with night hours, and we see a few yellow spans when the user’s viewing distance exceeded the device’s range (e.g. presumably when no object was within 100&nbsp;cm, such as when the user looked into the distance). At this stage, the Clouclip dataset&nbsp;<span class="in">`dataCC`</span>&nbsp;is fully preprocessed: all timestamps are regular 5-second intervals, missing data are explicitly marked, extraneous partial days are removed, and sentinel codes are handled via the status columns. The data are ready for calculating daily distance and light exposure metrics (as done in the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](https://tscnlab.github.io/ZaunerEtAl_JVis_2025/)</span>’s Results).</span>
<span id="cb41-225"><a href="#cb41-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-226"><a href="#cb41-226" aria-hidden="true" tabindex="-1"></a><span class="fu">## `VEET` Data: Ambient Light (Illuminance) Processing</span></span>
<span id="cb41-227"><a href="#cb41-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-228"><a href="#cb41-228" aria-hidden="true" tabindex="-1"></a>The&nbsp;**VEET**&nbsp;device [@Sah2025OphtalmicPhysiolOpt, @Sullivan2024] is a more complex logger that records multiple data modalities in one combined file. Its raw data file contains interleaved records for different sensor types, distinguished by a “modality” field. We focus first on the ambient light sensor modality (abbreviated&nbsp;**ALS**), which provides broad-spectrum illuminance readings (lux) and related information like sensor gains and flicker, recorded every 2 seconds. Later we will import the spectral sensor modality (**PHO**) for spectral irradiance data, and the time-of-flight modality (**TOF**) for distance data.</span>
<span id="cb41-229"><a href="#cb41-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-230"><a href="#cb41-230" aria-hidden="true" tabindex="-1"></a>In the <span class="in">`VEET`</span>’s export file, each line includes a timestamp and a modality code, followed by fields specific to that modality. Importantly, this means that the <span class="in">`VEET`</span> export is not rectangular, i.e., tabular. This makes it challenging for many import functions that expect the equal number of columns in every row, which is not the case in this instance. For the&nbsp;**ALS modality**, the relevant columns include a high-resolution timestamp (in Unix epoch format), integration time, UV/VIS/IR sensor gain settings, raw UV/VIS/IR sensor counts, a flicker measurement, and the computed illuminance in lux. For example, the ALS data columns are named:&nbsp;<span class="in">`time_stamp`</span>,&nbsp;<span class="in">`integration_time`</span>,&nbsp;<span class="in">`uvGain`</span>,&nbsp;<span class="in">`visGain`</span>,&nbsp;<span class="in">`irGain`</span>,&nbsp;<span class="in">`uvValue`</span>,&nbsp;<span class="in">`visValue`</span>,&nbsp;<span class="in">`irValue`</span>,&nbsp;<span class="in">`Flicker`</span>, and&nbsp;<span class="in">`Lux`</span>.</span>
<span id="cb41-231"><a href="#cb41-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-232"><a href="#cb41-232" aria-hidden="true" tabindex="-1"></a>For the&nbsp;**PHO (spectral) modality**, the columns include a timestamp, integration time, a general&nbsp;<span class="in">`Gain`</span>&nbsp;factor, and nine sensor channel readings covering different wavelengths (with names like&nbsp;<span class="in">`s415, s445, ..., s680, s940`</span>) as well as a&nbsp;<span class="in">`Dark`</span>&nbsp;channel and two broadband channels&nbsp;<span class="in">`ClearL`</span>&nbsp;and&nbsp;<span class="in">`ClearR`</span>. In essence, the <span class="in">`VEET`</span>’s spectral sensor captures light in several wavelength bands (from \~415&nbsp;nm up to 940&nbsp;nm, plus an infrared and two “clear” channels) rather than outputting a single lux value like the ambient light sensor does (*PHO*).</span>
<span id="cb41-233"><a href="#cb41-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-234"><a href="#cb41-234" aria-hidden="true" tabindex="-1"></a>To import the <span class="in">`VEET`</span> ambient light data, we again use the LightLogR import function, specifying the <span class="in">`ALS`</span> modality. The raw VEET data in our example is provided as a zip file (<span class="in">`01_VEET_L.csv.zip`</span>) containing the logged data for one week. We do the following:</span>
<span id="cb41-235"><a href="#cb41-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-238"><a href="#cb41-238" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-239"><a href="#cb41-239" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-VEET-overview</span></span>
<span id="cb41-240"><a href="#cb41-240" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Overview plot of imported VEET data"</span></span>
<span id="cb41-241"><a href="#cb41-241" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb41-242"><a href="#cb41-242" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 6</span></span>
<span id="cb41-243"><a href="#cb41-243" aria-hidden="true" tabindex="-1"></a><span class="co"># Import VEET Ambient Light Sensor (ALS) data</span></span>
<span id="cb41-244"><a href="#cb41-244" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="cb41-245"><a href="#cb41-245" aria-hidden="true" tabindex="-1"></a>tz   <span class="ot">&lt;-</span> <span class="st">"US/Central"</span>    <span class="co"># assuming device clock was set to US Central, for consistency</span></span>
<span id="cb41-246"><a href="#cb41-246" aria-hidden="true" tabindex="-1"></a>dataVEET <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"ALS"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>)</span>
<span id="cb41-247"><a href="#cb41-247" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-248"><a href="#cb41-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-249"><a href="#cb41-249" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb41-250"><a href="#cb41-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-251"><a href="#cb41-251" aria-hidden="true" tabindex="-1"></a>We get one warning as a single time stamp could not be parsed into a datetime - with \~300k observations, this is not an issue.</span>
<span id="cb41-252"><a href="#cb41-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-253"><a href="#cb41-253" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb41-254"><a href="#cb41-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-255"><a href="#cb41-255" aria-hidden="true" tabindex="-1"></a>This call reads in only the lines corresponding to the <span class="in">`ALS`</span> modality from the <span class="in">`VEET`</span> file. The result&nbsp;<span class="in">`dataVEET`</span>&nbsp;is a tibble<span class="ot">[^3]</span> with columns such as&nbsp;<span class="in">`Datetime`</span>&nbsp;(parsed from the&nbsp;<span class="in">`time_stamp`</span>&nbsp;epoch to POSIXct in US/Central time),&nbsp;<span class="in">`Lux`</span>&nbsp;(illuminance in lux),&nbsp;<span class="in">`Flicker`</span>, and the various sensor gains/values. Unneeded columns like the modality code or file name are also included but can be ignored or removed. After import, LightLogR again provides an overview of the data. We learn that the <span class="in">`VEET`</span> light data, like the <span class="in">`Clouclip`</span>, also exhibits irregularities and gaps. (The device nominally records every 2 seconds, but timing may drift or pause when not worn.)</span>
<span id="cb41-256"><a href="#cb41-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-257"><a href="#cb41-257" aria-hidden="true" tabindex="-1"></a><span class="ot">[^3]: </span>**tibble** are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <span class="co">[</span><span class="ot">tibbles</span><span class="co">](https://tibble.tidyverse.org/index.html)</span></span>
<span id="cb41-258"><a href="#cb41-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-259"><a href="#cb41-259" aria-hidden="true" tabindex="-1"></a>To make the VEET light data comparable to the Clouclip’s and to simplify analysis, we choose to&nbsp;**aggregate the VEET illuminance data to 5-second intervals**. This slight downsampling will both reduce data volume and help align with the Clouclip’s timeline for any combined analysis. Before aggregation, we will also explicitly mark gaps in the VEET data so that missing intervals are not overlooked.</span>
<span id="cb41-260"><a href="#cb41-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-263"><a href="#cb41-263" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-264"><a href="#cb41-264" aria-hidden="true" tabindex="-1"></a><span class="co"># Aggregate VEET light data to 5-second intervals and mark gaps</span></span>
<span id="cb41-265"><a href="#cb41-265" aria-hidden="true" tabindex="-1"></a>dataVEET <span class="ot">&lt;-</span> dataVEET <span class="sc">|&gt;</span></span>
<span id="cb41-266"><a href="#cb41-266" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 seconds"</span>) <span class="sc">|&gt;</span>  <span class="co"># resample to 5-sec bins (e.g. average Lux over 2-sec readings)</span></span>
<span id="cb41-267"><a href="#cb41-267" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>          <span class="co"># fill in implicit gaps with NA rows</span></span>
<span id="cb41-268"><a href="#cb41-268" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>          <span class="co"># add Date column for daily grouping</span></span>
<span id="cb41-269"><a href="#cb41-269" aria-hidden="true" tabindex="-1"></a>  <span class="fu">remove_partial_data</span>(Lux, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span>
<span id="cb41-270"><a href="#cb41-270" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-271"><a href="#cb41-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-272"><a href="#cb41-272" aria-hidden="true" tabindex="-1"></a>First, <span class="in">`aggregate_Datetime(unit = "5 seconds")`</span> combines the high-frequency 2-second observations into 5-second slots. By default, this function will average numeric columns like&nbsp;<span class="in">`Lux`</span>&nbsp;over each 5-second period (and handle counts or categorical appropriately). All of these *data type handlers* can be changed with the function call. The result is that&nbsp;<span class="in">`dataVEET`</span>&nbsp;now has a reading every 5 seconds (or an NA if none were present in that window). Next,&nbsp;<span class="in">`gap_handler(full.days = TRUE)`</span>&nbsp;inserts explicit NA entries for any 5-second timestamp that had no data within the continuous span of the recording. Then we add a&nbsp;<span class="in">`Date`</span>&nbsp;column for grouping, and finally we remove days with more than 1 hour of missing data (using a more strict criterion as we did for <span class="in">`Clouclip`</span>). According to the gap summary (@tbl-gaps2), this leaves six full days of <span class="in">`VEET`</span> light data with good coverage, after dropping the very incomplete start/end days.</span>
<span id="cb41-273"><a href="#cb41-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-274"><a href="#cb41-274" aria-hidden="true" tabindex="-1"></a>We can inspect the missing-data summary for the <span class="in">`VEET`</span> illuminance data:</span>
<span id="cb41-275"><a href="#cb41-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-278"><a href="#cb41-278" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-279"><a href="#cb41-279" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: tbl-gaps2</span></span>
<span id="cb41-280"><a href="#cb41-280" aria-hidden="true" tabindex="-1"></a><span class="co">#| tbl-cap: "Summary of missing and observed data for the VEET device"</span></span>
<span id="cb41-281"><a href="#cb41-281" aria-hidden="true" tabindex="-1"></a>dataVEET <span class="sc">|&gt;</span> <span class="fu">gap_table</span>(Lux, <span class="st">"Illuminance (lx)"</span>) <span class="sc">|&gt;</span>   </span>
<span id="cb41-282"><a href="#cb41-282" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cols_hide</span>(<span class="fu">contains</span>(<span class="st">"_n"</span>)) <span class="co">#remove the absolute number of data points</span></span>
<span id="cb41-283"><a href="#cb41-283" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-284"><a href="#cb41-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-285"><a href="#cb41-285" aria-hidden="true" tabindex="-1"></a>@tbl-gaps2 shows, for each retained day, the total recorded duration and the duration of gaps. The <span class="in">`VEET`</span> device, like the <span class="in">`Clouclip`</span>, was not worn continuously 24&nbsp;hours per day, so there are nightly gaps of wear time (when the device was likely off the participant), but not of recordings. After our preprocessing, any implicit gaps are represented as explicit missing intervals. The <span class="in">`VEET`</span>’s time sampling was originally more frequent, but by aggregating to 5&nbsp;s we have ensured a uniform timeline akin to the <span class="in">`Clouclip`</span>’s.</span>
<span id="cb41-286"><a href="#cb41-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-287"><a href="#cb41-287" aria-hidden="true" tabindex="-1"></a>At this point, the&nbsp;<span class="in">`dataVEET`</span>&nbsp;object (for illuminance) is cleaned and ready for computing light exposure metrics. For example, one could calculate daily mean illuminance or the duration spent above certain light thresholds (e.g. “outdoor light exposure” defined as <span class="sc">\&gt;</span>1000&nbsp;lx) using this dataset. Indeed, basic summary tables in the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](https://tscnlab.github.io/ZaunerEtAl_JVis_2025/)</span> illustrate the highly skewed nature of light exposure data and the calculation of outdoor light metrics. We will not repeat those metric calculations here in the supplement, as our focus is on data preprocessing; however, having a cleaned, gap-marked time series is crucial for those metrics to be accurate.</span>
<span id="cb41-288"><a href="#cb41-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-289"><a href="#cb41-289" aria-hidden="true" tabindex="-1"></a><span class="fu">## `VEET` Data: Spectral Data Processing</span></span>
<span id="cb41-290"><a href="#cb41-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-291"><a href="#cb41-291" aria-hidden="true" tabindex="-1"></a><span class="fu">### Import</span></span>
<span id="cb41-292"><a href="#cb41-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-293"><a href="#cb41-293" aria-hidden="true" tabindex="-1"></a>In addition to broad-band illuminance and distance, the <span class="in">`VEET`</span> provides&nbsp;**spectral sensor data**&nbsp;through its <span class="in">`PHO`</span> modality. Unlike illuminance, the spectral data are not directly given as directly interpretable radiometric metrics but rather as raw sensor counts across multiple wavelength channels, which require conversion to reconstruct a spectral power distribution. In our analysis, spectral data allow us to compute metrics like the relative contribution of short-wavelength (blue) light versus long-wavelength light in the participant’s environment. Processing this spectral data involves several necessary steps.</span>
<span id="cb41-294"><a href="#cb41-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-295"><a href="#cb41-295" aria-hidden="true" tabindex="-1"></a>First, we import the spectral modality from a second <span class="in">`VEET`</span> file. This device used the latest software version to allow for the best spectral reconstruction. This time we need to extract the lines marked as <span class="in">`PHO`</span>. We will store the spectral dataset in a separate object&nbsp;<span class="in">`dataVEET2`</span>&nbsp;so as not to overwrite the&nbsp;<span class="in">`dataVEET`</span>&nbsp;illuminance data in our R session:</span>
<span id="cb41-296"><a href="#cb41-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-299"><a href="#cb41-299" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-300"><a href="#cb41-300" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-VEET-overview2</span></span>
<span id="cb41-301"><a href="#cb41-301" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Overview plot of imported VEET data"</span></span>
<span id="cb41-302"><a href="#cb41-302" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb41-303"><a href="#cb41-303" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 6</span></span>
<span id="cb41-304"><a href="#cb41-304" aria-hidden="true" tabindex="-1"></a><span class="co"># Import VEET Spectral Sensor (PHO) data</span></span>
<span id="cb41-305"><a href="#cb41-305" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/02_VEET_L.csv.zip"</span></span>
<span id="cb41-306"><a href="#cb41-306" aria-hidden="true" tabindex="-1"></a><span class="co">#if a version of LightLogR ≤ 0.9.2 is used, this script needs to be imported, as the device data was collected with a newer firmware version that changes the output format.</span></span>
<span id="cb41-307"><a href="#cb41-307" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"scripts/VEET_import.R"</span>)</span>
<span id="cb41-308"><a href="#cb41-308" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="ot">&lt;-</span> import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"PHO"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>)</span>
<span id="cb41-309"><a href="#cb41-309" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-310"><a href="#cb41-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-311"><a href="#cb41-311" aria-hidden="true" tabindex="-1"></a>After import,&nbsp;<span class="in">`dataVEET2`</span>&nbsp;contains columns for the timestamp (<span class="in">`Datetime`</span>),&nbsp;<span class="in">`Gain`</span>&nbsp;(the sensor gain setting), and the nine spectral sensor channels plus a clear channel. These appear as numeric columns named&nbsp;<span class="in">`s415, s445, ..., s940, Dark, Clear`</span>. The spectral sensor was logging at a 2-second rate. Because this dataset is even denser and more high-dimensional, we will&nbsp;**aggregate it to a 5-minute interval** for computational efficiency. The assumption is that spectral composition does not need to be examined at every 2-second instant for our purposes, and 5-minute averages can capture the general trends while drastically reducing data size and downstream computational costs.</span>
<span id="cb41-312"><a href="#cb41-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-315"><a href="#cb41-315" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-316"><a href="#cb41-316" aria-hidden="true" tabindex="-1"></a><span class="co"># Aggregate spectral data to 5-minute intervals and mark gaps</span></span>
<span id="cb41-317"><a href="#cb41-317" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="ot">&lt;-</span> dataVEET2 <span class="sc">|&gt;</span></span>
<span id="cb41-318"><a href="#cb41-318" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 mins"</span>) <span class="sc">|&gt;</span>     <span class="co"># aggregate to 5-minute bins</span></span>
<span id="cb41-319"><a href="#cb41-319" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>           <span class="co"># explicit NA for any gaps in between</span></span>
<span id="cb41-320"><a href="#cb41-320" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb41-321"><a href="#cb41-321" aria-hidden="true" tabindex="-1"></a>  <span class="fu">remove_partial_data</span>(Clear, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span>
<span id="cb41-322"><a href="#cb41-322" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-323"><a href="#cb41-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-324"><a href="#cb41-324" aria-hidden="true" tabindex="-1"></a>We aggregate over 5-minute windows; within each 5-minute bin, multiple spectral readings (if present) are combined (averaged). We use one of the channels (here&nbsp;<span class="in">`Clear`</span>) as the reference variable for&nbsp;<span class="in">`remove_partial_data`</span>&nbsp;to drop incomplete days (the choice of channel is arbitrary as all channels share the same level of completeness).</span>
<span id="cb41-325"><a href="#cb41-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-326"><a href="#cb41-326" aria-hidden="true" tabindex="-1"></a>It is informative to look at a snippet of the imported spectral data before further processing. @tbl-PHO shows the first three rows of&nbsp;<span class="in">`dataVEET2`</span>&nbsp;after import (before calibration), with some technical columns omitted for brevity:</span>
<span id="cb41-327"><a href="#cb41-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-330"><a href="#cb41-330" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-331"><a href="#cb41-331" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: tbl-PHO</span></span>
<span id="cb41-332"><a href="#cb41-332" aria-hidden="true" tabindex="-1"></a><span class="co">#| tbl-cap: "Overview of the spectral sensor import from the VEET device (first 3 observations). Each row corresponds to a 5-minute timestamp (Datetime) and shows the raw sensor readings for the spectral channels (s415–s940, Dark, Clear). All values are in arbitrary sensor units (counts). Gain values and integration_time are also relevant for each interval, depending on the downstream computation."</span></span>
<span id="cb41-333"><a href="#cb41-333" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="dv">3</span>) <span class="sc">|&gt;</span> <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(modality, file.name, is.implicit, time_stamp)) <span class="sc">|&gt;</span> </span>
<span id="cb41-334"><a href="#cb41-334" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gt</span>() <span class="sc">|&gt;</span> <span class="fu">fmt_number</span>(s415<span class="sc">:</span>Clear) </span>
<span id="cb41-335"><a href="#cb41-335" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-336"><a href="#cb41-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-337"><a href="#cb41-337" aria-hidden="true" tabindex="-1"></a><span class="fu">### Spectral calibration</span></span>
<span id="cb41-338"><a href="#cb41-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-339"><a href="#cb41-339" aria-hidden="true" tabindex="-1"></a>Now we proceed with&nbsp;**spectral calibration**. The <span class="in">`VEET`</span>’s spectral sensor counts need to be converted to physical units (spectral irradiance) via a calibration matrix provided by the manufacturer. For this example, we assume we have a calibration matrix that maps all the channel readings to an estimated spectral power distribution (SPD). The LightLogR package provides a function&nbsp;<span class="in">`spectral_reconstruction()`</span>&nbsp;to perform this conversion. However, before applying it, we must ensure the sensor counts are in a normalized form. This procedure is laid out by the manufacturer. In our version, we refer to the *VEET SPD Reconstruction Guide.pdf*, version *06/05/2025*. Note that each manufacturer has to specify the method of count normalization (if any) and spectral reconstruction. In our raw data, each observation comes with a&nbsp;<span class="in">`Gain`</span>&nbsp;setting that indicates how the sensor’s sensitivity was adjusted; we need to divide the raw counts by the gain to get normalized counts. LightLogR offers&nbsp;<span class="in">`normalize_counts()`</span>&nbsp;for this purpose. We further need to scale by integration time (in milliseconds) and adjust depending on counts in the <span class="in">`Dark`</span> sensor channel. </span>
<span id="cb41-340"><a href="#cb41-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-343"><a href="#cb41-343" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-344"><a href="#cb41-344" aria-hidden="true" tabindex="-1"></a>count.columns <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"s415"</span>, <span class="st">"s445"</span>, <span class="st">"s480"</span>, <span class="st">"s515"</span>, <span class="st">"s555"</span>, <span class="st">"s590"</span>, <span class="st">"s630"</span>, </span>
<span id="cb41-345"><a href="#cb41-345" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"s680"</span>, <span class="st">"s910"</span>, <span class="st">"Dark"</span>, <span class="st">"Clear"</span>) <span class="co">#column names</span></span>
<span id="cb41-346"><a href="#cb41-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-347"><a href="#cb41-347" aria-hidden="true" tabindex="-1"></a>gain.ratios <span class="ot">&lt;-</span> <span class="co">#gain ratios as specified by the manufacturers reconstruction guide</span></span>
<span id="cb41-348"><a href="#cb41-348" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(</span>
<span id="cb41-349"><a href="#cb41-349" aria-hidden="true" tabindex="-1"></a>    <span class="at">gain =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">8</span>, <span class="dv">16</span>, <span class="dv">32</span>, <span class="dv">64</span>, <span class="dv">128</span>, <span class="dv">256</span>, <span class="dv">512</span>),</span>
<span id="cb41-350"><a href="#cb41-350" aria-hidden="true" tabindex="-1"></a>    <span class="at">gain.ratio =</span> <span class="fu">c</span>(<span class="fl">0.008</span>, <span class="fl">0.016</span>, <span class="fl">0.032</span>, <span class="fl">0.065</span>, <span class="fl">0.125</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="fl">3.95</span>, <span class="fl">7.75</span>)</span>
<span id="cb41-351"><a href="#cb41-351" aria-hidden="true" tabindex="-1"></a>  ) </span>
<span id="cb41-352"><a href="#cb41-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-353"><a href="#cb41-353" aria-hidden="true" tabindex="-1"></a><span class="co">#normalize data</span></span>
<span id="cb41-354"><a href="#cb41-354" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="ot">&lt;-</span></span>
<span id="cb41-355"><a href="#cb41-355" aria-hidden="true" tabindex="-1"></a>  dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="cb41-356"><a href="#cb41-356" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(s415<span class="sc">:</span>Clear), \(x) (x <span class="sc">-</span> Dark)<span class="sc">/</span>integration_time)) <span class="sc">|&gt;</span>  <span class="co">#remove dark counts &amp; scale by integration time</span></span>
<span id="cb41-357"><a href="#cb41-357" aria-hidden="true" tabindex="-1"></a>  <span class="fu">normalize_counts</span>( <span class="co">#function to normalize counts</span></span>
<span id="cb41-358"><a href="#cb41-358" aria-hidden="true" tabindex="-1"></a>    <span class="at">gain.columns =</span> <span class="fu">rep</span>(<span class="st">"Gain"</span>, <span class="dv">11</span>), <span class="co">#all sensor channels share the gain value</span></span>
<span id="cb41-359"><a href="#cb41-359" aria-hidden="true" tabindex="-1"></a>    <span class="at">count.columns =</span> count.columns, <span class="co">#sensor channels to normalize</span></span>
<span id="cb41-360"><a href="#cb41-360" aria-hidden="true" tabindex="-1"></a>    gain.ratios <span class="co">#gain ratios</span></span>
<span id="cb41-361"><a href="#cb41-361" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb41-362"><a href="#cb41-362" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(s415<span class="sc">:</span>Clear)) <span class="sc">|&gt;</span> <span class="co"># drop original raw count columns</span></span>
<span id="cb41-363"><a href="#cb41-363" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename_with</span>(<span class="sc">~</span> <span class="fu">str_remove</span>(.x, <span class="st">".normalized"</span>))</span>
<span id="cb41-364"><a href="#cb41-364" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-365"><a href="#cb41-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-366"><a href="#cb41-366" aria-hidden="true" tabindex="-1"></a>In this call, we specified <span class="in">`gain.columns = rep("Gain", 11)`</span> because we have 11 sensor columns that all use the same gain factor column (<span class="in">`Gain`</span>). This step will add new columns (with a suffix, e.g. <span class="in">`.normalized`</span>) for each spectral channel representing the count normalized by the gain. We then dropped the <span class="in">`raw`</span> count columns and renamed the normalized ones by dropping <span class="in">`.normalized`</span> from the names. After this, <span class="in">`dataVEET2`</span> contains the normalized sensor readings for <span class="in">`s415, s445, ..., s940, Dark, Clear`</span> for each 5-minute time point. The dataset is now ready for spectral reconstruction.</span>
<span id="cb41-367"><a href="#cb41-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-368"><a href="#cb41-368" aria-hidden="true" tabindex="-1"></a><span class="fu">### Spectral reconstruction</span></span>
<span id="cb41-369"><a href="#cb41-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-370"><a href="#cb41-370" aria-hidden="true" tabindex="-1"></a>For spectral reconstruction, we require a calibration matrix that corresponds to the VEET’s sensor channels. This matrix would typically be obtained from the device manufacturer or a calibration procedure. It defines how each channel’s count relates to intensity at various wavelengths. For demonstration, the calibration matrix was provided by the manufacturer and is specific to the make and model. It should not be used for research purposes without confirming its accuracy with the manufacturer.</span>
<span id="cb41-371"><a href="#cb41-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-374"><a href="#cb41-374" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-375"><a href="#cb41-375" aria-hidden="true" tabindex="-1"></a><span class="co">#import calibration matrix</span></span>
<span id="cb41-376"><a href="#cb41-376" aria-hidden="true" tabindex="-1"></a>calib_mtx <span class="ot">&lt;-</span> </span>
<span id="cb41-377"><a href="#cb41-377" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read_csv</span>(<span class="st">"data/VEET_calibration_matrix.csv"</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb41-378"><a href="#cb41-378" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"wavelength"</span>)</span>
<span id="cb41-379"><a href="#cb41-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-380"><a href="#cb41-380" aria-hidden="true" tabindex="-1"></a><span class="co"># Reconstruct spectral power distribution (SPD) for each observation</span></span>
<span id="cb41-381"><a href="#cb41-381" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="ot">&lt;-</span> dataVEET2 <span class="sc">|&gt;</span> <span class="fu">mutate</span>(</span>
<span id="cb41-382"><a href="#cb41-382" aria-hidden="true" tabindex="-1"></a>  <span class="at">Spectrum =</span> <span class="fu">spectral_reconstruction</span>(</span>
<span id="cb41-383"><a href="#cb41-383" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pick</span>(s415<span class="sc">:</span>s910),   <span class="co"># pick the normalized sensor columns</span></span>
<span id="cb41-384"><a href="#cb41-384" aria-hidden="true" tabindex="-1"></a>    <span class="at">calibration_matrix =</span> calib_mtx, </span>
<span id="cb41-385"><a href="#cb41-385" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"long"</span> <span class="co"># return a long-form list column (wavelength, intensity)</span></span>
<span id="cb41-386"><a href="#cb41-386" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb41-387"><a href="#cb41-387" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb41-388"><a href="#cb41-388" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-389"><a href="#cb41-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-390"><a href="#cb41-390" aria-hidden="true" tabindex="-1"></a>Here, we use&nbsp;<span class="in">`format = "long"`</span>&nbsp;so that the result for each observation is a&nbsp;**list-column**&nbsp;<span class="in">`Spectrum`</span>, where each entry is a tibble<span class="ot">[^4]</span> containing two columns:&nbsp;<span class="in">`wavelength`</span>&nbsp;and&nbsp;<span class="in">`irradiance`</span>&nbsp;(one row per wavelength in the calibration matrix). In other words, each row of&nbsp;<span class="in">`dataVEET2`</span>&nbsp;now holds a full reconstructed spectrum in the&nbsp;<span class="in">`Spectrum`</span>&nbsp;column. The long format is convenient for further calculations and plotting. (An alternative&nbsp;<span class="in">`format = "wide"`</span>&nbsp;would add each wavelength as a separate column, but that is less practical when there are many wavelengths.)</span>
<span id="cb41-391"><a href="#cb41-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-392"><a href="#cb41-392" aria-hidden="true" tabindex="-1"></a><span class="ot">[^4]: </span>**tibble** are data.tables with tweaked behavior, ideal for a tidy analysis workflow. For more information, visit the documentation page for <span class="co">[</span><span class="ot">tibbles</span><span class="co">](https://tibble.tidyverse.org/index.html)</span></span>
<span id="cb41-393"><a href="#cb41-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-394"><a href="#cb41-394" aria-hidden="true" tabindex="-1"></a>To visualize the data we will calculate the photopic illuminance based on the spectra and plot each spectrum color-scaled by their illuminance. For clarity, we reduce the data to observations within the first day.</span>
<span id="cb41-395"><a href="#cb41-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-398"><a href="#cb41-398" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-399"><a href="#cb41-399" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-VEET-spectra</span></span>
<span id="cb41-400"><a href="#cb41-400" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Overview of the reconstructed spectra, color-scaled by photopic illuminance (lx)"</span></span>
<span id="cb41-401"><a href="#cb41-401" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb41-402"><a href="#cb41-402" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb41-403"><a href="#cb41-403" aria-hidden="true" tabindex="-1"></a>dataVEET2 <span class="sc">|&gt;</span> </span>
<span id="cb41-404"><a href="#cb41-404" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_Date</span>(<span class="at">length =</span> <span class="st">"1 day"</span>) <span class="sc">|&gt;</span> <span class="co">#keep only observations for one day (from start)</span></span>
<span id="cb41-405"><a href="#cb41-405" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>( </span>
<span id="cb41-406"><a href="#cb41-406" aria-hidden="true" tabindex="-1"></a>    <span class="at">Illuminance =</span> Spectrum <span class="sc">|&gt;</span> <span class="co">#Use the spectrum,...</span></span>
<span id="cb41-407"><a href="#cb41-407" aria-hidden="true" tabindex="-1"></a>      <span class="fu">map_dbl</span>(spectral_integration, <span class="co">#... call the function spectral_integration for each,...</span></span>
<span id="cb41-408"><a href="#cb41-408" aria-hidden="true" tabindex="-1"></a>              <span class="at">action.spectrum =</span> <span class="st">"photopic"</span>, <span class="co">#... use the brightness sensitivity function,...</span></span>
<span id="cb41-409"><a href="#cb41-409" aria-hidden="true" tabindex="-1"></a>              <span class="at">general.weight =</span> <span class="st">"auto"</span>) <span class="co">#... and apply the appropriate efficacy weight.</span></span>
<span id="cb41-410"><a href="#cb41-410" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb41-411"><a href="#cb41-411" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(Spectrum) <span class="sc">|&gt;</span> <span class="co">#create a long format of the data where the spectrum is unnested</span></span>
<span id="cb41-412"><a href="#cb41-412" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> wavelength, <span class="at">y =</span> irradiance<span class="sc">*</span><span class="dv">1000</span>, <span class="at">group =</span> Datetime)) <span class="sc">+</span></span>
<span id="cb41-413"><a href="#cb41-413" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">col =</span> Illuminance)) <span class="sc">+</span></span>
<span id="cb41-414"><a href="#cb41-414" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Irradiance (mW/m²/nm)"</span>, <span class="at">x =</span> <span class="st">"Wavelength (nm)"</span>, <span class="at">col =</span> <span class="st">"Photopic illuminance (lx)"</span>)<span class="sc">+</span></span>
<span id="cb41-415"><a href="#cb41-415" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_b</span>() <span class="sc">+</span></span>
<span id="cb41-416"><a href="#cb41-416" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">5.3</span>)) <span class="sc">+</span></span>
<span id="cb41-417"><a href="#cb41-417" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb41-418"><a href="#cb41-418" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-419"><a href="#cb41-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-420"><a href="#cb41-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-421"><a href="#cb41-421" aria-hidden="true" tabindex="-1"></a>At this stage, the&nbsp;<span class="in">`dataVEET2`</span>&nbsp;dataset has been processed to yield time-series of spectral power distributions. We can use these to compute biologically relevant light metrics. For instance, one possible metric is the proportion of power in short wavelengths versus long wavelengths. In the main analysis, we defined&nbsp;**short-wavelength (blue light) content**&nbsp;as the integrated intensity in the 400–500&nbsp;nm range, and&nbsp;**long-wavelength content**&nbsp;as the integrated intensity in a longer range (e.g. 600–700&nbsp;nm), then computed the&nbsp;*short-to-long ratio*&nbsp;(“sl ratio”). Calculating these metrics is the first step of spectrum analysis in the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](https://tscnlab.github.io/ZaunerEtAl_JVis_2025/#spectrum)</span>.</span>
<span id="cb41-422"><a href="#cb41-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-423"><a href="#cb41-423" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-424"><a href="#cb41-424" aria-hidden="true" tabindex="-1"></a><span class="in">dataVEET2 |&gt; </span></span>
<span id="cb41-425"><a href="#cb41-425" aria-hidden="true" tabindex="-1"></a><span class="in">  select(Id, Date, Datetime, Spectrum) |&gt;    # focus on ID, date, time, and spectrum</span></span>
<span id="cb41-426"><a href="#cb41-426" aria-hidden="true" tabindex="-1"></a><span class="in">  mutate(</span></span>
<span id="cb41-427"><a href="#cb41-427" aria-hidden="true" tabindex="-1"></a><span class="in">    short = Spectrum |&gt; map_dbl(spectral_integration, wavelength.range = c(400, 500)),</span></span>
<span id="cb41-428"><a href="#cb41-428" aria-hidden="true" tabindex="-1"></a><span class="in">    long  = Spectrum |&gt; map_dbl(spectral_integration, wavelength.range = c(600, 700)),</span></span>
<span id="cb41-429"><a href="#cb41-429" aria-hidden="true" tabindex="-1"></a><span class="in">    `sl ratio` = short / long   # compute short-to-long ratio</span></span>
<span id="cb41-430"><a href="#cb41-430" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb41-431"><a href="#cb41-431" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-432"><a href="#cb41-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-433"><a href="#cb41-433" aria-hidden="true" tabindex="-1"></a>*(The cutoff of 500&nbsp;nm here is hypothetical for demonstration; actual definitions might vary.)*&nbsp;We would then have columns&nbsp;<span class="in">`short`</span>,&nbsp;<span class="in">`long`</span>, and&nbsp;<span class="in">`sl_ratio`</span>&nbsp;for each observation, which could be averaged per day or analyzed further. The cleaned spectral data in&nbsp;<span class="in">`dataVEET2`</span>&nbsp;makes it straightforward to calculate such metrics or apply spectral weighting functions (for melatonin suppression, circadian stimulus, etc., if one has the spectral sensitivity curves).</span>
<span id="cb41-434"><a href="#cb41-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-435"><a href="#cb41-435" aria-hidden="true" tabindex="-1"></a>With the <span class="in">`VEET`</span> spectral preprocessing complete, we emphasize that these steps – normaliziing by gain, applying calibration, and perhaps simplifying channels – are&nbsp;**device-specific**&nbsp;requirements. They ensure that the raw sensor counts are translated into meaningful physical measures (like spectral irradiance). Researchers using other spectral devices would follow a similar procedure, adjusting for their device’s particulars (some may output spectra directly, whereas others, like <span class="in">`VEET`</span>, require reconstruction.</span>
<span id="cb41-436"><a href="#cb41-436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-437"><a href="#cb41-437" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb41-438"><a href="#cb41-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-439"><a href="#cb41-439" aria-hidden="true" tabindex="-1"></a>Some devices may output normalized counts instead of raw counts. For example, the <span class="in">`ActLumus`</span> device outputs normalized counts, while the <span class="in">`VEET`</span> device records raw counts and the gain. Manufacturers will be able to speficy exact outputs for a given model and software version.</span>
<span id="cb41-440"><a href="#cb41-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-441"><a href="#cb41-441" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb41-442"><a href="#cb41-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-443"><a href="#cb41-443" aria-hidden="true" tabindex="-1"></a><span class="fu">## `VEET` Data: Time of flight (distance)</span></span>
<span id="cb41-444"><a href="#cb41-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-445"><a href="#cb41-445" aria-hidden="true" tabindex="-1"></a>In this last section, the distance data of the <span class="in">`VEET`</span> device will be imported, analogous to the other modalities. The <span class="in">`TOF`</span> modality contains information for up to two objects in a 8x8 grid of measurements, spanning a total of about 52° vertically and 41° horizontally. Because the <span class="in">`VEET`</span> device can detect up to two objects in a given grid point, and there is a confidence value assigned to every measurement, each observation contains $2*2*8*8 = 256$ measurements.</span>
<span id="cb41-446"><a href="#cb41-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-449"><a href="#cb41-449" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-450"><a href="#cb41-450" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-VEET-overview3</span></span>
<span id="cb41-451"><a href="#cb41-451" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Overview plot of imported VEET data"</span></span>
<span id="cb41-452"><a href="#cb41-452" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb41-453"><a href="#cb41-453" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 6</span></span>
<span id="cb41-454"><a href="#cb41-454" aria-hidden="true" tabindex="-1"></a><span class="co"># Import VEET Spectral Sensor (TOF) data</span></span>
<span id="cb41-455"><a href="#cb41-455" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/01_VEET_L.csv.zip"</span></span>
<span id="cb41-456"><a href="#cb41-456" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="ot">&lt;-</span> LightLogR<span class="sc">::</span>import<span class="sc">$</span><span class="fu">VEET</span>(path, <span class="at">tz =</span> tz, <span class="at">modality =</span> <span class="st">"TOF"</span>, <span class="at">manual.id =</span> <span class="st">"VEET"</span>)</span>
<span id="cb41-457"><a href="#cb41-457" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-458"><a href="#cb41-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-459"><a href="#cb41-459" aria-hidden="true" tabindex="-1"></a>In a first step, we condition the data similarly to the other <span class="in">`VEET`</span> modalities.</span>
<span id="cb41-460"><a href="#cb41-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-463"><a href="#cb41-463" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-464"><a href="#cb41-464" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: VEET-cleaning</span></span>
<span id="cb41-465"><a href="#cb41-465" aria-hidden="true" tabindex="-1"></a><span class="co"># Aggregate spectral data to 5-minute intervals and mark gaps</span></span>
<span id="cb41-466"><a href="#cb41-466" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="ot">&lt;-</span> dataVEET3 <span class="sc">|&gt;</span></span>
<span id="cb41-467"><a href="#cb41-467" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aggregate_Datetime</span>(<span class="at">unit =</span> <span class="st">"5 secs"</span>) <span class="sc">|&gt;</span>     <span class="co"># aggregate to 5-second bins</span></span>
<span id="cb41-468"><a href="#cb41-468" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gap_handler</span>(<span class="at">full.days =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span>           <span class="co"># explicit NA for any gaps in between</span></span>
<span id="cb41-469"><a href="#cb41-469" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_Date_col</span>(<span class="at">group.by =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb41-470"><a href="#cb41-470" aria-hidden="true" tabindex="-1"></a>  <span class="fu">remove_partial_data</span>(dist1_0, <span class="at">threshold.missing =</span> <span class="st">"1 hour"</span>)</span>
<span id="cb41-471"><a href="#cb41-471" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-472"><a href="#cb41-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-473"><a href="#cb41-473" aria-hidden="true" tabindex="-1"></a>In the next step, we need to transform the <span class="in">`wide`</span> format of the imported dataset into a <span class="in">`long`</span> format, where each row contains exactly one observation for one grid-point. </span>
<span id="cb41-474"><a href="#cb41-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-477"><a href="#cb41-477" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-478"><a href="#cb41-478" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: VEET-long</span></span>
<span id="cb41-479"><a href="#cb41-479" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb41-480"><a href="#cb41-480" aria-hidden="true" tabindex="-1"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb41-481"><a href="#cb41-481" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb41-482"><a href="#cb41-482" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(Datetime, file.name, Id, is.implicit, time_stamp, modality, Date),</span>
<span id="cb41-483"><a href="#cb41-483" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">".value"</span>, <span class="st">"position"</span>),</span>
<span id="cb41-484"><a href="#cb41-484" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_pattern =</span> <span class="st">"(conf1|conf2|dist1|dist2)_(</span><span class="sc">\\</span><span class="st">d+)"</span></span>
<span id="cb41-485"><a href="#cb41-485" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb41-486"><a href="#cb41-486" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-487"><a href="#cb41-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-488"><a href="#cb41-488" aria-hidden="true" tabindex="-1"></a>In a final step before we can use the data in the analysis, we need to assign <span class="in">`x`</span> and <span class="in">`y`</span> coordinates based on the <span class="in">`position`</span> column that was created when pivoting longer. Positions are counted from 0 to 63 starting at the top left and increasing towards the right, before continuing on the left in the next row below. <span class="in">`y`</span> positions thus depend on the row count, i.e., how often a row of 8 values fits into the <span class="in">`position`</span> column. <span class="in">`x`</span> positions consequently depend on the <span class="in">`position`</span> within each 8-value row. We also add an <span class="in">`observation`</span> variable that increases by <span class="in">`+1`</span> every time, the <span class="in">`position`</span> column hits <span class="in">`0`</span>. We then center both <span class="in">`x`</span> and <span class="in">`y`</span> coordinates to receive meaningful values, i.e., 0° indicates the center of the overall measurement cone. Lastly, we convert both confidence columns, which are scaled from 0-255 into percentages by dividing them by <span class="in">`255`</span>. Empirical data from the manufacturer points to a threshold of about <span class="in">`10%`</span>, under which the respective distance data is not reliable. </span>
<span id="cb41-489"><a href="#cb41-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-492"><a href="#cb41-492" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-493"><a href="#cb41-493" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: VEET-distance preparation</span></span>
<span id="cb41-494"><a href="#cb41-494" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="ot">&lt;-</span> </span>
<span id="cb41-495"><a href="#cb41-495" aria-hidden="true" tabindex="-1"></a>  dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb41-496"><a href="#cb41-496" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ungroup() |&gt; </span></span>
<span id="cb41-497"><a href="#cb41-497" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">position =</span> <span class="fu">as.numeric</span>(position),</span>
<span id="cb41-498"><a href="#cb41-498" aria-hidden="true" tabindex="-1"></a>         <span class="at">y.pos =</span> (position <span class="sc">%/%</span> <span class="dv">8</span>)<span class="sc">+</span><span class="dv">1</span>,</span>
<span id="cb41-499"><a href="#cb41-499" aria-hidden="true" tabindex="-1"></a>         <span class="at">y.pos =</span> <span class="fu">scale</span>(y.pos, <span class="at">scale =</span> <span class="cn">FALSE</span>)<span class="sc">*</span><span class="dv">52</span><span class="sc">/</span><span class="dv">8</span>,</span>
<span id="cb41-500"><a href="#cb41-500" aria-hidden="true" tabindex="-1"></a>         <span class="at">x.pos =</span> (position <span class="sc">%%</span> <span class="dv">8</span>)<span class="sc">+</span><span class="dv">1</span>,</span>
<span id="cb41-501"><a href="#cb41-501" aria-hidden="true" tabindex="-1"></a>         <span class="at">x.pos =</span> <span class="fu">scale</span>(x.pos, <span class="at">scale =</span> <span class="cn">FALSE</span>)<span class="sc">*</span><span class="dv">41</span><span class="sc">/</span><span class="dv">8</span>,</span>
<span id="cb41-502"><a href="#cb41-502" aria-hidden="true" tabindex="-1"></a>         <span class="at">observation =</span> <span class="fu">cumsum</span>(position <span class="sc">==</span> <span class="dv">0</span>),</span>
<span id="cb41-503"><a href="#cb41-503" aria-hidden="true" tabindex="-1"></a>         <span class="fu">across</span>(<span class="fu">starts_with</span>(<span class="st">"conf"</span>), \(x) x<span class="sc">/</span><span class="dv">255</span>)</span>
<span id="cb41-504"><a href="#cb41-504" aria-hidden="true" tabindex="-1"></a>         )</span>
<span id="cb41-505"><a href="#cb41-505" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-506"><a href="#cb41-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-507"><a href="#cb41-507" aria-hidden="true" tabindex="-1"></a>Now this dataset is ready for further analysis. We finish by visualizing the same observation time on different days. Note that we replace zero distance values with <span class="in">`5 meters`</span>, which is the maximum distance the <span class="in">`VEET`</span> can measure.</span>
<span id="cb41-508"><a href="#cb41-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-511"><a href="#cb41-511" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-512"><a href="#cb41-512" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-VEET-distance</span></span>
<span id="cb41-513"><a href="#cb41-513" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Example observations of the measurement grid"</span></span>
<span id="cb41-514"><a href="#cb41-514" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb41-515"><a href="#cb41-515" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 12</span></span>
<span id="cb41-516"><a href="#cb41-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-517"><a href="#cb41-517" aria-hidden="true" tabindex="-1"></a><span class="co">#set visualization parameters</span></span>
<span id="cb41-518"><a href="#cb41-518" aria-hidden="true" tabindex="-1"></a>extras <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb41-519"><a href="#cb41-519" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(),</span>
<span id="cb41-520"><a href="#cb41-520" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb41-521"><a href="#cb41-521" aria-hidden="true" tabindex="-1"></a>                       <span class="at">oob =</span> scales<span class="sc">::</span>oob_squish_any),</span>
<span id="cb41-522"><a href="#cb41-522" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"white"</span>)),</span>
<span id="cb41-523"><a href="#cb41-523" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>(),</span>
<span id="cb41-524"><a href="#cb41-524" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="st">"none"</span>),</span>
<span id="cb41-525"><a href="#cb41-525" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> (dist1<span class="sc">/</span><span class="dv">10</span>) <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">0</span>), <span class="at">colour =</span> dist1<span class="sc">&gt;</span><span class="dv">1000</span>), <span class="at">size =</span> <span class="fl">2.5</span>),</span>
<span id="cb41-526"><a href="#cb41-526" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(),</span>
<span id="cb41-527"><a href="#cb41-527" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"X position (°)"</span>, <span class="at">y =</span> <span class="st">"Y position (°)"</span>, </span>
<span id="cb41-528"><a href="#cb41-528" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="st">"Distance (cm)"</span>, <span class="at">alpha =</span> <span class="st">"Confidence (0-255)"</span>))</span>
<span id="cb41-529"><a href="#cb41-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-530"><a href="#cb41-530" aria-hidden="true" tabindex="-1"></a>slicer <span class="ot">&lt;-</span> <span class="cf">function</span>(x){<span class="fu">seq</span>(<span class="fu">min</span>((x<span class="dv">-1</span>)<span class="sc">*</span><span class="dv">64</span><span class="sc">+</span><span class="dv">1</span>), <span class="fu">max</span>(x<span class="sc">*</span><span class="dv">64</span>, <span class="at">by =</span> <span class="dv">1</span>))} <span class="co">#allows to choose an observation</span></span>
<span id="cb41-531"><a href="#cb41-531" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-532"><a href="#cb41-532" aria-hidden="true" tabindex="-1"></a>dataVEET3 <span class="sc">|&gt;</span> </span>
<span id="cb41-533"><a href="#cb41-533" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="fu">slicer</span>(<span class="dv">9530</span>)) <span class="sc">|&gt;</span>  <span class="co">#choose a particular observation</span></span>
<span id="cb41-534"><a href="#cb41-534" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">dist1 =</span> <span class="fu">ifelse</span>(dist1 <span class="sc">==</span> <span class="dv">0</span>, <span class="cn">Inf</span>, dist1)) <span class="sc">|&gt;</span> <span class="co">#replace 0 distances with 5m</span></span>
<span id="cb41-535"><a href="#cb41-535" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(conf1 <span class="sc">&gt;=</span> <span class="fl">0.1</span> <span class="sc">|</span> dist1 <span class="sc">==</span> <span class="cn">Inf</span>) <span class="sc">|&gt;</span> <span class="co">#remove data that has less than 10% confidence</span></span>
<span id="cb41-536"><a href="#cb41-536" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x.pos, <span class="at">y=</span>y.pos, <span class="at">fill =</span> dist1<span class="sc">/</span><span class="dv">10</span>))<span class="sc">+</span> extras <span class="sc">+</span> <span class="co">#plot the data</span></span>
<span id="cb41-537"><a href="#cb41-537" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="sc">~</span>Datetime) <span class="co">#show one plot per day</span></span>
<span id="cb41-538"><a href="#cb41-538" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-539"><a href="#cb41-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-540"><a href="#cb41-540" aria-hidden="true" tabindex="-1"></a>As we can see from the figure, different days have a vastly different distribution of distance data, and measurement confidence (values with confidence &lt; 10% are removed)</span>
<span id="cb41-541"><a href="#cb41-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-542"><a href="#cb41-542" aria-hidden="true" tabindex="-1"></a><span class="fu">## Saving the Cleaned Data</span></span>
<span id="cb41-543"><a href="#cb41-543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-544"><a href="#cb41-544" aria-hidden="true" tabindex="-1"></a>After executing all the above steps, we have three cleaned data frames in our R session:</span>
<span id="cb41-545"><a href="#cb41-545" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-546"><a href="#cb41-546" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**`dataCC`**&nbsp;– the processed <span class="in">`Clouclip`</span> dataset (5-second intervals, with distance and lux, including NA for gaps and sentinel statuses).</span>
<span id="cb41-547"><a href="#cb41-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-548"><a href="#cb41-548" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**`dataVEET`**&nbsp;– the processed <span class="in">`VEET`</span> ambient light dataset (5-second intervals, illuminance in lux, with gaps filled).</span>
<span id="cb41-549"><a href="#cb41-549" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-550"><a href="#cb41-550" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**`dataVEET2`**&nbsp;– the processed <span class="in">`VEET`</span> spectral dataset (5-minute intervals, each entry containing a spectrum or derived spectral metrics).</span>
<span id="cb41-551"><a href="#cb41-551" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-552"><a href="#cb41-552" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**`dataVEET3`**&nbsp;– the processed <span class="in">`VEET`</span> distance dataset (5-second intervals, each entry containing the distance of up to two objects in the 8x8 grid).</span>
<span id="cb41-553"><a href="#cb41-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-554"><a href="#cb41-554" aria-hidden="true" tabindex="-1"></a>For convenience and future reproducibility, we will save these combined results to a single R data file. Storing all cleaned data together ensures that any analysis can reload the exact same data state without re-running the import and cleaning (which can be time-consuming for large raw files).</span>
<span id="cb41-555"><a href="#cb41-555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-558"><a href="#cb41-558" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb41-559"><a href="#cb41-559" aria-hidden="true" tabindex="-1"></a><span class="co"># Create directory for cleaned data if it doesn't exist</span></span>
<span id="cb41-560"><a href="#cb41-560" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">"data/cleaned"</span>)) <span class="fu">dir.create</span>(<span class="st">"data/cleaned"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb41-561"><a href="#cb41-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-562"><a href="#cb41-562" aria-hidden="true" tabindex="-1"></a><span class="co"># Save all cleaned datasets into one .RData file</span></span>
<span id="cb41-563"><a href="#cb41-563" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(dataCC, dataVEET, dataVEET2, dataVEET3,</span>
<span id="cb41-564"><a href="#cb41-564" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="st">"data/cleaned/data.RData"</span>)</span>
<span id="cb41-565"><a href="#cb41-565" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb41-566"><a href="#cb41-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-567"><a href="#cb41-567" aria-hidden="true" tabindex="-1"></a>The above code creates (if necessary) a folder&nbsp;<span class="in">`data/cleaned/`</span>&nbsp;and saves a single RData file (<span class="in">`data.RData`</span>) containing the three objects. To retrieve them later, one can use&nbsp;<span class="in">`&lt;- load("data/cleaned/data.RData")`</span>, which will return the objects into the environment. This single-file approach simplifies sharing and keeps the cleaned data together.</span>
<span id="cb41-568"><a href="#cb41-568" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-569"><a href="#cb41-569" aria-hidden="true" tabindex="-1"></a>In summary, this supplement has walked through the full preprocessing pipeline for two example devices. We began by describing the raw data format for each device and then demonstrated how to import the data with correct time zone settings. We handled device-specific quirks like sentinel codes (for <span class="in">`Clouclip`</span>) and multiple modalities with gain normalization (for <span class="in">`VEET`</span>). We showed how to detect and address irregular sampling, how to explicitly mark missing data gaps to avoid analytic pitfalls, and how to reduce data granularity via rounding or aggregation when appropriate. Throughout, we used functions from LightLogR in a tidyverse workflow, aiming to make the steps clear and modular. By saving the final cleaned datasets, we set the stage for the computation of visual experience metrics such as working distance, time in bright light, spectral composition ratios, as presented in the <span class="co">[</span><span class="ot">main tutorial</span><span class="co">](https://tscnlab.github.io/ZaunerEtAl_JVis_2025/)</span>. We hope this detailed tutorial empowers researchers to adopt similar pipelines for their own data, facilitating reproducible and accurate analyses of visual experience.</span>
<span id="cb41-570"><a href="#cb41-570" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-571"><a href="#cb41-571" aria-hidden="true" tabindex="-1"></a><span class="fu">## References</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>  </div> <!-- /content -->  <script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script> 
  
</body></html>